---
title: "Parsing_QC"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

knitr document van Steensel lab

# Parsing and starcode quality control
# Introduction
After demultiplexing all the samples from the sequencing data, the data has been processed by the CRISPR-TRIP snakemake script. This script maps the barcodes from the iPCR (iPCR from 5 feb 2018). It calls the true barcodes with the starcode script for both the mutation and mapping reads. It call the mutations on all the mutation samples and spits them in mapped, unmapped and non genuine.


Here we want to do a QC of the parsing of the mapping, the barcodes and mutations. We will load the statistics data of all the files. We will also look at barcode counts from the table files (those list the barcodes from the starcode stript). I obtained the counts though shell with 

Pasted the result in a text editor and changes all `^ ` and changed the spaces to `\t`, removed the last total row, removed `.table` at the end and saved as text file. 
I will export a bed file also containing the broad integrations for the generation of mean chip values over the integration sites. 
```{r render processing pipeline}
# Render Parsing QC
render_processing = TRUE

# Render Validation Processing
render_validations = TRUE

# Render Synergies 
render_synergies = TRUE
```

```{r render figures, eval=F}
# Render all figure scripts
render_all_figures = TRUE

# Individual figures
fig1 = TRUE
fig2 = TRUE
fig3 = TRUE
fig4 = TRUE
fig5 = TRUE
```


# Data loading and processing 

## Path, Libraries, Parameters and Useful Functions  

```{r libraries validations, message=FALSE}
StartTime <-Sys.time()

# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

# libraries:
library(tidyverse)
library(GenomicRanges)
library(RColorBrewer)
library(ggplot2)
library(magrittr)
library(platetools)
library(grid)
library(gridExtra)
library(purrr)
library(cowplot)
library(data.table)
library(gghighlight)
library(PerformanceAnalytics)
library(corrr)

## In directory
in.dir = "/DATA/projects/DSBrepair/git/EpiScreen/files_scripts/"
## Select out directory
out.dir = paste0("/DATA/projects/DSBrepair/data/R/rs", Date, "_episcreen")
dir.create(out.dir, showWarnings = FALSE)

## Initials (for file naming)
initials = "rs"
```

## Custom functions  

Functions used in this script:   

```{r functions}
MakeGranges <- function(x) {
  x$end_pos <- as.numeric(as.character(x$start_pos+3))
  colnames(x) <- c("name",
                   "seqname", 
                   "strand", 
                   "start",
                   "total_mapped", 
                   "mapq_sum1", 
                   "reads1", 
                   "mapq_sum2", 
                   "reads2", 
                   "seq", 
                   "end")
  gr <- makeGRangesFromDataFrame(x, keep.extra.columns = T)
  return(gr)
}

BarcodeOverlap <- function(read1, read2) {
  counts <- c(nrow(read1), 
              nrow(read2), 
              sum(read2$barcode %in% read1$barcode), 
              sum(!read1$barcode %in% read2$barcode),
              sum(!read2$barcode %in% read1$barcode))
}

serialNext = function(prefix, extension){
  i=0
  repeat {
    f = paste0(prefix, "_", i , ".", extension)
    if(!file.exists(f)){return(f)}
    i=i+1
  }
}

SetFileName <- function(filename, initials, extension) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  filename <- paste0(out.dir, "/", initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename <- serialNext(filename, extension)
  filename
}

CallTrueBarcodes <- function(df) {
  df %<>% filter(barcode %in% barcodes.clone5)
  df
}
```

## Data loading  

These are data from the crispr-trip.snake script, and a text file that has has been obtained as explained above.  

```{r Data import}
# Import metadata 
file = "/DATA/projects/DSBrepair/git/EpiScreen/config/rs20210713_E177_E1504_EpiScreen_metadata.txt"
screen_meta_data = fread(file, header = TRUE) 

# load the statistics data from the indels.
file.list <- list.files(path = "/DATA/projects/DSBrepair/data/rs20210628_EpiScreen/parsed", pattern="indelPCR[.]E1.*statistics.txt", full.names = TRUE)
indel.statistics.list <- lapply(file.list, read.table, 
                                header = TRUE)
names(indel.statistics.list)<- gsub(".*//indelPCR.indel_(.*?).statistics.txt", 
                                    "\\1", 
                                    file.list)

# Format the statistics and bind the meta data
indel.statistics.tib <- do.call(rbind.data.frame, 
                                indel.statistics.list) %>% 
  rownames_to_column(var = "ID") %>%
  as_tibble() %>%
  mutate(ID = gsub(".*indelPCR.(.*).statistics.txt", "\\1", ID)) %>%
  left_join(screen_meta_data)

# Calculate fraction written
indel.statistics.tib %<>% 
  mutate(r_written_pct = reads_written / reads)

# Import the cell viability data
cell.viab.data = fread(paste0(in.dir, "mt20200618_viability_rep1-2.csv"), 
                       sep = ";", col.names = c("ID", "E177", "E1504"))

# Process the cell viability and add it to the indel stat table. 
# Rename the ID to match this pattern : E1504_01_100_1_1_A01
cell.viab.data %<>% mutate(ID = gsub("nm|um|plate|rep", "", ID)) %>% 
  separate(ID, into = c("concentration", "tech", "drug_plate", "well", "drug_no"), sep = "_") %>% 
  # change the class of the variables to integers
  mutate(concentration = as.integer(concentration),
         tech = as.integer(tech),
         drug_plate = as.integer(drug_plate),
         drug_no = as.integer(drug_no),
         well = str_replace_all(well, "(\\D)(?=\\d(\\D|\\b))", "\\10")) %>% # add leading 0 to the well string
  gather(E177, E1504, key = "replicate", value = "viability") # make long

# calculate the % of the DMSO control (here we take the mean of DMSO per replicate)
indel.statistics.tib %<>% left_join(cell.viab.data) %>%
  group_by(replicate) %>%
  mutate(viability = viability - min(viability)) %>%
  mutate(ctrl_mean = mean(viability[drug == "DMSO"]),
         viab_norm = viability / ctrl_mean,
         conc_char = ifelse(drug %in% c("DNA-PKi","PAO", "Mirin", "DMSO"), 
                            "control", 
                            ifelse(
                              concentration == 100, 
                              paste0(concentration, " nM"), 
                              paste0(concentration, " µM"))),
         conc_char = factor(conc_char, 
                            levels = c("control", "100 nM", "1 µM", "10 µM")),
         concentration = ifelse(drug %in% c("DNA-PKi","PAO", "Mirin", "DMSO"), 
                           0, concentration)) %>%
  select(-ctrl_mean, -bp, -bp_written) # remove this column
```


# Data plotting

Couple of plots to see how the reads are distributed. This will tell us if there are any biases. 

## Plotting the general read counts per index, barcode, column and row 

```{r general plots}
ggplot(indel.statistics.tib, aes(well, log10(reads))) + 
  geom_bar(stat = "identity", aes(fill = as.character(concentration))) + 
  facet_grid(replicate ~ plate, scales = "free")

# How many reads per index?
indel.statistics.tib %>% 
  group_by(replicate, plate, seq_barcode) %>% 
  summarise(reads = sum(reads)) %>%
  ggplot(., aes(plate, reads, fill = seq_barcode)) + 
  geom_bar(stat = "identity") + 
  ggtitle("reads per index") +
  facet_grid(. ~ replicate)
# How many reads per row?
indel.statistics.tib %>% 
  separate(well, c("row", "column"), sep = 1) %>% 
  group_by(replicate, row) %>% summarise(reads = sum(reads)) %>%
  ggplot(., aes(row, reads, fill = row)) + 
  geom_bar(stat = "identity") + 
  ggtitle("reads per row") +
  facet_grid(. ~ replicate)
# How many reads per column?
indel.statistics.tib %>% 
  separate(well, c("row", "column"), sep = 1) %>% 
  group_by(replicate, column) %>% summarise(reads = sum(reads)) %>%
  ggplot(., aes(column, reads, fill = column)) + 
  geom_bar(stat = "identity") + 
  ggtitle("reads per column") +
  facet_grid(. ~ replicate)

```

```{r reads written per compound plate and replicate}
# Find reasonable limits 
limits <- c(10000, 200000)

# How many reads per well (for the 96 index) replicate 1 compound plate 1?
well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E177" & drug_plate == 1) %>%
  group_by(well) %>% 
  summarise(reads = sum(reads))

p1 = raw_map(well.statistics$reads, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", name = "reads", limits = limits) + 
  ggtitle("total reads: replicate 1 drug plate 1")

# How many reads per well (for the 96 index) replicate 1 compound plate 2?
well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E177" & drug_plate == 2) %>%
  group_by(well) %>% 
  summarise(reads = sum(reads))

p2 = raw_map(well.statistics$reads, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", name = "reads", limits = limits) + 
  ggtitle("total reads: replicate 1 drug plate 2")

# Find reasonable limits 
limits <- c(20000, 2000000)

# How many reads per well (for the 96 index) replicate 2 compound plate 1?
well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E1504" & drug_plate == 1) %>%
  group_by(well) %>% 
  summarise(reads = sum(reads))

p3 = raw_map(well.statistics$reads, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", name = "reads", limits = limits) + 
  ggtitle("total reads: replicate 2 drug plate 1")

# How many reads per well (for the 96 index) replicate 2 compound plate 2?
well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E1504" & drug_plate == 2) %>%
  group_by(well) %>% 
  summarise(reads = sum(reads))

p4 = raw_map(well.statistics$reads, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", name = "reads", limits = limits) + 
  ggtitle("total reads: replicate 2 drug plate 2")

# Compound plate 1 replicate 1
p1
# Compound plate 2 replicate 1
p2 
# Compound plate 1 replicate 2
p3
# Compound plate 2 replicate 2
p4
```


## Plate plots - total reads per well
```{r reads per well replicate 1, fig.width = 12, fig.height = 4}
for (i in 1:18) {
  
  # Select the flowjo data set
  data <- filter(indel.statistics.tib, plate == i, replicate == "E177")
  
  # Find reasonable limits 
  limits <- c(2000, 35000)
  
  # Plot this, first the read cound density
  y <- density(data$reads)
  plt1 <- ggplot(data.frame(x = y$x, y = y$y),
                 aes(x = x, y = y)) +
    geom_line() +
    geom_segment(aes(xend = x, yend = 0, colour = x)) +
    xlim(limits) +
    scale_color_distiller(type = "div", limits = limits, name = "reads") +
    theme_bw() +
    theme(legend.position="none")
  
  # Then the plate with color-coded wells
  plt2 <- raw_map(data$reads, 
                  data$well) +
    ggtitle(unique(gsub("(.*)_...$", "\\1", data$ID))) +
    scale_fill_distiller(type = "div", limits = limits, name = "reads")
  
  grid.arrange(plt1, plt2, ncol = 2, widths = c(1, 1.3))
}
```

```{r reads per well replicate 2, fig.width = 12, fig.height = 4}
for (i in 1:18) {
  
  # Select the flowjo data set
  data <- filter(indel.statistics.tib, plate == i, replicate == "E1504")
  
  # Find reasonable limits 
  limits <- c(2000, 300000)
  
  # Plot this, first the read cound density
  y <- density(data$reads)
  plt1 <- ggplot(data.frame(x = y$x, y = y$y),
                 aes(x = x, y = y)) +
    geom_line() +
    geom_segment(aes(xend = x, yend = 0, colour = x)) +
    xlim(limits) +
    scale_color_distiller(type = "div", limits = limits, name = "reads") +
    theme_bw() +
    theme(legend.position="none")
  
  # Then the plate with color-coded wells
  plt2 <- raw_map(data$reads, 
                  data$well) +
    ggtitle(unique(gsub("(.*)_...$", "\\1", data$ID))) +
    scale_fill_distiller(type = "div", limits = limits, name = "reads")
  
  grid.arrange(plt1, plt2, ncol = 2, widths = c(1, 1.3))
}
```

## Checking how many reads are written after the snakemake script per plate.
### Combined plates

```{r percent reads written per compound plate and replicate}
# % of reads written from the total reads :
# How many % written per well (for the 96 index) replicate 1 compound plate 1?
# Find reasonable limits 
limits <- c(0.6, 1)

well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E177" & drug_plate == 1) %>%
  group_by(well) %>% 
  summarise(r_written_pct = mean(r_written_pct))

p1 = raw_map(well.statistics$r_written_pct, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", limits = limits, name = "% written") + 
  ggtitle("% of reads written from the total reads: replicate 1 drug plate 1")

# How many % written per well (for the 96 index) replicate 1 compound plate 2?
well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E177" & drug_plate == 2) %>%
  group_by(well) %>% 
  summarise(r_written_pct = mean(r_written_pct))

p2 = raw_map(well.statistics$r_written_pct, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", limits = limits, name = "% written") + 
  ggtitle("% of reads written from the total reads: replicate 1 drug plate 2")

# How many % written per well (for the 96 index) replicate 2 compound plate 1?
well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E1504" & drug_plate == 1) %>%
  group_by(well) %>% 
  summarise(r_written_pct = mean(r_written_pct))

p3 = raw_map(well.statistics$r_written_pct, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", limits = limits, name = "% written") + 
  ggtitle("% of reads written from the total reads: replicate 2 drug plate 1")

# How many % written per well (for the 96 index) replicate 2 compound plate 2?
well.statistics <- indel.statistics.tib %>% 
  filter(replicate == "E1504" & drug_plate == 2) %>%
  group_by(well) %>% 
  summarise(r_written_pct = mean(r_written_pct))

p4 = raw_map(well.statistics$r_written_pct, 
             well.statistics$well) + 
  scale_fill_distiller(type = "div", limits = limits, name = "% written") + 
  ggtitle("% of reads written from the total reads: replicate 2 drug plate 2")

# Compound plate 1 replicate 1
p1
# Compound plate 2 replicate 1
p2 
# Compound plate 1 replicate 2
p3
# Compound plate 2 replicate 2
p4
```

For replicate 1, most reads did pass the filters, replicate 2 E1504 was not as good. We can see that a few wells go below the average. Let's see how that turns out in the next plots when we split that.

### Numbers per individual plate
This time split per plate to find the wells that didn't pass the filters.

```{r percent reads written per well replicate 1 and 2, fig.width = 12, fig.height = 4}
for (i in seq(1, 18, 2)) {
  
  # Select the flowjo data set
  data_dp1_rep1 <- filter(indel.statistics.tib, plate == i, replicate == "E177")
  data_dp2_rep1 <- filter(indel.statistics.tib, plate == i+1, replicate == "E177")  
  data_dp1_rep2 <- filter(indel.statistics.tib, plate == i, replicate == "E1504")
  data_dp2_rep2 <- filter(indel.statistics.tib, plate == i+1, replicate == "E1504")
  # Find reasonable limits 
  limits <- c(0, 1)
  
  # Then the plate with color-coded wells compound plate 1 rep 1
  plt1 <- raw_map(data_dp1_rep1$r_written_pct, 
                  data_dp1_rep1$well) +
    ggtitle(unique(gsub("(.*)_...$", "\\1", data_dp1_rep1$ID))) +
    scale_fill_distiller(type = "div", limits = limits, name = "r_written_pct")
  
  # Then the plate with color-coded wells compound plate 2 rep 1
  plt2 <- raw_map(data_dp2_rep1$r_written_pct, 
                  data_dp2_rep1$well) +
    ggtitle(unique(gsub("(.*)_...$", "\\1", data_dp2_rep1$ID))) +
    scale_fill_distiller(type = "div", limits = limits, name = "r_written_pct")
  
  # Then the plate with color-coded wells compound plate 1 rep 2
  plt3 <- raw_map(data_dp1_rep2$r_written_pct, 
                  data_dp1_rep2$well) +
    ggtitle(unique(gsub("(.*)_...$", "\\1", data_dp1_rep2$ID))) +
    scale_fill_distiller(type = "div", limits = limits, name = "r_written_pct")
  
  # Then the plate with color-coded wells compound plate 2 rep 2
  plt4 <- raw_map(data_dp2_rep2$r_written_pct, 
                  data_dp2_rep2$well) +
    ggtitle(unique(gsub("(.*)_...$", "\\1", data_dp2_rep2$ID))) +
    scale_fill_distiller(type = "div", limits = limits, name = "r_written_pct")
  
  grid.arrange(plt1, plt2, plt3, plt4, ncol = 2)
}
```
I these plots we see that a couple of plates have low percentage reads written generally - average ~ 50% reads written (E1504 plates 11 & 13) and some other plates have some sparse wells with 0 reads written (E1504 plates 1, 3, 7, 12, & 15) and one plates that has many wells that have pat1 missing (E1504 plate 2).

### Identification of the low reads written
```{r low reads written}
low_pct_reads_written = indel.statistics.tib %>%
  filter(r_written_pct < .7) %>% # filter for less than 50% reads written 
  mutate(pat1_missing = pat1 / reads, 
         pat2_missing = pat2 / reads) %>%
  gather(key = "string", value = "pct", pat1_missing, pat2_missing)

low_pct_reads_written %>% 
  filter(replicate == "E1504" & plate %in% c(11, 13)) %>%
  ggplot(., aes(as.character(well), pct, color = string)) + 
  geom_jitter(width = 0.3, height = 0) + 
  facet_wrap(replicate ~ plate, scales = "free_x") +
  ylim(0, 1)

low_pct_reads_written %>%
  filter(!plate %in% c(11, 13, 2)) %>%
  ggplot(., aes(as.character(well), pct, color = string)) + 
  geom_jitter(width = 0.3, height = 0) + 
  facet_wrap(replicate ~ plate, scales = "free_x") +
  ylim(0, 1)

low_pct_reads_written %>% 
  filter(replicate == "E1504" & plate == 2) %>%
  ggplot(., aes(as.character(well), pct, color = string)) + 
  geom_jitter(width = 0.3, height = 0) + 
  facet_wrap(replicate ~ plate, scales = "free_x")
```

Plates 11 and 13 from replicate E1504 seem to have a rather poor quality. 

E1504 plate 12 is missing a complete column. It also has very low reads, it seems that the robot made a mistake for that set and missed a whole column while pipetting PCR1. 

E1504 plate 2 seems to be missing a whole part of the plate, why I don't know.

## What is up with these low read wells
Some of these wells have very low reads, part is due to sequence quality as seen from above, missing patterns in the sequence.. I will check the raw reads to try to understand what is going on. But I also want to know why some wells have in general low reads, not due to the pipeline filtering them out.

```{r viability}
viab_plot_tib = indel.statistics.tib %>%
  distinct(drug, conc_char,
           tech,
           replicate, target, viab_norm) %>%
  group_by(drug, conc_char, replicate) %>%
  mutate(viab_var = var(viab_norm, na.rm = T),
         viab_mean = mean(viab_norm, na.rm = T)) %>%
  select(-viab_norm) %>%
  distinct() %>%
  pivot_wider(names_from = replicate, values_from = c(viab_mean, viab_var)) %>%
  filter(!is.na(conc_char)) %>%
  mutate(viab_diff = abs(viab_mean_E1504-viab_mean_E177)) %>%
  distinct(drug,
           target,
           conc_char,
           viab_diff,
           viab_mean_E1504,
           viab_mean_E177,
           viab_var_E1504,
           viab_var_E177)

viab_plot_tib 

PAO_data =viab_plot_tib %>% filter(drug == "PAO")
DMSO_data = viab_plot_tib %>% filter(drug == "DMSO")
Mirin_data = viab_plot_tib %>% filter(drug == "Mirin")

odd_data = viab_plot_tib %>% 
  filter(viab_diff > 0.25)

drug_conc_odd = odd_data %>% 
  mutate(viab_reproducibility = FALSE) %>% 
  distinct(drug, conc_char, viab_reproducibility)

control.colors <- c("PAO" = "red", "DMSO" = "yellow", "Mirin" = "brown", "Not reproducible" = "orange")


# Highlighting the controls, concentration and outlyers. 
ggplot(viab_plot_tib, 
       aes(viab_mean_E1504, viab_mean_E177, color = viab_var_E1504)) + 
  geom_point() +
  geom_point(data=odd_data,
             aes(x=viab_mean_E1504,y=viab_mean_E177),
             color='orange',
             size=3) +
  geom_point(data=PAO_data,
             aes(x=viab_mean_E1504,y=viab_mean_E177),
             color='red',
             size=3) +
  geom_point(data=DMSO_data,
             aes(x=viab_mean_E1504,y=viab_mean_E177),
             color='yellow',
             size=3) +
  geom_point(data=Mirin_data,
             aes(x=viab_mean_E1504,y=viab_mean_E177),
             color='brown',
             size=3) +
  geom_abline(slope = 1, intercept = 0.2, linetype = 6) +
  geom_abline(slope = 1, intercept = -0.2, linetype = 6) +
  geom_abline(slope = 1, intercept = 0) +
  coord_fixed() +
  ylim(0, 1.2) +
  xlim(0, 1.2) +
  theme_bw(base_size = 16) + 
  facet_grid(. ~ conc_char) +
  labs(x = "Mean viability E1504",
       y = "Mean viability E177",
       color = "Technical variance")
```

### Check which drugs have low reproducibility in cell viability. 
This might be due to different reasons, mainly unknown... Let's also keep a list of these just in case. Here I arbitrarily use a 2x increase or decrease in cell viability ratio to DMSO. 
```{r drugs with low viability reproducibility}
for (i in unique(viab_plot_tib$target)) {
  p = filter(viab_plot_tib, target == i) %>%
    ggplot(.,
           aes(viab_mean_E1504, viab_mean_E177)) + 
    geom_point() +
    gghighlight(viab_diff > 0.20 &
                  (viab_mean_E1504 > .5 |
                     viab_mean_E177 > .5),
                label_key = drug,
                max_highlight = 100L) +
    geom_abline(slope = 1, intercept = 0.2, 
                linetype = "dashed", alpha = 0.3) +
    geom_abline(slope = 1, intercept = -0.2, 
                linetype = "dashed", alpha = 0.3) +
    geom_vline(xintercept = 0.5, 
               linetype = "dashed", alpha = 0.3) +
    geom_hline(yintercept = 0.5, 
               linetype = "dashed", alpha = 0.3) +
    coord_fixed() +
    ylim(0, 1.2) +
    xlim(0, 1.2) +
    # scale_color_brewer(palette = "YlGnBu") +
    theme_bw(base_size = 16) + 
    facet_grid(. ~ conc_char) +
    theme(legend.title = element_blank()) +
    ggtitle(i)
  print(p)  
}
```


```{r low read wells}
# Filter all the wells for less than 2000 reads, this would be equal to about 100 reads per 
# barcode per well. This is a rough check, we might go for this or for 30 reads min for 
# +1/-7.
# That ratio is the minimal ratio we want for our balance calculations
low.reads = indel.statistics.tib %>% filter(reads < 2000)

# How many wells are there? 
low_r = indel.statistics.tib %>% filter(reads < 2000) %>% 
  distinct(ID, concentration) %>% 
  group_by(concentration, replicate) %>% 
  summarise(count_low = n())
hi_r = indel.statistics.tib %>% filter(reads >= 2000) %>% 
  distinct(ID, concentration) %>% 
  group_by(concentration, replicate) %>% 
  summarise(count_high = n()) %>% left_join(low_r)

# Fraction of total:
hi_r %>% mutate(fraction = count_low / (count_low+count_high))

# Quick plot to check if this is related to the cell viability
indel.statistics.tib %>% 
  mutate(viab_threshhold = ifelse(viab_norm < 0.5 | reads < 2000, "discarded", "viable & sufficient reads")) %>% 
  ggplot(., aes(viab_norm, log10(reads), color = viab_threshhold)) +  
  geom_point() + 
  facet_grid(replicate ~ conc_char) 
```

The proportion of wells that we cannot use for the balance is in general very low ~ 5%. The one plate in E1504 100 nM lost 25% of the wells, makes sense with the plots above.

In general there are fewer reads in the low cell viability samples but the lowest ones are more due to library prep, missing primers (for the 8 samples in E1504). Let's keep it simple. However we should still make sure the viability correlates between replicates. Let's plot that in Supplemental figure 1. 
```{r correlation between experiments}
viab_matrix = indel.statistics.tib %>%
  mutate(ID = gsub("^E.*_.._(.*_)[123]_([12]_...)$", "\\1\\2", ID)) %>% 
  distinct(ID, tech, replicate, viab_norm) %>%
  pivot_wider(., names_from = c("replicate", "tech"), values_from = "viab_norm") %>%
  column_to_rownames(., var = "ID")


cor_df = viab_matrix %>% 
  correlate() %>% 
  shave() # Shave off the upper triangle for a clean result

cor_df

rplot(cor_df)

# 
chart.Correlation(viab_matrix, histogram=TRUE, pch=19)

```

The two replicates correlate quite well except for some of these outliers in E1504 that were less deadly to the cells.


```{r data export}
viab_plot_tib %<>% 
  pivot_longer(., cols = viab_mean_E1504:viab_var_E177, 
               names_to = c("var", "replicate"), 
               names_pattern = "(viab_.*)_(E.+)", 
               values_to = "value") %>%
  pivot_wider(., 
              names_from = var, 
              values_from = value, 
              names_repair = "check_unique") %>%
  unique() %>%
  left_join(drug_conc_odd) %>% 
  mutate(viab_reproducibility = ifelse(is.na(viab_reproducibility), TRUE, FALSE))

filename <- SetFileName("Episcreen_Reads_Viab_Metadata", initials = initials, extension = "RDS")

indel.statistics.tib %>%
  select(-reads_written, -n_tooshort,
         -index, -pat1, -barcode,  -pat2) %>%
  left_join(viab_plot_tib)  %>%
  write_rds(.,
            file = filename)
```


# Conclusion
- In general, the data looks very good
- Plates 11 and 13 from replicate E1504 seem to have a rather poor quality. 
- E1504 plate 2 seems to be missing a whole part of the plate, why I don't know.
- E1504 plate 12 is missing a complete column. It also has very low reads, it seems that the robot made a mistake for that set and missed a whole column while pipetting PCR1. 


```{r render processing scripts}
# Render screen processing
if(render_processing) {
rmarkdown::render("2_Preprocessing_EpiScreen_Screen.Rmd")
}

# Render validations
if(render_validations) {
rmarkdown::render("2_Preprocessing_Validations.Rmd")
}

# Render validations
if(render_synergies) {
rmarkdown::render("3_Synergies_Processing.Rmd")
}
```


# Session Info  
```{r session info QC}
sessionInfo()
getwd()
date()
paste("Run time: ",format(Sys.time()-StartTime))
rm(list = ls(all.names = TRUE)) #will clear all objects includes hidden objects.
gc() #free up memrory and report the memory usage.
```