---
title: "Parsing_QC"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  pdf_document:
    toc: yes
  editor_options:
    chunk_output_type: console
  html_document:
    theme: journal
    highlight: monochrome
    toc: yes
    toc_float: yes
    code_folding: show
---

knitr document van Steensel lab

# Data loading and processing 

## Path, Libraries, Parameters and Useful Functions  

```{r message=FALSE}
StartTime <-Sys.time()

# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

# libraries:
library(plyr)
library(dplyr)
library(GenomicRanges)
library(RColorBrewer)
library(ggplot2)
library(magrittr)
library(platetools)
library(grid)
library(gridExtra)
```
 
## Custom functions  

Functions used in this script:   

```{r}
MakeGranges <- function(x) {
  x$end_pos <- as.numeric(as.character(x$start_pos+3))
  colnames(x) <- c("name",
                   "seqname", 
                   "strand", 
                   "start",
                   "total_mapped", 
                   "mapq_sum1", 
                   "reads1", 
                   "mapq_sum2", 
                   "reads2", 
                   "seq", 
                   "end")
  gr <- makeGRangesFromDataFrame(x, keep.extra.columns = T)
  return(gr)
}

BarcodeOverlap <- function(read1, read2) {
  counts <- c(nrow(read1), 
              nrow(read2), 
              sum(read2$barcode %in% read1$barcode), 
              sum(!read1$barcode %in% read2$barcode),
              sum(!read2$barcode %in% read1$barcode))
}

SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), filename)
  filename
}
```

## Data loading  

These are data from the crispr-trip.snake script, and a text file that has has been obtained as explained above.  

```{r}
# First I load the statistics data from the mapping.
setwd("/DATA/projects/DSBrepair/data/rs20210526_EpiScreen/parsed/")

file.list <- list.files(pattern='indelPCR.(.*?).statistics.txt')
indel.statistics.list <- lapply(file.list, 
                                read.table, 
                                header = TRUE)
names(indel.statistics.list)<- gsub('indelPCR.(.*?).statistics.txt', 
                                    '\\1', 
                                    file.list)
# we need to remove the index column for the time being
indel.statistics.list <- lapply(indel.statistics.list, function(x) { x["index"] <- NULL; x })
indel.statistics.df <- do.call(rbind.data.frame, 
                               indel.statistics.list)
indel.statistics.df$sample.well <- rownames(indel.statistics.df)
indel.statistics.df$concentration <- gsub("^(.*)_rep.*", "\\1", indel.statistics.df$sample.well)
indel.statistics.df$replicate <- gsub("^.*_(.*?)_.*", "\\1", indel.statistics.df$sample.well)
#indel.statistics.df$plate <- as.character(rep(c(1:18), each = 96))
indel.statistics.df$sample <- gsub("(.*_.*_.*?)_.*_.*$", "\\1", indel.statistics.df$sample.well)
indel.statistics.df$well <- gsub(".*_.*_(.*?)_[0-999]*", "\\1", indel.statistics.df$sample.well)
indel.statistics.df$row <- gsub("[0-999]", "\\1", indel.statistics.df$well)
indel.statistics.df$column <- gsub("[A-Z]", "\\1", indel.statistics.df$well)
indel.statistics.df$drug <- gsub(".*_([0-999])", "\\1", indel.statistics.df$sample.well)
```

### Indel parsing  


```{r}
low.reads.indel.stat.df <- indel.statistics.df[order(indel.statistics.df[, "reads"]), ][1:100, ]
mypalette5<-brewer.pal(5,"Greens")

barplot(as.matrix(t(low.reads.indel.stat.df[c(1:20),c(1,3,6:8)])), 
        col = mypalette5,
        beside = T, 
        main = "Read parsing statistics", 
        ylab = "reads", 
        legend = colnames(indel.statistics.df[,c(1,3,6:8)]), 
        las = 2,
        cex.names = 0.4,
        cex.axis = 0.6,
        args.legend = list(x = "topright", cex = 0.6))

barplot(as.matrix(t(low.reads.indel.stat.df[c(1:50),c(1,3)])), 
        col = c("Red", "Orange"),
        beside = T, 
        main = "Read parsing statistics", 
        ylab = "reads", 
        legend = colnames(indel.statistics.df[,c(1,3)]), 
        las = 2,
        cex.names = 0.4,
        cex.axis = 0.6,
        args.legend = list(x = "topright", cex = 0.6))

barplot(as.matrix(t(low.reads.indel.stat.df[c(51:100),c(1,3)])), 
        col = c("Red", "Orange"),
        beside = T, 
        main = "Read parsing statistics ", 
        ylab = "reads", 
        legend = colnames(indel.statistics.df[,c(1,3)]), 
        las = 2,
        cex.names = 0.4,
        cex.axis = 0.6,
        args.legend = list(x = "topright", cex = 0.6))


ggplot(indel.statistics.df, aes(sample, reads)) + 
  geom_bar(stat = "identity", aes(fill = concentration)) +
  theme_bw() +
  coord_flip()

ggplot(indel.statistics.df[indel.statistics.df$replicate=="rep1" & indel.statistics.df$concentration=="1um", ], aes(sample, reads)) + 
  geom_bar(stat = "identity", aes(fill = replicate)) +
  theme_bw() +
  coord_flip()

# How many reads per sample?
indel.statistics.df %>% group_by(sample) %>% summarise(reads = sum(reads))
# How many reads per row?
indel.statistics.df %>% group_by(row) %>% summarise(reads = sum(reads))
# How many reads per column?
indel.statistics.df %>% group_by(column) %>% summarise(reads = sum(reads))
# How many reads per well (for the 96 index)?
well.statistics <- indel.statistics.df %>% group_by(well) %>% summarise(reads = sum(reads))
well.statistics
raw_map(well.statistics$reads, 
                 well.statistics$well) + scale_fill_distiller(type = "div", name = "reads")

# % of reads written from the total reads :
indelreadswritten.vec <- indel.statistics.df$reads_written/indel.statistics.df$reads*100
names(indelreadswritten.vec) <- rownames(indel.statistics.df)
barplot(indelreadswritten.vec, 
        las=2, 
        cex.names = 0.6)

```

```{r, fig.width = 12, fig.height = 4}
for (i in unique(indel.statistics.df$sample)) {
  
  # Select the flowjo data set
  data <- indel.statistics.df[indel.statistics.df$sample == i, ]
  
  # Find reasonable limits 
  limits <- c(2000, 25000)

  # Plot this, first the PI staining with a custom color-coded density
  y <- density(data$reads)
  plt1 <- ggplot(data.frame(x = y$x, y = y$y),
                 aes(x = x, y = y)) +
    geom_line() +
    geom_segment(aes(xend = x, yend = 0, colour = x)) +
    xlim(limits) +
    ggtitle(paste("Reads per well -",
                unique(data$sample))) +
    scale_color_distiller(type = "div", limits = limits, name = "PI_area") +
    theme_bw() +
    theme(legend.position="none")
  
  # Then the plate with color-coded wells
  plt2 <- raw_map(data$reads, 
                 data$well) +
    ggtitle(unique(data$sample)) +
    scale_fill_distiller(type = "div", limits = limits, name = "reads")
  plt2
  grid.arrange(plt1, plt2, ncol = 2, widths = c(1, 1.3))
}
```

# Conclusion
- In general, the data looks very good
- Some plates have wells with 0 reads - other wells take over the reads in this plate - we have to investigate further whether this can be a problem.

# Session Info  
```{r}
sessionInfo()
getwd()
date()
paste("Run time: ",format(Sys.time()-StartTime))

```