---
title: "rs20181010_IndelSpectraPlotting"
output: html_document
---

# knitr document van Steensel lab

# CRISPR - TRIP project
## Ruben Schep, from 20181010

### Introduction
In this project we want to combine TRIP and CRISPR to test the kinetics, fidelity and 
pathways of the DNA damage response. We use a CRISPR guide that creates a +1 and -7 indel
in a consistent way. We found that these indels reflect the relative activity of NHEJ and
MMEJ respectively. By putting this sequence in different locations throughout the genome,
we hope to be able to discover how these two pathways react to their local chromatin 
environment.

The integrations were mapped by iPCR and the indelspectra were obtained by sequencing the
PCR over the breaksites. The mapping links a barcode to a region in the genome, the 
integrations can be found in the table folder of the TRIP output folder. The indels were
seperated by index and filtered on genuine (starcode) and mapped (overlap with the 
mapping) and can be found in the mutations table (.genuine.mapped.table files). 

The goal of this script is to create various plots of indel spectra, in forms of 
histograms. 

We are going to plot the total indel spectra of each sample, meaning the total of all
integrations.

Then we will plot all the spectra of all the barcodes within one sample. In this file
there will only be a preview, the full list can be obtained by making the PDF.

Finally, we will plot the indel spectra on one barcode in the different samples, to get a
feeling of how the integration behaves in different samples (biological and technical
replicates).

### Description of Data

For this analysis we need the indel data of the TRIP integrations. These files are 
obtained with the crispr_trip.snake script that C. Leemans edited. This data contains the
indel frequencies at each integration.

The mutations were called by counting the distance between two constant regions. These
were separated by barcode. The barcodes were also filtered on the starcode, to pick out
the most abundant, and considered real, ones.

The latest data can be found in : 
~/mydata/projects/RSTP2/20171024_RSTP2_2000-4653/rs20171110_trip_hg38/


Mutations files : /mutation/*genuine_mapped.table

| barcode  | type | score | 
| ------- | --------- | ----- | 
| TTCTATTCGCACACAA | ins | 1 |
| TTTCCCACATCAGGAG | wt | 0 |
| CCATAGTAGTGATTAC | del | -4 |


### Path, Libraries, Parameters and Useful Functions

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

# libraries:
library(GenomicRanges)
library(rtracklayer)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(magrittr)
library(dplyr)
library(tibble)
```

### Custom functions
Functions used thoughout this script.

```{r}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), filename)
  filename
}


ReadsPerBarcode <- function(mutations) {
  #This function sums up all the mutations to count the number of reads per barcode.
  df <- data.frame(barcode = row.names(mutations),
             reads = rowSums(mutations))
  df
}

RemoveLargeDeletions <- function(x, lowerindels = as.character(c(-120:-11, 3:20))) {
 x <- x[ , -which(names(x) %in% lowerindels)]
}

SumLargeIndels <- function(x, maxshift = 10, maxinsert = 5) {
  min <- as.numeric(names(x)[1])

  if(-maxshift<=min){
    lowerindels <- NA
  } else{
    lowerindels = as.character(c(min:-(maxshift+1))) 
  }
  
  y <- (which(names(x) %in% lowerindels))
    
  if(length(y)>0){
    sum.y <- rowSums(as.matrix(x[,y]))
    x <- x[, -y]
    x <- cbind(sum.y, x)
    colnames(x)[1] <- as.character(-(maxshift+1))
  }
  
  max <- as.numeric(names(x)[length(names(x))])
  
  if(maxinsert>=max){
    higherindels <- NA
  } else{
    higherindels = as.character(c((maxinsert+1):max))
  }

  z <- (which(names(x) %in% higherindels))
  
  if(length(z)>0){
    sum.z <- rowSums(as.matrix(x[,z]))
    x <- x[, -z]
    x <- cbind(x, sum.z)
    colnames(x)[length(colnames(x))] <- as.character(maxinsert+1)
  }
  
  shiftrange <- c(-(maxshift+1): (maxinsert+1)) 
  if(length(colnames(x))< length(shiftrange)){
    missing <- shiftrange[!shiftrange%in%colnames(x)]
    df_new <- matrix(0, nrow=nrow(x), ncol=length(shiftrange) )
    colnames(df_new) <- as.character(shiftrange)
    for(m in 1:length(shiftrange)){
      ms <- shiftrange[m]
      if(ms%in%colnames(x)){
      df_new[,as.character(ms)] <- x[,as.character(ms)]
      }
    }
   df_new <- as.data.frame(df_new)
   rownames(df_new) <- rownames(x)
   x <- df_new
  }
  return(x)
}  

OverallIndels <- function(x) {
  colSums(x[,sapply(x, is.numeric)])
}


OverallIndelPlot <- function(indellist) {
  
  OverallIndels <- function(x) {
  colSums(x[,sapply(x, is.numeric)])
  }
  
  sum.list <- lapply(indellist, OverallIndels)
    
  par(mfrow=c(1,1), pty = "m")
  for (j in 1:length(sum.list)) { 

    y <- (sum.list[[j]][1:length(sum.list[[j]])] / sum(sum.list[[j]])*100)
  
    shiftrange <- names(y)
    
    sum(y) #should be 100
    indel.summary <- round(y,1)
  
    if(length(shiftrange)==0){
        shiftrange <- b_del <- b_ins <- 0
      }
    b_del <- abs(min(as.numeric(shiftrange)))
    b_ins <- abs(max(as.numeric(shiftrange)))
    range <- c(-b_del:b_ins)
  
  COL <- "black"
  COL[!as.numeric(names(y))==0] <- "black"
  COL[as.numeric(names(y))==0] <- "#808184"
  COL[as.numeric(names(y))==-7] <- "#3D4FA1"
  COL[as.numeric(names(y))==1] <- "#EE3F2A"
  
    if(length(range)>1){
       bp <- barplot(as.numeric(indel.summary), 
                     main=paste("Sample LBR2 2000", names(indellist)[j]),
                     col=COL, 
                     border = COL,
                     names.arg = names(indel.summary),
                     xlab="type of indel",
                     ylim=c(0, 100), 
                     ylab="% of sequences"
                     )
    
       #above each group of bars: show percentage (mean across four bases)
       text(bp[y>5],
            (y+5)[y>5], 
            as.character(((round(y,1))[y>5]))
            )
    
       #display Rsq values as an indication of the accuracy:
    
       eff <- round((100) - as.numeric(y["0"]),1)
    
       dl <- (as.numeric(names(y))<0)
       insert <- (as.numeric(names(y))>0)
       min7 <- (as.numeric(names(y))==-7)
       plus1 <- (as.numeric(names(y))==1)
       deletions <- round(sum(as.vector(y[dl]), na.rm = TRUE),1)
       insertions <- round(sum(as.vector(y[insert]), na.rm = TRUE),1)
       m7 <- round(sum(as.vector(y[min7])),1)
       p1 <- round(sum(as.vector(y[plus1])),1)

    
       legend("topleft", legend=c(paste("total eff. =", eff, "%"), 
                                  paste("total del =", deletions, "%"), 
                                  paste("total ins =", insertions, "%"), 
                                  paste("+1/-7 =", round((p1/m7),1)), 
                                  paste("total reads =", sum(sum.list[[j]])),
                                  paste("total barcodes = ",  nrow(indellist[[j]]))
                                  ), 
             bty="n") 
      }
  }
}


SampleIndelSpectra <- function(j, name="sample", barcodes.no=nrow(j)) {
  par(mfrow=c(2,3), pty = "m")
  if(barcodes.no>nrow(j)){
    stop(paste("you enter more barcodes than the data has"))
  }
     
  for (i in 1:barcodes.no) { 

  y <- (j[i, ][1:ncol(j[i, ])] / rowSums(j[i,]) *100)
  
  sum(y)
  shiftrange <- colnames(y)
  
  indel.summary <- round(y[1,],1)
  
  if(length(shiftrange)==0){
    shiftrange <- b_del <- b_ins <- 0
  }
  b_del <- abs(min(as.numeric(shiftrange)))
  b_ins <- abs(max(as.numeric(shiftrange)))
  range <- c(-b_del:b_ins)
  
  COL <- "black"
  COL[!as.numeric(names(y))==0] <- "black"
  COL[as.numeric(names(y))==0] <- "#808184"
  COL[as.numeric(names(y))==-7] <- "#3D4FA1"
  COL[as.numeric(names(y))==1] <- "#EE3F2A"


  
  
  if(length(range)>1){
    bp <- barplot(as.numeric(indel.summary), 
                  main=paste(name),
                  col=COL, 
                  border = COL,
                  names.arg = names(indel.summary),
                  xlab="type of indel",
                  ylim=c(0, 100), 
                  ylab="% of sequences"
                  )
    
    #above each group of bars: show percentage (mean across four bases)
    text(bp[y>5], 
         (y+5)[y>5], 
         as.character(((round(y,1))[y>5]))
         )
    
    #display Rsq values as an indication of the accuracy:
    
    eff <- round((100) - as.numeric(y["0"]),1)
    
    dl <- (as.numeric(names(y))<0)
    insert <- (as.numeric(names(y))>0)
    min7 <- (as.numeric(names(y))==-7)
    plus1 <- (as.numeric(names(y))==1)
    deletions <- round(sum(as.vector(y[, dl]), na.rm = TRUE),1)
    insertions <- round(sum(as.vector(y[, insert]), na.rm = TRUE),1)
    m7 <- round(sum(as.vector(y[, min7])),1)
    p1 <- round(sum(as.vector(y[, plus1])),1)

    
    legend("topleft", 
           legend=c(paste("Barcode: ", rownames(j)[i]), 
                    paste("total eff. =", eff, "%"),
                    paste("total del =", deletions, "%"),
                    paste("total ins =", insertions, "%"),
                    paste("+1/-7 =", round((p1/m7),1)),
                    paste("total reads =", round(as.numeric(rowSums(j[i,])), 2))
                    ), 
           bty="n")
    }
  }
}

BarcodeIndelSpectrum <- function(indellist, i) {
  par(mfrow=c(2,2), pty = "m")
  i <- as.character(i)
  for (j in 1:length(indellist)) { 
  
  if(!i %in% rownames(indellist[[j]])) {
    next
  }
  
  y <- (indellist[[j]][i, ][1:ncol(indellist[[j]][i, ])] / sum(indellist[[j]][i,]) *100)
  
  sum(y)
  shiftrange <- colnames(y)
  
  indel.summary <- round(y[1,],1)
  
  if(length(shiftrange)==0){
    shiftrange <- b_del <- b_ins <- 0
  }
  b_del <- abs(min(as.numeric(shiftrange)))
  b_ins <- abs(max(as.numeric(shiftrange)))
  range <- c(-b_del:b_ins)
  
  COL <- "black"
  COL[!as.numeric(names(y))==0] <- "#1874CD"
  COL[as.numeric(names(y))==0] <- "#98F5FF"
  
  if(length(range)>1){
    bp <- barplot(as.numeric(indel.summary), 
                  main=paste("Sample LBR2 2000", names(indellist)[j]),
                  col=COL, 
                  border = COL,
                  names.arg = names(indel.summary),
                  xlab="type of indel",
                  ylim=c(0, 100), 
                  ylab="% of sequences"
                  )
    
    #above each group of bars: show percentage (mean across four bases)
    text(bp[y>5], 
         (y+5)[y>5], 
         as.character(((round(y,1))[y>5])))
    
    #display Rsq values as an indication of the accuracy:
    
    eff <- round((100) - as.numeric(y["0"]),1)
    
    dl <- (as.numeric(names(y))<0)
    insert <- (as.numeric(names(y))>0)
    min7 <- (as.numeric(names(y))==-7)
    plus1 <- (as.numeric(names(y))==1)
    deletions <- round(sum(as.vector(y[, dl]), na.rm = TRUE),1)
    insertions <-  round(sum(as.vector(y[, insert]), na.rm = TRUE),1)
    m7 <-  round(sum(as.vector(y[, min7])),1)
    p1 <-  round(sum(as.vector(y[, plus1])),1)

    
    legend("topleft", 
           legend=c(paste("Barcode: ", rownames(indellist[[j]][i, ])), 
                    paste("total eff. =", eff, "%"), 
                    paste("total del =", deletions, "%"), 
                    paste("total ins =", insertions, "%"), 
                    paste("+1/-7 =", round((p1/m7),1)), 
                    paste("total reads =", round(sum(indellist[[j]][i, ]), 2))
                    ), 
           bty="n")
  }
}
}

FrequencyReadsMut <- function(indellist, experiments_name){
  mutations.regular.list <- indellist[experiments_name]
  names(mutations.regular.list)
  dat <- do.call(rbind, setNames(mutations.regular.list, NULL)) # This is a big data.frame with all data 
  repl <- length(mutations.regular.list)
  
  # make rowname row 1
  dat <- tibble::rownames_to_column(dat, "barcode")
  
  # This creates a dataframe with different numbers after each barcodes we do not want that.
  dat$barcode <- gsub('[0-9]+', '', dat$barcode)
  reads.mut.barcode <- dat %>%
    group_by(barcode) %>%
    summarise_all(funs(sum))
  
  #counts of each barcode in the various replicates
  reads.mut.barcode_counts <- dat %>%
    group_by(barcode) %>%
    tally()
  
  reads.mut.barcode.freq <- as.data.table(reads.mut.barcode) # Make a data.table
  reads.mut.barcode.freq[,2:length(names(reads.mut.barcode))] <-
    reads.mut.barcode[,2:length(names(reads.mut.barcode))]/rowSums(reads.mut.barcode[,2:length(names(reads.mut.barcode))])*100
  reads.mut.barcode.freq <- melt(reads.mut.barcode.freq, id = "barcode")
  return(list(reads.mut.barcode.freq=reads.mut.barcode.freq))
  }

BeeswarmIndelSpectrum <- function(readfrequency){
   # Plot it in a beeswarm plot
  ggplot(readfrequency$reads.mut.barcode.freq, aes(variable, value)) + geom_quasirandom() + theme_bw() + ylim(0, 100)
}

BarcodesToRownames <- function(x) {
  tmp <- x %>% remove_rownames %>% column_to_rownames(var="V2")
  tmp
}

```

### Data import
The mutation data was produced on 2017.10.10 with as minimal read count per barcode per
sample set at 1000. This means that every barcode will have at least 1000 reads. 

The files can be found in the /mutation folder of the trip results.

```{r}
# Set directory to the mutation output folder of the CRISPR-TRIP snakemake script
setwd("/DATA/projects/DSBrepair/data/R/")
load("rs20181206_All.IndelFreqData.RData")
file.list <- list.files(pattern = "*_All.IndelFreqData.RData")
RData <- sort(file.list, decreasing = TRUE)[1] #take the most recent.
load(RData)

setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/data/files_scripts/")
mutations.list <- get(load("mt20190510_mutations.list"))
drugs.list <- read.delim("mt20190326_DrugNames.txt")
```

### Some data pre-processing
We need to combine all the mutations, that for the moment are in a data.frame that looks 
like this: 

| barcode  | type | score | 
| ------- | --------- | ----- | 
| TTCTATTCGCACACAA | ins | 1 |
| TTTCCCACATCAGGAG | wt | 0 |
| CCATAGTAGTGATTAC | del | -4 |

We use the MutationsToDatafame function that counts all the barcodes and lists them in a
dataframe (one per sample) where each column represents the indel size and each row is 
the barcode. The dataframes are listed together, so that we have one list of dataframes
(samples) that has two inputs (barcodes and indel size)

We also count the sum of de indels in each barcode in each sample to get the total amount
of reads per barcode. 

```{r}
# Make the sums of the reads per barcode
mutations.list <- lapply(mutations.list, BarcodesToRownames)
barcodesum <- lapply(mutations.list, ReadsPerBarcode)

# Let's keep the total mutations list also.
mutations.list.large <- mutations.list
mutations.list <- lapply(mutations.list, SumLargeIndels, maxshift = 15)
```

Conclusion : We now have a dataframes with somewhat reduced (max -15 deletion) indels.
We can use this for plot the indel spectra in various ways. 

### Plotting the overall indel spectrum for each sample.
Here we start by processing the data, so that we get only the sums of all the indels in
each sample. We can then plot these in a simple way, a seperate PDF can also be generated.

```{r overall}
# I want to check the overall indel spectra. Here we calculate the total sums of all the
# indels, and keep them in a seperate list, it was too laborious to create a single 
# dataframe. This works as wel, because the other plots also rely on dataframe lists.

# Apply the sum of the columns over all the dataframes in the list.
sum.list <- lapply(mutations.list, OverallIndels)

# We can plot all these
###OverallIndelPlot(sum.list)
OverallIndelPlot(mutations.list)

# To create a PDF of these figures, for the small mutations.
setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/data/Results/")
filename <- SetFileName("_Overall_IndelSpectra.pdf", "mt")
pdf(filename, width = 11, height = 8)
OverallIndelPlot(mutations.list)
dev.off()
```




### Plotting the indel spectrum for one specific well.
We want to be able to have a quick look at how all the barcodes look in one sample.
moreover, we want to see how they vary (i.e. in what extent the location creates
variation)

```{r samples, echo=FALSE}
# Preview of a few wells in a sample.
names(mutations.list)

# View the sample that you want to call from this list
View(drugs.list)

# Name the sample:
plot_sample <- c("100nm_rep1_plate1_F4_064.count.table")
####### Function that creates the PDF ######
# Generate file name with date and sample
if (plot_sample %in% names(mutations.list)) {
  # Date is called at the beginning of the script.
  filename <- paste("mt", Date, "Overall_IndelSpectra" , plot_sample, ".pdf", sep = "_")
  filename <- gsub("-", "", filename)
  filename <- gsub("_[.]", ".", filename)
  ### Run PDF
  setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/data/Results/")
  pdf(filename, width = 11, height = 8)
  ### make the plots
  #SampleIndelSpectra(plot_sample)
  SampleIndelSpectra(j=mutations.list[[plot_sample]], name=plot_sample, barcodes.no = 19)

  dev.off()
} else {
  print(paste(plot_sample, " is not a correct sample name"))
}
```





## Plotting the indel sprectra for one barcode in all the samples.
Again this can also be saved as a PDF.

```{r barcodes}
setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/data/Results/")
# This script produces histograms of the indels of a specific barcode in all the samples.
# Preview :
# #BarcodeIndelSpectrum("ACTGTCGAGTTGTCCG)
# BarcodeIndelSpectrum(indellist = mutations.list, i="ACTGTCGAGTTGTCCG")

# To make a PDF of these plots
filename <- SetFileName("_ACTGTCGAGTTGTCCG.pdf", "mt")
pdf(filename, width = 11, height = 8)
#BarcodeIndelSpectrum("ACTGTCGAGTTGTCCG")
BarcodeIndelSpectrum(indellist = mutations.list, i="ACTGTCGAGTTGTCCG")
dev.off()
```


# Group experiments
```{r sample names}
samples_names <- unique(names(mutations.list))

DMSO_samples <- grep("_B1_|_B12_|_F1_|_F12_", samples_names, value=TRUE)

Mirin_samples <- grep("_C1_|_C12_|_G1_|_G12_", samples_names, value=TRUE)

DNAPK_samples <- grep("_D1_|_D12_|_H1_|_H12_", samples_names,  value=TRUE)

All_samples <- grep("", samples_names,  value=TRUE)
```

## Beeswarm plots of all the integrations in one. 
```{r plotting}
### Beeswarm of all the indel for all barcodes of multiple experiments
mutations.list.dmso <- mutations.list[DMSO_samples]
names(mutations.list.dmso)
dat <- do.call(rbind, setNames(mutations.list.dmso, NULL)) # This is a big data.frame with all data 
repl <- length(mutations.list.dmso)

# make rowname row 1
dat <- tibble::rownames_to_column(dat, "barcode")

# This creates a dataframe with different numbers after each barcodes, we do not want that.
dat$barcode <- gsub('[0-9]+', '', dat$barcode)
head(dat)
reads.mut.barcode <- dat %>%
  group_by(barcode) %>%
  summarise_all(funs(sum))

# Test the new data frame with the sums of the barcodes 
reads.mut.barcode[reads.mut.barcode$barcode=="ACTGTCGAGTTGTCCG", -1] == colSums(dat[dat$barcode=="ACTGTCGAGTTGTCCG", -1])

#counts of each barcode in the various replicates
reads.mut.barcode_counts <- dat %>%
  group_by(barcode) %>%
  tally()

reads.mut.barcode.freq <- as.data.table(reads.mut.barcode)
reads.mut.barcode.freq[,2:length(names(reads.mut.barcode))] <- reads.mut.barcode[,2:length(names(reads.mut.barcode))]/rowSums(reads.mut.barcode[,2:length(names(reads.mut.barcode))])*100
reads.mut.barcode.freq <- melt(reads.mut.barcode.freq, id = "barcode")

# Plot it in a beeswarm plot
ggplot(reads.mut.barcode.freq, aes(variable, value)) + geom_quasirandom() + theme_bw() + ylim(0, 100)

Freq.DMSO_samples <- FrequencyReadsMut(mutations.list, DMSO_samples)
BeeswarmIndelSpectrum(Freq.DMSO_samples)

Freq.Mirin_samples <- FrequencyReadsMut(mutations.list, Mirin_samples)
BeeswarmIndelSpectrum(Freq.Mirin_samples)

Freq.DNAPK_samples <- FrequencyReadsMut(mutations.list, DNAPK_samples)
BeeswarmIndelSpectrum(Freq.DNAPK_samples)

Freq.All_samples <- FrequencyReadsMut(mutations.list, All_samples)
BeeswarmIndelSpectrum(Freq.All_samples)
```
  
### Conclusions 

### Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

