---
title: "DSB Episcreen data analysis & plotting"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
$(document).ready(function() {
$items = $('div#TOC li');
$items.each(function(idx) {
num_ul = $(this).parentsUntil('#TOC').length;
$(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
});

});
</script>

```{css, echo = FALSE}
div.sourceCode {
overflow-x: hidden;
}
```
### Libraries
### Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(knitr)
library(tidyverse)
library(outliers)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
# library(ggrepel)
library(GGally)
library(MASS)
# library(DESeq2)
# library(gghighlight)
library(ggrepel)
# library(platetools)
library(ggpubr)
# library(Laurae)
library(pheatmap)
library(RColorBrewer)
# # library(ggrastr)
library(broom)
# library(ggsignif)
# library(rstatix)
library(colorspace)
library(analogue)

## Select outdir
out.dir = paste0("./figures/", Date, "/")
dir.create(out.dir)
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

### Functions

```{r load_functions, echo = FALSE, warning = FALSE}

p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
  ## Methods 'Hommel', 'BH', 'BY' and speed improvements
  ## contributed by Gordon Smyth
  method <- match.arg(method)
  if(method == "fdr") method <- "BH"	# back compatibility
  nm <- names(p)
  p <- as.numeric(p)
  p0 <- setNames(p, nm)
  if(all(nna <- !is.na(p))) nna <- TRUE
  p <- p[nna]
  lp <- length(p)
  # stopifnot(n >= lp)
  if (n <= 1) return(p0)
  if (n == 2 && method == "hommel") method <- "hochberg"
  
  p0[nna] <-
    switch(method,
           bonferroni = pmin(1, n * p),
           holm = {
             i <- seq_len(lp)
             o <- order(p)
             ro <- order(o)
             pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
           },
           hommel = { ## needs n-1 >= 2 in for() below
             if(n > lp) p <- c(p, rep.int(1, n-lp))
             i <- seq_len(n)
             o <- order(p)
             p <- p[o]
             ro <- order(o)
             q <- pa <- rep.int( min(n*p/i), n)
             for (j in (n-1):2) {
               ij <- seq_len(n-j+1)
               i2 <- (n-j+2):n
               q1 <- min(j*p[i2]/(2:j))
               q[ij] <- pmin(j*p[ij], q1)
               q[i2] <- q[n-j+1]
               pa <- pmax(pa,q)
             }
             pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
           },
           hochberg = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
           },
           BH = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             pmin(1, cummin( n / i * p[o] ))[ro]
           },
           BY = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             q <- sum(1L/(1L:n))
             pmin(1, cummin(q * n / i * p[o]))[ro]
           },
           none = p)
  p0
}

SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {
  
  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)
  
  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }
  
  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())
  
  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]
  
  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }
  
  p <- p +
    theme(panel.background = element_rect(fill = corCol))
  
  return(p)
}


theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
    theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
          strip.background = element_rect(fill = "white", color = "white")
    )
}
theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
    theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
          strip.background = element_rect(fill = "white", color = "white")
    )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
    theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
          strip.background = element_rect(fill = "white", color = "white")
    )
}

theme_classic_lines_90_noborder <- function() {
  theme_pubr(border = F, legend = "top", x.text.angle = 90) +
    theme(strip.background = element_rect(fill = "white", color = "white")
    )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
# colors_drugs <- c("DMSO (neg control)" = "#65BFA4", "DNA-PKi (NHEJ control)" = "#8CA0C4", "drugs" = "#808184", "Mirin (MMEJ control)" = "#5D79AC")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 
## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

colors_targets <-  c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
                     HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
                     HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
                     PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43")

conc_control_colors = c("CTRL" = colors_drugs[2], "100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

```


### Data import

```{r data import, echo = FALSE, warning = FALSE}
# Import data from preprocessing script
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
file = list.files(in.dir, pattern = "episcreen_ratios", full.names = T) %>% tail(n = 1)
indel.data <- readRDS(file) # %>% filter(pathway_bal_filter == TRUE & viab_reproducibility == TRUE)

# Import data from preprocessing validation script
file.validation = list.files(in.dir, pattern = "validation_ratio", full.names = T) %>% tail(n = 1)
indel.data.validation <- readRDS(file.validation)

chromatin.features <- colnames(indel.data)[grepl("[.]zscore", colnames(indel.data))]
chrom_data = indel.data %>% dplyr::select(barcode, all_of(chromatin.features)) %>% distinct()  %>%
  rename_with(., ~gsub(".zscore", "", .x))

chromatin.features.short = gsub(".zscore", "", chromatin.features)
```


```{r}
#data frame to save data
mean_sd_fits <- tibble(term = NA, estimate = NA, std.error = NA, barcode = NA, rep = NA)


# First calculate means and sd of the DMSO efficiency per replicate and barcode
dmso.eff.score <- indel.data %>%
  filter(drug == "DMSO") %>%
  dplyr::select('DMSO' = MMEJratio, well, barcode, replicate, tech, plate) %>%
  mutate(DMSO = ave(DMSO, well, replicate, tech, plate, FUN = function(x) mean(x, na.rm=T))) %>%
  distinct(replicate, tech, well, plate, DMSO) %>% 
  pull(DMSO) %>%
  fitdistr(., "normal") %>% tidy()
```

```{r Compounds that affect the balance at all the sites}
## Which compounds change the balance at all the sites in both screens?
all_balance = indel.data %>% 
  distinct(drug, conc_char, replicate, zscore_bal_comb_rep, target, 
           barcode, viab_reproducibility, pathway_bal_filter) %>% 
  filter(!is.na(zscore_bal_comb_rep) & 
           !between(zscore_bal_comb_rep, -1.96, 1.96) & 
           viab_reproducibility == T & 
           pathway_bal_filter == T) %>% 
  group_by(drug, conc_char, target) %>% 
  count() %>% 
  filter(n == 36) # 2 x 18

kable(all_balance)
all_balance %>% ungroup() %>% distinct(drug, target)

## Now with 16/18 (~ 90%) of the sites with an effect. Near global
near_all_balance = indel.data %>% 
  distinct(drug, conc_char, replicate, zscore_bal_comb_rep, target, 
           barcode, viab_reproducibility, pathway_bal_filter) %>% 
  filter(!is.na(zscore_bal_comb_rep) & 
           !between(zscore_bal_comb_rep, -1.96, 1.96) & 
           viab_reproducibility == T & 
           pathway_bal_filter == T) %>% 
  group_by(drug, conc_char, target) %>% 
  count() %>% 
  filter(n >= 32) # 2 x 16


kable(near_all_balance)
near_all_balance %>% ungroup() %>% distinct(drug, target)
```

## A: MMEJ-z-score per drug
Are the compounds (per concentration) that significantly alter the MMEJ score?
```{r warning = FALSE}
# Plot raw MMEJ ratio
dmso_bal <- indel.data %>%
  filter(drug == "DMSO") %>%
  dplyr::select('dmso_bal' = MMEJratio, IPR, replicate, tech, well, plate) %>%
  group_by(replicate, tech, well, plate) %>%
  summarise(dmso_bal = mean(dmso_bal)) %>%
  ungroup() %>%
  mutate(dmso_mean = mean(dmso_bal),
         dsmo_sd = sd(dmso_bal)) %>%
  distinct(dmso_mean, dsmo_sd)



volcano.df = indel.data %>%
  filter(sample == "Epigenetic Compound" & pathway_bal_filter == TRUE & viab_reproducibility == TRUE) %>%
  distinct(MMEJratio, drug, conc_char, zscore_bal_global_comb_rep) %>%
  mutate(GlobalMMEJratio = ave(MMEJratio, drug, conc_char, FUN = mean)) %>%
  distinct() %>%
  mutate(MMEJ_pval_global_comb = 2*pnorm(q = abs(zscore_bal_global_comb_rep), lower.tail = F)) %>%
  left_join(near_all_balance) %>%
  mutate(MMEJ_pval_adj_global_comb = log10(p.adjust(MMEJ_pval_global_comb, method = 'fdr')),
         col_bal = ifelse(MMEJ_pval_adj_global_comb < 0.01 & 
                            GlobalMMEJratio > (dmso_bal$dmso_mean + dmso_bal$dsmo_sd), "MMEJ increase", 
                          ifelse(MMEJ_pval_adj_global_comb < 0.01 & 
                                   GlobalMMEJratio < (dmso_bal$dmso_mean - dmso_bal$dsmo_sd), "NHEJ increase","blue"))) %>%
  distinct(MMEJ_pval_adj_global_comb, GlobalMMEJratio, drug, conc_char, target, n, col_bal)


ggplot(volcano.df, 
       aes(y = MMEJ_pval_adj_global_comb, 
           x = GlobalMMEJratio, 
           color = col_bal)) +
  geom_point() +
  geom_hline(linetype = "dashed", yintercept = log10(0.01)) +
  geom_vline(linetype = "solid", xintercept = (dmso_bal$dmso_mean)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_bal$dmso_mean + dmso_bal$dsmo_sd)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_bal$dmso_mean - dmso_bal$dsmo_sd)) +
  scale_color_manual(values = c("#C2C2C2", "#3d405b", "#e07a5f")) +
  facet_wrap(~conc_char) +
  annotate("rect",
           xmin=(dmso_bal$dmso_mean - dmso_bal$dsmo_sd),
           xmax=(dmso_bal$dmso_mean + dmso_bal$dsmo_sd),
           ymin=-Inf,ymax=Inf, alpha=0.1, fill="black") +
  annotate("rect",
           xmin=-Inf,
           xmax=Inf,
           ymin=-Inf,
           ymax= -log10(0.01), 
           alpha=0.1, 
           fill="black") +
  coord_flip() +
  ylim(0, -300) + 
  geom_label_repel(data = volcano.df, 
                   aes(y = MMEJ_pval_adj_global_comb, 
                       x = GlobalMMEJratio, 
                       label = ifelse(n == 36, drug,'')),
                   box.padding   = 1, 
                   point.padding = 0.5,
                   force = 210,
                   segment.color = 'black', 
                   inherit.aes = F) +
  geom_label_repel(aes(label = ifelse(between(n, 32, 35), drug,'')),
                   box.padding   = 1, 
                   force = 110,
                   point.padding = 0.5,
                   segment.color = 'gray50') 



indel.data %>%
  filter(sample == "Epigenetic Compound" & pathway_bal_filter == TRUE & viab_reproducibility == TRUE) %>%
  distinct(MMEJratio, drug, conc_char, zscore_bal_global_comb_rep) %>%
  mutate(GlobalMMEJratio = ave(MMEJratio, drug, conc_char, FUN = mean)) %>%
  distinct() %>%
  mutate(zscore_bal_global_comb_rep = 2*pnorm(q = abs(zscore_bal_global_comb_rep), lower.tail = F)) %>%
  mutate(MMEJ_pval_adj_global_comb = log10(p.adjust(zscore_bal_global_comb_rep, method = 'fdr')),
         col_bal = ifelse(MMEJ_pval_adj_global_comb < 0.01 & 
                            GlobalMMEJratio > (dmso_bal$dmso_mean + dmso_bal$dsmo_sd), "green", 
                          ifelse(MMEJ_pval_adj_global_comb < 0.01 & 
                                   GlobalMMEJratio < (dmso_bal$dmso_mean - dmso_bal$dsmo_sd), "yellow","blue"))) %>%  
  filter(col_bal != "blue") %>%
  distinct(drug, conc_char)
```


*Figure 3A: Compound-induced change in MMEJ ratio of all tested compounds. Mean MMEJ z-score per drug & concentration.* 

**Conclusion: Yes, there are a lot of interesting hits at all tested concentrations.**

---

## B: MMEJ ratio change per target group
Are there any target groups that are prone to changes in MMEJ ratio?
```{r warning = FALSE}
# C: Bar plot
eff.data <- indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  distinct(zscore_bal_global_comb_rep, drug, target, conc_char, MMEJratio) %>%
  mutate(MMEJratio = ave(MMEJratio, drug, conc_char, FUN = mean)) %>%
  distinct() %>%
  mutate(MMEJ_pval_global_comb = 2*pnorm(q = abs(zscore_bal_global_comb_rep), lower.tail = F)) %>%
  mutate(MMEJ_pval_adj_global_comb = p.adjust(MMEJ_pval_global_comb, method = 'fdr'))

eff.data$sign <- "2_No"
eff.data$sign[eff.data$MMEJratio > (dmso_bal$dmso_mean + dmso_bal$dsmo_sd) & eff.data$MMEJ_pval_adj_global_comb < 0.01] <- "1_Up"
eff.data$sign[eff.data$MMEJratio < (dmso_bal$dmso_mean - dmso_bal$dsmo_sd) & eff.data$MMEJ_pval_adj_global_comb < 0.01] <- "0_Down"

eff.data <- eff.data %>%
  mutate(number = ave(target, target, sign, conc_char, FUN = length)) %>%
  mutate(number = as.numeric(number)) %>%
  distinct(conc_char, target, sign, number) %>%
  spread(sign, number, fill = 0) %>%
  mutate(fraction = (`1_Up` + `0_Down`) / (`2_No` + (`1_Up` + `0_Down`))) %>%
  pivot_longer(`0_Down`:`2_No`, names_to = "sign", values_to = "number")


ggplot(eff.data,
       aes(y = reorder(target, number), x = number, fill = sign)) +
  geom_bar(stat = "identity") +
  facet_wrap(~conc_char)+
  scale_fill_manual(values = c("#e07a5f", "#3d405b", "#C2C2C2")) 
```

*Figure 2B: Number of compounds with significant changes in MMEJ ratio per target group.*

**Conclusion: Some target groups have more changes. But there are no groups significantly sticking out.**

---
```{r 2C Top TIF boosting, , fig.width = 12, fig.height = 10}
rank_dt = indel.data %>% 
  filter(drug != "DMSO" & pathway_bal_filter == TRUE & viab_reproducibility == TRUE) %>%
  distinct(drug, conc_char, zscore_bal_global_comb_rep , target) %>% 
  arrange(desc(zscore_bal_global_comb_rep )) %>%
  mutate(rank = seq(1:n()),
         drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " "))

p1 = rank_dt %>% 
  # filter(rank < 100 ) %>%
  ggplot(., aes(rank, zscore_bal_global_comb_rep )) +
  geom_point()  +
  scale_color_manual(values = colors_targets) +
  geom_label_repel(aes(label = ifelse(drug == "DNA-PKi", drug_conc,'')),
                   box.padding   = 1, 
                   point.padding = 0.5,
                   segment.color = 'gray50') +
  geom_hline(linetype = "dashed", yintercept = 2.58) +
  geom_hline(linetype = "dashed", yintercept = -2.58) +
  # scale_x_reverse(limits = c(100, 0)) +
  scale_x_reverse() +
  ylab("Global combined MMEJ ratio z-score")

p2 = rank_dt %>%
  filter(rank <= 100 & drug != "DNA-PKi") %>%
  ggplot(., aes(rank, zscore_bal_global_comb_rep , color = target)) +
  geom_point()  +
  scale_color_manual(values = colors_targets) +
  gghighlight(rank <= 8, label_key = drug_conc) +
  geom_hline(linetype = "dashed", yintercept = 2.58) +
  scale_x_reverse(limits = c(100, 0)) +
  # scale_x_reverse() +
  ylab("Global combined MMEJ ratio z-score")

maxrank = rank_dt %>% nrow()

p3 = rank_dt %>%
  filter(rank > nrow(.)-100) %>%
  ggplot(., aes(rank, zscore_bal_global_comb_rep , color = target)) +
  geom_point()  +
  scale_color_manual(values = colors_targets) +
  gghighlight(rank > maxrank-10, label_key = drug_conc) +
  geom_hline(linetype = "dashed", yintercept = -2.58) +
  scale_x_reverse(limits = c(maxrank, maxrank-100)) +
  # scale_x_reverse() +
  ylab("Global combined MMEJ ratio z-score")


left_col_zoom <- plot_grid(p2, p3, labels = c('B', 'C'), label_size = 12, ncol = 1)

plot_grid(p1, left_col_zoom, labels = c('A', ''), label_size = 12, ncol = 2, rel_widths = c(3, 2))

top_hits = indel.data  %>% 
  filter(conc_char != "control" & pathway_bal_filter == T & viab_reproducibility == TRUE) %>% 
  distinct(drug, conc_char, zscore_bal_global_comb_rep , target) %>%
  arrange(desc(zscore_bal_global_comb_rep )) %>% slice_head(n = 10) %>% distinct(drug, conc_char, target, zscore_bal_global_comb_rep ) %>% 
  mutate(rank = seq(1:n()),
         drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " "))
```


```{r}
# Data table and compute log2_ratio (This line needs to be adjusted)
filtered_drug_conc = indel.data %>%
  filter(abs(zscore_TIF_comb_rep) > 2.58 | abs(zscore_TIF_comb_rep) > 2.58 | drug == "DMSO") %>%
  mutate(drug_conc = paste(drug, conc_char,sep = " "),
         effect_on_indelrate = ifelse(abs(zscore_TIF_comb_rep) > 2.58, TRUE, FALSE),
         effect_on_ratio = ifelse(abs(zscore_bal_comb_rep) > 2.58, TRUE, FALSE)) %>%
  distinct(drug, conc_char, drug_conc, effect_on_indelrate, effect_on_ratio) %>%
  mutate(counts = ave(drug_conc, drug, conc_char, FUN = function(x) length(x)),
         keep = ifelse(counts > 1 & !effect_on_indelrate | counts > 1 & !effect_on_ratio, FALSE, TRUE)) %>%
  filter(keep) %>% 
  distinct(drug, conc_char, drug_conc, effect_on_indelrate, effect_on_ratio)


ratios_tib <- indel.data %>%
  mutate(drug_conc = paste(drug, conc_char, sep = " ")) %>%
  semi_join(filtered_drug_conc) %>%
  distinct(MMEJratio, freqCut, drug, conc_char, drug_conc, target, barcode) %>%
  mutate(MMEJratio = ave(MMEJratio, drug, conc_char, barcode, FUN = mean)) %>%
  mutate(MMEJratio.log2 = log2(MMEJratio)) %>%
  mutate(freqCut = (ave(freqCut, drug, conc_char, barcode, FUN = mean))) %>%
  mutate(freqCut.log2 = log2(freqCut)) %>%
  distinct()

dmso_ratios_tib <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(MMEJratio, freqCut, barcode) %>%
  mutate(DMSO.ratio = ave(MMEJratio, barcode, FUN = mean)) %>%
  mutate(DMSO.ratio.log2 = log2(DMSO.ratio)) %>%
  mutate(DMSO.freqCut = ave(freqCut, barcode, FUN = mean)) %>%
  mutate(DMSO.freqCut.log2 = log2(DMSO.freqCut)) %>%
  dplyr::select(-MMEJratio, -freqCut) %>%
  distinct()

ratios_tib <- merge(ratios_tib, dmso_ratios_tib)
```

The log2 distance ratio in our case is positive if it promotes NHEJ and negative if it promotes NHEJ


# 2nd step check compute distances and plot them
```{r}
# Each replicate has a different value (compute differences by replicate)
ratios_tib_dist <- ratios_tib %>% 
  dplyr::group_by(barcode, drug, conc_char) %>% 
  mutate(log2.dist.ratio = MMEJratio.log2 - DMSO.ratio.log2, 
         log2.dist.freqCut = freqCut.log2 - DMSO.freqCut.log2) %>%
  left_join(chrom_data)
```


# Extract slopes for all genes
```{r}
# Perform analysis across all features for three test genes
ratios_tib_dist %<>% mutate(drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  filter(!is.na(MMEJratio))

slope.protein.features <- tibble(drug_conc = NA, feature = NA,  term = NA, 
                                 indelrate = NA, p.value.indelrate = NA, 
                                 ratio = NA, p.value.ratio = NA)

for (i in unique(ratios_tib_dist$drug_conc)) {
  for (j in chromatin.features.short) {
    model_tib <- ratios_tib_dist %>% filter(drug_conc == i) 
    
    # Indel rates
    model.indelrate.log2 <- lm(formula = log2.dist.freqCut ~ unlist(model_tib[j]), 
                               data = model_tib) %>% 
      tidy()
    
    # Pathway balance
    model.ratio.log2 <- lm(formula = log2.dist.ratio ~ unlist(model_tib[j]), 
                           data = model_tib) %>% 
      tidy()
    
    slope.protein.features <- slope.protein.features %>% 
      add_row(drug_conc = i, 
              feature = j, 
              term = model.indelrate.log2 %>%
                pull(term),
              indelrate = model.indelrate.log2 %>% 
                pull(estimate), 
              p.value.indelrate = model.indelrate.log2 %>% 
                pull(p.value),
              ratio = model.ratio.log2 %>% 
                pull(estimate), 
              p.value.ratio = model.ratio.log2 %>% 
                pull(p.value))
  }
}

# Remove the 1st NA column
slope.protein.features %<>% 
  filter(!is.na(feature)) %>%
  mutate(term = ifelse(term == "(Intercept)", "intercept", "slope"))
```

```{r}
CCD_tib = ratios_tib_dist %>% dplyr::select(barcode, drug_conc, log2.dist.freqCut, log2.dist.ratio, all_of(chromatin.features.short)) %>% ungroup()

#Create an empty dt with CCDs of DDR proteins
drug_conc_CCDs_dt <- tibble(var = NA, drug_conc = NA, num_comp = NA, r.squared = NA, adj.r.squared = NA,p.value = NA)
# chrom_formula = as.formula(paste("y ~ ", paste(chromatin.features, collapse= "+")))

for (i in unique(CCD_tib$drug_conc)){
  for (j in c("log2.dist.freqCut", "log2.dist.ratio")) {
    drug_conc_dt <- filter(CCD_tib, drug_conc == i)
    # Run a model per drug_conc
    set.seed(1)
    chrom_formula = reformulate(chromatin.features.short, response = j)
    PCR_model_DDR_test <- pcr(chrom_formula,
                              data=drug_conc_dt , 
                              validation="CV")
    
    pcr_pred <- predict(PCR_model_DDR_test, 
                        drug_conc_dt, ncomp = 3)
    
    combined.dt <- tibble(measured = drug_conc_dt %>% pull(j), 
                          predicted = as.numeric(pcr_pred))
    
    pred_vs_estim <- lm(formula = measured ~ predicted, 
                        data = combined.dt) %>% 
      glance()
    
    drug_conc_CCDs_dt <- drug_conc_CCDs_dt %>% 
      add_row(var = j,
              drug_conc = i, 
              r.squared = pred_vs_estim %>% 
                pull(r.squared), 
              adj.r.squared = pred_vs_estim %>% 
                pull(adj.r.squared), 
              p.value = pred_vs_estim %>% 
                pull(p.value))
  }
}

#Correct model to adjust for multiple testing correction
adj_p.value_KO_model <- drug_conc_CCDs_dt %>% 
  dplyr::select(var, num_comp, p.value, drug_conc) %>% 
  group_by(var) %>%
  mutate(p.adj = p.adjust(p.value, method = "BH")) %>% 
  dplyr::select(var, drug_conc,p.value,p.adj)

signif_hits = adj_p.value_KO_model %>% filter(p.adj < 0.01) %>% 
  distinct(drug_conc, var) %>% 
  ungroup() %>% 
  mutate(val = TRUE,
         var = gsub("log2.dist", "signif", var)) %>%
  pivot_wider(names_from = var, 
              values_from = val, 
              values_fill = FALSE) 
```


```{r epistasis filtering}
# Calculate postition where the linear model crosses y = 0, we do this by -intercept / slope
epistasis_tib = slope.protein.features %>% 
  distinct(drug_conc, term, feature, indelrate, ratio) %>% 
  pivot_wider(., names_from = term, values_from = c(indelrate, ratio)) %>%
  mutate(feature = gsub(".zscore", "", feature),
         x_intercept_indelrate = -indelrate_intercept / indelrate_slope,
         x_intercept_ratio = -ratio_intercept / ratio_slope)

effect_compounds = ratios_tib_dist %>%
  group_by(drug_conc)  %>% 
  mutate(indelrate_pos = sum(log2.dist.freqCut > 0),
         indelrate_effect = ifelse(indelrate_pos >= 17, "indel_rate_increase", 
                                   ifelse(indelrate_pos <= 1, "indel_rate_decrease", "mixed")),
         ratio_pos = sum(log2.dist.ratio > 0),
         ratio_effect = ifelse(ratio_pos >= 17, "NHEJ_increase", 
                               ifelse(ratio_pos <= 1, "MMEJ_increase", "mixed"))) %>%
  distinct(drug_conc, indelrate_effect, ratio_effect) %>% 
  left_join(distinct(filtered_drug_conc, drug_conc, effect_on_indelrate, effect_on_ratio))


epistasis_tib %<>% left_join(effect_compounds)  %>% 
  left_join(signif_hits) %>%
  mutate(
    # x_intercept_indelrate = ifelse(p.value.indelrate < 0.05,
    #                                x_intercept_indelrate, NA),
    # x_intercept_ratio = ifelse(p.value.ratio < 0.05,
    #                            x_intercept_ratio, NA),
    indelrate_slope_plot = ifelse(x_intercept_indelrate < 1 & effect_on_indelrate & signif.freqCut, indelrate_slope, NA),
    ratio_slope_plot = ifelse(x_intercept_ratio < 1 & effect_on_ratio & signif.ratio, ratio_slope, NA)) %>%
  distinct(drug_conc, feature, indelrate_slope_plot, ratio_slope_plot)
```


```{r prepare heatmaps, fig.width= 10, fig.height= 20}
clone5.matrix = indel.data %>% 
  dplyr::select(IPR, ends_with(".zscore")) %>% 
  distinct() %>% 
  rename_with(., ~gsub(".zscore", "", .x)) %>%
  column_to_rownames(var="IPR")

clone5_map = hclust(dist(t(clone5.matrix)))
clone5_map_ipr = dendextend::rotate(hclust(dist(clone5.matrix)), c(1:2,5:8,3:4, 15:14, 17, 16, 18 ,13,10:12,9))

chromatin <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(indel.data, barcode, IPR)) 

chromatin <- chromatin %>%
  filter(IPR != "<NA>") %>% 
  dplyr::select(-barcode) %>%
  column_to_rownames(var="IPR")

# Set annotation for heatmap
viab_drug = indel.data %>% filter(drug != "PAO") %>% 
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  distinct(viab_mean, drug_conc, replicate) %>% group_by(drug_conc) %>%
  summarise(viab = mean(viab_mean))

# Row annotation: annotate drugs with the target group
target <- indel.data %>% 
  # filter(!target %in% c("Negative Control", "MRN", "DNA-PK")) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>%
  left_join(viab_drug) %>%
  distinct(drug_conc, conc_char, target, viab) %>% 
  column_to_rownames(var="drug_conc")

TIF_IPR = indel.data %>% filter(drug == "DMSO") %>%
  distinct(replicate, tech, freqCut,MMEJratio, IPR, well, plate) %>%
  group_by(IPR) %>%
  summarise(TIF = mean(freqCut),
            MMEJratio = mean(MMEJratio))

chromatin_TIF = chromatin %>% rownames_to_column(var = "IPR") %>%
  left_join(TIF_IPR) %>%
  column_to_rownames("IPR")
```



```{r vorinostat synergy}
vorinostat_ratios =  indel.data.validation %>% 
  filter(replicate == "mean" & plasmid == "LBR2" & pool == "AB") %>% 
  distinct(barcode, drug, NHEJ_MMEJ , efficiency) %>% 
  pivot_wider(names_from = drug, values_from = c(efficiency, NHEJ_MMEJ)) %>%
  mutate(diff_efficiency = efficiency_Vorinostat - efficiency_DMSO,
         ratio_efficiency = efficiency_Vorinostat/efficiency_DMSO,
         logFC_NHEJ_MMEJ = log2(NHEJ_MMEJ_Vorinostat / NHEJ_MMEJ_DMSO))  %>%
  filter(!is.na(logFC_NHEJ_MMEJ) & !is.na(ratio_efficiency) & !is.infinite(logFC_NHEJ_MMEJ))

vorinostat_ratios %<>% left_join(unique(indel.data.validation[, c(1, 42:75)]), by = "barcode") %>% distinct(barcode, .keep_all = TRUE)

chromatin.features <- colnames(indel.data.validation)[42:65]

slope.vorinostat.features <- tibble(feature = NA,  term = NA, 
                                    ratio = NA, p.value = NA)

for (j in chromatin.features) {
  model_tib <- vorinostat_ratios
  
  # Pathway balance
  model.ratio.log2 <- lm(formula = logFC_NHEJ_MMEJ ~ unlist(model_tib[j]), 
                         data = model_tib) %>% 
    tidy()
  
  slope.vorinostat.features <- slope.vorinostat.features %>% 
    add_row(feature = j, 
            term = model.indelrate.log2 %>%
              pull(term),
            ratio = model.ratio.log2 %>% 
              pull(estimate), 
            p.value = model.ratio.log2 %>% 
              pull(p.value))
}

# Remove the 1st NA column
slope.vorinostat.features %<>% 
  filter(!is.na(feature)) %>%
  mutate(term = ifelse(term == "(Intercept)", "intercept", "slope"))

# Calculate postition where the linear model crosses y = 0, we do this by -intercept / slope
synergy_tib = slope.vorinostat.features %>% 
  distinct(term, feature, ratio, p.value) %>% 
  pivot_wider(., names_from = term, values_from = c(ratio, p.value)) %>%
  mutate(x_intercept_ratio = -ratio_intercept / ratio_slope,
         p.adj.slope = p.adjust(p.value_slope, method = "fdr"))

synergy_tib %>% 
  mutate(x_intercept_ratio = ifelse(p.adj.slope < 0.05, x_intercept_ratio, NA),
         ratio_slope_plot = ifelse(x_intercept_ratio < 1, ratio_slope, NA)) %>%
  filter(!is.na(ratio_slope_plot))


# pdf(paste0(figure_out, "rs20220210_vorinostat_synergy.pdf"), width=8, height=4)
p1 = vorinostat_ratios %>% ggplot(., aes(H3K27me3, logFC_NHEJ_MMEJ)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16)
p2 = vorinostat_ratios %>% ggplot(., aes(EZH2, logFC_NHEJ_MMEJ)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16)
plot_grid(p1, p2)
# dev.off()
```



```{r}
vorino_screen = ratios_tib %>% filter(drug == "Vorinostat (SAHA, MK0683)") %>% 
  ungroup() %>%
  dplyr::select(all_of(chromatin.features.short), barcode, log2.FC.ratio, NHEJratio, DMSO.ratio) %>% 
  dplyr::rename(screen_FC = log2.FC.ratio, DMSO_screen = DMSO.ratio) %>%
  distinct(barcode, screen_FC, H3K27me3, NHEJratio, DMSO_screen) %>%
  filter(!is.na(screen_FC))

vorino_pool_screen = vorinostat_ratios %>% 
  mutate(barcode = gsub(".[AB]$", "", barcode)) %>% 
  filter(barcode %in% vorino_screen$barcode) %>% 
  distinct(barcode, logFC_NHEJ_MMEJ, H3K27me3, NHEJ_MMEJ_Vorinostat, NHEJ_MMEJ_DMSO) %>%
  dplyr::rename(H3K27me3_pool = H3K27me3,
                pool_FC = logFC_NHEJ_MMEJ, 
                NHEJratio_pool = NHEJ_MMEJ_Vorinostat, 
                DMSO_pool = NHEJ_MMEJ_DMSO) %>%
  left_join(vorino_screen)

ggplot(vorino_pool_screen, aes(pool_FC, screen_FC)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16) + xlab("Log2 FC pool") + ylab("Log2 FC screen")

ggplot(vorino_pool_screen, aes(H3K27me3_pool, H3K27me3)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16) + xlab("Log2 FC pool") + ylab("Log2 FC screen")


ggplot(vorino_pool_screen, aes(NHEJratio_pool, NHEJratio)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16) + xlab("NHEJ/MMEJ pool") + ylab("NHEJ/MMEJ screen")

ggplot(vorino_pool_screen, aes(DMSO_pool, DMSO_screen)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16) + xlab("NHEJ/MMEJ pool") + ylab("NHEJ/MMEJ screen")
```

```{r}
FC_ttest = indel.data %>% 
  filter(pathway_bal_filter == T) %>%
  distinct(MMEJ_log2_FC, IPR, conc_char, drug, chromatin, replicate, tech) %>% 
  group_by(IPR, drug, conc_char) %>%
  do(tidy(t.test(.$MMEJ_log2_FC))) %>% 
  ungroup() %>%
  mutate(p.adj = p.adjust(p.value, method = "fdr"))


```


```{r}
Vorino_rep_check = indel.data %>% 
  filter(drug == "Vorinostat (SAHA, MK0683)") %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore, IPR, tech, replicate, chromatin) %>% 
  mutate(tech_rep = paste(replicate, tech, sep = "_"))

p1 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_log2_FC)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("Vorinostat MMEJ log2 fold change")
p2 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_zscore)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("Vorinostat MMEJ zscores")

plot_grid(p1, p2, ncol = 1)
```
```{r}
indel.data %>% pull(drug) %>% unique() %>% .[grepl("MC-1293", .)]

Vorino_rep_check = indel.data %>% 
  filter(drug == "PCI-24781 (Abexinostat)") %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore, IPR, tech, replicate, chromatin) %>% 
  mutate(tech_rep = paste(replicate, tech, sep = "_"))

p1 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_log2_FC)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("PCI-24781 (Abexinostat) MMEJ log2 fold change")
p2 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_zscore)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("PCI-24781 (Abexinostat) MMEJ zscores") +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96)

plot_grid(p1, p2, ncol = 1)


Vorino_rep_check = indel.data %>% 
  filter(drug == "CI994 (Tacedinaline)") %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore, IPR, tech, replicate, chromatin) %>% 
  mutate(tech_rep = paste(replicate, tech, sep = "_"))

p1 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_log2_FC)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("CI994 (Tacedinaline) MMEJ log2 fold change")
p2 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_zscore)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("CI994 (Tacedinaline) MMEJ zscores") +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96)

plot_grid(p1, p2, ncol = 1)

Vorino_rep_check = indel.data %>% 
  filter(drug == "Scriptaid") %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore, IPR, tech, replicate, chromatin) %>% 
  mutate(tech_rep = paste(replicate, tech, sep = "_"))

p1 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_log2_FC)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("Scriptaid MMEJ log2 fold change")
p2 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_zscore)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("Scriptaid MMEJ zscores") +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96)

plot_grid(p1, p2, ncol = 1)

Vorino_rep_check = indel.data %>% 
  filter(drug == "Resminostat") %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore, IPR, tech, replicate, chromatin) %>% 
  mutate(tech_rep = paste(replicate, tech, sep = "_"))

p1 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_log2_FC)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("Resminostat MMEJ log2 fold change")
p2 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_zscore)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("Resminostat MMEJ zscores") +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96)

plot_grid(p1, p2, ncol = 1)

Vorino_rep_check = indel.data %>% 
  filter(drug == "MC-1293") %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore, IPR, tech, replicate, chromatin) %>% 
  mutate(tech_rep = paste(replicate, tech, sep = "_"))

p1 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_log2_FC)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("MC-1293 MMEJ log2 fold change")
p2 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_zscore)) + 
  geom_quasirandom(aes(shape = tech_rep, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("MC-1293 MMEJ zscores") +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96)

plot_grid(p1, p2, ncol = 1)
```

```{r}
indel.data %>% pull(drug) %>% unique() %>% .[grepl("MC-1293", .)]

HDAC_followup = indel.data %>% 
  filter(drug %in% c("PCI-24781 (Abexinostat)", "MC-1293", "Resminostat", "Scriptaid", 
                     "CI994 (Tacedinaline)", "PCI-24781 (Abexinostat)", "Vorinostat (SAHA, MK0683)")) %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore_comb, IPR, chromatin, drug, conc_char) %>% 
  mutate(mean_MMEJ_FC = ave(MMEJ_log2_FC, IPR, conc_char, drug, FUN = mean)) %>%
  distinct(mean_MMEJ_FC, MMEJ_zscore_comb, IPR, chromatin, drug)

p1 = ggplot(HDAC_followup, aes(IPR, mean_MMEJ_FC)) + 
  geom_quasirandom(aes(shape = drug, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("PCI-24781 (Abexinostat) MMEJ log2 fold change")
p2 = ggplot(HDAC_followup, aes(IPR, MMEJ_zscore_comb)) + 
  geom_quasirandom(aes(shape = drug, color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("PCI-24781 (Abexinostat) MMEJ zscores") +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96)

plot_grid(p1, p2, ncol = 1)
```


```{r}
Vorino_rep_check = indel.data %>% 
  filter(target == "HDAC") %>% 
  distinct(MMEJ_log2_FC, MMEJ_zscore, IPR, tech, replicate, chromatin) %>% 
  mutate(tech_rep = paste(replicate, tech, sep = "_"))

p1 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_log2_FC)) + 
  geom_quasirandom(aes(color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("HDAC MMEJ log2 fold change")
p2 = ggplot(Vorino_rep_check, aes(IPR, MMEJ_zscore)) + 
  geom_quasirandom(aes(color = chromatin)) + 
  geom_boxplot(aes(color = chromatin), alpha = 0.2) + ggtitle("HDAC MMEJ zscores")

plot_grid(p1, p2, ncol = 1)
```


```{r}
GSK_BIX_molcell_tib <- readRDS("/DATA/projects/DSBrepair/data/R/rs20210311/RSTP2_IndelRatios_Chromatin_2kb.RDS") %>% 
  filter(exp_pool %in% c("mean_RSTP2_2000_LBR2_DMSOGSK_64", 
                         "mean_RSTP2_2000_LBR2_GSK126_64")) 

GSK_fc = GSK_BIX_molcell_tib %>% 
  distinct(barcode, drug, NHEJ, MMEJ) %>%
  mutate(NHEJ_MMEJ = NHEJ/MMEJ) %>%
  dplyr::select(-NHEJ, -MMEJ) %>%
  pivot_wider(values_from = NHEJ_MMEJ, names_from = drug) %>%
  filter(complete.cases(.)) %>%
  mutate(logFC_GSK= log2(GSK126 / DMSO_GSK))  %>%
  filter(!is.na(logFC_GSK) & !is.infinite(logFC_GSK)) %>%
  dplyr::select(barcode, logFC_GSK)

chrom_data_pool = GSK_BIX_molcell_tib %>% dplyr::select(barcode, all_of(chromatin.features)) %>% distinct()

GSK_fc %<>% left_join(chrom_data_pool)

slope.compound.features <- tibble(feature = NA,  term = NA,
                                  ratio = NA, p.value = NA)

for (j in chromatin.features) {
  model_tib <- GSK_fc
  
  # Pathway balance
  model.ratio.log2 <- lm(formula = logFC_GSK ~ unlist(model_tib[j]), 
                         data = model_tib) %>% 
    tidy()
  
  slope.compound.features <- slope.compound.features %>% 
    add_row(feature = j, 
            term = model.ratio.log2 %>%
              pull(term),
            ratio = model.ratio.log2 %>% 
              pull(estimate), 
            p.value = model.ratio.log2 %>% 
              pull(p.value))
}


# Remove the 1st NA column
slope.compound.features %<>% 
  filter(!is.na(feature)) %>%
  mutate(term = ifelse(term == "(Intercept)", "intercept", "slope"))

# Calculate postition where the linear model crosses y = 0, we do this by -intercept / slope
synergy_tib = slope.compound.features %>% 
  distinct(term, feature, ratio, p.value) %>% 
  pivot_wider(., names_from = term, values_from = c(ratio, p.value)) %>%
  mutate(x_intercept_ratio = -ratio_intercept / ratio_slope,
         p.adj.slope = p.adjust(p.value_slope, method = "fdr"))

synergy_tib_filtered = synergy_tib %>% 
  mutate(x_intercept_ratio = ifelse(p.adj.slope < 0.05, x_intercept_ratio, NA),
         ratio_slope_plot = ifelse(x_intercept_ratio < 1, ratio_slope, NA)) %>%
  filter(!is.na(ratio_slope_plot))

synergy_tib_filtered


# pdf(paste0(figure_out, "rs20220210_GSK_schepetal.pdf"), width=8, height=4)
p1 = GSK_fc %>% ggplot(., aes(H3K27me3, logFC_GSK)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16)
p2 = GSK_fc  %>% ggplot(., aes(EZH2, logFC_GSK)) + geom_point() + stat_cor() + geom_smooth(method = "lm") + theme_bw(base_size = 16)
plot_grid(p1, p2)
# dev.off()
```


# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

