---
title: "DSB epigenetic drug screen - Figure 3"
author:
  - name: "Max Trauernicht & Ruben Schep"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output:
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
  pdf_document:
    toc: yes
    toc_depth: '4'
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
$(document).ready(function() {
  $items = $('div#TOC li');
  $items.each(function(idx) {
    num_ul = $(this).parentsUntil('#TOC').length;
    $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
  });
});
</script>

```{css, echo = FALSE}
div.sourceCode {
overflow-x: hidden;
}
```
### Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(knitr)
library(tidyverse)
library(outliers)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
# library(ggrepel)
library(GGally)
# library(DESeq2)
library(gghighlight)
# library(platetools)
library(ggpubr)
# library(Laurae)
library(pheatmap)
library(RColorBrewer)
# # library(ggrastr)
library(broom)
# library(ggsignif)
library(cowplot)
library(rstatix)
library(colorspace)
library(Hmisc)
library(MASS)

## Select outdir
out.dir = paste0("./figures/", Date, "/")
dir.create(out.dir)
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

### Functions

```{r functions, message=FALSE, warnings=FALSE}
CorrelationDFMean <- function(data, feature_name, targets, p.value=0.001){
  hist.corr.list <- data %>% dplyr::select(all_of(feature_name), targets) %>% as.matrix() %>% rcorr(type = "spearman")
  
  histone.corr.df <- data.frame(rowname = names(hist.corr.list$r[1, -1]),
                                corr = hist.corr.list$r[1, -1],
                                pvalue = hist.corr.list$P[1, -1])
  
  return(histone.corr.df=histone.corr.df)
}


CorrelationPlotMean <- function(data, feature_name, targets, title = "with treatment", p.value=0.001, alfa = c(0.5, 1)){
  cor.df = data.frame()
  for (i in feature_name){
    corr.list <- data %>% dplyr::select(i, as.character(targets)) %>% as.matrix() %>% rcorr(type = "spearman")
    
    df <- data.frame(rowname = names(corr.list$r[1, -1]),
                     corr = corr.list$r[1, -1],
                     pvalue = corr.list$P[1, -1],
                     feature = i)
    cor.df <- rbind(cor.df, df)
  }
  p <- cor.df %>%
    transform(rowname = factor(rowname, levels = levels(targets)),
              corr = corr,
              pvalue = pvalue) %>%
    ggplot(aes(x = rowname, y = corr, fill = pvalue < p.value)) +
    geom_bar(stat = "identity", position = "dodge") +
    ylab(paste("Spearman correlation ",title)) +
    xlab("Chromatin Feature") +
    # ggtitle(paste("Repair pathways correlation with chromatin features in", condition_name, "exps")) +
    theme_bw(base_size = 16) +
    scale_y_continuous(breaks=seq(-1,1, .1)) +
    coord_flip()
  return(list(dataframe=cor.df, plot = p))
}

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2',
                                    'H3K27me3', 'EZH2', 'CTCF', 'SMC3',
                                    'HDAC3', 'HDAC2', 'HDAC1', 'H3K4me1',
                                    'H3K4me2', 'H3K4me3', 'H3K27ac',
                                    'H4K5acK8ac', 'H2AFZ', 'DNAse', 'Dam', 
                                    'm5C', 'H3K79me2', 'TTseq', 'H3K36me3', 
                                    'POL2AS2', 'POL2'),
                        group=factor(c(rep('repressive', 5), rep('insulator', 2),
                                       rep('HDAC', 3), rep('euchromatin', 6),
                                       rep('accessibility', 2), 'methylation',
                                       rep('transcribing', 5)),
                                     levels=c('transcribing', 'accessibility',
                                              'methylation', 'euchromatin',
                                              'HDAC', 'insulator', 'repressive')))

cor_test <- function(y, X, left, right){
  cor_list = lapply(X, function(x){
    cor.test(x,y)
  })
  p_vec = unlist(lapply(cor_list, function(x){x$p.value}))
  cor_vec = unlist(lapply(cor_list, function(x){x$estimate}))
  p_adj = p.adjust(p_vec)
  color = ifelse(p_adj < 0.05, ifelse(cor_vec > 0, left, right),
                 'None')
  importance = abs(cor_vec)
  target=factor(colnames(X), levels=rownames(clustering))
  feature_group = clustering[colnames(X), ]
  return(list(correlation=cor_vec,importance = importance, p=p_vec,
              p_adj=p_adj, target=target, color=color,
              feature_group=feature_group))
}

get_combinations <- function(domain_vec){
  n_domain = length(domain_vec)
  get_ivec <-function (last_x, n_domain){
    if (last_x < n_domain){
      i_vec = c(0, c((last_x+1):n_domain))
    } else{
      i_vec = c(0)
    }
  }
  
  combine <- function(x_new, x_vec, n_domain){
    new_vec = c(x_vec, x_new)
    if (length(new_vec)==n_domain){
      return(new_vec)
    } else {
      i_vec = get_ivec(max(new_vec), n_domain)
      x_list = lapply(i_vec, combine, x_vec=new_vec,
                      n_domain=n_domain)
      return(do.call(rbind, x_list))
    }
  }
  
  comb_list = lapply(c(0:n_domain), combine, x_vec=c(), n_domain=n_domain)
  
  comb = do.call(rbind, comb_list)
  comb_df = unique(t(apply(comb, 1, function(x){x[order(x)]})))
  
  name_list = apply(comb_df, 1, function(x){
    name = paste(domain_vec[x], collapse='-')
    if (name == ''){
      name = 'iDomain'
    }
    return(name)
  })
  
  opt_list = lapply(domain_vec, function(d){
    apply(comb_df, 1, function(x){
      opt = domain_vec[x]
      return(d%in%opt)
    })
  })
  
  opt_df = do.call(cbind, opt_list)
  colnames(opt_df) = domain_vec
  rownames(opt_df) = name_list
  
  return(opt_df)
}

plot_comb_grid <- function(indel.dt, opt_df, domains, y_name, min_count=0,
                           max_count=Inf, group_by=NULL, color=NULL, y_plabel=c(0.15, 0.25)){
  
  opt_melt = reshape2::melt(opt_df)
  colnames(opt_melt) = c('group', 'domain', 'is_in')
  
  opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
  opt_melt$domain = factor(opt_melt$domain, levels=domains)
  
  
  group_list = lapply(rownames(opt_df), function(opt_name){
    opt = opt_df[opt_name, ]
    in_vec = domains[opt]
    out_vec = domains[!opt]
    
    in_sum = rowSums(indel.dt[,..in_vec])
    out_sum = rowSums(indel.dt[,..out_vec])
    if (length(in_vec)==0){
      in_group = which(out_sum == 0)
    } else if (length(out_vec) == 0){
      in_group = which(in_sum==length(in_vec))
    } else  {
      in_group = which(in_sum == length(in_vec) & out_sum==0)
    }
    return(in_group)
  })
  
  names(group_list) = rownames(opt_df)
  for (opt_name in names(group_list)){
    i_vec = group_list[[opt_name]]
    indel.dt[i_vec, group := opt_name]
  }
  if (is.null(group_by)){
    group_count_df = data.frame(indel.dt[,table(group)])
  } else {
    group_count_df = data.frame(indel.dt[,table(group)/length(unique(get(group_by)))])
  }
  group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
  # print(group_count_df)
  group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
  group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
  group_levels =  group_count_sel$group
  opt_sel = opt_melt[opt_melt$group %in% group_levels, ]
  
  opt_sel$group = factor(opt_sel$group, levels=group_levels)
  # print(opt_sel)
  
  indel_selection = indel.dt[group%in%group_levels, ]
  indel_selection[,group:=factor(group, levels=group_levels)]
  
  group_count_sel$group = factor(group_count_sel$group, levels=group_levels)
  
  count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
    geom_label() +
    theme_void()
  
  
  
  dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    geom_point(size=2) +
    geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    theme_bw() +
    scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    guides(color="none")  +
    geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          panel.grid.major = element_blank())
  
  symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                symbols = c("****", "***", "**", "*", ""))
  
  if (is.null(group_by)){
    level_vec = indel_selection[, levels(group)]
    wilcox_list = lapply(level_vec[level_vec!='iDomain'],
                         function(g){
                           x = indel_selection[group==g, y_name, with=F]
                           y = indel_selection[group=='iDomain', y_name,
                                               with=F]
                           w = wilcox.test(unlist(x),unlist(y))
                           return(data.table('group1'='iDomain', 'group2'=g,
                                             'p'=w$p.value))
                         })
    wilcox_dt = do.call(rbind, wilcox_list)
    wilcox_dt[,p_adj := p.adjust(p)]
    wilcox_dt[,p_label := sprintf('%.3g', p_adj)]
    wilcox_dt$y.position = seq(y_plabel[1], y_plabel[2],
                               length.out=nrow(wilcox_dt))
    
    comparisons = lapply(level_vec[level_vec!='iDomain'],
                         function(x){return(c('iDomain', x))})
    beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
      geom_quasirandom(width=0.3) +
      stat_summary(fun=median, fun.min = median,
                   fun.max = median,
                   geom="crossbar", width = 0.5,
                   color='red') +
      stat_pvalue_manual(wilcox_dt, label = "p_label")
  } else {
    wilcox_dt = indel_selection[,test_wilcox(.SD, group_by, y_name),
                                by='group']
    wilcox_dt[,p_adj:=p.adjust(p_value)]
    wilcox_dt[,p_signif:=symnum(p_adj, cutpoints=symnum$cutpoints,
                                symbols=symnum$symbols)]
    beeswarm = ggplot(indel_selection,
                      aes_string(x='group', y=y_name, color=group_by)) +
      geom_quasirandom(dodge.width=1) +
      scale_color_manual(values=color) +
      stat_summary(aes_string(group=group_by), fun.=median,
                   fun.min = median, fun.max = median,
                   geom="crossbar", width = 0.5,
                   color='red',
                   position= position_dodge(width =1)) +
      geom_text(data=wilcox_dt, aes(x=group, y=0.8, label=p_signif),
                inherit.aes=F)
  }
  
  beeswarm = beeswarm +
    theme_bw() +
    geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text.x=element_blank(),
          panel.grid.major.x = element_blank(),
          axis.ticks.x=element_blank())
  
  count = data.frame(domain=gsub('_domain', '', domains),
                     count=colSums(indel.dt[,..domains], na.rm=T))
  
  count$domain = factor(count$domain, levels=domain_levels)
  
  
  histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    geom_label() +
    theme_bw() +
    theme(plot.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          axis.title=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks=element_blank())
  
  empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
    geom_point() +
    theme_void()
  
  count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                             align='v', axis='lr')
  return(plot_grid(empty, beeswarm, empty, count_plot,
                   histogram, dots, nrow=3,
                   rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                   axis='tlbr'))
}

quantiles_95 <- function(x) {
  r <- quantile(x, probs=c(0.05, 0.25, 0.5, 0.75, 0.95))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

plot_comb_grid_bothplot <- function(bix.dt, opt_df, domains, y_name, min_count=0,
                                   max_count=Inf, group_by=NULL, color=NULL){
  
  opt_melt = reshape2::melt(opt_df)
  colnames(opt_melt) = c('group', 'domain', 'is_in')
  
  opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
  opt_melt$domain = factor(opt_melt$domain, levels=domains)
  
  
  group_list = lapply(rownames(opt_df), function(opt_name){
    opt = opt_df[opt_name, ]
    in_vec = domains[opt]
    out_vec = domains[!opt]
    
    in_sum = rowSums(bix.dt[,..in_vec])
    out_sum = rowSums(bix.dt[,..out_vec])
    if (length(in_vec)==0){
      in_group = which(out_sum == 0)
    } else if (length(out_vec) == 0){
      in_group = which(in_sum==length(in_vec))
    } else  {
      in_group = which(in_sum == length(in_vec) & out_sum==0)
    }
    return(in_group)
  })
  
  names(group_list) = rownames(opt_df)
  for (opt_name in names(group_list)){
    i_vec = group_list[[opt_name]]
    bix.dt[i_vec, group := opt_name]
  }
  if (is.null(group_by)){
    group_count_df = data.frame(bix.dt[,table(group)])
  } else {
    group_count_df = data.frame(bix.dt[,table(group)/length(unique(get(group_by)))])
  }
  group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
  # print(group_count_df)
  group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
  group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
  group_levels =  group_count_sel$group
  opt_sel = opt_melt[opt_melt$group %in% group_levels, ]
  
  opt_sel$group = factor(opt_sel$group, levels=group_levels)
  # print(opt_sel)
  
  indel_selection = bix.dt[group%in%group_levels, ]
  indel_selection[,group:=factor(group, levels=group_levels)]

  group_count_sel$group = factor(group_count_sel$group, levels=group_levels)
  
  count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
    geom_label() +
    theme_void()
  
  
  
  dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    geom_point(size=2) +
    geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    theme_bw() +
    scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    guides(color="none")  +
    geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          panel.grid.major = element_blank())
  
  symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                symbols = c("****", "***", "**", "*", " "))
  
  if (is.null(group_by)){
    boxpl = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
      geom_quasirandom(width=0.3) +
      stat_summary(fun=median, fun.min = median,
                   fun.max = median,
                   geom="crossbar", width = 0.5,
                   color='red') +
      stat_compare_means(aes(label = ..p.signif..),
                         symnum.args = symnum,
                         method = 'wilcox.test',
                         ref.group='iDomain')
  } else {
    
    boxpl = ggplot(indel_selection,
                   aes_string(x='group', y=y_name, color=group_by)) +
      guides(fill="none") +
      # scale_fill_manual(values=color) +
      scale_color_manual(values=color) +
      # geom_violin(position=position_dodge(width=1)) +
      geom_quasirandom(dodge.width = 1) +
      # stat_summary(fun.data = quantiles_95,
      #              geom="boxplot",
      #              position=position_dodge(width=1),
      #              width = 0.2) +
      stat_summary(fun=median, fun.min = median,
                   fun.max = median,
                   geom="crossbar", width = 0.6,
                   position=position_dodge(width=1))
  }
  boxpl = boxpl +
    theme_bw() +
    geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text.x=element_blank(),
          panel.grid.major.x = element_blank(),
          axis.ticks.x=element_blank()) +
    ylim(0, 1)
  
  
  count = data.frame(domain=gsub('_domain', '', domains),
                     count=colSums(bix.dt[,..domains], na.rm=T))
  
  count$domain = factor(count$domain, levels=domain_levels)
  
  
  histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    geom_label() +
    theme_bw() +
    theme(plot.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          axis.title=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks=element_blank())
  
  empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
    geom_point() +
    theme_void()
  
  count_beeswarm = plot_grid(boxpl, count_plot, ncol=1, rel_heights=c(20,1),
                             align='v', axis='lr')
  return(plot_grid(empty, boxpl, empty, count_plot,
                   histogram, dots, nrow=3,
                   rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                   axis='tlbr'))
}

plot_comb_grid_beeswarm <- function(dt, opt_df, domains, y_name, min_count=0,
                                    max_count=Inf, group_by=NULL, color=NULL, y_plabel=c(-1.5, 1.5)){
  
  opt_melt = reshape2::melt(opt_df)
  colnames(opt_melt) = c('group', 'domain', 'is_in')
  
  opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
  opt_melt$domain = factor(opt_melt$domain, levels=domains)
  
  
  group_list = lapply(rownames(opt_df), function(opt_name){
    opt = opt_df[opt_name, ]
    in_vec = domains[opt]
    out_vec = domains[!opt]
    
    in_sum = rowSums(dt[,..in_vec])
    out_sum = rowSums(dt[,..out_vec])
    if (length(in_vec)==0){
      in_group = which(out_sum == 0)
    } else if (length(out_vec) == 0){
      in_group = which(in_sum==length(in_vec))
    } else  {
      in_group = which(in_sum == length(in_vec) & out_sum==0)
    }
    return(in_group)
  })
  
  names(group_list) = rownames(opt_df)
  for (opt_name in names(group_list)){
    i_vec = group_list[[opt_name]]
    dt[i_vec, group := opt_name]
  }
  if (is.null(group_by)){
    group_count_df = data.frame(dt[,table(group)])
  } else {
    group_count_df = data.frame(dt[,table(group)/length(unique(get(group_by)))])
  }
  group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
  # print(group_count_df)
  group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
  group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
  group_levels =  group_count_sel$group
  opt_sel = opt_melt[opt_melt$group %in% group_levels, ]
  
  opt_sel$group = factor(opt_sel$group, levels=group_levels)
  # print(opt_sel)
  
  indel_selection = dt[group%in%group_levels, ]
  indel_selection[,group:=factor(group, levels=group_levels)]
  
  # wilcox_list = lapply(group_levels[group_levels!='iDomain'],
  #                      function(g){
  #                          x = indel_selection[group==g, y_name, with=F]
  #                          y = indel_selection[group=='iDomain', y_name,
  #                                              with=F]
  #                          wilcox.test(unlist(x),unlist(y))
  #                      })
  
  group_count_sel$group = factor(group_count_sel$group, levels=group_levels)
  
  count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
    geom_label() +
    theme_void()
  
  
  
  dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    geom_point(size=2) +
    geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    theme_bw() +
    scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    guides(color="none")  +
    geom_vline(xintercept=seq(1.5, length(unique(dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          panel.grid.major = element_blank())
  
  symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                symbols = c("****", "***", "**", "*", " "))
  
  
  if (is.null(group_by)){
    beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
      geom_quasirandom(width=0.3) +
      stat_summary(fun=median, fun.min = median,
                   fun.max = median,
                   geom="crossbar", width = 0.5,
                   color='red') +
      stat_compare_means(aes(label = ..p.signif..),
                         symnum.args = symnum,
                         method = 'wilcox.test',
                         ref.group='iDomain')
    
  } else {
    wilcox_dt = indel_selection[,test_wilcox(.SD, group_by, y_name),
                                by='group']
    wilcox_dt[,p_adj:=p.adjust(p_value)]
    wilcox_dt[,p_signif:=symnum(p_adj, cutpoints=symnum$cutpoints,
                                symbols=symnum$symbols)]
    
    
    beeswarm = ggplot(indel_selection,
                      aes_string(x='group', y=y_name, color=group_by)) +
      geom_quasirandom(dodge.width=1) +
      scale_color_manual(values=color) +
      stat_summary(aes_string(group=group_by), fun=median,
                   fun.min = median, fun.max = median,
                   geom="crossbar", width = 0.5,
                   color='red',
                   position= position_dodge(width =1)) +
      geom_text(data=wilcox_dt, aes(x=group, y=0.8, label=p_signif),
                inherit.aes=F)
    print(wilcox_dt)
  }
  beeswarm = beeswarm +
    theme_bw() +
    geom_vline(xintercept=seq(1.5, length(unique(dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text.x=element_blank(),
          panel.grid.major.x = element_blank(),
          axis.ticks.x=element_blank()) +
    ylim(y_plabel)
  
  count = data.frame(domain=gsub('_domain', '', domains),
                     count=colSums(dt[,..domains], na.rm=T))
  
  count$domain = factor(count$domain, levels=domain_levels)
  
  histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    geom_label() +
    theme_bw() +
    theme(plot.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          axis.title=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks=element_blank())
  
  empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
    geom_point() +
    theme_void()
  
  count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                             align='v', axis='lr')
  return(plot_grid(empty, beeswarm, empty, count_plot,
                   histogram, dots, nrow=3,
                   rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                   axis='tlbr'))
}

p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
    ## Methods 'Hommel', 'BH', 'BY' and speed improvements
    ## contributed by Gordon Smyth
    method <- match.arg(method)
    if(method == "fdr") method <- "BH"	# back compatibility
    nm <- names(p)
    p <- as.numeric(p)
    p0 <- setNames(p, nm)
    if(all(nna <- !is.na(p))) nna <- TRUE
    p <- p[nna]
    lp <- length(p)
    # stopifnot(n >= lp)
    if (n <= 1) return(p0)
    if (n == 2 && method == "hommel") method <- "hochberg"

    p0[nna] <-
	switch(method,
	       bonferroni = pmin(1, n * p),
	       holm = {
		   i <- seq_len(lp)
		   o <- order(p)
		   ro <- order(o)
		   pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
	       },
	       hommel = { ## needs n-1 >= 2 in for() below
		   if(n > lp) p <- c(p, rep.int(1, n-lp))
		   i <- seq_len(n)
		   o <- order(p)
		   p <- p[o]
		   ro <- order(o)
		   q <- pa <- rep.int( min(n*p/i), n)
		   for (j in (n-1):2) {
		       ij <- seq_len(n-j+1)
		       i2 <- (n-j+2):n
		       q1 <- min(j*p[i2]/(2:j))
		       q[ij] <- pmin(j*p[ij], q1)
		       q[i2] <- q[n-j+1]
		       pa <- pmax(pa,q)
		   }
		   pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
	       },
	       hochberg = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
	       },
	       BH = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( n / i * p[o] ))[ro]
	       },
	       BY = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   q <- sum(1L/(1L:n))
		   pmin(1, cummin(q * n / i * p[o]))[ro]
	       },
	       none = p)
    p0
}

SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}

theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "white", color = "white")
            )
}
theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "white", color = "white")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "white", color = "white")
            )
}

theme_classic_lines_90_noborder <- function() {
  theme_pubr(border = F, legend = "top", x.text.angle = 90) +
            theme(strip.background = element_rect(fill = "white", color = "white")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
# colors_drugs <- c("DMSO (neg control)" = "#65BFA4", "DNA-PKi (NHEJ control)" = "#8CA0C4", "drugs" = "#808184", "Mirin (MMEJ control)" = "#5D79AC")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 
## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

colors_targets <-  c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
                     HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
                     HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
                     PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43")

conc_control_colors = c("CTRL" = colors_drugs[2], "100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

```


### Data import

```{r data import, echo = FALSE, warning = FALSE}
# # Import data from preprocessing script
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
file = list.files(in.dir, pattern = "episcreen_ratios", full.names = T) %>% tail(n = 1)
indel.data <- readRDS(file) # %>% filter(pathway_bal_filter == TRUE & viab_reproducibility == TRUE)

# # Import data validation data
file = list.files(in.dir, pattern = "validation_ratios", full.names = T) %>% tail(n = 1)
indel.validation.data <- readRDS(file) # %>% filter(pathway_bal_filter == TRUE & viab_reproducibility == TRUE)

# List chromatin name columns from this table
chromatin_cols = colnames(indel.validation.data)[42:75]
# Import data from DSB TRIP paper
RSTP2_IndelRatios_Chromatin_2kb <- readRDS("~/DSBrepair/data/R/rs20210311/RSTP2_IndelRatios_Chromatin_2kb.RDS") %>% 
  filter(replicate == "mean" & drug %in% c("DMSO_GSK", "GSK126") & plasmid == "LBR2") %>%
  dplyr::select(barcode, drug, MMEJ, NHEJ, efficiency, all_of(chromatin_cols)) %>%
  distinct() %>%
  mutate(MMEJ_NHEJ = MMEJ/NHEJ) %>%
  dplyr::select(-NHEJ, -MMEJ)
```

---

```{r heatmap annotation}
# Order IPRs by their starting freqCut value
dmso_freq <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(IPR, freqCut) %>%
  mutate(freqCut = ave(freqCut, IPR, FUN = function(x) mean(x, na.rm = TRUE))*100) %>%
  dplyr::select('ctrl_TIF' = freqCut, IPR) %>%
  distinct() %>%
  arrange(desc(ctrl_TIF))

iprs <- dmso_freq$IPR
iprs <- as.character(iprs)

# Set annotation for heatmap
# viability
viab_drug = indel.data %>% filter(drug != "PAO") %>% 
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  distinct(viab_mean, drug_conc, replicate) %>% 
  group_by(drug_conc) %>%
  summarise(viab = mean(viab_mean))
# Row annotation: annotate drugs with the target group
target_set_drugs <- indel.data %>% 
  filter(!target %in% c("Negative Control", "MRN", "DNA-PK")) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>%
  left_join(viab_drug) %>%
  distinct(drug, drug_conc, conc_char, target, viab) %>% 
  column_to_rownames(var="drug_conc")

# Set annotation for heatmap
# viability
viab_drug = indel.data %>% filter(drug != "PAO") %>% 
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  distinct(viab_mean, drug_conc, replicate) %>% 
  group_by(drug_conc) %>%
  summarise(viab = mean(viab_mean))
# Row annotation: annotate drugs with the target group
target_set_drugs <- indel.data %>% 
  filter(!target %in% c("Negative Control", "MRN", "DNA-PK")) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>%
  left_join(viab_drug) %>%
  distinct(drug, drug_conc, conc_char, target, viab) %>% 
  column_to_rownames(var="drug_conc")

ctrl.vals <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(IPR, freqCut, MMEJratio) %>%
  mutate(ctrl_TIF = ave(freqCut, IPR, FUN = function(x) mean(x, na.rm = TRUE))*100,
         ctrl_balance = ave(MMEJratio, IPR, FUN = function(x) mean(x, na.rm = TRUE))*100) %>%
  distinct(ctrl_TIF, 
           ctrl_balance,
           IPR)

clone5.matrix = indel.data %>% 
  dplyr::select(IPR, ends_with(".zscore")) %>% 
  distinct() %>% 
  rename_with(., ~gsub(".zscore", "", .x)) %>%
  # left_join(ctrl.vals) %>%
  column_to_rownames(var="IPR")

clone5_map_ipr = hclust(dist(clone5.matrix), method = "ward.D") %>% 
  # Rotate the middle node to math better the TIF in CTRL conditions
  dendextend::rotate(c(1:6, 10:7, 11:18))

# Chromatin annotation
chromatin <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(indel.data, barcode, IPR)) %>%
  left_join(dmso_freq)

# chromatin$IPR[is.na(chromatin$IPR)] <- "IPR1"
chromatin <- chromatin %>%
  filter(IPR != "<NA>") %>% 
  dplyr::select(-barcode) %>%
  column_to_rownames(var="IPR")

## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")


viab_colors = sequential_hcl(11, "Peach", rev = TRUE)
names(viab_colors) = seq(0, 1, by = 0.1)

annotation_colors_scr = list(
  chromatin = group_colors,
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43"),
  conc_char = conc_colors,
  clusters_drugs = sequential_hcl(7, "Plasma"),
  # viab_rep1 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  # viab_rep2 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  viab = viab_colors,
  ctrl_TIF = colorRampPalette(c("white", "black"))(10))

# Order the annotation legend
ordered_marks <- c("Euchromatin", 
                   "H3K27me3", 
                   "other-heterochromatin",  
                   "Triple Heterochromatin")

annotation_colors_scr$chromatin <- annotation_colors_scr$chromatin[ordered_marks]


# First calculate means and sd of the DMSO efficiency per replicate and barcode
dmso.eff.score <- indel.data %>%
  filter(drug == "DMSO") %>%
  dplyr::select('DMSO' = freqCut, well, barcode, replicate, tech, plate) %>%
  mutate(DMSO = ave(DMSO, well, replicate, tech, plate, FUN = function(x) mean(x, na.rm=T))) %>%
  distinct(replicate, tech, well, plate, DMSO) %>% 
  pull(DMSO) %>%
  fitdistr(., "normal") %>% tidy()
```

```{r filter heatmap}
# Generate heatmap: select top drugs (highest change at any IPR) per concentration
eff.data <- indel.data %>%
  #filter(target != "Aurora Kinase") %>%
  filter(sample == "Epigenetic Compound", indel_freq_filter, viab_reproducibility, !(freqCut_diff_mean < -0.1 & viab_mean < 0.5)) %>%
  # filter(abs(zscore_TIF_global_comb_rep) >= 2.58) %>%  # we can choose to display all drugs that have significant global hits
  distinct(conc_char, drug, target, IPR, zscore_TIF_comb_rep, freqCut_diff_mean, replicate, tech, filter_zscore_TIF_comb_rep)

eff.data.heatmap <-  eff.data %>%
  group_by(drug, conc_char) %>%
  # ave to add IPR grouping level to calculate mean of the replicates
  mutate(freqCut_diff_mean = ave((freqCut_diff_mean*100), IPR, FUN = function(x) mean(x, na.rm = TRUE)),
  ## Select drugs that have a large effect size in at least one location 
         ## (some drugs might have a small global effect size, but large local)
         zscore_TIF_comb_rep_temp = ifelse(!filter_zscore_TIF_comb_rep, 0, zscore_TIF_comb_rep)) %>%
  ungroup() %>%
  mutate(p_value = 2*pnorm(q = abs(zscore_TIF_comb_rep_temp), lower.tail = F),
         # Adjust p value for false discovery rate
         adj_pvalue = p.adjust(p_value, method = 'BH')) %>%
         # Pick the smallest per compound & concentration
  group_by(drug, conc_char) %>%
         mutate(min_score = min(adj_pvalue)) %>%
  # filter for adjusted p.val < 0.01 so we keep compounds with at least one signif effect.
  filter(min_score < 0.01) %>%
  distinct(drug, target, conc_char, freqCut_diff_mean, IPR, min_score) %>%
  ungroup()

#eff.data.heatmap2$freqCut_norm_mean[eff.data.heatmap2$mean_eff_zscore_comb > 0.05] <- NA
eff.data.heatmap2 <- eff.data.heatmap %>%
  dplyr::select(-min_score) %>%
  spread(IPR, freqCut_diff_mean) %>%
  filter(!is.nan(IPR3))

eff.data.heatmap2 <- eff.data.heatmap2[,c("drug", "target", "conc_char", iprs)]
eff.data.heatmap2$drug <- gsub(" [(].*[)]", "", eff.data.heatmap2$drug) #shorten name of drug
```


## Visualisation
```{r, fig.width = 8, fig.height = 4, eval = FALSE}
# rmarkdown::render("3_Processing_FACS.Rmd")
all.samples = read_csv2("/DATA/usr/m.trauernicht/projects/EpiScreen/FACS/Data/mt20220802_facs_all_drugs.csv")

all.samples$drug[all.samples$drug == "AZ-960"] <- "AZ 960"
all.samples$drug[all.samples$drug == "Suramin"] <- "Suramin·6Na"
all.samples$drug[all.samples$drug == "CX-6258"] <- "CX-6258 HCl"
all.samples$drug[all.samples$drug == "DZNeP"] <- "3-Deazaneplanocin A"

all.samples <- all.samples %>%
  mutate(drug_conc = paste(drug, conc, sep = "_")) %>%
  mutate(drug_conc_exp = paste(drug, conc, exp, sep = "_")) %>%
  mutate(mean_value = ave(value, drug_conc, variable, exp, FUN = mean)) %>%
  mutate(SD = ave(value, drug_conc, variable, exp, FUN = sd)) %>%
  filter(variable == "G2") %>%
  dplyr::select(-variable) %>%
  na.omit() %>%
  mutate(pval = "")

dmso.samples <- all.samples %>%
  filter(drug_conc == "DMSO_10um") %>%
  mutate(mean_value = ave(mean_value, drug_conc_exp, FUN = mean)) %>%
  distinct('mean_value_dmso' = mean_value, exp)

all.samples <- merge(all.samples, dmso.samples, by = "exp", all = T) %>%
  mutate(value = value / mean_value_dmso) %>%
  mutate(mean_value = ave(value, drug_conc, exp, FUN = mean)) %>%
  mutate(SD = ave(value, drug_conc, exp, FUN = sd))

pval_list <- list()
for (i in unique(all.samples$drug_conc_exp)) {
  if (all.samples$exp[all.samples$drug_conc_exp == i] == "five_drugs") {
    all.samples$pval[all.samples$drug_conc_exp == i] <- t.test(all.samples$value[all.samples$drug_conc == "DMSO_10um" & all.samples$exp == "five_drugs"], all.samples$value[all.samples$drug_conc_exp == i])$p.value
  } else {
    all.samples$pval[all.samples$drug_conc_exp == i] <- t.test(all.samples$value[all.samples$drug_conc == "DMSO_10um" & all.samples$exp == "effective_drugs"], all.samples$value[all.samples$drug_conc_exp == i])$p.value
  }
}

target_set_drugs2 <- target_set_drugs %>% distinct(drug, target)

all.samples <- all.samples %>%
  dplyr::select(-value) %>%
  mutate(pval_adj = p.adjust(pval, method = "fdr")) %>%
  left_join(target_set_drugs2) %>%
  distinct()

# vadj.G2 = all.samples %>% filter(variable %in% c("G2", "S", "G1")) %>% group_by(drug_conc) %>% summarise(vadj = sum(mean_value)) %>% mutate(variable = "G2")
# vadj.S = all.samples %>% filter(variable %in% c("S", "G1")) %>% group_by(drug_conc) %>% summarise(vadj = sum(mean_value)) %>% mutate(variable = "S")
# vadj.G1 = all.samples %>% filter(variable %in% c("G1")) %>% group_by(drug_conc) %>% summarise(vadj = sum(mean_value)) %>% mutate(variable = "G1")
# 
# vadj.tib = rbind(vadj.G1, vadj.S, vadj.G2)
# 
# all.samples.plot = all.samples  %>% 
#   as_tibble() %>% 
#   filter(variable != "Apoptotic cells") %>% 
#   left_join(vadj.tib) %>%
#   left_join(target_set_drugs) %>%
#   mutate(variable = factor(variable, levels = c("G2", "S", "G1")),
#          target = ifelse(drug == "CX-6258", "PIM",
#                          ifelse(drug == "DMSO", "Vehicle",
#                                 ifelse(drug == "DZNeP", "HMT",
#                                        ifelse(drug == "Suramin", "Sirtuin",target)))),
#          ID = gsub("GFP", "-sgRNA", ID))
# 
# 
# aurk.inhibitors <- c("Alisertib 100 nM", "Aurora A Inhibitor I 1 µM", "Barasertib 1 µM", "CYC116 1 µM", "SNS-314 Mesylate 10 µM")
# 
# all.samples.plot %>% filter(drug_conc %in% c("DMSO 10 µM", aurk.inhibitors), ID == "LBR2") %>%
#   mutate(drug_conc = factor(drug_conc, levels = c("DMSO 10 µM", aurk.inhibitors))) %>%
#   ggplot(., aes(y = mean_value, x = drug_conc, fill = variable)) +
#   geom_bar(stat="identity") + 
#   geom_errorbar(aes(ymin = vadj-SD, ymax = vadj+SD, color = variable), width = 0.2, size = 1.5)+
#   labs(title = "cell cycle arrest upon treatment with Aurora Kinase inhibitors") +
#   scale_fill_manual(values = c("grey80","grey60","grey40")) +
#   scale_color_manual(values = c("grey80","grey60","grey40")) +
#   theme_pubr(border = T, x.text.angle = 45)

all.samples %>%
  ggplot(., aes(y = mean_value, x = conc, color = target)) +
  geom_line()+
  geom_pointrange(aes(ymin=mean_value, ymax=mean_value+SD)
                  #, position = position_dodge(width = 0.1)
                  ) +
  geom_hline(yintercept = 0.08419397, lty = "dashed") + 
  ylim(0,0.7) +
  gghighlight(pval < 0.05, label_key = drug_conc) +
  scale_color_manual(values = colors_targets) 

all.samples %>%
  distinct(drug_conc_exp, target, mean_value, pval, drug, drug_conc) %>%
  mutate(pval = -log10(as.numeric(pval))) %>%
  filter(drug != "DMSO", drug != "PAO") %>%
  ggplot(., aes(y = log2(mean_value), x = target, 
                fill = target, size = pval, 
                color = ifelse(pval > -log10(0.05), "black", "white"))) +
  geom_point(shape = 21) +
  #gghighlight(mean_value > 2, label_key = drug_conc) +
  scale_fill_manual(values = colors_targets) +
  scale_color_manual(values = c("black", "white")) +
  theme_classic_lines_45()

other_drugs <- unique(all.samples.plot$drug_conc[all.samples.plot$drug_conc != "DMSO 10 µM" & all.samples.plot$target != "Aurora Kinase"])
all.samples.plot %>% filter(!target %in% c("Aurora Kinase"), ID == "LBR2") %>%
mutate(drug_conc = factor(drug_conc, levels = c("DMSO 10 µM", other_drugs))) %>%
ggplot(., aes(y = mean_value, x = drug_conc, fill = variable)) +
  geom_bar(stat="identity") + 
   geom_errorbar(aes(ymin = vadj-SD, ymax = vadj+SD, color = variable), width = 0.2, size = 1.5)+
  labs(title = "cell cycle arrest upon treatment with Aurora Kinase inhibitors", 
       x = "GFP/LBR2 transfection", y = "G1, S, and G2 fractions") + 
  scale_fill_manual(values = c("grey80","grey60","grey40")) +
  scale_color_manual(values = c("grey80","grey60","grey40")) +
  theme_pubr(border = T, x.text.angle = 45)
```

```{r, fig.width = 4, fig.height = 4, eval = FALSE}
# Barplot of each drug next to each other
all.samples.plot %>% filter(target %in% c("Vehicle")) %>%
ggplot(., aes(y = mean_value, x = ID, fill = variable)) +
  geom_bar(stat="identity") + 
   geom_errorbar(aes(ymin = vadj-SD, ymax = vadj+SD, color = variable), width = 0.2)+
  labs(title = "Cll cycle distribution in control setting", 
       x = "GFP/LBR2 transfection", y = "G1, S, and G2 fractions") + 
  scale_fill_manual(values = c("#DED89C","#BF7538","#749075"))
```


## HDACi only heatmap
```{r}
HDAC_tib = read.table("rs20220708_HDAC_targets.txt", sep = "\t", fill = NA, header = T) %>%
  dplyr::select(-target_main, -target_second) %>%
  as_tibble()

eff.data.heatmap.cluster1 = eff.data.heatmap2 %>% 
  filter(target == "HDAC" & drug != "Nullscript") %>%
  mutate(drug_conc = paste(drug, conc_char, sep = " "))

# Add cluster annotation
cluster_HDAC <- hclust(dist(eff.data.heatmap.cluster1), method = "ward.D")
cluster_HDAC <- cutree(tree = cluster_HDAC, k = 5)
cluster_HDAC <- data.frame(cluster_HDAC)

eff.data.heatmap.cluster1 <- cbind(eff.data.heatmap.cluster1, cluster_HDAC) %>% 
  as_tibble()

breaks <- quantile(unlist(eff.data.heatmap.cluster1 %>% 
                            dplyr::select(contains('IPR'))), c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 100

myBreaks <- c(seq(-breaks[2], 0, length.out=ceiling(palette_length/2) + 1), 
              seq(breaks[2]/palette_length, breaks[2], length.out=floor(palette_length/2)))


cluster_HDAC <- eff.data.heatmap.cluster1 %>%
  distinct(drug_conc, cluster_HDAC)

target.cluster1 <- target_set_drugs %>%
  rownames_to_column("drug_conc") %>%
  left_join(HDAC_tib) %>%
  left_join(cluster_HDAC) %>%
  filter(!is.na(cluster_HDAC)) %>%
  replace(is.na(.), 0)

rownames(target.cluster1) <- target.cluster1$drug_conc
target.cluster1 <- target.cluster1 %>%
  dplyr::select(cluster_HDAC, conc_char, pan_HDAC, HDAC1,HDAC2, HDAC3,
                HDAC4, HDAC6, HDAC8, HDAC10, viab)

target.cluster1

HDAC_matrix = eff.data.heatmap.cluster1 %>%
  arrange(target.cluster1) %>%
  column_to_rownames('drug_conc') %>%
  dplyr::select(contains('IPR')) %>%
  t()

# pdf(paste0(figure_out, "rs20220623_indelfrequency_heatmap_HDAC.pdf"), width=16, height=8)
plt3 = pheatmap(HDAC_matrix[clone5_map_ipr$labels,],
                border_color = F,
                cellwidth = 8,
                cellheight = 8,
                breaks = myBreaks,
                clustering_method = "ward.D",
                cluster_rows = clone5_map_ipr,
                cluster_cols = T,
                angle_col = 90,
                na_col = "white",
                main = 'efficiency change, all concentrations',
                annotation_col = target.cluster1,
                annotation_row = chromatin,
                color = diverge_hcl(palette_length, "Blue-Red"),
                annotation_colors = annotation_colors_scr,
                annotation_legend = T,
                cutree_cols = 5)
```


## 2F Plot HDACi effects on different chromatin types
```{r 2F HDACi in different clusters}
chromatin_ipr <- chromatin %>%
  rownames_to_column("IPR")

eff.data.heatmap3 <- eff.data.heatmap.cluster1 %>%
  pivot_longer(contains("IPR"), names_to = "IPR", values_to = "freqCut_dif")

chromatin_effect <- merge(eff.data.heatmap3, chromatin_ipr) %>%
  #filter(target == "HDAC") %>%
  mutate(freqCut_dif = ave(freqCut_dif, drug, chromatin, conc_char, FUN = function(x) mean(x, na.rm = TRUE))) %>%
  distinct(freqCut_dif, drug, chromatin, conc_char, target, cluster_HDAC)

sigFunc = function(x){
  if(x < 0.001){"***"} 
  else if(x < 0.01){"**"}
  else if(x < 0.05){"*"}
  else{NA}}

chromatin_effect %>%
  filter(target == "HDAC")  %>%
  ggplot(.,
         aes(x = chromatin, y = freqCut_dif, fill = chromatin)) +
  geom_boxplot(outlier.shape = NA) +
  geom_quasirandom() +
  geom_signif(comparisons = combn(sort(unique(chromatin_effect$chromatin)), 2, simplify = F), 
              map_signif_level=sigFunc, 
              step_increase = 0.1) +
  theme_classic_lines_45() +
  scale_fill_manual(values = group_colors) +
  theme(legend.position = "none") +
  facet_grid(. ~ cluster_HDAC) +
  ylab("Total indel frequency change") +
  xlab("Chromatin cluster")
```

# Synergies

## Load data
```{r}
file = list.files(in.dir, pattern = "CCD_table", full.names = T) %>% tail(n = 1)
CCD_tib <- readRDS(file)
file = list.files(in.dir, pattern = "IPR_annotation_heatmap", full.names = T) %>% tail(n = 1)
chromatin_TIF <- readRDS(file)
file = list.files(in.dir, pattern = "compound_annotation_heatmap", full.names = T) %>% tail(n = 1)
target <- readRDS(file)
file = list.files(in.dir, pattern = "chrom_cluster_heatmap", full.names = T) %>% tail(n = 1)
TRIP_chrom_clust <- readRDS(file)
```

```{r heatmap annotation CCD}
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
conc_control_colors = c("control" = colors_drugs[2], "100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")

group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

viab_colors = sequential_hcl(11, "Peach", rev = TRUE)
names(viab_colors) = as.numeric(seq(0, 1, by = 0.1))

# Annotation colors heatmaps
annotation_colors_scr = list(
  chromatin = group_colors,
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43", 
             `Negative Control` = "#65BFA4", `MRN` = "#5D79AC"),
  conc_char = conc_colors,
  conc_char_ctrl = conc_control_colors,
  clusters_drugs = sequential_hcl(7, "Plasma"),
  # viab_rep1 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  # viab_rep2 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  viab = viab_colors,
  TIF = colorRampPalette(c("white", "black"))(20),
  MMEJratio = c('#e1251b', '#d32c2d', '#c5313c', '#b6354a', '#a73857', '#963a65', '#843d72', '#6e3e7f', '#53408d', '#26419a'),
  group = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
            transcribing="#FBB040", HDAC="#aac9e0", accessibility='#cebc85',
            methylation='#7dc98f'))


ordered_marks2 <- c("Negative Control", "MRN", "DNA-PK","Epigenetic Reader Domain", "HDAC",  "Sirtuin", "HAT", 
                    "Histone Demethylase", "HMT", "DNMT", 
                    "HIF", "JAK", "PIM",
                    "Aurora Kinase", "PARP")

#Epistatic interaction colors
slope.colors <- tibble(color = c("#8c510a", "#f5f5f5","#01665e"),label = c("negative","none","positive"), feature = "epistasis")

annotation_colors_scr$target <- annotation_colors_scr$target[ordered_marks2]

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2', 'H3K9me3', 'H3K27me3', 'EZH2', 
                                    'CTCF', 'SMC3',
                                    'HDAC3', 
                                    'H3K4me1', 'H3K4me2', 'H3K4me3', 'H3K27ac', 'H4K5acK8ac', 'H2AFZ', 
                                    'DNAse', 'Dam', 
                                    'm5C',
                                    'H3K79me2', 'TTseq',  'H3K36me3', 'POL2AS2','POL2'),
                        group=factor(c(rep('repressive', 6), 
                                       rep('insulator', 2),
                                       rep('HDAC', 1), 
                                       rep('euchromatin', 6),
                                       rep('accessibility', 2), 
                                       'methylation',
                                       rep('transcribing', 5)),
                        levels=c('transcribing', 'accessibility',
                                 'methylation', 'euchromatin',
                                 'HDAC', 'insulator', 'repressive')))
```


## Synergies Heatmaps

```{r Synergies chromatin - TIF all, fig.width= 10, fig.height= 20}
all_compounds_epistasis_TIF = CCD_tib %>%
  dplyr::select(drug_conc, indelrate_slope_plot, feature) %>%
  group_by(drug_conc) %>%
  mutate(onlyNA = sum(is.na(indelrate_slope_plot))) %>% 
  filter(onlyNA != 25) %>%
  reshape2::dcast(drug_conc ~ feature, value.var = "indelrate_slope_plot", fill = 0) %>% 
  column_to_rownames(var = "drug_conc")

target_ctrl = rename(target, conc_char_ctrl = conc_char)


# change these numbers according to where you want to set the cutoff
quant_breaks <- quantile(all_compounds_epistasis_TIF, c(.01,.99), na.rm = T)
palette_length <- 100
breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
            seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
colors[which(quant_breaks == 0)-1] = "gray90"

pheatmap(all_compounds_epistasis_TIF[, TRIP_chrom_clust$labels],
         border_color = F,
         cellwidth = 8,
         cellheight = 8,
         breaks = breaks,
         cluster_cols = TRIP_chrom_clust,
         clustering_method = "ward.D",
         annotation_row = target_ctrl,
         main = "Chromatin synergies - TIF all",
         color = colors,
         annotation_colors = annotation_colors_scr, 
         annotation_legend = T)
```

```{r Synergies chromatin - TIF per target, fig.width= 10, fig.height= 10}
targets = target %>% 
  pull(target) %>% 
  unique()

for (i in targets) {
  target_drugs = rownames(filter(target, target %in% i))
  
  # Make matrix with drugs from selected target
  TIF_synergy_matrix = all_compounds_epistasis_TIF %>% 
    rownames_to_column(var = "drug_conc") %>%
    filter(drug_conc %in% target_drugs) %>%
    column_to_rownames(var = "drug_conc")
  
  cluster_rows = ifelse(nrow(TIF_synergy_matrix) > 1, TRUE, FALSE)
  
  if(nrow(TIF_synergy_matrix) > 0) {
    # change these numbers according to where you want to set the cutoff
    quant_breaks <- quantile(TIF_synergy_matrix, c(.01,.99), na.rm = T) 
    palette_length <- 100
    breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
                seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
    slope.colors.temp = slope.colors
    if(all(TIF_synergy_matrix <= 0)) {
      breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length)))
      slope.colors.temp = filter(slope.colors, label != "positive")
    }
    if(all(TIF_synergy_matrix >= 0)) {
      breaks <- c(seq(0, quant_breaks[2], length.out=ceiling(palette_length)))
      slope.colors.temp = filter(slope.colors, label != "negative")
    }
    colors = colorRampPalette(slope.colors.temp %>% pull(color))(length(breaks))
    colors[which(quant_breaks == 0)-1] = "gray90"
    
    # heatmap
    p = pheatmap(TIF_synergy_matrix[, TRIP_chrom_clust$labels],
                 border_color = F,
                 cellwidth = 8,
                 cellheight = 8,
                 cluster_rows = cluster_rows,
                 breaks = breaks,
                 cluster_cols = TRIP_chrom_clust,
                 annotation_row = target_ctrl,
                 clustering_method = "ward.D",
                 main = paste("Chromatin synergies - TIF per target :" , i, sep = " "),
                 color = colors,
                 annotation_colors = annotation_colors_scr, 
                 annotation_legend = T)
    print(p)
  }
}
```

# Screen z-scores

```{r Z-score - TIF all, fig.width= 10, fig.height= 20}
ritios_tib_epi_effects = indel.data %>%
  mutate(drug_conc_ctrl = paste(drug, conc_char, sep = " ")) %>%
  filter(drug_conc_ctrl %in% rownames(all_compounds_epistasis_TIF)) %>%
  distinct(IPR, zscore_TIF_comb_rep, drug_conc_ctrl, drug, conc_char) 

mmejzscore_matrix = ritios_tib_epi_effects %>%
  distinct(drug_conc_ctrl, IPR, zscore_TIF_comb_rep) %>% 
  pivot_wider(names_from = IPR, values_from = zscore_TIF_comb_rep) %>%
  column_to_rownames(var = "drug_conc_ctrl")


quant_breaks <- quantile(mmejzscore_matrix, c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 1000
breaks <- breaks <- c(seq(quant_breaks[1], -1, 
                          length.out=ceiling(palette_length/2) + 1), 
                      seq(1, 
                          quant_breaks[2], 
                          length.out=floor(palette_length/2)))
colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
colors[-which(breaks < -2.58 | breaks > 2.58 )] = "gray90"

pheatmap(mmejzscore_matrix[, clone5_map_ipr$labels],
         border_color = F,
         cellwidth = 8,
         cellheight = 8,
         breaks = breaks,
         cluster_cols = clone5_map_ipr,
         annotation_row = target_ctrl,
         main = "Z-score [< 0.01 p.val] - TIF all",
         color = colors,
         annotation_colors = annotation_colors_scr, 
         annotation_legend = T,
         annotation_col = chromatin_TIF)
```

```{r Z-score - TIF per target, fig.width= 10, fig.height= 10}
targets = target %>% 
  pull(target) %>% 
  unique()

HDAC_tib = read.table("rs20220708_HDAC_targets.txt", sep = "\t", fill = NA, header = T) %>%
  dplyr::select(-target_main, -target_second) %>%
  as_tibble()

for (i in targets) {
  target_drugs = rownames(filter(target, target %in% i))
  
  # Make matrix with drugs from selected target
  mmejzscore_matrix = ritios_tib_epi_effects %>%
    distinct(drug_conc_ctrl, IPR, zscore_TIF_comb_rep) %>% 
    filter(drug_conc_ctrl %in% target_drugs) %>%
    pivot_wider(names_from = IPR, values_from = zscore_TIF_comb_rep) %>%
    column_to_rownames(var = "drug_conc_ctrl")
  
  cluster_rows = ifelse(nrow(mmejzscore_matrix) > 1, TRUE, FALSE)
  if(nrow(mmejzscore_matrix) > 0) {
    # change these numbers according to where you want to set the cutoff
    quant_breaks <- quantile(mmejzscore_matrix, c(.01,.99), na.rm = T) 
    palette_length <- 100
    breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
                seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
    colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
    colors[which(quant_breaks == 0)-1] = "gray90"
    
    # heatmap
    p = pheatmap(mmejzscore_matrix[, clone5_map_ipr$labels],
                 border_color = F,
                 cellwidth = 8,
                 cellheight = 8,
                 cluster_cols = clone5_map_ipr,
                 cluster_rows = cluster_rows,
                 breaks = breaks,
                 annotation_row = target_ctrl,
                 main = paste("Z-score [< 0.01 p.val] - TIF per target :" , i, sep = " "),
                 color = colors,
                 annotation_colors = annotation_colors_scr, annotation_legend = T,
                 annotation_col = chromatin_TIF)
    print(p)
  }
}
```


# Vorinostat & GSK126 validations

```{r}
vorinostat_validation = indel.validation.data %>% 
  filter(replicate == "mean" & plasmid == "LBR2", drug %in% c("DMSO", "Vorinostat"), sum_bc_reads > 1000)%>% 
  dplyr::select(barcode, drug, MMEJ_NHEJ, efficiency, all_of(chromatin_cols)) %>%
  distinct() 

barcde_all_reps_vorino = vorinostat_validation %>% 
  distinct(barcode, drug) %>% 
  group_by(barcode) %>%
  summarise(bccount = n()) %>% 
  filter(bccount == 2) %>%
  pull(barcode)

vorinostat_ratios = vorinostat_validation  %>% 
  filter(barcode %in% barcde_all_reps_vorino) %>%
  pivot_wider(names_from = drug, values_from = c(efficiency, MMEJ_NHEJ)) %>%
  mutate(diff_TIF = efficiency_Vorinostat - efficiency_DMSO,
         ratio_TIF = efficiency_Vorinostat/efficiency_DMSO,
         log2_FC_TIF = log2(ratio_TIF),
         diff_MMEJ_NHEJ = MMEJ_NHEJ_Vorinostat - MMEJ_NHEJ_DMSO,
         ratio_MMEJ_NHEJ = MMEJ_NHEJ_Vorinostat/MMEJ_NHEJ_DMSO,
         log2_FC_MMEJ_NHEJ = log2(ratio_MMEJ_NHEJ)) 

barcde_all_reps_gsk = RSTP2_IndelRatios_Chromatin_2kb %>% 
  distinct(barcode, drug) %>% 
  group_by(barcode) %>%
  summarise(bccount = n()) %>% 
  filter(bccount == 2) %>%
  pull(barcode)

GSK_ratios = RSTP2_IndelRatios_Chromatin_2kb %>% 
  filter(barcode %in% barcde_all_reps_gsk) %>%
  # distinct(barcode, drug, MMEJ_NHEJ , efficiency) %>% 
  pivot_wider(names_from = drug, values_from = c(efficiency, MMEJ_NHEJ)) %>%
  mutate(diff_TIF = efficiency_GSK126 - efficiency_DMSO_GSK,
         ratio_TIF = efficiency_GSK126/efficiency_DMSO_GSK,
         log2_FC_TIF = log2(ratio_TIF),
         diff_MMEJ_NHEJ = MMEJ_NHEJ_GSK126 - MMEJ_NHEJ_DMSO_GSK,
         ratio_MMEJ_NHEJ = MMEJ_NHEJ_GSK126/MMEJ_NHEJ_DMSO_GSK,
         log2_FC_MMEJ_NHEJ = log2(ratio_MMEJ_NHEJ))
```

```{r}
## Vorinostat
chrom.mods.cols <- seq(grep("CTCF", names(vorinostat_ratios)), grep("H3K27me3_domain", names(vorinostat_ratios))-1)

vorinostat.hist.corr <- CorrelationDFMean(vorinostat_ratios, "log2_FC_TIF", targets = chrom.mods.cols) %>%
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)

lims = 1

CorrelationPlotMean(vorinostat_ratios, "log2_FC_TIF", targets = vorinostat.hist.corr, title = "with log2 FC TIF after Vorinostat treatment")

## GSK126
chrom.mods.cols <- seq(grep("CTCF", names(GSK_ratios)), grep("H3K27me3_domain", names(GSK_ratios))-1)

GSK.hist.corr <- CorrelationDFMean(GSK_ratios, "log2_FC_TIF", targets = chrom.mods.cols) %>%
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)

lims = 1

CorrelationPlotMean(GSK_ratios, "log2_FC_TIF", targets = GSK.hist.corr, title = "with log2 FC TIF after GSK126 treatment")
```

```{r}
## Vorinostat
chrom.mods.cols <- seq(grep("CTCF", names(vorinostat_ratios)), grep("H3K27me3_domain", names(vorinostat_ratios))-1)

vorinostat.hist.corr <- CorrelationDFMean(vorinostat_ratios, "log2_FC_MMEJ_NHEJ", targets = chrom.mods.cols) %>%
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)

lims = 1

CorrelationPlotMean(vorinostat_ratios, "log2_FC_MMEJ_NHEJ", targets = vorinostat.hist.corr, title = "with log2 FC balance after Vorinostat treatment")

## GSK126
chrom.mods.cols <- seq(grep("CTCF", names(GSK_ratios)), grep("H3K27me3_domain", names(GSK_ratios))-1)

GSK.hist.corr <- CorrelationDFMean(GSK_ratios, "log2_FC_MMEJ_NHEJ", targets = chrom.mods.cols) %>%
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)

lims = 1

CorrelationPlotMean(GSK_ratios, "log2_FC_MMEJ_NHEJ", targets = GSK.hist.corr, title = "with log2 FC balance after GSK126 treatment")
```

```{r Vorino upset, fig.height=5, fig.width=7}
domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

vorino.dt = vorinostat_ratios %>%
  filter(!is.na(diff_TIF)) %>%
  dplyr::select(barcode, diff_TIF, ratio_MMEJ_NHEJ, log2_FC_MMEJ_NHEJ, log2_FC_TIF, domains) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()

dim(vorino.dt)


vorino.dt[vorino.dt=='iDomain'] <- 0
vorino.dt[vorino.dt=='Domain'] <- 1
vorino.dt[vorino.dt=='DHS'] <- 1

vorino.dt[is.na(LAD_domain), LAD_domain:=0]
# vorino.dt[is.na(DHS_domain), DHS_domain:=0]

vorino.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(vorino.dt) = gsub('_domain', '', colnames(vorino.dt))

opt_df = get_combinations(domain_levels)

# Ratio of SSTR vs MME
p1 = plot_comb_grid_beeswarm(vorino.dt, opt_df,
                        domain_levels, 'diff_TIF', min_count = 10, y_plabel=c(-.2, .2))
p2 = plot_comb_grid_beeswarm(vorino.dt, opt_df,
                        domain_levels, 'ratio_MMEJ_NHEJ', min_count = 10, y_plabel=c(-2.5, 2.5))
p3 = plot_comb_grid_beeswarm(vorino.dt, opt_df,
                        domain_levels, 'log2_FC_MMEJ_NHEJ', min_count = 10, y_plabel=c(-2.5, 2.5))
p4 = plot_comb_grid_beeswarm(vorino.dt, opt_df,
                        domain_levels, 'log2_FC_TIF', min_count = 10, y_plabel=c(-.5, .5))

plot_grid(p1, p2, p3, p4, labels = c("ΔTIF", "M/N", "LogBal", "LogTIF"))

k27me3 = vorino.dt %>% filter(late_replicating == 0, 
                             LAD == 0,
                             H3K9me2 == 0,
                             H3K27me3 == 1) %>% 
  pull(diff_TIF)

any.k27me3 = vorino.dt %>% filter(H3K27me3 == 1) %>% 
  pull(diff_TIF)

not.k27me3 = vorino.dt %>% filter(H3K27me3 != 1) %>% 
  pull(diff_TIF)

hetcomb = vorino.dt %>% filter(late_replicating == 1, 
                              LAD == 1,
                              H3K9me2 == 1,
                              H3K27me3 == 0) %>% 
  pull(diff_TIF)

idomain = vorino.dt %>% filter(late_replicating == 0, 
                              LAD == 0,
                              H3K9me2 == 0,
                              H3K27me3 == 0) %>% 
  pull(diff_TIF)

# DHS = vorino.dt %>% filter(late_replicating == 0,
#                           LAD == 0,
#                           H3K9me2 == 0,
#                           H3K27me3 == 0,
#                           DHS == 1) %>%
#   pull(diff_TIF)

wilcox.test(idomain, k27me3, alternative = "two.sided")

wilcox.test(idomain, any.k27me3, alternative = "two.sided")
wilcox.test(not.k27me3, any.k27me3, alternative = "two.sided")

wilcox.test(idomain, hetcomb, alternative = "two.sided")

vorino.dt %>% mutate(state = ifelse(late_replicating == 1 & 
                                     LAD == 1 &
                                     H3K9me2 == 1 &
                                     H3K27me3 == 0, "triple", ifelse(
                                       late_replicating == 0 & 
                                         LAD == 0 & 
                                         H3K9me2 == 0 & 
                                         H3K27me3 == 0, "iDomain", ifelse(late_replicating == 0 &  
                                                                            LAD == 0 & 
                                                                            H3K9me2 == 0 & 
                                                                            H3K27me3 == 1, "K27Me3", "other")))) %>%
  ggplot(., aes(diff_TIF, color = state)) +
  geom_density()
```

```{r GSK upset, fig.height=5, fig.width=7}
domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

GSK.dt = GSK_ratios %>%
  filter(!is.na(diff_TIF)) %>%
  dplyr::select(barcode, diff_TIF, log2_FC_MMEJ_NHEJ, ratio_MMEJ_NHEJ, log2_FC_TIF, domains) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()

dim(GSK.dt)


GSK.dt[GSK.dt=='iDomain'] <- 0
GSK.dt[GSK.dt=='Domain'] <- 1
GSK.dt[GSK.dt=='DHS'] <- 1

GSK.dt[is.na(LAD_domain), LAD_domain:=0]
# GSK.dt[is.na(DHS_domain), DHS_domain:=0]

GSK.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(GSK.dt) = gsub('_domain', '', colnames(GSK.dt))

opt_df = get_combinations(domain_levels)

# Ratio of SSTR vs MME
p1 = plot_comb_grid_beeswarm(GSK.dt, opt_df,
                        domain_levels, 'diff_TIF', min_count = 10, y_plabel=c(-.25, .25))
p1 = plot_comb_grid_beeswarm(GSK.dt, opt_df,
                        domain_levels, 'ratio_MMEJ_NHEJ', min_count = 10, y_plabel=c(-2, 2))
p1 = plot_comb_grid_beeswarm(GSK.dt, opt_df,
                        domain_levels, 'log2_FC_MMEJ_NHEJ', min_count = 10, y_plabel=c(-1.5, 1.5))
p1 = plot_comb_grid_beeswarm(GSK.dt, opt_df,
                        domain_levels, 'log2_FC_TIF', min_count = 10, y_plabel=c(-1.5, 1.5))

plot_grid(p1, p2, p3, p4, labels = c("ΔTIF", "M/N", "LogBal", "LogTIF"))

k27me3 = GSK.dt %>% filter(late_replicating == 0, 
                             LAD == 0,
                             H3K9me2 == 0,
                             H3K27me3 == 1) %>% 
  pull(diff_TIF)

any.k27me3 = GSK.dt %>% filter(H3K27me3 == 1) %>% 
  pull(diff_TIF)

not.k27me3 = GSK.dt %>% filter(H3K27me3 != 1) %>% 
  pull(diff_TIF)

hetcomb = GSK.dt %>% filter(late_replicating == 1, 
                              LAD == 1,
                              H3K9me2 == 1,
                              H3K27me3 == 0) %>% 
  pull(diff_TIF)

idomain = GSK.dt %>% filter(late_replicating == 0, 
                              LAD == 0,
                              H3K9me2 == 0,
                              H3K27me3 == 0) %>% 
  pull(diff_TIF)

# DHS = GSK.dt %>% filter(late_replicating == 0,
#                           LAD == 0,
#                           H3K9me2 == 0,
#                           H3K27me3 == 0,
#                           DHS == 1) %>%
#   pull(diff_TIF)

wilcox.test(idomain, k27me3, alternative = "two.sided")

wilcox.test(idomain, any.k27me3, alternative = "two.sided")
wilcox.test(not.k27me3, any.k27me3, alternative = "two.sided")

wilcox.test(idomain, hetcomb, alternative = "two.sided")

GSK.dt %>% mutate(state = ifelse(late_replicating == 1 & 
                                     LAD == 1 &
                                     H3K9me2 == 1 &
                                     H3K27me3 == 0, "triple", ifelse(
                                       late_replicating == 0 & 
                                         LAD == 0 & 
                                         H3K9me2 == 0 & 
                                         H3K27me3 == 0, "iDomain", ifelse(late_replicating == 0 &  
                                                                            LAD == 0 & 
                                                                            H3K9me2 == 0 & 
                                                                            H3K27me3 == 1, "K27Me3", "other")))) %>%
  ggplot(., aes(diff_TIF, color = state)) +
  geom_density()
```

```{r}
indel.data %>% filter(!between(zscore_TIF_comb_rep, -3.2905, 3.2905) & conc_char != "control" & viab_reproducibility & cor_TIF_filter & filter_zscore_TIF_comb_rep) %>% 
  # distinct(IPR, drug, replicate, target, tech, conc_char, zscore_TIF, zscore_TIF_comb_tech, zscore_TIF_comb_rep) %>% 
  distinct(IPR, drug, conc_char, target, zscore_TIF_comb_rep) %>% 
  ggplot(., aes(IPR, zscore_TIF_comb_rep, color = target)) + 
  geom_quasirandom() + 
  facet_grid(. ~ conc_char) + 
  scale_color_manual(values = colors_targets) +
  theme_classic2()
```


## 2G Aurora A vs. Aurora B

```{r 2G TIF AurK inhibitor heatmap, warning = FALSE, fig.width= 8, fig.height=10}
AurK_tib = read.table("rs20220711_AurK_inhibitors.txt", sep = "\t", fill = NA, header = T) %>%
  as_tibble()

cluster_AurK = target_set_drugs %>% filter(target == "Aurora Kinase") %>% rownames()

eff.data.heatmap.aurK = eff.data.heatmap2 %>% 
  mutate(drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  filter(drug_conc %in% cluster_AurK)

# Add cluster annotation
cluster_AurK <- hclust(dist(eff.data.heatmap.aurK), method = "complete")
cluster_AurK <- cutree(tree = cluster_AurK, k = 4)
cluster_AurK <- data.frame(cluster_AurK)

eff.data.heatmap.aurK <- cbind(eff.data.heatmap.aurK, cluster_AurK) %>% 
  as_tibble()

cluster_AurK <- eff.data.heatmap.aurK %>%
  distinct(drug_conc, cluster_AurK)

target.aurK <- target_set_drugs %>%
  rownames_to_column("drug_conc") %>%
  left_join(AurK_tib) %>%
  left_join(cluster_AurK) %>%
  filter(!is.na(cluster_AurK)) %>%
  replace(is.na(.), 0)

rownames(target.aurK) <- target.aurK$drug_conc
target.AurK.selection <- target.aurK %>%
  dplyr::select(conc_char,pan_AurK, AurC, 
                AurB,	AurA, viab)

target.AurK.selection

AurK_matrix = eff.data.heatmap.aurK %>%
  arrange(target.AurK.selection) %>%
  column_to_rownames('drug_conc') %>%
  dplyr::select(contains('IPR')) %>%
  t()

# pdf(paste0(figure_out, "rs20220623_indelfrequency_heatmap_HDAC.pdf"), width=16, height=8)
plt3 = pheatmap(AurK_matrix[clone5_map_ipr$labels, ],
                border_color = F,
                cellwidth = 8,
                cellheight = 8,
                breaks = myBreaks,
                clustering_method = "ward.D",
                cluster_rows = clone5_map_ipr,
                cluster_cols = T,
                angle_col = 90,
                na_col = "white",
                main = 'efficiency change, all concentrations',
                annotation_col = target.AurK.selection,
                annotation_row = chromatin,
                color = diverge_hcl(palette_length, "Blue-Red"),
                annotation_colors = annotation_colors_scr,
                annotation_legend = T,
                cutree_cols = 4)
```


## 2H Aurora A vs. Aurora B global 

```{r}
## Plot mean DMSO cutting efficiency in middle, and difference from that to the left and right in percent
indel.data.freq <- indel.data %>%
  filter((sample == "Epigenetic Compound" | drug == "DMSO") & 
           indel_freq_filter & viab_reproducibility &
           cor_TIF_filter & !is.na(freqCut)) %>%
  distinct(IPR, drug, conc_char, sample, target, viab_mean, replicate,
           freqCut_global, freqCut, freqCut_mean, IPR,
           freqCut_diff_global, freqCut_diff_global_mean,
           zscore_TIF_global_comb_rep , filter_zscore_TIF_global_comb_rep, 
           zscore_TIF_comb_rep , filter_zscore_TIF_comb_rep) %>%
  filter(!(freqCut_diff_global < -0.1 & viab_mean < 0.5)) %>%
  mutate(freqCut_mean = ave(freqCut_global, drug, conc_char),
         freqCut_diff_mean =  ave(freqCut_diff_global, drug, conc_char))

eff.data <- indel.data.freq %>%
  filter(sample == "Epigenetic Compound") %>%
  distinct(IPR, zscore_TIF_global_comb_rep , filter_zscore_TIF_global_comb_rep, 
           zscore_TIF_comb_rep , filter_zscore_TIF_comb_rep, 
           drug, target, conc_char, freqCut_diff_global_mean) %>%
  mutate(mean_eff_pval_comb = 2*pnorm(q = abs(zscore_TIF_comb_rep ), lower.tail = F),
         mean_eff_pval_adj_comb = p.adjust(mean_eff_pval_comb, method = 'fdr'),
         sign_local = ifelse((mean_eff_pval_adj_comb < 0.01 & 
                                filter_zscore_TIF_comb_rep &
                                 zscore_TIF_comb_rep > 2.58), "TIF_Increase", 
                             ifelse((mean_eff_pval_adj_comb < 0.01 &
                                       filter_zscore_TIF_global_comb_rep &
                                 zscore_TIF_comb_rep < -2.58), "TIF_Decrease", 
                                    "No_Change"))) %>%
  group_by(drug, conc_char) %>%
  mutate(sign_global = ifelse(length(table(sign_local) == 1),  names(table(sign_local)), "No_Change")) %>%
  distinct(IPR, drug, conc_char, target, sign_local, sign_global)
```

```{r 2H AurK inhib global effects, fig.width= 4.5, fig.height= 5}
CTRL.mean = dmso.eff.score$estimate[1]
CTRL.SD = dmso.eff.score$estimate[2]

# # Alternative: label drugs and color targets
indel.data.freq2  <- indel.data.freq %>%
  filter(sample == "Epigenetic Compound") %>%
  distinct(freqCut_mean, freqCut_diff_mean, drug, conc_char, zscore_TIF_global_comb_rep , target) %>%
  mutate(mean_eff_pval_global_comb = 2*pnorm(q = abs(zscore_TIF_global_comb_rep ), lower.tail = F)) %>%
  mutate(mean_eff_pval_adj_global_comb = p.adjust(mean_eff_pval_global_comb, method = 'fdr'))

labels <- indel.data.freq2 %>%
  mutate(freqCut_diff_mean = abs(freqCut_diff_mean)) %>%
  group_by(conc_char) %>%
  top_n(3, freqCut_diff_mean) %>%
  ungroup() %>%
  mutate(flag = "TRUE") %>%
  distinct(drug, conc_char, flag)

indel.data.freq2 <- merge(indel.data.freq2, labels, all = T)
indel.data.freq2$flag[is.na(indel.data.freq2$flag)] <- "FALSE"
indel.data.freq2$drug <- gsub(" [(].*[)]", "", indel.data.freq2$drug) #shorten name of drug

indel.data.freq2 %>% filter(mean_eff_pval_adj_global_comb < 0.01) %>%
  distinct(drug, conc_char, mean_eff_pval_adj_global_comb) %>%
  group_by(conc_char) %>%
  summarise(count = n())

indel.data.freq2 %>%
  filter(target == "Aurora Kinase") %>%
  left_join(AurK_tib) %>%
  filter(!is.na(mean_eff_pval_global_comb)) %>%
  mutate(AurK_Target = ifelse((AurB == 1 & AurA == 0), "B", ifelse(
    (AurB == 0 & AurA == 1), "A", "A+B")),
    mean_eff_pval_global_comb = -log10(p.adjust(mean_eff_pval_global_comb, method = 'fdr')),
    mean_eff_pval_global_comb = ifelse(mean_eff_pval_global_comb > 100, 100, mean_eff_pval_global_comb)) %>%
  ggplot(.,
         aes(y = mean_eff_pval_global_comb, x = freqCut_mean, 
             color = AurK_Target, size = conc_char)) +
  geom_point() +
  #geom_label_repel(data = indel.data.freq2 %>% filter(flag == "TRUE") %>% unique(), 
  #aes(label = drug)) +
  geom_hline(linetype = "dashed", yintercept = -log10(0.01)) +
  geom_vline(linetype = "solid", xintercept = (CTRL.mean)) +
  geom_vline(linetype = "dashed", xintercept = (CTRL.mean + CTRL.SD)) +
  geom_vline(linetype = "dashed", xintercept = (CTRL.mean - CTRL.SD)) +
  #geom_vline(linetype = "twodash", xintercept = unique(indel.data.freq$max_dmso_freq))+
  scale_color_manual(values = c("#6d6875", "#b5838d", "#ffb4a2")) +
  # scale_x_continuous(labels = scales::percent) +
  coord_cartesian(xlim = c(0, 100)) +
  coord_flip() +
  theme_classic_lines_90()
```


## 2I Aurora A vs. Aurora B
```{r 2I AurA vs AurB inhibitors, fig.width= 3, fig.height= 4}
viab_tib = indel.data %>% filter(viab_reproducibility & indel_freq_filter & cor_TIF_filter) %>% 
  distinct(replicate, drug, conc_char, viab_mean) %>%
  group_by(drug, conc_char) %>%
  summarise(viab_mean = mean(viab_mean, na.rm = TRUE)) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug))

aurora_df <- eff.data.heatmap2 %>%
  filter(target == "Aurora Kinase") %>%
  pivot_longer(contains("IPR"), names_to = "IPR", values_to = "freqCut_dif") %>%
  mutate(freqCut_dif = ave(freqCut_dif, drug, conc_char, FUN = function(x) mean(x, na.rm = TRUE))) %>%
  dplyr::select(-IPR) %>%
  left_join(viab_tib) %>%
  distinct() 

aurora_df <- merge(aurora_df, AurK_tib) %>%
  filter(!is.na(viab_mean)) %>%
  mutate(AurK_Target = ifelse((AurB == 1 & AurA == 0), "B", ifelse(
    (AurB == 0 & AurA == 1), "A", "A+B")))

ggplot(aurora_df, 
       aes(x = AurK_Target, y = freqCut_dif, fill = viab_mean)) + 
  geom_quasirandom(size = 2, shape = 21) +
  geom_signif(comparisons = combn(sort(unique(aurora_df$AurK_Target)), 2, simplify = F), 
              map_signif_level=sigFunc, 
              step_increase = 0.1) +
  scale_fill_continuous_sequential(palette = "Peach", rev = TRUE) +
  scale_color_manual(values= "black") +
  theme_classic_lines() +
  ylab("global TIF change")
```


# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

