---
title: "DSB Episcreen data analysis & plotting"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```
### Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(dplyr)
library(outliers)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(ggrepel)
library(GGally)
library(DESeq2)
library(gghighlight)
library(platetools)
library(ggpubr)
library(tidyr)
library(Laurae)
library(tibble)
library(pheatmap)
library(RColorBrewer)
library(ggrastr)
```

### Functions

```{r load_functions, echo = FALSE, warning = FALSE}

p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
    ## Methods 'Hommel', 'BH', 'BY' and speed improvements
    ## contributed by Gordon Smyth
    method <- match.arg(method)
    if(method == "fdr") method <- "BH"	# back compatibility
    nm <- names(p)
    p <- as.numeric(p)
    p0 <- setNames(p, nm)
    if(all(nna <- !is.na(p))) nna <- TRUE
    p <- p[nna]
    lp <- length(p)
    # stopifnot(n >= lp)
    if (n <= 1) return(p0)
    if (n == 2 && method == "hommel") method <- "hochberg"

    p0[nna] <-
	switch(method,
	       bonferroni = pmin(1, n * p),
	       holm = {
		   i <- seq_len(lp)
		   o <- order(p)
		   ro <- order(o)
		   pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
	       },
	       hommel = { ## needs n-1 >= 2 in for() below
		   if(n > lp) p <- c(p, rep.int(1, n-lp))
		   i <- seq_len(n)
		   o <- order(p)
		   p <- p[o]
		   ro <- order(o)
		   q <- pa <- rep.int( min(n*p/i), n)
		   for (j in (n-1):2) {
		       ij <- seq_len(n-j+1)
		       i2 <- (n-j+2):n
		       q1 <- min(j*p[i2]/(2:j))
		       q[ij] <- pmin(j*p[ij], q1)
		       q[i2] <- q[n-j+1]
		       pa <- pmax(pa,q)
		   }
		   pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
	       },
	       hochberg = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
	       },
	       BH = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( n / i * p[o] ))[ro]
	       },
	       BY = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   q <- sum(1L/(1L:n))
		   pmin(1, cummin(q * n / i * p[o]))[ro]
	       },
	       none = p)
    p0
}

SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

# From Fede:
# ggpairs custom functions
boundaries <- c(0.7,0.75,0.8,0.85)
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(techn_repl %>% filter(large_effect == "0_yes") %>% dplyr::select(contains("E1")), mapping$x)
  y   <- eval_data_col(techn_repl %>% filter(large_effect == "0_yes") %>% dplyr::select(contains("E1")), mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}


# stat_smooth_func <- function(mapping = NULL, data = NULL,
#                         geom = "smooth", position = "identity",
#                         ...,
#                         method = "auto",
#                         formula = y ~ x,
#                         se = TRUE,
#                         n = 80,
#                         span = 0.75,
#                         fullrange = FALSE,
#                         level = 0.95,
#                         method.args = list(),
#                         na.rm = FALSE,
#                         show.legend = NA,
#                         inherit.aes = TRUE,
#                         xpos = NULL,
#                         ypos = NULL) {
#   layer(
#     data = data,
#     mapping = mapping,
#     stat = StatSmoothFunc,
#     geom = geom,
#     position = position,
#     show.legend = show.legend,
#     inherit.aes = inherit.aes,
#     params = list(
#       method = method,
#       formula = formula,
#       se = se,
#       n = n,
#       fullrange = fullrange,
#       level = level,
#       na.rm = na.rm,
#       method.args = method.args,
#       span = span,
#       xpos = xpos,
#       ypos = ypos,
#       ...
#     )
#   )
# }
# 
# StatSmoothFunc <- ggproto("StatSmooth", Stat,
#                       
#                       setup_params = function(data, params) {
#                         # Figure out what type of smoothing to do: loess for small datasets,
#                         # gam with a cubic regression basis for large data
#                         # This is based on the size of the _largest_ group.
#                         if (identical(params$method, "auto")) {
#                           max_group <- max(table(data$group))
#                           
#                           if (max_group < 1000) {
#                             params$method <- "loess"
#                           } else {
#                             params$method <- "gam"
#                             params$formula <- y ~ s(x, bs = "cs")
#                           }
#                         }
#                         if (identical(params$method, "gam")) {
#                           params$method <- mgcv::gam
#                         }
#                         
#                         params
#                       },
#                       
#                       compute_group = function(data, scales, method = "auto", formula = y~x,
#                                                se = TRUE, n = 80, span = 0.75, fullrange = FALSE,
#                                                xseq = NULL, level = 0.95, method.args = list(),
#                                                na.rm = FALSE, xpos=NULL, ypos=NULL) {
#                         if (length(unique(data$x)) < 2) {
#                           # Not enough data to perform fit
#                           return(data.frame())
#                         }
#                         
#                         if (is.null(data$weight)) data$weight <- 1
#                         
#                         if (is.null(xseq)) {
#                           if (is.integer(data$x)) {
#                             if (fullrange) {
#                               xseq <- scales$x$dimension()
#                             } else {
#                               xseq <- sort(unique(data$x))
#                             }
#                           } else {
#                             if (fullrange) {
#                               range <- scales$x$dimension()
#                             } else {
#                               range <- range(data$x, na.rm = TRUE)
#                             }
#                             xseq <- seq(range[1], range[2], length.out = n)
#                           }
#                         }
#                         # Special case span because it's the most commonly used model argument
#                         if (identical(method, "loess")) {
#                           method.args$span <- span
#                         }
#                         
#                         if (is.character(method)) method <- match.fun(method)
#                         
#                         base.args <- list(quote(formula), data = quote(data), weights = quote(weight))
#                         model <- do.call(method, c(base.args, method.args))
#                         
#                         m = model
#                         eq <- substitute(italic(y) == a + b %.% italic(x)*","~~italic(r)^2~"="~r2, 
#                                          list(a = format(coef(m)[1], digits = 3), 
#                                               b = format(coef(m)[2], digits = 3), 
#                                               r2 = format(summary(m)$r.squared, digits = 3)))
#                         func_string = as.character(as.expression(eq))
#                         
#                         if(is.null(xpos)) xpos = min(data$x)*0.9
#                         if(is.null(ypos)) ypos = max(data$y)*0.9
#                         data.frame(x=xpos, y=ypos, label=func_string)
#                         
#                       },
#                       
#                       required_aes = c("x", "y")
# )

theme_classic_lines <- function() {
  theme_pubr(border = F, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}
theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

```


### Data import

```{r data import, echo = FALSE, warning = FALSE}
# Import data from preprocessing script
#indel.data <- readRDS("/DATA/projects/DSBrepair/data/R/rs20220120_episcreen2/rs20220120_episcreen_ratios.RDS")
indel.data <- readRDS("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/mt20220124_episcreen.RDS")
```

---

## A: Cutting efficiency change per drug
Are there any drugs (per concentration) that significantly change the cutting efficiency significantly across all IPRs?
```{r warning = FALSE}
# A: Cutting efficiency per drug - beeswarm plot
ggplot_custom(indel.data %>%
                filter(sample == "drug") %>%
                distinct(mean_eff_zscore_global_comb, drug, conc_char) %>%
                mutate(mean_eff_pval_global_comb = 2*pnorm(q = abs(mean_eff_zscore_global_comb), lower.tail = F)) %>%
                mutate(mean_eff_pval_adj_global_comb = p.adjust(mean_eff_pval_global_comb, method = 'fdr')),
              aes(x = conc_char, y = mean_eff_zscore_global_comb, 
                  color = ifelse(mean_eff_pval_adj_global_comb < 0.05 & mean_eff_zscore_global_comb > 1, "green", 
                               ifelse(mean_eff_pval_adj_global_comb < 0.05 & mean_eff_zscore_global_comb < -1, "yellow","blue")))) +
  geom_quasirandom() +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96) +
  ggtitle("first z-score per tech. rep., then mean of z-scores") +
  scale_color_manual(values = c(colors_diverse[1],colors_diverse[2], colors_diverse[4]))
```

*Figure 2A: Compound-induced change in cutting efficiency of all tested compounds. Mean per drug & concentration vs. p-value from t-test.*

**Conclusion: Yes, some drugs do alter cutting efficiency significantly. Changes go in both directions, at lower concentration the efficiency is mostly increased (at higher concentrations the decrease might be due to toxicity)**

---

## B: Cutting efficiency change per target group
Are there any target groups that are prone to changes in cutting efficiency?
```{r warning = FALSE}
# B: Cutting efficiency change per target group, all concentrations together
ggplot_custom(indel.data %>%
                filter(conc_char == "100 nM") %>%
                filter(sample == "drug") %>%
                distinct(mean_eff_zscore_global_comb, drug, conc_char, target) %>%
                mutate(mean_eff_pval_global_comb = 2*pnorm(q = abs(mean_eff_zscore_global_comb), lower.tail = F)) %>%
                mutate(mean_eff_pval_adj_global_comb = p.adjust(mean_eff_pval_global_comb, method = 'fdr')) %>%
                mutate(mean_eff_target = ave(mean_eff_zscore_global_comb, target, conc_char, FUN = mean)), 
              aes(y = mean_eff_zscore_global_comb, x = reorder(target, -mean_eff_target, FUN = median))) + 
    geom_boxplot(position = position_dodge(0.75), alpha = 0.4, outlier.shape = NA) +
  geom_quasirandom(aes(color = ifelse(mean_eff_pval_adj_global_comb < 0.05 & mean_eff_zscore_global_comb > 1, "green", 
                               ifelse(mean_eff_pval_adj_global_comb < 0.05 & mean_eff_zscore_global_comb < -1, "yellow","blue")))) +
  geom_hline(linetype = "dashed", yintercept = 2.58) +
  geom_hline(linetype = "dashed", yintercept = -2.58) +
  theme_classic_lines_90()+
  scale_color_manual(values = c(colors_diverse[1],colors_diverse[2], colors_diverse[4]))
```

*Figure 2B: Compound-induced change in cutting efficiency per compound target groups.*

**Conclusion: Yes, specifically HDACs seem to increase cutting efficiency. Some other interesting observations can also be made from this plot.**

---

## C: Cutting efficiency change per target group
Are there any target groups that are prone to changes in cutting efficiency?
```{r warning = FALSE}
# C: Bar plot
eff.data <- indel.data %>%
  filter(sample == "drug") %>%
  distinct(mean_eff_zscore_global_comb, drug, target, conc_char) %>%
  mutate(mean_eff_pval_global_comb = 2*pnorm(q = abs(mean_eff_zscore_global_comb), lower.tail = F)) %>%
  mutate(mean_eff_pval_adj_global_comb = p.adjust(mean_eff_pval_global_comb, method = 'fdr'))

eff.data$sign <- "No"
eff.data$sign[eff.data$mean_eff_pval_adj_global_comb < -1 & eff.data$mean_eff_pval_adj_global_comb < 0.05] <- "Up"
eff.data$sign[eff.data$mean_eff_pval_adj_global_comb > 1 & eff.data$mean_eff_pval_adj_global_comb < 0.05] <- "Down"

eff.data <- eff.data %>%
  mutate(number = ave(target, target, sign, conc_char, FUN = length)) %>%
  mutate(number = as.numeric(number)) %>%
  distinct(conc_char, target, sign, number) %>%
  spread(sign, number, fill = 0) %>%
  mutate(fraction = (Up + Down) / (No + (Up + Down))) %>%
  pivot_longer(Yes:No, names_to = "sign", values_to = "number")

ggplot_custom(eff.data ,
              aes(y = reorder(target, -number), x = number, fill = sign)) +
  geom_bar(stat = "identity") +
  facet_wrap(~conc_char)+
  scale_color_manual(values = c(colors_diverse[1],colors_diverse[2], colors_diverse[4]))
```

*Figure 2C: Number of compounds with significant changes in cutting efficiency per target group.*

**Conclusion: Same as B)**

---

## D: Efficiency heatmap
Are the effects on cutting efficiency global or local? Can we plot all interesting effects in one heatmap per concentration?
```{r warning = FALSE}
# Set annotation for heatmap
# Row annotation: annotate drugs with the target group
target <- indel.data %>% 
  filter(!target %in% c("Negative Control", "MRN", "DNA-PK")) %>% 
  distinct(drug, target) %>% 
  column_to_rownames(var="drug")

file_list = list.files("/DATA/projects/DSBrepair/data/R/rs20210813_episcreen_2", full.names = TRUE)
file_list
episcreen_ratios = readRDS(file_list[[3]])

chromatin <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(episcreen_ratios, barcode, IPR)) %>%
  filter(IPR != "<NA>") %>% 
  select(-barcode) %>%
  column_to_rownames(var="IPR")

# Change colors of annotations
chrom_colors = c("Euchromatin" = "#EAEEE8",
                  "other-heterochromatin" = "#CAD6C2",  
                     "H3K27me3" = "#a5a58d", 
                  "Triple Heterochromatin" = "#6b705c")

# colors <- brewer.pal(length(unique(barcode.annotation$cluster)), "Pastel2")
# names(colors) <- unique(barcode.annotation$cluster)

annotation_colors_scr = list(
  chromatin = chrom_colors,
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43"
             ))



# Order the annotation legend
ordered_marks <- c("Euchromatin", 
                     "other-heterochromatin",  
                     "H3K27me3", 
                     "Triple Heterochromatin")

annotation_colors_scr$chromatin <- annotation_colors_scr$chromatin[ordered_marks]

ordered_marks2 <- c("DNA-PK","Epigenetic Reader Domain", "HDAC",  "Sirtuin", "HAT", 
                    "Histone Demethylase", "HMT", "DNMT", 
                    "HIF", "JAK", "PIM",
                    "Aurora Kinase", "PARP")
annotation_colors_scr$target <- annotation_colors_scr$target[ordered_marks2]


# Generate heatmap: select top drugs (highest change at any IPR) per concentration
eff.data.heatmap <- indel.data %>%
  #filter(target != "Aurora Kinase") %>%
  filter(sample == "drug") %>%
  #filter(Mod(mean_eff_zscore_comb) >= 2.58) %>%  # we can choose to display all drugs that have significant global hits
  distinct(conc_char, drug, target, IPR, mean_eff_zscore_comb, freqCut_norm_mean) %>%
  mutate(freqCut_norm_mean = log2(ave(freqCut_norm_mean, drug, conc_char, IPR, FUN = mean))) %>%
  distinct()

## Select drugs that have a large effect size in at least one location (some drugs might have a small global effect size, but large local)
eff.data.heatmap2 <- eff.data.heatmap %>%
  mutate(mean_eff_zscore_comb = 2*pnorm(q = abs(mean_eff_zscore_comb), lower.tail = F)) %>%
  mutate(mean_eff_zscore_comb = p.adjust(mean_eff_zscore_comb, method = 'fdr')) %>%
  mutate(min = ave(mean_eff_zscore_comb, drug, conc_char, FUN = min)) %>%
  filter(min < 0.000001)

eff.data.heatmap2$freqCut_norm_mean[eff.data.heatmap2$mean_eff_zscore_comb > 0.001] <- NA
eff.data.heatmap2 <- eff.data.heatmap2 %>%
  dplyr::select(-min, -mean_eff_zscore_comb) %>%
  spread(IPR, freqCut_norm_mean)

chromatin_ord <- chromatin %>% 
  rownames_to_column('IPR') %>%
  mutate(chromatin = gsub("Euc", "0_Euc", chromatin)) %>%
  mutate(chromatin = gsub("other-heterochromatin", "1_other-heterochromatin", chromatin)) %>%
  mutate(chromatin = gsub("H3K27me3", "2_H3K27me3", chromatin)) %>%
  mutate(chromatin = gsub("Triple Heterochromatin", "3_Triple Heterochromatin", chromatin)) %>%
  arrange(chromatin)

eff.data.heatmap2 <- eff.data.heatmap2[,c("drug","conc_char", "target", chromatin_ord$IPR)]

breaks <- quantile(unlist(eff.data.heatmap2 %>% dplyr::select(contains('IPR'))), c(.05,.95), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 100

myBreaks <- c(seq(breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
              seq(breaks[2]/palette_length, breaks[2], length.out=floor(palette_length/2)))

for (i in unique(eff.data.heatmap2$conc_char)) {
  pheatmap(eff.data.heatmap2[eff.data.heatmap2$conc_char == i,] %>%
                       arrange(target) %>%
                       column_to_rownames('drug') %>%
                       dplyr::select(contains('IPR')) %>%
                       t(),
           border_color = "white",
           cellwidth = 12,
           cellheight = 12,
           breaks = myBreaks,
           cluster_rows = F,
           cluster_cols = F,
           na_col = "white",
           main = paste('efficiency change, concentration =', i),
           annotation_col = target,
           annotation_row = chromatin,
           color = colorRampPalette(c("#e07a5f", "#e5e5e5", "#3d405b"))(palette_length),
           annotation_colors = annotation_colors_scr, annotation_legend = T)
}
```

*Figure 2D: Heatmap of cutting efficiency change induced by HDAC inhibitors at 19 genomic locations.*

**Conclusion: Many different HDAC inhibitors can change cutting efficiency locally. As expected, heterochromatin is more prone to an increase in cutting efficiency.**



## SA: Correlation of cutting efficiency between replicates of significant hits
We know that the cutting efficiency does not correlate well if all data is taken. However, because of the high noise level and the low effect size of the efficiency changes, it makes more sense to only investigate whether significant hits are robust throughout all replicates. 
```{r warning = FALSE}
# SA: Correlation between technical replicates efficiency: only significant hits
# significant_hits_efficiency <- eff.data %>%
#   filter(pval_adj < 0.05) %>%
#   distinct(drug, conc_char, IPR) %>%
#   mutate(drug_conc = paste(drug, conc_char, IPR, paste = "_"))

# techn_repl <- indel.data %>%
#   mutate(drug_conc = paste(drug, conc_char, IPR, paste = "_")) %>%
#   #filter(drug_conc %in% significant_hits_efficiency$drug_conc) %>%
#   dplyr::distinct(freqCut_norm, drug, tech, replicate, conc_char, sample, IPR) %>%
#   mutate(freqCut_norm = log2(ave(freqCut_norm, drug, tech, replicate, conc_char, IPR, FUN = mean))) %>%
#   unique() %>%
#   mutate(rep = paste(replicate, tech, sep = "_")) %>%
#   dplyr::select(-replicate, -tech) %>%
#   mutate(max = ave(freqCut_norm, drug, conc_char, IPR, FUN = max),
#          min = ave(freqCut_norm, drug, conc_char, IPR, FUN = min)) %>%
#   distinct() %>%
#   spread(rep, freqCut_norm) %>%
#   na.omit() 
#   
# techn_repl$large_effect <- "1_no"
# techn_repl$large_effect[techn_repl$max > 0.3 | techn_repl$min < -0.3] <- "0_yes"
# 
# ggpairs(techn_repl %>% 
#        #filter(conc_char == "1 µM") %>% #1 µM
#        # filter(drug %in% c("Vorinostat (SAHA, MK0683)", "Scriptaid",   "Decitabine" , "(+)-JQ1" ,  "PCI-24781 (Abexinostat)")) %>%
#           dplyr::select(contains("E1")),
#        mapping = aes(color = techn_repl$large_effect),
#                upper = list(continuous = corColor),
#                lower = list(continuous = function(data, mapping, ...) {
#                    ggally_points(data = data, mapping = mapping, alpha = 0.5, size = 0.2) +
#                    geom_abline(slope = 1, lty = "dashed", col = "red")}),
#                diag = list(continuous = function(data, mapping, ...) {
#                    ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
#   # xlim(.4, .7) + 
#   xlab("Cutting efficiency") +
#   ylab("Cutting efficiency") +
#   scale_color_manual(values = colors_diverse)
```

*Figure S2A: Correlation between cutting efficiency of significant hits.*

**Conclusion: The correlation between 'strong' changers of efficiency is OK, not very good, but OK.**



# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

