---
title: "DSB epigenetic drug screen - Figure 1"
author: 
  - name: "Max Trauernicht & Ruben Schep"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
  pdf_document:
    toc: yes
    toc_depth: '4'
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```
### Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 


library(knitr)
library(tidyverse)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(GGally)
library(pheatmap)
library(RColorBrewer)
library(ggrastr)
library(cowplot)
library(ggpubr)
library(scales)

## Select outdir
out.dir = paste0("./figures/", Date, "/")
dir.create(out.dir)
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

### Functions

```{r load_functions, results = 'hide', warning = FALSE}
p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
    ## Methods 'Hommel', 'BH', 'BY' and speed improvements
    ## contributed by Gordon Smyth
    method <- match.arg(method)
    if(method == "fdr") method <- "BH"	# back compatibility
    nm <- names(p)
    p <- as.numeric(p)
    p0 <- setNames(p, nm)
    if(all(nna <- !is.na(p))) nna <- TRUE
    p <- p[nna]
    lp <- length(p)
    # stopifnot(n >= lp)
    if (n <= 1) return(p0)
    if (n == 2 && method == "hommel") method <- "hochberg"

    p0[nna] <-
	switch(method,
	       bonferroni = pmin(1, n * p),
	       holm = {
		   i <- seq_len(lp)
		   o <- order(p)
		   ro <- order(o)
		   pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
	       },
	       hommel = { ## needs n-1 >= 2 in for() below
		   if(n > lp) p <- c(p, rep.int(1, n-lp))
		   i <- seq_len(n)
		   o <- order(p)
		   p <- p[o]
		   ro <- order(o)
		   q <- pa <- rep.int( min(n*p/i), n)
		   for (j in (n-1):2) {
		       ij <- seq_len(n-j+1)
		       i2 <- (n-j+2):n
		       q1 <- min(j*p[i2]/(2:j))
		       q[ij] <- pmin(j*p[ij], q1)
		       q[i2] <- q[n-j+1]
		       pa <- pmax(pa,q)
		   }
		   pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
	       },
	       hochberg = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
	       },
	       BH = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( n / i * p[o] ))[ro]
	       },
	       BY = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   q <- sum(1L/(1L:n))
		   pmin(1, cummin(q * n / i * p[o]))[ro]
	       },
	       none = p)
    p0
}

serialNext = function(prefix, extension){
    i=0
    repeat {
       f = paste(prefix, i , extension, sep=".")
       if(!file.exists(f)){return(f)}
       i=i+1
     }
  }

SetFileName <- function(filename, initials, extension) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  filename <- paste0(out.dir, "/", initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename <- serialNext(filename, extension)
  filename
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}


theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", 
                                                  size = 0.25),
                  strip.background = element_rect(fill = "white", 
                                                  color = "white")
            )
}
theme_classic_lines_45 <- function() {
  theme_pubr(border = T, 
             legend = "top", 
             x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", 
                                                  size = 0.25),
                  strip.background = element_rect(fill = "white", 
                                                  color = "white")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, 
             legend = "top", 
             x.text.angle = 90) +
    theme(strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}

theme_classic_lines_90_noborder <- function() {
  theme_pubr(border = F, 
             legend = "top", 
             x.text.angle = 90) +
    theme(strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}

<<<<<<< HEAD
theme_classic_lines_90_noborder <- function() {
  theme_pubr(border = F, legend = "top", x.text.angle = 90) +
            theme(strip.background = element_rect(fill = "white", color = "white")
            )
}

=======

theme_classic_nolines_90 <- function() {
  theme_pubr(border = T, 
             legend = "top", 
             x.text.angle = 90,) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank(),
          strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}


>>>>>>> b55fd7ec8e6377114d8f6ec3b0b181ca26dec7ea
theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
# colors_drugs <- c("DMSO (neg control)" = "#65BFA4", "DNA-PKi (NHEJ control)" = "#8CA0C4", "drugs" = "#808184", "Mirin (MMEJ control)" = "#5D79AC")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 
## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")
colors_pathways = c("wt" = "#808184", "other" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "SSTR" = "#007A4C")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

changeLevels <- function(object, column = "chr_r", levels = c(paste0("chr", 1:22), "chrX")) {
  object[, column] <- as.character(object[, column])
  object[, column] <- factor(object[, column], levels)
  object
}
```


### Data import

```{r data import, results = 'hide', warning = FALSE}
# Import data from preprocessing script
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
file = list.files(in.dir, pattern = "ratios", full.names = T) %>% tail(n = 1)
indel.data <- readRDS(file)

# Import data from preprocessing script
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
file.muts = list.files(in.dir, pattern = "mutations", full.names = T) %>% tail(n = 1)

IPR_set = c("IPR5", "IPR7")

mutation.data <- readRDS(file.muts) %>% filter(IPR %in% IPR_set)

# Complete data without exlcluding wells <0.5 viability (I need this for plotting)
meta.viab.file = list.files(in.dir, full.names = T)  %>% .[grepl("Viab_Meta", .)] %>% tail(n = 1)
indel.statistics.dt = readRDS(meta.viab.file) %>%
  dplyr::select(ID, drug, target, replicate, viability, conc_char, tech)

chromatin <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(indel.data, barcode))

map_clones = read_table("/DATA/projects/DSBrepair/data/cl20220321_cured_mapping_clones.txt", col_names = c("chr", "start", "barcode", "clone")) %>%
  filter(chr != "chrY" & clone == "RSTP2_5") %>%
  mutate(chr = factor(chr, levels = c("chr1", "chr2", "chr3", "chr4", "chr5", 
                                      "chr6", "chr7", "chr8", "chr9", "chr10", 
                                      "chr11", "chr12", "chr13", "chr14", "chr15", 
                                      "chr16", "chr17", "chr18", "chr19", "chr20", 
                                      "chr21", "chr22", "chrX")),
         start = start / 1e6) %>%
  left_join(chromatin) %>%
  filter(barcode != "ACCCCTAAAGGCGCTG")

chromsizes = read_table("/DATA/projects/DSBrepair/data/rs20220602_chromsizes_hg38_UCSC.txt", col_names = T) %>% 
  filter(chr != "chrY") %>%
  mutate(chr = factor(chr, levels = c("chr1", "chr2", "chr3", "chr4", "chr5", 
                                      "chr6", "chr7", "chr8", "chr9", "chr10", 
                                      "chr11", "chr12", "chr13", "chr14", "chr15", 
                                      "chr16", "chr17", "chr18", "chr19", "chr20", 
                                      "chr21", "chr22", "chrX")),
         start = start / 1e6)

centromeres = read_table("/DATA/projects/DSBrepair/git/EpiScreen/rs20220714_UCSC_centromeres_24501022.txt") %>%
  group_by(chr) %>%
  filter(chr != "chrY") %>%
  mutate(chr = factor(chr, levels = c("chr1", "chr2", "chr3", "chr4", "chr5", 
                                      "chr6", "chr7", "chr8", "chr9", "chr10", 
                                      "chr11", "chr12", "chr13", "chr14", "chr15", 
                                      "chr16", "chr17", "chr18", "chr19", "chr20", 
                                      "chr21", "chr22", "chrX")),
         start = start / 1e6,
         end = end / 1e6) %>%
  summarise(start = min(start), # earliest start of centromere
            end = max(end)) # furthest end of centromere

```

---

## A: Schematic

![Scheme of the experimental setup](figures/rs20220707_figure1A-01.png)

***Figure 1C:*** *Experimental procedure. Clonal K562 cell line with 18 integrated pathway reporters (IPRs) is transfected with a gRNA targeting the reporter and treated with a 180 compounds at three concentrations (100 nM, 1µM and 10µM) and incubated for 72 hours. Cell viability is then measured per well and the sequence of the reporters are then amplified for high throughput sequencing. The sequencing then reveals the indel frequency and pathway balance at the 18 genomic sites per compound and concentration.*

### Figure Legend


## B: Drugs per target group
Which drugs did we use for the screen and which targets do they have?
```{r 1B drug targets, warning = FALSE, fig.height=4, fig.width=4}
# B: Drugs per target group
## Plot from viability data table (because there all drugs are included)
targets <- indel.statistics.dt %>% 
  ungroup() %>%
  filter(!drug %in% c("Mirin", "DNA-PKi", "DMSO", "PAO")) %>%
  dplyr::distinct(drug, target) %>% 
  count(target, sort = TRUE)

ggplot(targets, 
       aes(y = reorder(target, n), x = n)) + 
  geom_bar(stat = "identity", fill = "black")  + 
  ylab("") +
  theme_pubr(border = T) 
```

***Figure 1B:*** *Number of probed compounds per target group. HDAC = Histone Deacetylase, JAK = Janus Kinase, PARP = Poly ADP Ribose Polymerase, HMT = Histone methyltransferase, PMT = Protein Methyltransferase, DNMT = DNA Methyltransferase,  HAT = Histone Acetyltransferase, HIF = Hypoxia Inducible Factor*

**Conclusion: Most drugs are HDACs, a lot of compounds do not really target chromatin directly.**

---
```{r FigS1A chromideograms, fig.height=5, fig.width=6}
p1 = ggplot() +
  geom_segment(data = chromsizes,
               aes(x = chr, xend = chr, y = 0, yend = start),
               lineend = "round", color = "lightgrey", size = 7) +
  geom_segment(data = map_clones,
               aes(x = as.integer(chr) - 0.3, xend = as.integer(chr) + 0.3,
                   y = start, yend = start, color = chromatin),
               size = 3) +
  geom_segment(data = centromeres,
               aes(x = as.integer(chr) - 0.2, xend = as.integer(chr) + 0.2,
                   y = start, yend = start),
               size = 2) +
  geom_segment(data = centromeres,
               aes(x = as.integer(chr) - 0.2, xend = as.integer(chr) + 0.2,
                   y = end, yend = end),
               size = 2) +
  theme_classic_lines_90_noborder() +
  scale_color_manual(values = group_colors) + 
  xlab("Chromosome") + 
<<<<<<< HEAD
  ylab("Position [Mb]")
=======
  ylab("Position [Mb]") +
  coord_flip()

label_ratio <- filter(indel.data, IPR %in% IPR_set , drug == "DMSO") %>% 
  group_by(IPR) %>%
  summarise(MMEJratio_mean = mean(MMEJratio),
            MMEJratio_sd = sd(MMEJratio),
            freqCut_mean = mean(freqCut),
            freqCut_sd = sd(freqCut)) %>%
  mutate(label_efficiency = paste0("Total indel frequency = ", round(freqCut_mean*100, 1),"% ±", round(freqCut_sd*100, 1),"%"), 
         label_ratio = paste0("MMEJ ratio = ", round(MMEJratio_mean, 2)," ±", round(MMEJratio_sd, 2)))

bc_mut_tib_mean <- mutation.data %>% 
  distinct(IPR, drug, freq_mean,freq_sd, color, mutation, type) %>%
  filter(IPR %in% IPR_set , drug == "DMSO", type %in% c("core", "grouped_indel")) %>% 
  left_join(label_ratio, by = "IPR")

p2 = ggplot(bc_mut_tib_mean, 
       aes(mutation, freq_mean)) + 
  geom_bar(stat = "identity", aes(fill = color)) + 
  # geom_quasirandom(data = bc_mut_tib, aes(mutation, freq_split, fill = color),  color = "black", shape = 21, ) + 
  scale_fill_manual("legend", values = colors_pathways) + 
  scale_color_manual("legend", values = colors_pathways) + 
  scale_y_continuous(name="frequency", labels = percent) +
  scale_x_discrete(name="indel size", breaks = c("<-14", -7, 0, 1, ">2")) + 
  facet_wrap( ~ IPR, ncol = 1) +
  geom_text(aes(x = 7, y = .7, label = label_efficiency)) +
  geom_text(aes(x = 7, y = .6, label = label_ratio)) + 
  geom_errorbar(aes(ymin=freq_mean-freq_sd, ymax=freq_mean+freq_sd), width=.2,
                position=position_dodge(.9)) +
  theme_classic_lines()

plot_grid(p1, p2, nrow = 1, rel_widths = c(3, 2))
>>>>>>> b55fd7ec8e6377114d8f6ec3b0b181ca26dec7ea
```

## C: Viability filtering
How many drugs are filtered out because they are too toxic?
```{r 1C Viability filter, warning = FALSE, fig.height = 4, fig.width = 4}
# C: Viability filtering
indel.data %>% 
  filter(sample == "Epigenetic Compound") %>%
  mutate(viab = ave(viab_mean, conc_char, drug, FUN = function(x) mean(x, na.rm = TRUE)),
         filter_viab = ifelse(pathway_bal_filter, "Balance filter", 
                              ifelse(indel_freq_filter,"TIF filter" , "Discarded")),
         filter_viab = factor(filter_viab, levels = c("Discarded", "TIF filter" , "Balance filter"))) %>% # average the two replicates for plotting
  distinct(conc_char, drug, viab, filter_viab) %>%
  ggplot_custom(.,
              aes(x = conc_char, y = viab)) +
    geom_quasirandom(aes(fill = filter_viab), color = "black", shape = 21, size = 2) +
    theme_pubr(border = T) +
  xlab("Concentration") +
  ylab("Relative viability to control (DMSO)")+ 
  scale_fill_manual(values = hcl.colors(4, palette = "Lajolla", rev = F)[2:4])
  
```

***Figure 1C:*** *Mean viability for each compound at the three concentrations.Compounds that decreased viability by >50% in more than one technical replicate in one or both biological replicate were removed (marked in yellow).*

**Conclusion: The higher the concentration, the more toxic the compounds are to the cells.**

---
<<<<<<< HEAD
```{r 1D Example BCs indel pattern, warning = FALSE, fig.height=4, fig.width=3}
label_ratio <- filter(indel.data, IPR %in% c("IPR5", "IPR7"), drug == "DMSO") %>% 
  group_by(IPR) %>%
  summarise(MMEJratio_mean = mean(MMEJratio),
            MMEJratio_sd = sd(MMEJratio),
            freqCut_mean = mean(freqCut),
            freqCut_sd = sd(freqCut)) %>%
  mutate(label_efficiency = paste0("Total indel frequency = ", round(freqCut_mean*100, 1),"% ±", round(freqCut_sd*100, 1),"%"), 
         label_ratio = paste0("MMEJ ratio = ", round(MMEJratio_mean, 2)," ±", round(MMEJratio_sd, 2)))

bc_mut_tib_mean <- mutation.data %>% 
  distinct(IPR, drug, freq_mean,freq_sd, color, mutation, type) %>%
  filter(IPR %in% c("IPR5", "IPR7"), drug == "DMSO", type %in% c("core", "grouped_indel")) %>% 
  left_join(label_ratio, by = "IPR")

ggplot(bc_mut_tib_mean, 
       aes(mutation, freq_mean)) + 
  geom_bar(stat = "identity", aes(fill = color)) + 
  # geom_quasirandom(data = bc_mut_tib, aes(mutation, freq_split, fill = color),  color = "black", shape = 21, ) + 
  scale_fill_manual("legend", values = colors_pathways) + 
  scale_color_manual("legend", values = colors_pathways) + 
  theme_bw() +
  scale_y_continuous(name="frequency", labels = percent) +
  scale_x_discrete(name="indel size", breaks = c("<-14", -7, 0, 1, ">2")) + 
  facet_wrap( ~ IPR, ncol = 1) +
  geom_text(aes(x = 7, y = .7, label = label_efficiency)) +
  geom_text(aes(x = 7, y = .6, label = label_ratio)) + 
  geom_errorbar(aes(ymin=freq_mean-freq_sd, ymax=freq_mean+freq_sd), width=.2,
                position=position_dodge(.9)) 
```
=======
>>>>>>> b55fd7ec8e6377114d8f6ec3b0b181ca26dec7ea

## D: Cutting efficiency per IPR
How do the drugs impact cutting efficiency in general? Does the genomic location & concentration of the drugs matter?
```{r 1DE TIF, warning = FALSE, fig.height=5, fig.width=3} 
chromatin_tib <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin)

ordered_marks <- c("Euchromatin", 
                     "other-heterochromatin",  
                     "H3K27me3", 
                     "Triple Heterochromatin")

chromatin_tib  = chromatin_tib %>% mutate(chromatin = factor(chromatin, levels = ordered_marks))

conc_control_colors = c("CTRL" = colors_drugs[2], "100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")

# D: Cutting efficiency per IPR
CTRL.data = indel.data %>%
 filter(sample == "DMSO (neg control)" & indel_freq_filter == T & IPR %in% IPR_set ) %>% 
  distinct(freqCut, IPR, drug, conc_char, tech, sample, replicate, 
           barcode, MMEJratio, MMEJ_zscore_comb, MMEJ_zscore_mean, mean_eff_zscore_comb,  plate, well) %>%
  mutate(sample_conc = "CTRL",
         freqCut = ave(freqCut, well, plate, conc_char, IPR, FUN = function(x) mean(x, na.rm = TRUE)),
         MMEJratio = ave(MMEJratio, well, plate, conc_char, IPR, FUN = function(x) mean(x, na.rm = TRUE))) %>% 
  left_join(chromatin_tib) %>% 
  select(-well, -tech)

indel.data2 <- indel.data %>%
 filter(sample %in% c("Epigenetic Compound") & indel_freq_filter == T & IPR %in% IPR_set ) %>%
  distinct(freqCut, IPR, drug, conc_char, tech, sample, replicate, 
           barcode, MMEJratio, MMEJ_zscore_comb, mean_eff_zscore_comb) %>%
  left_join(chromatin_tib) %>%
  mutate(freqCut = ave(freqCut, IPR, drug, conc_char, tech, replicate), 
         MMEJratio = ave(MMEJratio, IPR, drug, conc_char, tech, replicate),
         sample_conc = as.character(conc_char),
         IPR = factor(IPR, levels = IPR_set))  %>%
  distinct()
         
sample_order <- c("CTRL", "100 nM", "1 µM", "10 µM")
# indel.data2 <- indel.data2 %>%
#   mutate(IPR = gsub("IPR", "", IPR))
#ipr_order <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19")


plt1 = indel.data2 %>%
  mutate(freqCut = ave(freqCut, drug, conc_char, IPR, FUN = function(x) mean(x, na.rm = TRUE))) %>%
  bind_rows(CTRL.data) %>%
  mutate(sample_conc = factor(sample_conc, levels = sample_order)) %>%
  distinct(freqCut, drug, conc_char, IPR, chromatin, sample_conc) %>% 
  ggplot(.,
         aes(x = sample_conc, y = freqCut*100)) +
  geom_quasirandom_rast(aes(color = sample_conc), dodge.width = 0.75, raster.dpi = 300, size = 1) +
  scale_color_manual(values = conc_control_colors, guide = "none") +
  facet_wrap(~ IPR, nrow = 2) +
  ylim(0,100)  +
  theme_classic_nolines_90() + 
  ylab("Total indel frequency [TIF]")

plt2 =  indel.data2 %>%
  distinct(mean_eff_zscore_comb, drug, conc_char, IPR, chromatin, sample_conc) %>%
  mutate(sample_conc = factor(sample_conc, levels = sample_order),
         IPR = factor(IPR, levels = )) %>%
  ggplot(.,
         aes(x = sample_conc, y = mean_eff_zscore_comb)) +
  geom_quasirandom_rast(aes(color = sample_conc), dodge.width = 0.75, raster.dpi = 300, size = 1) +
<<<<<<< HEAD
  scale_color_manual(values = conc_colors, guide= "none") +
  facet_wrap(~chromatin, scales = "free_x", nrow = 1) +
=======
  scale_color_manual(values = conc_colors, guide = "none") +
  facet_wrap(~ IPR, nrow = 2) +
>>>>>>> b55fd7ec8e6377114d8f6ec3b0b181ca26dec7ea
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96) +
  theme_classic_nolines_90() + 
  ylab("TIF z-score")

plot_grid(plt1, plt2, labels = c("D", "E"), ncol = 2)
```

*Figure 1D: Compound-induced changes in cutting efficiency on the 19 genomic locations. The three probed concentrations are plotted next to each other.*

**Conclusion: Cutting efficiency changes can go both directions, either up or down. The concentration of the compounds does not matter that much.**

---

## E: MMEJ score per IPR
How do the drugs impact MMEJ score in general? Does the genomic location & concentration of the drugs matter?
```{r 1FG MMEJ ratio, warning = FALSE, fig.height=5, fig.width=3}
# E: MMEJ score per IPR
plt1 = 
  indel.data2 %>%
  mutate(MMEJratio = ave(MMEJratio, drug, conc_char, IPR, FUN = function(x) mean(x, na.rm = TRUE))) %>%
  bind_rows(CTRL.data)   %>%
    mutate(sample_conc = factor(sample_conc, levels = sample_order)) %>%
  distinct(MMEJratio , drug, conc_char, IPR, chromatin, sample_conc) %>%
  ggplot(., aes(x = sample_conc, y = MMEJratio)) +
  geom_quasirandom_rast(aes(color = sample_conc), dodge.width = 0.75, raster.dpi = 600, size = 0.5) +
  scale_color_manual(values = conc_control_colors, guide= "none") +
  facet_wrap(~ IPR, nrow = 2) +
  #geom_hline(linetype = "dashed", yintercept = 1.96) +
  #geom_hline(linetype = "dashed", yintercept = -1.96)   +
  theme_classic_nolines_90()

plt2 = indel.data2 %>%
        distinct(MMEJ_zscore_comb, drug, conc_char, IPR, chromatin, sample_conc) %>%
  ggplot(.,
       aes(x = sample_conc, y = MMEJ_zscore_comb)) +
  geom_quasirandom_rast(aes(color = sample_conc), dodge.width = 0.75, raster.dpi = 600, size = 0.5) +
  scale_color_manual(values = conc_colors, guide= "none") +
  facet_wrap(~ IPR, nrow = 2) +
  geom_hline(linetype = "dashed", yintercept = 1.96) +
  geom_hline(linetype = "dashed", yintercept = -1.96) +
  theme_classic_nolines_90()

plot_grid(plt1, plt2, labels = c("F", "G"), ncol = 2)
```

*Figure 1E: Compound-induced changes in MMEJ score on the 19 genomic locations. The three probed concentrations are plotted next to each other.*

**Conclusion: MMEJ score is mostly reduced and not increased. Increasing the concentration amplifies this phenotype. IPRs with a high MMEJ score to begin with have a higher reduction of MMEJ score in general.**

---

## S1A: Clone 5 heatmap of chromatin values
What is the chromatin environment of each IPR?
```{r S1A clone 5 heatmap, warning = FALSE, fig.height=6, fig.width=8}
domain_dt <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin)

bc_IPR = indel.data %>% 
  distinct(IPR, barcode)

domain_matrix = bc_IPR %>% left_join(domain_dt) %>% select(IPR, chromatin) %>% column_to_rownames(., var = "IPR")

chrom_tib = indel.data %>% 
  select(IPR, barcode, contains(".zscore")) %>%
  distinct()

oldnames = colnames(chrom_tib)
newnames = gsub(".zscore", "", colnames(chrom_tib))

chrom_tib = chrom_tib %>% rename_at(all_of(oldnames), ~ newnames)

chrom_matrix = chrom_tib %>% select(-barcode) %>% column_to_rownames(., var = "IPR") %>% t()

## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 
'H3K9me2', 'H3K9me3',
'H3K27me3', 'EZH2', 
'CTCF', 'SMC3',
'HDAC3', 'HDAC2', 'HDAC1', 
'H3K4me1', 'H3K4me2', 'H3K4me3', 'H3K27ac',
'H4K5acK8ac', 'H2AFZ', 
'DNAse', 'Dam', 'm5C',
'H3K79me2', 
'TTseq', 
'H3K36me3', 'POL2AS2',
'POL2'),
group=factor(c(rep('repressive', 6), 
  # rep('repressive', 4), 
  rep('insulator', 2),
  rep('HDAC', 3), 
  rep('euchromatin', 6),
  rep('accessibility', 2), 'methylation',
  rep('transcribing', 5)
  # rep('transcribing', 4)
  ),
  levels=c('transcribing', 'accessibility',
           'methylation', 'euchromatin',
           'HDAC', 'insulator', 'repressive')))
feature_fill = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
                 transcribing="#FBB040", HDAC="#aac9e0", accessibility='#cebc85',
                 methylation='#7dc98f')

order_vec=rownames(clustering)[nrow(clustering):1]

### Colors
quant_breaks <- quantile(chrom_matrix, c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 100
max_quant = max(abs(quant_breaks[1]), abs(quant_breaks[2]))

breaks.chip <- c(seq(-quant_breaks[2], 0, length.out=ceiling(palette_length/2) + 1), 
                 seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))


colors = colorRampPalette(rev(brewer.pal(n = 11, name="RdBu")))(palette_length)

pheatmap(chrom_matrix, annotation_row=clustering,
         annotation_colors=list(chromatin=group_colors, group=feature_fill),
         color = colors, breaks = breaks.chip,
         annotation_col = domain_matrix)
```


## S1B: Correlate viability data from the 2 replicates
How well does the viability data correlate between the two biological replicates? Did the drugs become less potent after 3 years in the freezer?
```{r S1B viability correlation, warning = FALSE, fig.height=4, fig.width=8}
# A: Correlate viability data from the 2 replicates
indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  mutate(filter_viab = ifelse(pathway_bal_filter, "Balance filter", 
                              ifelse(indel_freq_filter,"TIF filter" , "Discarded")),
         filter_viab = factor(filter_viab, levels = c("Discarded", "TIF filter" , "Balance filter"))) %>%
  distinct(viab_mean, sample, replicate, conc_char, drug, viab_reproducibility,
           filter_viab, indel_freq_filter, pathway_bal_filter) %>%
  pivot_wider(names_from = replicate, values_from = viab_mean) %>%
  ggplot(., aes(x = E177, y = E1504, color = filter_viab,  shape = viab_reproducibility)) +
  geom_point(size = 3) +
  scale_shape_manual(values=c(18, 16))+
  geom_smooth(aes(group = 1), method = "lm", color = colors_diverse[1]) +
  # geom_segment(x = 0.5, xend = 0.5, y = 0.5, yend = 1.2, linetype = "dashed") +
  # geom_segment(x = 0.5, xend = 1.2, y = 0.5, yend = 0.5, linetype = "dashed") +
  theme_classic_lines_90() +
  scale_color_manual(values = hcl.colors(4, palette = "Lajolla", rev = F)[2:4]) +
  facet_grid(. ~ conc_char)

```

*Figure S1A: Correlation of the cell viability between the two biological replicates. Marked in yellow are the compounds that were removed from the analysis.*

**Conclusion: Most drugs that were toxic in rep 1 were also toxic in rep 2, so the trends are similar. However, some drugs were especially more toxic in rep 2. These drugs were removed from the analysis.**


## S1C: Correlation between viability and TIF
Is lower viability causing a decrease in TIF? Can we justify our 0.50 viability cutoff based on the correlation between viability and TIF?

```{r S1C TIF correlation with viability, warning=F, fig.height=5, fig.width=5}
# F: Correlation between viability and MMEJ ratio
# load indel data that has no filter on it.
indel.data.nofilter <- readRDS("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/mt20220126_episcreen.RDS")

viab_eff_cor <- indel.data.nofilter %>%
  # filter(sample == "Epigenetic Compound") %>%
  dplyr::distinct(viab_norm_split, freqCut, drug, replicate, conc_char, tech, sample) %>%
  mutate(freqCut = ave(freqCut, drug, replicate, conc_char, tech, FUN = mean)) %>%
  distinct()

ggplot_custom(viab_eff_cor,
              aes(x = viab_norm_split, y = freqCut, color = sample)) +
  geom_point() +
  geom_vline(xintercept = 0.25, color = "red") +
  annotate("rect",xmin=0,xmax=0.25,ymin=-Inf,ymax=Inf, alpha=0.1, fill="red") +
  ylim(0, 1) +
  xlim(0, 1.5)

indel.data  %>% 
  filter(is.na(freqCut)) %>% # freqCut is NA if sample didn't pass filter for viability.
  distinct(drug, conc_char) %>% # unique compound and concentration combination
  pull(conc_char) %>% 
  table() # count compounds per concentration that are removed
```

## S1D: Correlation between viability and MMEJ ratio
Is lower viability causing a decrease in MMEJ? Can we justify our 0.50 viability cutoff based on the correlation between viability and MMEJ ratio?

```{r S1D MMEJ ratio correlation with viability, warning=F, fig.height=5, fig.width=5}
viab_mmej_cor <- indel.data.nofilter %>%
  # filter(sample == "Epigenetic Compound") %>%
  dplyr::distinct(viab_norm_split, MMEJratio, drug, replicate, conc_char, tech, sample) %>%
  mutate(MMEJratio = ave(MMEJratio, drug, replicate, conc_char, tech, FUN = mean)) %>%
  distinct()

ggplot_custom(viab_mmej_cor,
              aes(x = viab_norm_split, y = log2(MMEJratio), color = sample)) +
  geom_point() +
  geom_vline(xintercept = 0.50, color = "red") +
  annotate("rect",xmin=0,xmax=0.50,ymin=-Inf,ymax=Inf, alpha=0.1, fill="red") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.25), limits = c(0.1, 1.2)) +
  xlab("Normalised viability") +
  ylab("MMEJ ratio (Log2)")
```

---

## S1E: Comparison between replicates of efficiency in DMSO

```{r S1E control distribution of indel rates across wells and replicates, warning = FALSE, fig.height=3, fig.width=5}
ggplot(indel.data %>%
         filter(drug == "DMSO") %>%
         dplyr::distinct(freqCut, IPR, exp, replicate),
       aes(x = IPR, y = freqCut)) +
  geom_quasirandom_rast(aes(color = replicate), dodge.width = 0.75, raster.dpi = 600) +
  scale_color_manual(values = c(colors_drugs[8], colors_drugs[7])) +
  theme_classic_lines_90() +
  ylim(0,1)
```


## S1F: Comparison between replicates MMEJ ratio in DMSO

```{r S1F control distribution of MMEJ ratio across wells and replicates, warning = FALSE, fig.height=3, fig.width=5}
ggplot(indel.data %>%
         filter(drug == "DMSO") %>%
         dplyr::distinct(MMEJratio, IPR, exp, replicate),
       aes(x = IPR, y = MMEJratio)) +
  geom_quasirandom_rast(aes(color = replicate), dodge.width = 0.75, raster.dpi = 600) +
  scale_color_manual(values = c(colors_drugs[8], colors_drugs[7])) +
  theme_classic_lines_90()
```

## S1G: Comparison between technical replicates of efficiency in DMSO

```{r S1G comparison for mean indel rate across replicates, warning = FALSE, fig.height=3, fig.width=2}
ggplot(indel.data %>%
         filter(drug == "DMSO") %>%
         dplyr::distinct(freqCut, well, drug_plate, conc_char, tech, replicate) %>%
         mutate(freqCut = ave(freqCut, well, drug_plate, conc_char, tech, replicate, FUN = mean)) %>%
         distinct() %>%
         mutate(rep = paste(replicate, "tech", tech, sep = "_")),
       aes(x = rep, y = freqCut)) +
  geom_quasirandom_rast(aes(color = replicate), dodge.width = 0.75, raster.dpi = 600) +
  scale_color_manual(values = c(colors_drugs[8], colors_drugs[7])) +
  theme_classic_lines_90() +
  ylim(0,1)
```


## S1H: Correlation between technical replicates MMEJ score

How well does the MMEJ score correlate between the six technical replicates?
```{r S1H normalized MMEJ ratio correlation, warning = FALSE, fig.height=5, fig.width=5}
# B: Correlation between technical replicates MMEJ score
techn_repl <- indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  dplyr::distinct(MMEJ_ratio_norm, drug, tech, replicate, conc_char, sample) %>%
  mutate(MMEJ_ratio_norm = log2(ave(MMEJ_ratio_norm, drug, tech, replicate, conc_char, FUN = mean))) %>%
  unique() %>%
  mutate(rep = paste(replicate, tech, sep = "_")) %>%
  dplyr::select(-replicate, -tech) %>%
  spread(rep, MMEJ_ratio_norm) %>%
  na.omit()

boundaries <- c(0.7,0.75,0.8,0.85)
ggpairs(techn_repl %>% dplyr::select(contains("E1")),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data, mapping = mapping, alpha = 0.6, size = 0.5) +
                   scale_y_continuous(breaks = seq(-2, 1, by = 1), limits = c(-2, 0.5)) +
                   scale_x_continuous(breaks = seq(-2, 1, by = 1), limits = c(-2, 0.5)) +
                   geom_abline(slope = 1, lty = "twodash", col = "grey") +
                   geom_smooth(method = "lm", color = "red", lty = "dashed", size = 0.5)}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("MMEJ ratio change (log2)") +
  ylab("MMEJ ratio change (log2)")
```

*Figure S1B: Correlation of the MMEJ score between all technical replicates.*

**Conclusion: The MMEJ scores correlate relatively well, especially within biological replicates. Even the correlation between biological replicates is good.**

---

## S1I: Correlation between technical replicates efficiency
How well does the cutting efficiency correlate between the six technical replicates
```{r S1I normalized TIF correlation, warning = FALSE, fig.height=5, fig.width=5}
# C: Correlation between technical replicates efficiency
techn_repl <- indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  dplyr::distinct(freqCut_norm, drug, conc_char, replicate, tech) %>%
  mutate(freqCut_norm = ave(freqCut_norm, drug, conc_char, replicate, tech, FUN = mean)) %>%
  distinct() %>%
  mutate(rep = paste(replicate, tech, sep = "_")) %>%
  dplyr::select(-replicate, -tech) %>%
  spread(rep, freqCut_norm) %>%
  na.omit()

ggpairs(techn_repl %>% dplyr::select(contains("E1")),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data, mapping = mapping, alpha = 0.6, size = 0.5) +
                   scale_y_continuous(breaks = seq(0.8, 1.2, by = 0.2), limits = c(0.7, 1.2)) +
                   scale_x_continuous(breaks = seq(0.8, 1.2, by = 0.2), limits = c(0.7, 1.2)) +
                   geom_abline(slope = 1, lty = "twodash", col = "grey") +
                   geom_smooth(method = "lm", color = "red", lty = "dashed", size = 0.5)}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("Indel rate change") +
  ylab("Indel rate change")
```

*Figure S1C: Correlation of the cutting efficiency between all technical replicates.*

**Conclusion: Only for replicate two the cutting efficiency correlates well. We need to further investigate why replicate 1 does not correlate well.**

---

## S1H: Correlation in MMEJ z-score between biological replicates
How well does the MMEJ z-score correlate between the biologcial replicates before combining them?
```{r S1H MMEJ z-score correlation, warning = FALSE, fig.height=5, fig.width=5}
# D: Correlation between biological replicates MMEJ score
biol_repl <- indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  dplyr::distinct(MMEJ_zscore_global_mean, drug, replicate, conc_char, sample) %>%
  spread(replicate, MMEJ_zscore_global_mean) %>%
  na.omit()

ggplot(biol_repl,
       aes(x = E1504, y = E177)) +
  geom_point() +
  geom_abline(lty = "twodash") +
  geom_smooth(method = "lm", color = "red", lty = "dashed") +
  stat_cor(method = "pearson") +
  xlab("Screen 2 MMEJ ratio z-score") +
  ylab("Screen 1 MMEJ ratio z-score")
```



*Figure S1D: Correlation of the MMEJ z-scores between the two biological replicates.*

Conclusion: MMEJ z-scores correlate well.


## S1I: Correlation in cutting efficiencies between biological replicates
How well does the cutting efficiency correlate between the biologcial replicates?
```{r S1I TIF z-score correlation, warning = FALSE, fig.height=5, fig.width=5}
# E: Correlation between biological replicates efficiency
biol_repl <- indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  dplyr::distinct(mean_eff_zscore_global, drug, replicate, conc_char, sample) %>%
  spread(replicate, mean_eff_zscore_global) %>%
  na.omit()

ggplot(biol_repl,
       aes(x = E1504, y = E177)) +
  geom_point() +
  geom_abline(lty = "twodash") +
  geom_smooth(method = "lm", color = "red", lty = "dashed") +
  stat_cor(method = "pearson") +
  xlab("Screen 2 TIF z-score") +
  ylab("Screen 1 TIF z-score")
```

*Figure S1E: Correlation of the cutting efficiencies between the two biological replicates.*

**Conclusion: Cutting efficiencies do not correlate well.**

```{r export}
# setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/")
# filename <- SetFileName("_indel.data", "mt")
# save(indel.data, file = filename)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

