---
title: "DSB Episcreen data analysis & plotting"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```
### Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(dplyr)
library(outliers)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(ggrepel)
library(GGally)
library(DESeq2)
library(gghighlight)
library(platetools)
library(ggpubr)
library(tidyr)
library(Laurae)
library(tibble)
library(pheatmap)
library(RColorBrewer)
library(ggrastr)
```

### Functions

```{r load_functions, echo = FALSE, warning = FALSE}

p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
    ## Methods 'Hommel', 'BH', 'BY' and speed improvements
    ## contributed by Gordon Smyth
    method <- match.arg(method)
    if(method == "fdr") method <- "BH"	# back compatibility
    nm <- names(p)
    p <- as.numeric(p)
    p0 <- setNames(p, nm)
    if(all(nna <- !is.na(p))) nna <- TRUE
    p <- p[nna]
    lp <- length(p)
    # stopifnot(n >= lp)
    if (n <= 1) return(p0)
    if (n == 2 && method == "hommel") method <- "hochberg"

    p0[nna] <-
	switch(method,
	       bonferroni = pmin(1, n * p),
	       holm = {
		   i <- seq_len(lp)
		   o <- order(p)
		   ro <- order(o)
		   pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
	       },
	       hommel = { ## needs n-1 >= 2 in for() below
		   if(n > lp) p <- c(p, rep.int(1, n-lp))
		   i <- seq_len(n)
		   o <- order(p)
		   p <- p[o]
		   ro <- order(o)
		   q <- pa <- rep.int( min(n*p/i), n)
		   for (j in (n-1):2) {
		       ij <- seq_len(n-j+1)
		       i2 <- (n-j+2):n
		       q1 <- min(j*p[i2]/(2:j))
		       q[ij] <- pmin(j*p[ij], q1)
		       q[i2] <- q[n-j+1]
		       pa <- pmax(pa,q)
		   }
		   pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
	       },
	       hochberg = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
	       },
	       BH = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( n / i * p[o] ))[ro]
	       },
	       BY = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   q <- sum(1L/(1L:n))
		   pmin(1, cummin(q * n / i * p[o]))[ro]
	       },
	       none = p)
    p0
}

SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}


# stat_smooth_func <- function(mapping = NULL, data = NULL,
#                         geom = "smooth", position = "identity",
#                         ...,
#                         method = "auto",
#                         formula = y ~ x,
#                         se = TRUE,
#                         n = 80,
#                         span = 0.75,
#                         fullrange = FALSE,
#                         level = 0.95,
#                         method.args = list(),
#                         na.rm = FALSE,
#                         show.legend = NA,
#                         inherit.aes = TRUE,
#                         xpos = NULL,
#                         ypos = NULL) {
#   layer(
#     data = data,
#     mapping = mapping,
#     stat = StatSmoothFunc,
#     geom = geom,
#     position = position,
#     show.legend = show.legend,
#     inherit.aes = inherit.aes,
#     params = list(
#       method = method,
#       formula = formula,
#       se = se,
#       n = n,
#       fullrange = fullrange,
#       level = level,
#       na.rm = na.rm,
#       method.args = method.args,
#       span = span,
#       xpos = xpos,
#       ypos = ypos,
#       ...
#     )
#   )
# }
# 
# StatSmoothFunc <- ggproto("StatSmooth", Stat,
#                       
#                       setup_params = function(data, params) {
#                         # Figure out what type of smoothing to do: loess for small datasets,
#                         # gam with a cubic regression basis for large data
#                         # This is based on the size of the _largest_ group.
#                         if (identical(params$method, "auto")) {
#                           max_group <- max(table(data$group))
#                           
#                           if (max_group < 1000) {
#                             params$method <- "loess"
#                           } else {
#                             params$method <- "gam"
#                             params$formula <- y ~ s(x, bs = "cs")
#                           }
#                         }
#                         if (identical(params$method, "gam")) {
#                           params$method <- mgcv::gam
#                         }
#                         
#                         params
#                       },
#                       
#                       compute_group = function(data, scales, method = "auto", formula = y~x,
#                                                se = TRUE, n = 80, span = 0.75, fullrange = FALSE,
#                                                xseq = NULL, level = 0.95, method.args = list(),
#                                                na.rm = FALSE, xpos=NULL, ypos=NULL) {
#                         if (length(unique(data$x)) < 2) {
#                           # Not enough data to perform fit
#                           return(data.frame())
#                         }
#                         
#                         if (is.null(data$weight)) data$weight <- 1
#                         
#                         if (is.null(xseq)) {
#                           if (is.integer(data$x)) {
#                             if (fullrange) {
#                               xseq <- scales$x$dimension()
#                             } else {
#                               xseq <- sort(unique(data$x))
#                             }
#                           } else {
#                             if (fullrange) {
#                               range <- scales$x$dimension()
#                             } else {
#                               range <- range(data$x, na.rm = TRUE)
#                             }
#                             xseq <- seq(range[1], range[2], length.out = n)
#                           }
#                         }
#                         # Special case span because it's the most commonly used model argument
#                         if (identical(method, "loess")) {
#                           method.args$span <- span
#                         }
#                         
#                         if (is.character(method)) method <- match.fun(method)
#                         
#                         base.args <- list(quote(formula), data = quote(data), weights = quote(weight))
#                         model <- do.call(method, c(base.args, method.args))
#                         
#                         m = model
#                         eq <- substitute(italic(y) == a + b %.% italic(x)*","~~italic(r)^2~"="~r2, 
#                                          list(a = format(coef(m)[1], digits = 3), 
#                                               b = format(coef(m)[2], digits = 3), 
#                                               r2 = format(summary(m)$r.squared, digits = 3)))
#                         func_string = as.character(as.expression(eq))
#                         
#                         if(is.null(xpos)) xpos = min(data$x)*0.9
#                         if(is.null(ypos)) ypos = max(data$y)*0.9
#                         data.frame(x=xpos, y=ypos, label=func_string)
#                         
#                       },
#                       
#                       required_aes = c("x", "y")
# )

theme_classic_lines <- function() {
  theme_pubr(border = F, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}
theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

```


### Data import

```{r data import, echo = FALSE, warning = FALSE}
# Import data from preprocessing script
#indel.data <- readRDS("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/mt20220105_episcreen.RDS")
indel.data <- readRDS("/DATA/projects/DSBrepair/data/R/rs20220117_episcreen/rs20220117_episcreen_ratios.RDS")

# Complete data without exlcluding wells <0.45 viability (I need this for plotting)
viability.data = as.data.table(readRDS("/DATA/projects/DSBrepair/git/EpiScreen/files_scripts/rs20211221_Episcreen_Reads_Viab_Metadata.RDS")) %>%
  dplyr::select(ID, drug, target, replicate, viability, conc_char, tech)
```

---

## B: Drugs per target group
Which drugs did we use for the screen and which targets do they have?
```{r}
# B: Drugs per target group
## Plot from viability data table (because there all drugs are included)
targets <- viability.data %>% 
  filter(!drug %in% c("Mirin", "DNA-PKi", "DMSO", "PAO")) %>%
  dplyr::distinct(drug, target) %>% 
  mutate(number = ave(target, target, FUN = function(x) length(x))) %>% 
  mutate(number = as.numeric(number)) %>%
  dplyr::distinct(target, number)

ggplot_custom(targets, aes(y = reorder(target, +number), x = number)) + 
  geom_bar(stat = "identity")  + 
  ylab("")
```
*Figure 1B: Number of probed compounds per target group.*

Conclusion: Most drugs are HDACs, a lot of compounds do not really target chromatin directly.  

---

## C: Viability filtering
How many drugs are filtered out because they are too toxic?
```{r}
# C: Viability filtering
## Calculate normalized viability (same as in script 1)
viability.data$ID <- gsub("^E[0-9]{3,4}_", "", viability.data$ID)
viability.data <- viability.data %>%
  group_by(replicate) %>%
  mutate(viability = viability - min(viability)) %>%
  mutate(ctrl_mean = mean(viability[drug == "DMSO"]),
         viab_norm = viability / ctrl_mean) %>%
  dplyr::select(-viability, -ctrl_mean) %>%
  unique() 

remove_drugs <- viability.data %>%
  filter(viab_norm < 0.45) %>%
  dplyr::distinct(conc_char, drug, tech, replicate) %>%
  mutate(conc_char_drug = paste(conc_char, drug, sep = " "),
         tech_n = ave(conc_char_drug, conc_char_drug, replicate, FUN = length)) %>%
  filter(tech_n > 1) %>%
  ungroup() %>%
  dplyr::select(-tech_n, -tech, -replicate) %>%
  unique() %>%
  mutate(removed = "removed") %>%
  dplyr::select(-conc_char_drug)

viability.data <- merge(viability.data, remove_drugs, all = T)
viability.data$removed[is.na(viability.data$removed)] <- "kept"

viability.data <- viability.data %>%
  spread(key = replicate, value = viab_norm) %>%
  mutate(mean_viab = (E1504+E177) / 2)

viability.data$group <- "drug"
viability.data$group[viability.data$drug == "DMSO"] <- "DMSO"
viability.data$group[viability.data$drug == "PAO"] <- "PAO"

## Viability mean across the two biological replicates - color highlight removed drugs
ggplot_custom(viability.data %>%
                filter(group == "drug") %>%
                mutate(mean_viab = ave(mean_viab, drug, conc_char, FUN = mean)) %>%
                dplyr::distinct(mean_viab, drug, conc_char, removed),
       aes(x = conc_char, y = mean_viab, color = removed)) +
  geom_quasirandom() +
  scale_color_manual(values = c(colors_diverse[1], colors_diverse[4]))
```
*Figure 1C: Filtering of toxic compounds. Compounds that decreased viability by >55% in more than one technical replicate in one or both biological replicate were removed (marked in yellow).*

Conclusion: The higher the concentration, the more toxic the compounds are to the cells.  

---

## D: Cutting efficiency per IPR
How do the drugs impact cutting efficiency in general? Does the genomic location & concentration of the drugs matter?
```{r}
# D: Cutting efficiency per IPR
indel.data2 <- indel.data

indel.data2$sample[indel.data2$sample == "drug" & indel.data2$concentration == 100] <- "drug_100nM"
indel.data2$sample[indel.data2$sample == "drug" & indel.data2$concentration == 10] <- "drug_10uM"
indel.data2$sample[indel.data2$sample == "drug" & indel.data2$concentration == 1] <- "drug_1uM"
sample_order <- c("DMSO (neg control)" , "DNA-PKi (NHEJ control)", "Mirin (MMEJ control)", "drug_100nM", "drug_1uM", "drug_10uM")
indel.data2$sample <- factor(indel.data2$sample, levels = sample_order)
indel.data2 <- indel.data2 %>%
  mutate(IPR = gsub("IPR", "", IPR))
ipr_order <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19")
indel.data2$IPR <- factor(indel.data2$IPR, levels = ipr_order)


ggplot(indel.data2 %>%
         filter(sample != "DNA-PKi (NHEJ control)", sample != "Mirin (MMEJ control)") %>%
         dplyr::distinct(freqCut, IPR, tech, conc_char, drug, sample) %>%
           mutate(freqCut = ave(freqCut, IPR, tech, conc_char, drug, FUN = mean)),
       aes(x = IPR, y = freqCut)) +
  geom_violin(aes(fill = sample), position = position_dodge(0.75), draw_quantiles = 0.5, width = 3) +
  scale_fill_manual(values = c(colors_drugs[2], colors_drugs[6], colors_drugs[7], colors_drugs[4]))+
  coord_polar()+
  ylim(0,1)

ggplot(indel.data2 %>%
         filter(sample != "DNA-PKi (NHEJ control)", sample != "Mirin (MMEJ control)", replicate == "E1504") %>%
         dplyr::distinct(freqCut, IPR, tech, conc_char, drug, sample) %>%
           mutate(freqCut = ave(freqCut, IPR, tech, conc_char, drug, FUN = mean)),
       aes(x = IPR, y = freqCut)) +
  geom_quasirandom_rast(aes(color = sample), dodge.width = 0.75, raster.dpi = 600) +
  scale_color_manual(values = c(colors_drugs[2], colors_drugs[7], colors_drugs[4], "#000000")) +
  coord_polar()+
  ylim(0,1)
```
*Figure 1D: Compound-induced changes in cutting efficiency on the 19 genomic locations. The three probed concentrations are plotted next to each other.*

Conclusion: Cutting efficiency changes can go both directions, either up or down. The concentration of the compounds does not matter that much. 

---

## E: MMEJ score per IPR
How do the drugs impact MMEJ score in general? Does the genomic location & concentration of the drugs matter?
```{r}
# E: MMEJ score per IPR
ggplot(indel.data2 %>%
         filter(sample != "DNA-PKi (NHEJ control)", sample != "Mirin (MMEJ control)") %>%
         dplyr::distinct(MMEJscore, IPR, tech, conc_char, drug, sample) %>%
           mutate(MMEJscore = ave(MMEJscore, IPR, tech, conc_char, drug, FUN = mean)),
       aes(x = IPR, y = MMEJscore)) +
  geom_quasirandom_rast(aes(color = sample), dodge.width = 0.75, raster.dpi = 600) +
  scale_color_manual(values = c(colors_drugs[2], colors_drugs[7], colors_drugs[4], "#000000")) +
  coord_polar() +
  ylim(-0.4,0.6)
```
*Figure 1E: Compound-induced changes in MMEJ score on the 19 genomic locations. The three probed concentrations are plotted next to each other.*

Conclusion: MMEJ score is mostly reduced and not increased. Increasing the concentration amplifies this phenotype. IPRs with a high MMEJ score to begin with have a higher reduction of MMEJ score in general. 

---

## S_A: Correlate viability data from the 2 replicates
How well does the viability data correlate between the two biological replicates? Did the drugs become less potent after 3 years in the freezer?
```{r}
# A: Correlate viability data from the 2 replicates
ggplot_custom(viability.data %>%
                filter(group == "drug"), aes(x = E177, y = E1504)) +
  geom_point(aes(color = removed), size = 1) +
  geom_smooth(method = "lm", color = colors_diverse[1]) +
  geom_segment(x = 0.5, xend = 0.5, y = 0.5, yend = 1.2, linetype = "dashed") +
  geom_segment(x = 0.5, xend = 1.2, y = 0.5, yend = 0.5, linetype = "dashed") +
  scale_color_manual(values = c(colors_diverse[1], colors_diverse[4])) +
  facet_wrap(~conc_char)
```
*Figure S1A: Correlation of the cell viability between the two biological replicates. Marked in yellow are the compounds that were removed from the analysis.*

Conclusion: Most drugs that were toxic in rep 1 were also toxic in rep 2, so the trends are similar. However, some drugs were especially more toxic in rep 2. These drugs were removed from the analysis. 

---

## S_B: Correlation between technical replicates MMEJ score
How well does the MMEJ score correlate between the six technical replicates?
```{r}
# B: Correlation between technical replicates MMEJ score
techn_repl <- indel.data %>%
  filter(sample == "drug") %>%
  dplyr::distinct(MMEJscore, drug, tech, replicate, conc_char, sample) %>%
  mutate(MMEJscore = ave(MMEJscore, drug, tech, replicate, conc_char, FUN = mean)) %>%
  unique() %>%
  mutate(rep = paste(replicate, tech, sep = "_")) %>%
  dplyr::select(-replicate, -tech) %>%
  spread(rep, MMEJscore) %>%
  na.omit()

boundaries <- c(0.7,0.75,0.8,0.85)
ggpairs(techn_repl %>% dplyr::select(contains("E1")),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data, mapping = mapping, alpha = 0.6, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red")}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("MMEJ score") +
  ylab("MMEJ score")
```
*Figure S1B: Correlation of the MMEJ score between all technical replicates.*

Conclusion: The MMEJ scores correlate relatively well, especially within biological replicates. Even the correlation between biological replicates is good.

---

## S_C: Correlation between technical replicates efficiency
How well does the cutting efficiency correlate between the six technical replicates
```{r}
# C: Correlation between technical replicates efficiency
techn_repl <- indel.data %>%
  filter(sample == "drug" | drug == "DMSO") %>%
  dplyr::distinct(freqCut, drug, tech, replicate, conc_char, sample) %>%
  mutate(freqCut = ave(freqCut, drug, tech, replicate, conc_char, FUN = mean)) %>%
  unique() %>%
  mutate(rep = paste(replicate, tech, sep = "_")) %>%
  dplyr::select(-replicate, -tech) %>%
  spread(rep, freqCut) %>%
  na.omit()

ggpairs(techn_repl %>% dplyr::select(contains("E1")),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data, mapping = mapping, alpha = 0.6, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red")}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("Cutting efficiency") +
  ylab("Cutting efficiency")
```
*Figure S1C: Correlation of the cutting efficiency between all technical replicates.*

Conclusion: Only for replicate two the cutting efficiency correlates well. We need to further investigate why replicate 1 does not correlate well.

---

## S_D: Correlation in MMEJ z-score between biological replicates
How well does the MMEJ z-score correlate between the biologcial replicates before combining them?
```{r}
# D: Correlation between biological replicates MMEJ score
biol_repl <- indel.data %>%
  filter(sample == "drug") %>%
  dplyr::distinct(MMEJ_zscore_drug, drug, replicate, conc_char, sample) %>%
  spread(replicate, MMEJ_zscore_drug) %>%
  na.omit()

ggpairs(biol_repl %>% dplyr::select(contains("E1")),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.6, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red")}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("MMEJ score") +
  ylab("MMEJ score")
```
*Figure S1D: Correlation of the MMEJ z-scores between the two biological replicates.*

Conclusion: MMEJ z-scores correlate well.

## S_E: Correlation in cutting efficiencies between biological replicates
How well does the cutting efficiency correlate between the biologcial replicates?
```{r}
# E: Correlation between biological replicates efficiency
biol_repl <- indel.data %>%
  filter(sample == "drug") %>%
  dplyr::distinct(freqCut, drug, replicate, conc_char, sample) %>%
  mutate(freqCut = ave(freqCut, drug, conc_char, replicate, FUN = mean)) %>%
  unique() %>%
  spread(replicate, freqCut) %>%
  na.omit()

ggpairs(biol_repl %>% dplyr::select(contains("E1")),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.6, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red")}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("Cutting efficiency") +
  ylab("Cutting efficiency")
```
*Figure S1E: Correlation of the cutting efficiencies between the two biological replicates.*

Conclusion: Cutting efficiencies do not correlate well.


```{r export}
setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/")
filename <- SetFileName("_indel.data", "mt")
save(indel.data, file = filename)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

