---
title: "DSB epigenetic drug screen - Figure 4"
author:
  - name: "Max Trauernicht & Ruben Schep"
    email: "m.trauernicht@nki.nl; r.schep@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output:
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
  pdf_document:
    toc: yes
    toc_depth: '4'
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
$(document).ready(function() {
$items = $('div#TOC li');
$items.each(function(idx) {
num_ul = $(this).parentsUntil('#TOC').length;
$(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
});
});
</script>

```{css, echo = FALSE}
div.sourceCode {
overflow-x: hidden;
}
```
### Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(tidyverse)
library(outliers)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(ggrepel)
library(GGally)
library(gghighlight)
library(ggpubr)
library(pheatmap)
library(RColorBrewer)
library(broom)
library(cowplot)
library(rstatix)
library(colorspace)
library(Hmisc)
library(pls)
library(MASS)
library(corrr)
library(ggstatsplot)
library(ggpmisc)


## Select outdir
out.dir = paste0("./figures/", Date, "/")
dir.create(out.dir)
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
knitr::opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))

pdf.options(useDingbats = FALSE)
```

### Functions

```{r functions, message=FALSE, warnings=FALSE}
p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
  ## Methods 'Hommel', 'holm', 'BY' and speed improvements
  ## contributed by Gordon Smyth
  method <- match.arg(method)
  if(method == "fdr") method <- "BH"	# back compatibility
  nm <- names(p)
  p <- as.numeric(p)
  p0 <- setNames(p, nm)
  if(all(nna <- !is.na(p))) nna <- TRUE
  p <- p[nna]
  lp <- length(p)
  # stopifnot(n >= lp)
  if (n <= 1) return(p0)
  if (n == 2 && method == "hommel") method <- "hochberg"
  
  p0[nna] <-
    switch(method,
           bonferroni = pmin(1, n * p),
           holm = {
             i <- seq_len(lp)
             o <- order(p)
             ro <- order(o)
             pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
           },
           hommel = { ## needs n-1 >= 2 in for() below
             if(n > lp) p <- c(p, rep.int(1, n-lp))
             i <- seq_len(n)
             o <- order(p)
             p <- p[o]
             ro <- order(o)
             q <- pa <- rep.int( min(n*p/i), n)
             for (j in (n-1):2) {
               ij <- seq_len(n-j+1)
               i2 <- (n-j+2):n
               q1 <- min(j*p[i2]/(2:j))
               q[ij] <- pmin(j*p[ij], q1)
               q[i2] <- q[n-j+1]
               pa <- pmax(pa,q)
             }
             pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
           },
           hochberg = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
           },
           BH = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             pmin(1, cummin( n / i * p[o] ))[ro]
           },
           BY = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             q <- sum(1L/(1L:n))
             pmin(1, cummin(q * n / i * p[o]))[ro]
           },
           none = p)
  p0
}

# From Fede:
# ggpairs custom functions
theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
    theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
          strip.background = element_rect(fill = "white", color = "white")
    )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
# colors_drugs <- c("DMSO (neg control)" = "#65BFA4", "DNA-PKi (NHEJ control)" = "#8CA0C4", "drugs" = "#808184", "Mirin (MMEJ control)" = "#5D79AC")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 
## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

colors_targets <-  c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
                     HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
                     HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
                     PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43")

conc_control_colors = c("CTRL" = colors_drugs[2], "100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

```

---

# Data import

## Screen and validation data

#### Import
```{r data import, echo = FALSE, warning = FALSE}
# Import data from preprocessing script
file = list.files(in.dir, pattern = "episcreen_ratios", full.names = T) %>% tail(n = 1)
indel.data <- readRDS(file) # %>% filter(pathway_bal_filter == TRUE & viab_reproducibility == TRUE)
```

---

#### Heatmap annotation
```{r heatmap annotation}
# Order IPRs by their starting freqCut value
dmso_freq <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(IPR, freqCut) %>%
  mutate(freqCut = ave(freqCut, IPR, FUN = function(x) mean(x, na.rm = TRUE))*100) %>%
  dplyr::select('ctrl_TIF' = freqCut, IPR) %>%
  distinct() %>%
  arrange(desc(ctrl_TIF))

iprs <- dmso_freq$IPR
iprs <- as.character(iprs)

# Set annotation for heatmap
# viability
viab_drug = indel.data %>% filter(drug != "PAO") %>% 
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  distinct(viab_mean, drug_conc, replicate) %>% 
  group_by(drug_conc) %>%
  summarise(viab = mean(viab_mean))
# Row annotation: annotate drugs with the target group
target_set_drugs <- indel.data %>% 
  filter(drug != "PAO") %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>%
  left_join(viab_drug) %>%
  distinct(drug_conc, drug, conc_char, target, viab) %>% 
  column_to_rownames(var="drug_conc")


ctrl.vals <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(IPR, freqCut, MMEJratio) %>%
  mutate(ctrl_TIF = ave(freqCut, IPR, FUN = function(x) mean(x, na.rm = TRUE))*100,
         ctrl_balance = ave(MMEJratio, IPR, FUN = function(x) mean(x, na.rm = TRUE))*100) %>%
  distinct(ctrl_TIF, 
           ctrl_balance,
           IPR)

HDAC_tib = read.table("rs20220708_HDAC_targets.txt", sep = "\t", fill = NA, header = T) %>%
  dplyr::select(-target_main, -target_second) %>%
  as_tibble()

clone5.matrix = indel.data %>% 
  dplyr::select(IPR, ends_with(".zscore")) %>% 
  distinct() %>% 
  rename_with(., ~gsub(".zscore", "", .x)) %>%
  # left_join(ctrl.vals) %>%
  column_to_rownames(var="IPR")

clone5_map_ipr = hclust(dist(clone5.matrix), method = "ward.D") %>% 
  # Rotate the middle node to math better the TIF in CTRL conditions
  dendextend::rotate(c(1:6, 10:7, 11:18))

# Chromatin annotation
chromatin <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(indel.data, barcode, IPR)) %>%
  left_join(dmso_freq)

# chromatin$IPR[is.na(chromatin$IPR)] <- "IPR1"
chromatin <- chromatin %>%
  filter(IPR != "<NA>") %>% 
  dplyr::select(-barcode) %>%
  column_to_rownames(var="IPR")

## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")


viab_colors = sequential_hcl(11, "Peach", rev = TRUE)
names(viab_colors) = seq(0, 1, by = 0.1)

annotation_colors_scr = list(
  chromatin = group_colors,
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43"),
  conc_char = conc_colors,
  clusters_drugs = sequential_hcl(7, "Plasma"),
  # viab_rep1 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  # viab_rep2 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  viab = viab_colors,
  ctrl_TIF = colorRampPalette(c("white", "black"))(10))

# Order the annotation legend
ordered_marks <- c("Euchromatin", 
                   "H3K27me3", 
                   "other-heterochromatin",  
                   "Triple Heterochromatin")

annotation_colors_scr$chromatin <- annotation_colors_scr$chromatin[ordered_marks]


# First calculate means and sd of the DMSO efficiency per replicate and barcode
dmso.eff.score <- indel.data %>%
  filter(drug == "DMSO") %>%
  dplyr::select('DMSO' = freqCut, well, barcode, replicate, tech, plate) %>%
  mutate(DMSO = ave(DMSO, well, replicate, tech, plate, FUN = function(x) mean(x, na.rm=T))) %>%
  distinct(replicate, tech, well, plate, DMSO) %>% 
  pull(DMSO) %>%
  fitdistr(., "normal") %>% tidy()
```

#### processing screen data for heatmaps
```{r filter heatmap}
# Generate heatmap: select top drugs (highest change at any IPR) per concentration
eff.data <- indel.data %>%
  #filter(target != "Aurora Kinase") %>%
  filter(sample == "Epigenetic Compound", indel_freq_filter, viab_reproducibility, !(freqCut_diff_mean < -0.1 & viab_mean < 0.5)) %>%
  # filter(abs(zscore_TIF_global_comb_rep) >= 2.58) %>%  # we can choose to display all drugs that have significant global hits
  distinct(conc_char, drug, target, IPR, zscore_TIF_comb_rep, freqCut_diff_mean, replicate, tech, filter_zscore_TIF_comb_rep)

eff.data.heatmap <-  eff.data %>%
  group_by(drug, conc_char) %>%
  # ave to add IPR grouping level to calculate mean of the replicates
  mutate(freqCut_diff_mean = ave((freqCut_diff_mean*100), IPR, FUN = function(x) mean(x, na.rm = TRUE)),
         ## Select drugs that have a large effect size in at least one location 
         ## (some drugs might have a small global effect size, but large local)
         zscore_TIF_comb_rep_temp = ifelse(!filter_zscore_TIF_comb_rep, 0, zscore_TIF_comb_rep)) %>%
  ungroup() %>%
  mutate(p_value = 2*pnorm(q = abs(zscore_TIF_comb_rep_temp), lower.tail = F),
         # Adjust p value for false discovery rate
         adj_pvalue = p.adjust(p_value, method = 'BH')) %>%
  # Pick the smallest per compound & concentration
  group_by(drug, conc_char) %>%
  mutate(min_score = min(adj_pvalue)) %>%
  # filter for adjusted p.val < 0.01 so we keep compounds with at least one signif effect.
  filter(min_score < 0.01) %>%
  distinct(drug, target, conc_char, freqCut_diff_mean, IPR, min_score) %>%
  ungroup()

#eff.data.heatmap2$freqCut_norm_mean[eff.data.heatmap2$mean_eff_zscore_comb > 0.05] <- NA
eff.data.heatmap2 <- eff.data.heatmap %>%
  dplyr::select(-min_score) %>%
  spread(IPR, freqCut_diff_mean) %>%
  filter(!is.nan(IPR3))

eff.data.heatmap2 <- eff.data.heatmap2[,c("drug", "target", "conc_char", iprs)]
eff.data.heatmap2$drug <- gsub(" [(].*[)]", "", eff.data.heatmap2$drug) #shorten name of drug
```

#### HDAC only heatmap
```{r HDAC heatmap processing}
HDACs = indel.data %>% 
  distinct(drug, conc_char, target) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  filter(target == "HDAC" & drug != "Nullscript") %>% 
  pull(drug_conc)

target.HDAC <- target_set_drugs %>%
  rownames_to_column("drug_conc") %>%
  left_join(HDAC_tib) %>%
  filter(drug_conc %in% HDACs) %>%
  replace(is.na(.), 0)

rownames(target.HDAC) <- target.HDAC$drug_conc
target.HDAC <- target.HDAC %>%
  dplyr::select(pan_HDAC, HDAC1,HDAC2, HDAC3,
                HDAC4, HDAC6, HDAC8, HDAC10, viab)

```

# CCD Data Screen

###Load data
```{r}
file = list.files(in.dir, pattern = "CCD_table", full.names = T) %>% tail(n = 1)
CCD_tib <- readRDS(file)
file = list.files(in.dir, pattern = "IPR_annotation_heatmap", full.names = T) %>% tail(n = 1)
chromatin_TIF <- readRDS(file)
file = list.files(in.dir, pattern = "compound_annotation_heatmap", full.names = T) %>% tail(n = 1)
target <- readRDS(file)
file = list.files(in.dir, pattern = "chrom_cluster_heatmap", full.names = T) %>% tail(n = 1)
TRIP_chrom_clust <- readRDS(file)
file = list.files(in.dir, pattern = "FC_ratios", full.names = T) %>% tail(n = 1)
ratios_tib_fc <- readRDS(file)
```

### Annotation

```{r heatmap annotation CCD}
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
conc_control_colors = c("control" = colors_drugs[2], "100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")

group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

viab_colors = sequential_hcl(11, "Peach", rev = TRUE)
names(viab_colors) = as.numeric(seq(0, 1, by = 0.1))

# Annotation colors heatmaps
annotation_colors_scr = list(
  chromatin = group_colors,
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43", 
             `Negative Control` = "#65BFA4", `MRN` = "#5D79AC"),
  conc_char = conc_colors,
  conc_char_ctrl = conc_control_colors,
  clusters_drugs = sequential_hcl(7, "Plasma"),
  # viab_rep1 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  # viab_rep2 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  viab = viab_colors,
  ctrl_TIF = colorRampPalette(c("white", "black"))(20),
  MMEJratio = c('#e1251b', '#d32c2d', '#c5313c', '#b6354a', '#a73857', '#963a65', '#843d72', '#6e3e7f', '#53408d', '#26419a'),
  group = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
            transcribing="#FBB040", HDAC="#aac9e0", accessibility='#cebc85',
            methylation='#7dc98f'))


ordered_marks2 <- c("Negative Control", "MRN", "DNA-PK","Epigenetic Reader Domain", "HDAC",  "Sirtuin", "HAT", 
                    "Histone Demethylase", "HMT", "DNMT", 
                    "HIF", "JAK", "PIM",
                    "Aurora Kinase", "PARP")

#Epistatic interaction colors
slope.colors <- tibble(color = c("grey60", "#f5f5f5","#84a98c"),label = c("negative","none","positive"), feature = "epistasis")

annotation_colors_scr$target <- annotation_colors_scr$target[ordered_marks2]

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2', 'H3K9me3', 'H3K27me3', 'EZH2', 
                                    'CTCF', 'SMC3',
                                    'HDAC3', 
                                    'H3K4me1', 'H3K4me2', 'H3K4me3', 'H3K27ac', 'H4K5acK8ac', 'H2AFZ', 
                                    'DNAse', 'Dam', 
                                    'm5C',
                                    'H3K79me2', 'TTseq',  'H3K36me3', 'POL2AS2','POL2'),
                        group=factor(c(rep('repressive', 6), 
                                       rep('insulator', 2),
                                       rep('HDAC', 1), 
                                       rep('euchromatin', 6),
                                       rep('accessibility', 2), 
                                       'methylation',
                                       rep('transcribing', 5)),
                                     levels=c('transcribing', 'accessibility',
                                              'methylation', 'euchromatin',
                                              'HDAC', 'insulator', 'repressive')))
```


### Synergies Heatmaps

```{r Synergies chromatin prep, fig.width= 10, fig.height= 20}
all_compounds_epistasis_TIF = CCD_tib %>%
  dplyr::select(drug_conc, indelrate_slope_plot, feature) %>%
  group_by(drug_conc) %>%
  mutate(onlyNA = sum(is.na(indelrate_slope_plot))) %>% 
  filter(onlyNA != 25) %>%
  reshape2::dcast(drug_conc ~ feature, value.var = "indelrate_slope_plot", fill = 0) %>% 
  column_to_rownames(var = "drug_conc")

target_ctrl = dplyr::select(target, -conc_char)

quant_breaks <- quantile(all_compounds_epistasis_TIF, c(.01,.99), na.rm = T)
palette_length <- 100
breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
            seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
colors = rev(diverge_hcl(length(breaks), palette = "Cork"))
colors[which(breaks == 0)-1] = "white"
```

---

# Main figures
## 4A HDAC CCD Heatmap
```{r 4A HDAC CCD, fig.width= 10, fig.height= 10}
# Make matrix with drugs from selected target
TIF_synergy_matrix = all_compounds_epistasis_TIF %>% 
  rownames_to_column(var = "drug_conc") %>%
  filter(drug_conc %in% HDACs) %>%
  column_to_rownames(var = "drug_conc")

cluster_rows = ifelse(nrow(TIF_synergy_matrix) > 1, TRUE, FALSE)

# # change these numbers according to where you want to set the cutoff
# quant_breaks_ <- quantile(TIF_synergy_matrix, c(.01,.99), na.rm = T) 
# max_quant = max(abs(quant_breaks))
# palette_length <- 100
# breaks <- c(seq(-max_quant, 0, length.out=ceiling(palette_length/2) + 1), 
#             seq(max_quant/palette_length, max_quant, length.out=floor(palette_length/2)))
# slope.colors.temp = slope.colors
# colors = colorRampPalette(slope.colors.temp %>% pull(color))(length(breaks))
# colors[which(quant_breaks == 0)-1] = "gray90"

p = pheatmap(TIF_synergy_matrix[, TRIP_chrom_clust$labels],
             border_color = F,
             cellwidth = 8,
             cellheight = 8,
             cluster_rows = cluster_rows,
             breaks = breaks,
             cluster_cols = TRIP_chrom_clust,
             annotation_row = target.HDAC,
             clustering_distance_rows = "correlation",
             main = paste("Chromatin synergies - TIF HDAC cluster"),
             color = colors,
             annotation_colors = annotation_colors_scr, 
             annotation_legend = T)

TIF_synergy_cluster <- TIF_synergy_matrix %>%
  rownames_to_column("drug_conc")
```


## 4B K27me3 top CCDs
```{r 4B H3K27me3 synergies, fig.width = 8, fig.height = 8}
rank_dt = CCD_tib %>% 
  left_join(rownames_to_column(target, var = "drug_conc")) %>%
  filter(feature == "H3K27me3", !is.na(indelrate_slope_plot)) %>%
  distinct() %>%
  arrange(desc(indelrate_slope_plot)) %>%
  mutate(rank = seq(1:n()))

maxrank = nrow(rank_dt)

quant_breaks <- quantile(all_compounds_epistasis_TIF, c(.01,.99), na.rm = T)
palette_length <- 100
breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
            seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
colors[which(quant_breaks == 0)-1] = "gray90"

max_slope = max(abs(rank_dt$indelrate_slope_plot))

p2 = rank_dt %>% 
  ggplot(., aes(x = rank, fill = target)) +
  geom_bar() + 
  scale_x_reverse() +
  coord_flip() +
  xlab("Rank") +
  theme_classic2(base_size = 16) +
  scale_fill_manual(values = colors_targets) +
  # geom_label(aes(label = ifelse(width == 2, drug_conc,''))) +
  theme(legend.position = "none",
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  # add horizontal line for 0.01 slope cutoff which is highlighted in the heatmap
  geom_vline(xintercept = rank_dt[rank_dt$indelrate_slope_plot  < -0.01, ]$rank[1] - 0.5)  +
  geom_vline(xintercept = max(rank_dt[rank_dt$indelrate_slope_plot  > 0.01, ]$rank) + 0.5)


p3 = rank_dt %>% 
  filter(!between(rank, 11, maxrank-10)) %>%
  mutate(rank = seq(1:n())) %>% 
  ggplot(., aes(x = rank, y = indelrate_slope_plot, fill = target)) +
  geom_bar(stat = "identity") + 
  scale_x_reverse() +
  coord_flip() +
  ylab("Synergy score") +
  theme_classic2(base_size = 16) +
  scale_fill_manual(values = colors_targets) +
  # geom_label(aes(label = ifelse(width == 2, drug_conc,''))) +
  geom_text(aes(label=drug_conc)) +
  theme(legend.position = "none",
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())


plot_grid(# p1, #comment out the slope per rank
  p2, p3, labels = c("E", "F"), rel_widths = c(1, 7), nrow = 1)

# Samples to highlight in the heatmap S3C
all_compounds_epistasis_TIF %>% 
  filter(!between(H3K27me3, -0.01, .01)) %>%
  rownames()
```




## 4C m5C top CCDs
```{r 4C m5C synergies, fig.width = 8, fig.height = 4}
rank_dt = CCD_tib %>% 
  left_join(rownames_to_column(target, var = "drug_conc")) %>%
  filter(feature == "m5C", !is.na(indelrate_slope_plot)) %>%
  distinct() %>%
  arrange(desc(indelrate_slope_plot)) %>%
  mutate(rank = seq(1:n()))

maxrank = nrow(rank_dt)

max_slope = max(abs(rank_dt$indelrate_slope_plot))

p2 = rank_dt %>% 
  ggplot(., aes(x = rank, fill = target)) +
  geom_bar() + 
  scale_x_reverse() +
  coord_flip() +
  xlab("Rank") +
  theme_classic2(base_size = 16) +
  scale_fill_manual(values = colors_targets) +
  # geom_label(aes(label = ifelse(width == 2, drug_conc,''))) +
  theme(legend.position = "none",
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  # add horizontal line for 0.01 slope cutoff which is highlighted in the heatmap
  geom_vline(xintercept = rank_dt[rank_dt$indelrate_slope_plot  < -0.01, ]$rank[1] - 0.5)  +
  geom_vline(xintercept = max(rank_dt[rank_dt$indelrate_slope_plot  > 0.01, ]$rank) + 0.5)

p3 = rank_dt %>% 
  filter(!between(rank, 6, maxrank-3)) %>%
  mutate(rank = seq(1:n())) %>% 
  ggplot(., aes(x = rank, y = indelrate_slope_plot, fill = target)) +
  geom_bar(stat = "identity") + 
  scale_x_reverse() +
  coord_flip() +
  ylab("Synergy score") +
  theme_classic2(base_size = 16) +
  scale_fill_manual(values = colors_targets) +
  # geom_label(aes(label = ifelse(width == 2, drug_conc,''))) +
  geom_text(aes(label=drug_conc)) +
  theme(legend.position = "none",
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())


plot_grid(# p1, # comment out the TIF slope 
  p2, p3, labels = c("E", "F"), rel_widths = c(1, 6), nrow = 1)

# Samples to highlight in the heatmap S3C
all_compounds_epistasis_TIF %>% 
  filter(!between(m5C, -0.01, .01)) %>%
  rownames()
```
# Supplementary Figures
## S4A HMT CCD Heatmap
```{r S4A HMT CCD, fig.width= 8, fig.height= 6}
DNMT_drugs = target_set_drugs %>% 
  rownames_to_column(var = "drug_conc") %>% 
  filter(target == "DNMT" &
           conc_char == "100 nM"
  ) %>% pull(drug_conc)

TIF_synergy_matrix_DNMT = CCD_tib %>%
  dplyr::select(drug_conc, indelrate_slope, feature) %>%
  group_by(drug_conc) %>%
  mutate(onlyNA = sum(is.na(indelrate_slope))) %>% 
  filter(onlyNA != 25 & drug_conc %in% DNMT_drugs) %>%
  reshape2::dcast(drug_conc ~ feature, value.var = "indelrate_slope", fill = 0) %>% 
  column_to_rownames(var = "drug_conc")

# Make matrix with drugs from selected target
cluster_rows = ifelse(nrow(TIF_synergy_matrix_DNMT) > 1, TRUE, FALSE)

# # change these numbers according to where you want to set the cutoff
quant_breaks <- quantile(TIF_synergy_matrix_DNMT, c(.01,.99), na.rm = T)
max_quant = max(abs(quant_breaks))
palette_length <- 100
breaks <- c(seq(-max_quant, 0, length.out=ceiling(palette_length/2) + 1),
            seq(max_quant/palette_length, max_quant, length.out=floor(palette_length/2)))
slope.colors.temp = slope.colors
colors =  rev(diverge_hcl(length(breaks), palette = "Cork"))
colors[which(breaks == 0)-1] = "white"

p = pheatmap(TIF_synergy_matrix_DNMT[, TRIP_chrom_clust$labels],
             border_color = F,
             cellwidth = 8,
             cellheight = 8,
             cluster_rows = cluster_rows,
             breaks = breaks,
             cluster_cols = TRIP_chrom_clust,
             clustering_distance_rows = "correlation",
             main = paste("Chromatin synergies - TIF DNMT 1 µM"),
             color = colors,
             annotation_colors = annotation_colors_scr, 
             annotation_legend = T)

DNMT_drugs = target_set_drugs %>% 
  rownames_to_column(var = "drug_conc") %>% 
  filter(target == "DNMT" &
           conc_char == "1 µM"
  ) %>% pull(drug_conc)

TIF_synergy_matrix_DNMT = CCD_tib %>%
  dplyr::select(drug_conc, indelrate_slope, feature) %>%
  group_by(drug_conc) %>%
  mutate(onlyNA = sum(is.na(indelrate_slope))) %>% 
  filter(onlyNA != 25 & drug_conc %in% DNMT_drugs) %>%
  reshape2::dcast(drug_conc ~ feature, value.var = "indelrate_slope", fill = 0) %>% 
  column_to_rownames(var = "drug_conc")

# Make matrix with drugs from selected target
cluster_rows = ifelse(nrow(TIF_synergy_matrix_DNMT) > 1, TRUE, FALSE)

# # change these numbers according to where you want to set the cutoff
quant_breaks <- quantile(TIF_synergy_matrix_DNMT, c(.01,.99), na.rm = T)
max_quant = max(abs(quant_breaks))
palette_length <- 100
breaks <- c(seq(-max_quant, 0, length.out=ceiling(palette_length/2) + 1),
            seq(max_quant/palette_length, max_quant, length.out=floor(palette_length/2)))
slope.colors.temp = slope.colors
colors =  rev(diverge_hcl(length(breaks), palette = "Cork"))
colors[which(breaks == 0)-1] = "white"

p = pheatmap(TIF_synergy_matrix_DNMT[, TRIP_chrom_clust$labels],
             border_color = F,
             cellwidth = 8,
             cellheight = 8,
             cluster_rows = cluster_rows,
             breaks = breaks,
             cluster_cols = TRIP_chrom_clust,
             clustering_distance_rows = "correlation",
             main = paste("Chromatin synergies - TIF DNMT 1 µM"),
             color = colors,
             annotation_colors = annotation_colors_scr, 
             annotation_legend = T)

DNMT_drugs = target_set_drugs %>% 
  rownames_to_column(var = "drug_conc") %>% 
  filter(target == "DNMT" &
           conc_char == "10 µM") %>% pull(drug_conc)

TIF_synergy_matrix_DNMT = CCD_tib %>%
  dplyr::select(drug_conc, indelrate_slope, feature) %>%
  group_by(drug_conc) %>%
  mutate(onlyNA = sum(is.na(indelrate_slope))) %>% 
  filter(onlyNA != 25 & drug_conc %in% DNMT_drugs) %>%
  reshape2::dcast(drug_conc ~ feature, value.var = "indelrate_slope", fill = 0) %>% 
  column_to_rownames(var = "drug_conc")

# Make matrix with drugs from selected target
cluster_rows = ifelse(nrow(TIF_synergy_matrix_DNMT) > 1, TRUE, FALSE)

# # change these numbers according to where you want to set the cutoff
quant_breaks <- quantile(TIF_synergy_matrix_DNMT, c(.01,.99), na.rm = T)
max_quant = max(abs(quant_breaks))
palette_length <- 100
breaks <- c(seq(-max_quant, 0, length.out=ceiling(palette_length/2) + 1),
            seq(max_quant/palette_length, max_quant, length.out=floor(palette_length/2)))
slope.colors.temp = slope.colors
colors = rev(diverge_hcl(length(breaks), palette = "Cork"))
colors[which(breaks == 0)-1] = "white"

p = pheatmap(TIF_synergy_matrix_DNMT[, TRIP_chrom_clust$labels],
             border_color = F,
             cellwidth = 8,
             cellheight = 8,
             cluster_rows = cluster_rows,
             breaks = breaks,
             cluster_cols = TRIP_chrom_clust,
             clustering_distance_rows = "correlation",
             main = paste("Chromatin synergies - TIF DNMT 10 µM"),
             color = colors,
             annotation_colors = annotation_colors_scr, 
             annotation_legend = T)
```

## S4B Scatter plot HMT inhibotors 5mC
```{r S4B HMT inhibitor synergy with 5mC,  fig.width = 10, fig.height = 6}
DNMT_drugs = target_set_drugs %>% 
  rownames_to_column(var = "drug_conc") %>% 
  filter(target == "DNMT") %>% pull(drug_conc)

ratios_tib_fc %>% 
  filter(drug_conc %in% DNMT_drugs)  %>% 
  ggplot(.,aes(LMNB1, log2.fc.freqCut)) + geom_point() +
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "pearson") +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*")),
               label.x = "left", label.y = "bottom") +
  facet_grid(drug ~ conc_char) +
  geom_vline(xintercept = 0, linetype = 2) + 
  geom_hline(yintercept = 0, linetype = 2) + 
  theme_classic(base_size = 16) 

ratios_tib_fc %>% 
  filter(drug_conc %in% DNMT_drugs)  %>% 
  ggplot(.,aes(m5C, log2.fc.freqCut)) + geom_point() +
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "pearson") +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*")),
               label.x = "left", label.y = "bottom") +
  facet_grid(drug ~ conc_char) +
  geom_vline(xintercept = 0, linetype = 2) + 
  geom_hline(yintercept = 0, linetype = 2) + 
  theme_classic(base_size = 16) 
```


## S4B Scatter plot EPZ & Vorniostat in K27me3
```{r S4C Bossing in H3K27me3 specifically,  fig.width = 6, fig.height = 4}
ratios_tib_fc %>% 
  filter(drug_conc %in% c("Vorinostat 1 µM", "EPZ-6438 10 µM"))  %>% 
  ggplot(.,aes(H3K27me3, log2.fc.freqCut)) + geom_point() +
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "pearson") +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*")),
               label.x = "left", label.y = "bottom") +
  geom_vline(xintercept = 0, linetype = 2) + 
  geom_hline(yintercept = 0, linetype = 2) + 
  facet_wrap(. ~ drug_conc) +
  theme_classic(base_size = 16) 
```


```{r eval=FALSE}
indel.data %>% filter(!between(zscore_TIF_comb_rep, -3.2905, 3.2905) & conc_char != "control" & viab_reproducibility & cor_TIF_filter & filter_zscore_TIF_comb_rep) %>% 
  # distinct(IPR, drug, replicate, target, tech, conc_char, zscore_TIF, zscore_TIF_comb_tech, zscore_TIF_comb_rep) %>% 
  distinct(IPR, drug, conc_char, target, zscore_TIF_comb_rep) %>% 
  ggplot(., aes(IPR, zscore_TIF_comb_rep, color = target)) + 
  geom_quasirandom() + 
  facet_grid(. ~ conc_char) + 
  scale_color_manual(values = colors_targets) +
  theme_classic2(base_size = 16)
```


# Top hits and heatmaps for all features and targets
```{r top CCDs for all chromatin features, fig.width= 6, fig.height= 6, eval = FALSE}
for(i in unique(CCD_tib$feature)) {
  rank_dt = CCD_tib %>% 
    left_join(rownames_to_column(target, var = "drug_conc"), by = "drug_conc") %>%
    filter(feature == i, !is.na(indelrate_slope_plot)) %>%
    distinct() %>%
    arrange(desc(indelrate_slope_plot))
  
  if(nrow(rank_dt) > 1) {
    rank_dt %<>%
      mutate(rank = seq(1:n()))
    
    maxrank = nrow(rank_dt)
    
    p1 = rank_dt %>% 
      ggplot(., aes(x = rank, fill = target)) +
      geom_bar() + 
      scale_x_reverse() +
      coord_flip() +
      xlab("Rank") +
      theme_classic2(base_size = 16) +
      scale_fill_manual(values = colors_targets) +
      # geom_label(aes(label = ifelse(width == 2, drug_conc,''))) +
      theme(legend.position = "none",
            axis.title.x=element_blank(),
            axis.text.x=element_blank(),
            axis.ticks.x=element_blank())+
      # add horizontal line for 0.01 slope cutoff which is highlighted in the heatmap
      geom_vline(xintercept = rank_dt[rank_dt$indelrate_slope_plot  < -0.01, ]$rank[1] - 0.5)  +
      geom_vline(xintercept = max(rank_dt[rank_dt$indelrate_slope_plot  > 0.01, ]$rank) + 0.5)
    
    
    p2 = rank_dt %>% 
      filter(!between(rank, 16, maxrank-15)) %>%
      mutate(rank = seq(1:n())) %>% 
      ggplot(., aes(x = rank, y = indelrate_slope_plot, fill = target)) +
      geom_bar(stat = "identity") + 
      scale_x_reverse() +
      coord_flip() +
      ylab("Synergy score") +
      theme_classic2(base_size = 16) +
      scale_fill_manual(values = colors_targets) +
      # geom_label(aes(label = ifelse(width == 2, drug_conc,''))) +
      geom_text(aes(label=drug_conc)) +
      theme(legend.position = "none",
            axis.title.y=element_blank(),
            axis.text.y=element_blank(),
            axis.ticks.y=element_blank()) +
      ggtitle(paste0("Top synergies with: ", i))
    
    
    p = plot_grid(p1, p2, labels = c("E", "F"), rel_widths = c(1, 5))
    print(p)
  }
}
```

```{r heatmaps per compound target, eval = FALSE}
target_set_target = dplyr::rename(target_set_drugs, conc_char_ctrl = conc_char) %>% dplyr::select(-drug)

quant_breaks <- quantile(all_compounds_epistasis_TIF, c(.01,.99), na.rm = T)
palette_length <- 100
breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
            seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
colors = rev(diverge_hcl(length(breaks), palette = "Cork"))
colors[which(quant_breaks == 0)-1] = "gray90"

for(i in unique(target$target)) {
  target_drugs = target %>% 
    rownames_to_column(var = "drug_conc") %>%
    filter(target == i) %>% 
    pull(drug_conc)
  
  all_compounds_epistasis_TIF = CCD_tib %>%
    dplyr::select(drug_conc, indelrate_slope, feature) %>%
    group_by(drug_conc) %>%
    mutate(onlyNA = sum(is.na(indelrate_slope))) %>% 
    filter(onlyNA != 25) %>%
    reshape2::dcast(drug_conc ~ feature, value.var = "indelrate_slope", fill = 0) %>% 
    column_to_rownames(var = "drug_conc")
  
  # Make matrix with drugs from selected target
  TIF_synergy_matrix = all_compounds_epistasis_TIF %>% 
    rownames_to_column(var = "drug_conc") %>%
    filter(drug_conc %in% target_drugs) %>%
    column_to_rownames(var = "drug_conc")
  
  if(nrow(TIF_synergy_matrix) > 0) {
    cluster_rows = ifelse(nrow(TIF_synergy_matrix) > 1, TRUE, FALSE)
    
    p = pheatmap(TIF_synergy_matrix[, TRIP_chrom_clust$labels],
                 border_color = F,
                 cellwidth = 8,
                 cellheight = 8,
                 cluster_rows = cluster_rows,
                 breaks = breaks,
                 cluster_cols = TRIP_chrom_clust,
                 annotation_row = target_set_target,
                 clustering_distance_rows = "correlation",
                 main = paste("Chromatin synergies - TIF ", i),
                 color = colors,
                 annotation_colors = annotation_colors_scr, 
                 annotation_legend = T)
  }
}
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

