---
title: "DSB epigenetic drug screen - Figure 2"
author:
  - name: "Max Trauernicht & Ruben Schep"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'

output:
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
  pdf_document:
    toc: yes
    toc_depth: '4'
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```
### Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(dplyr)
library(outliers)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(ggrepel)
library(GGally)
library(DESeq2)
library(gghighlight)
library(platetools)
library(ggpubr)
library(tidyr)
library(Laurae)
library(tibble)
library(pheatmap)
library(RColorBrewer)
library(ggrastr)
library(rstatix)
```

### Functions

```{r load_functions, echo = FALSE, warning = FALSE}

p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
    ## Methods 'Hommel', 'BH', 'BY' and speed improvements
    ## contributed by Gordon Smyth
    method <- match.arg(method)
    if(method == "fdr") method <- "BH"	# back compatibility
    nm <- names(p)
    p <- as.numeric(p)
    p0 <- setNames(p, nm)
    if(all(nna <- !is.na(p))) nna <- TRUE
    p <- p[nna]
    lp <- length(p)
    # stopifnot(n >= lp)
    if (n <= 1) return(p0)
    if (n == 2 && method == "hommel") method <- "hochberg"

    p0[nna] <-
	switch(method,
	       bonferroni = pmin(1, n * p),
	       holm = {
		   i <- seq_len(lp)
		   o <- order(p)
		   ro <- order(o)
		   pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
	       },
	       hommel = { ## needs n-1 >= 2 in for() below
		   if(n > lp) p <- c(p, rep.int(1, n-lp))
		   i <- seq_len(n)
		   o <- order(p)
		   p <- p[o]
		   ro <- order(o)
		   q <- pa <- rep.int( min(n*p/i), n)
		   for (j in (n-1):2) {
		       ij <- seq_len(n-j+1)
		       i2 <- (n-j+2):n
		       q1 <- min(j*p[i2]/(2:j))
		       q[ij] <- pmin(j*p[ij], q1)
		       q[i2] <- q[n-j+1]
		       pa <- pmax(pa,q)
		   }
		   pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
	       },
	       hochberg = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
	       },
	       BH = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   pmin(1, cummin( n / i * p[o] ))[ro]
	       },
	       BY = {
		   i <- lp:1L
		   o <- order(p, decreasing = TRUE)
		   ro <- order(o)
		   q <- sum(1L/(1L:n))
		   pmin(1, cummin(q * n / i * p[o]))[ro]
	       },
	       none = p)
    p0
}

SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

# From Fede:
# ggpairs custom functions
boundaries <- c(0.7,0.75,0.8,0.85)
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}


# stat_smooth_func <- function(mapping = NULL, data = NULL,
#                         geom = "smooth", position = "identity",
#                         ...,
#                         method = "auto",
#                         formula = y ~ x,
#                         se = TRUE,
#                         n = 80,
#                         span = 0.75,
#                         fullrange = FALSE,
#                         level = 0.95,
#                         method.args = list(),
#                         na.rm = FALSE,
#                         show.legend = NA,
#                         inherit.aes = TRUE,
#                         xpos = NULL,
#                         ypos = NULL) {
#   layer(
#     data = data,
#     mapping = mapping,
#     stat = StatSmoothFunc,
#     geom = geom,
#     position = position,
#     show.legend = show.legend,
#     inherit.aes = inherit.aes,
#     params = list(
#       method = method,
#       formula = formula,
#       se = se,
#       n = n,
#       fullrange = fullrange,
#       level = level,
#       na.rm = na.rm,
#       method.args = method.args,
#       span = span,
#       xpos = xpos,
#       ypos = ypos,
#       ...
#     )
#   )
# }
# 
# StatSmoothFunc <- ggproto("StatSmooth", Stat,
#                       
#                       setup_params = function(data, params) {
#                         # Figure out what type of smoothing to do: loess for small datasets,
#                         # gam with a cubic regression basis for large data
#                         # This is based on the size of the _largest_ group.
#                         if (identical(params$method, "auto")) {
#                           max_group <- max(table(data$group))
#                           
#                           if (max_group < 1000) {
#                             params$method <- "loess"
#                           } else {
#                             params$method <- "gam"
#                             params$formula <- y ~ s(x, bs = "cs")
#                           }
#                         }
#                         if (identical(params$method, "gam")) {
#                           params$method <- mgcv::gam
#                         }
#                         
#                         params
#                       },
#                       
#                       compute_group = function(data, scales, method = "auto", formula = y~x,
#                                                se = TRUE, n = 80, span = 0.75, fullrange = FALSE,
#                                                xseq = NULL, level = 0.95, method.args = list(),
#                                                na.rm = FALSE, xpos=NULL, ypos=NULL) {
#                         if (length(unique(data$x)) < 2) {
#                           # Not enough data to perform fit
#                           return(data.frame())
#                         }
#                         
#                         if (is.null(data$weight)) data$weight <- 1
#                         
#                         if (is.null(xseq)) {
#                           if (is.integer(data$x)) {
#                             if (fullrange) {
#                               xseq <- scales$x$dimension()
#                             } else {
#                               xseq <- sort(unique(data$x))
#                             }
#                           } else {
#                             if (fullrange) {
#                               range <- scales$x$dimension()
#                             } else {
#                               range <- range(data$x, na.rm = TRUE)
#                             }
#                             xseq <- seq(range[1], range[2], length.out = n)
#                           }
#                         }
#                         # Special case span because it's the most commonly used model argument
#                         if (identical(method, "loess")) {
#                           method.args$span <- span
#                         }
#                         
#                         if (is.character(method)) method <- match.fun(method)
#                         
#                         base.args <- list(quote(formula), data = quote(data), weights = quote(weight))
#                         model <- do.call(method, c(base.args, method.args))
#                         
#                         m = model
#                         eq <- substitute(italic(y) == a + b %.% italic(x)*","~~italic(r)^2~"="~r2, 
#                                          list(a = format(coef(m)[1], digits = 3), 
#                                               b = format(coef(m)[2], digits = 3), 
#                                               r2 = format(summary(m)$r.squared, digits = 3)))
#                         func_string = as.character(as.expression(eq))
#                         
#                         if(is.null(xpos)) xpos = min(data$x)*0.9
#                         if(is.null(ypos)) ypos = max(data$y)*0.9
#                         data.frame(x=xpos, y=ypos, label=func_string)
#                         
#                       },
#                       
#                       required_aes = c("x", "y")
# )

theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "white", color = "white")
            )
}
theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "white", color = "white")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "white", color = "white")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 
colors_targets <-  c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43"
             )


ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

```


### Data import

```{r data import, echo = FALSE, warning = FALSE}
# Figure out folder
figure_out = "/DATA/projects/DSBrepair/scratch/episcreen/"

# Import data from preprocessing script
indel.data <- readRDS("/DATA/projects/DSBrepair/data/R/rs20220216_episcreen/rs20220216_episcreen_ratios.3.RDS")
```

---

## A: Cutting efficiency change per drug
Are there any drugs (per concentration) that significantly change the cutting efficiency significantly across all IPRs?
```{r warning = FALSE}
# A: Global cutting efficiency change induced by all drugs 
## Plot mean DMSO cutting efficiency in middle, and difference from that to the left and right in percent
indel.data.freq <- indel.data %>%
  filter(sample == "drug" | drug == "DMSO") %>%
  distinct(freqCut_norm_global_mean, drug, conc_char, mean_eff_zscore_global_comb, freqCut, sample, IPR, target)

dmso_freq <- indel.data.freq %>%
  filter(drug == "DMSO") %>%
  dplyr::select('dmso_freq' = freqCut, IPR) %>%
  mutate(max_dmso_freq = ave(dmso_freq, IPR, FUN = mean),
         dmso_freq = mean(dmso_freq)) %>%
  mutate(max_dmso_freq = max(max_dmso_freq)) %>%
  #mutate(max_dmso_freq = max_dmso_freq - dmso_freq) %>%
  distinct(dmso_freq, max_dmso_freq)

indel.data.freq <- merge(indel.data.freq, dmso_freq) %>%
  mutate(freqCut_dif = freqCut - dmso_freq) %>%
  mutate(freqCut_dif = ave(freqCut_dif, drug, conc_char, FUN = mean)) %>%
  mutate(freqCut = ave(freqCut, drug, conc_char, FUN = mean))

  
ggplot(indel.data.freq %>%
                filter(sample == "drug") %>%
                distinct(freqCut, drug, conc_char, mean_eff_zscore_global_comb) %>%
                mutate(mean_eff_pval_global_comb = 2*pnorm(q = abs(mean_eff_zscore_global_comb), lower.tail = F)) %>%
                mutate(mean_eff_pval_adj_global_comb = p.adjust(mean_eff_pval_global_comb, method = 'fdr')),
              aes(y = -log10(mean_eff_pval_adj_global_comb), x = freqCut, 
                  color = ifelse(mean_eff_pval_adj_global_comb < 0.05 & freqCut > (dmso_freq$dmso_freq + 0.02), "green", 
                               ifelse(mean_eff_pval_adj_global_comb < 0.05 & freqCut < (dmso_freq$dmso_freq - 0.02), "yellow","blue")))) +
  geom_point() +
  geom_hline(linetype = "dashed", yintercept = -log10(0.05)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_freq$dmso_freq + 0.02)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_freq$dmso_freq - 0.02)) +
  geom_vline(linetype = "twodash", xintercept = unique(indel.data.freq$max_dmso_freq))+
  ggtitle("first z-score per tech. rep., then mean of z-scores") +
  scale_color_manual(values = c("#C2C2C2", "#3d405b", "#e07a5f")) +
  facet_wrap(~conc_char) +
  annotate("rect",xmin=(dmso_freq$dmso_freq - 0.02),xmax=(dmso_freq$dmso_freq + 0.02),ymin=-Inf,ymax=Inf, alpha=0.1, fill="black") +
  annotate("rect",xmin=unique(indel.data.freq$max_dmso_freq),xmax=Inf,ymin=-Inf,ymax=Inf, alpha=0.25, fill="black") +
  annotate("rect",xmin=-Inf,xmax=Inf,ymin=-Inf,ymax=-log10(0.05), alpha=0.1, fill="black") +
  scale_x_continuous(labels = scales::percent) +
  coord_flip()

# Alternative: label drugs and color targets
indel.data.freq2  <- indel.data.freq %>%
    filter(sample == "drug") %>%
    distinct(freqCut, freqCut_dif, drug, conc_char, mean_eff_zscore_global_comb, target) %>%
    mutate(mean_eff_pval_global_comb = 2*pnorm(q = abs(mean_eff_zscore_global_comb), lower.tail = F)) %>%
    mutate(mean_eff_pval_adj_global_comb = p.adjust(mean_eff_pval_global_comb, method = 'fdr'))

labels <- indel.data.freq2 %>%
  mutate(freqCut_dif = abs(freqCut_dif)) %>%
    group_by(conc_char) %>%
    top_n(3, freqCut_dif) %>%
    ungroup() %>%
    mutate(flag = "TRUE") %>%
  distinct(drug, conc_char, flag)

indel.data.freq2 <- merge(indel.data.freq2, labels, all = T)
indel.data.freq2$flag[is.na(indel.data.freq2$flag)] <- "FALSE"
indel.data.freq2$drug <- gsub(" [(].*[)]", "", indel.data.freq2$drug) #shorten name of drug


ggplot(indel.data.freq2,
              aes(y = -log10(mean_eff_pval_adj_global_comb), x = freqCut, 
                  color = target)) +
  geom_point(aes(alpha = ifelse(abs(freqCut_dif)>0.02, 1, 0.1)), size = 2) +
  #geom_label_repel(data = indel.data.freq2 %>% filter(flag == "TRUE") %>% unique(), 
                   #aes(label = drug)) +
  geom_hline(linetype = "dashed", yintercept = -log10(0.05)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_freq$dmso_freq + 0.02)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_freq$dmso_freq - 0.02)) +
  geom_vline(linetype = "twodash", xintercept = unique(indel.data.freq$max_dmso_freq))+
  ggtitle("first z-score per tech. rep., then mean of z-scores") +
  scale_color_manual(values = colors_targets) +
  facet_wrap(~conc_char) +
  annotate("rect",xmin=(dmso_freq$dmso_freq + 0.02),xmax=unique(indel.data.freq$max_dmso_freq),ymin=-log10(0.05),ymax=Inf, alpha=0.1, fill="#e07a5f") +
  annotate("rect",xmin=unique(indel.data.freq$max_dmso_freq),xmax=Inf,ymin=-Inf,ymax=Inf, alpha=0.25, fill="black") +
  annotate("rect",xmin=-Inf,xmax=(dmso_freq$dmso_freq - 0.02),ymin=-log10(0.05),ymax=Inf, alpha=0.1, fill="#3d405b") +
  scale_x_continuous(labels = scales::percent) +
  coord_flip()

aurora_targets <- data.frame("drug" = c("Alisertib","Aurora A Inhibitor I", "Barasertib","CCT129202", "CCT137690", "CYC116",
                                        "ENMD-2076","Hesperadin","JNJ-7706621","KW-2449","MK-5108","MLN8054","PHA-680632","SNS-314 Mesylate","ZM 447439"),
                             "A_B" = c("A", "A", "B", "A+B", "A+B","A+B", "A", "B", "A+B", "A", "A", "A", "A", "A+B", "A+B"),
                             "B" = c("No", "No", "Yes", "Yes", "Yes", "Yes", "No", "Yes", "Yes", "No", "No", "No", "No", "Yes", "Yes"),
                             "A" = c("Yes", "Yes", "No", "Yes", "Yes", "Yes", "Yes", "No", "Yes", "Yes", "Yes", "Yes", "Yes", "Yes", "Yes"))


pdf(paste0(figure_out, "rs20220623_aurora_kinase.pdf"), width=4, height=3)

ggplot(indel.data.freq2 %>%
         filter(target == "Aurora Kinase") %>%
         left_join(aurora_targets),
              aes(y = -log10(mean_eff_pval_adj_global_comb), x = freqCut, 
                  color = A_B)) +
  geom_point(size = 2) +
  #geom_label_repel(data = indel.data.freq2 %>% filter(flag == "TRUE") %>% unique(), 
                   #aes(label = drug)) +
  geom_hline(linetype = "dashed", yintercept = -log10(0.05)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_freq$dmso_freq + 0.02)) +
  geom_vline(linetype = "dashed", xintercept = (dmso_freq$dmso_freq - 0.02)) +
  #geom_vline(linetype = "twodash", xintercept = unique(indel.data.freq$max_dmso_freq))+
  scale_color_manual(values = c("#6d6875", "#b5838d", "#ffb4a2")) +
  #facet_wrap(~conc_char) +
  #annotate("rect",xmin=(dmso_freq$dmso_freq + 0.02),xmax=unique(indel.data.freq$max_dmso_freq),ymin=-log10(0.05),ymax=Inf, alpha=0.1, fill="#e07a5f") +
  #annotate("rect",xmin=unique(indel.data.freq$max_dmso_freq),xmax=Inf,ymin=-Inf,ymax=Inf, alpha=0.25, fill="black") +
  #nnotate("rect",xmin=-Inf,xmax=(dmso_freq$dmso_freq - 0.02),ymin=-log10(0.05),ymax=Inf, alpha=0.1, fill="#3d405b") +
  scale_x_continuous(labels = scales::percent) +
  coord_flip()
dev.off()
```

*Figure 2A: Compound-induced change in cutting efficiency of all tested compounds. Mean per drug & concentration vs. p-value from t-test.*

**Conclusion: Yes, some drugs do alter cutting efficiency significantly. Changes go in both directions, at lower concentration the efficiency is mostly increased (at higher concentrations the decrease might be due to toxicity)**


---

## B: Cutting efficiency change per target group
Are there any target groups that are prone to changes in cutting efficiency?
```{r warning = FALSE}
# C: Bar plot
eff.data <- indel.data.freq %>%
    filter(sample == "drug") %>%
    distinct(mean_eff_zscore_global_comb, drug, target, conc_char, freqCut_dif) %>%
    mutate(freqCut_dif = ave(freqCut_dif, drug, conc_char, FUN = mean)) %>%
    distinct() %>%
    mutate(mean_eff_pval_global_comb = 2*pnorm(q = abs(mean_eff_zscore_global_comb), lower.tail = F)) %>%
    mutate(mean_eff_pval_adj_global_comb = p.adjust(mean_eff_pval_global_comb, method = 'fdr'))

eff.data$sign <- "2_No"
eff.data$sign[eff.data$freqCut_dif > 0.02 & eff.data$mean_eff_pval_adj_global_comb < 0.05] <- "1_Up"
eff.data$sign[eff.data$freqCut_dif < -0.02 & eff.data$mean_eff_pval_adj_global_comb < 0.05] <- "0_Down"

eff.data <- eff.data %>%
    mutate(number = ave(target, target, sign, conc_char, FUN = length)) %>%
    mutate(number = as.numeric(number)) %>%
    distinct(conc_char, target, sign, number) %>%
    spread(sign, number, fill = 0) %>%
    mutate(fraction = (`1_Up` + `0_Down`) / (`2_No` + (`1_Up` + `0_Down`))) %>%
    pivot_longer(`0_Down`:`2_No`, names_to = "sign", values_to = "number")

pdf(paste0(figure_out, "rs20220623_indelfrequency_targets_barplot.pdf"), width=8, height=4)
ggplot(eff.data,
       aes(y = reorder(target, number), x = number, fill = sign)) +
    geom_bar(stat = "identity") +
    facet_wrap(~conc_char)+
    scale_fill_manual(values = c("#e07a5f", "#3d405b", "#C2C2C2")) +
    theme_pubr(border = T)
dev.off()
```

*Figure 2B: Number of compounds with significant changes in cutting efficiency per target group.*

**Conclusion: Many HDACis have increased indel rates, especially at lower concentrations already. AURK and ERD inhibitors decrease efficiency at higher concentrations.**

---

## C: Efficiency heatmap
Are the effects on cutting efficiency global or local? Can we plot all interesting effects in one heatmap per concentration?
```{r warning = FALSE}
# Set annotation for heatmap
# viability
viab_drug = indel.data %>% filter(drug != "PAO") %>% mutate(drug = gsub(" [(].*[)]", "", drug),
                        drug_conc = paste(drug, conc_char, sep = " ")) %>% distinct(viab_mean, drug_conc, replicate) %>% pivot_wider(values_from = viab_mean, names_from = replicate) %>%
  dplyr::rename(., viab_rep2 = E1504, viab_rep1 = E177)
# Row annotation: annotate drugs with the target group
target_set <- indel.data %>% 
  filter(!target %in% c("Negative Control", "MRN", "DNA-PK")) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>%
  left_join(viab_drug) %>%
  distinct(drug, drug_conc, conc_char, target, viab_rep1, viab_rep2) %>% 
  column_to_rownames(var="drug_conc")


chromatin <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(indel.data, barcode, IPR)) 
# chromatin$IPR[is.na(chromatin$IPR)] <- "IPR1"
chromatin <- chromatin %>%
  filter(IPR != "<NA>") %>% 
  dplyr::select(-barcode) %>%
  column_to_rownames(var="IPR")

# Change colors of annotations
chrom_colors = c("Euchromatin" = "#EAEEE8",
                  "other-heterochromatin" = "#a5a58d",  
                     "H3K27me3" = "#CAD6C2", 
                  "Triple Heterochromatin" = "#6b705c")

# colors <- brewer.pal(length(unique(barcode.annotation$cluster)), "Pastel2")
# names(colors) <- unique(barcode.annotation$cluster)

annotation_colors_scr = list(
  chromatin = chrom_colors,
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43"),
  conc_char = c(`100 nM` = "#78BF84", `1 µM`  = "#1E9532", `10 µM` = "#0F4A19"),
  clusters_drugs = c(`1` = "grey10", `2` = "grey20", `3` = "grey50"))



# Order the annotation legend
ordered_marks <- c("Euchromatin", 
                   "H3K27me3", 
                     "other-heterochromatin",  
                     "Triple Heterochromatin")

annotation_colors_scr$chromatin <- annotation_colors_scr$chromatin[ordered_marks]

# Order IPRs by their starting freqCut value
dmso_freq <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(IPR, freqCut) %>%
  mutate(freqCut = ave(freqCut, IPR, FUN = mean)*100) %>%
  dplyr::select('dmso_eff' = freqCut, IPR) %>%
  distinct() %>%
  arrange(desc(dmso_eff))

iprs <- dmso_freq$IPR
iprs <- as.character(iprs)


# Generate heatmap: select top drugs (highest change at any IPR) per concentration
eff.data.heatmap <- indel.data %>%
  #filter(target != "Aurora Kinase") %>%
  filter(sample == "drug") %>%
  #filter(Mod(mean_eff_zscore_comb) >= 2.58) %>%  # we can choose to display all drugs that have significant global hits
  distinct(conc_char, drug, target, IPR, mean_eff_zscore_comb, freqCut, replicate, tech)

dmso_eff <- indel.data %>%
  filter(drug == "DMSO") %>%
  dplyr::select('dmso_freq' = freqCut, IPR) %>%
  mutate(dmso_freq = ave(dmso_freq, IPR, FUN = mean)) %>%
  distinct()

eff.data.heatmap <- left_join(eff.data.heatmap, dmso_eff) %>%
  mutate(freqCut_dif = freqCut - dmso_freq,
         freqCut_FC = freqCut / dmso_freq) %>%
  mutate(freqCut_dif = ave(freqCut_dif, drug, conc_char, IPR, FUN = mean)*100,
         freqCut_FC = ave(freqCut_FC, drug, conc_char, IPR, FUN = mean))

## Select drugs that have a large effect size in at least one location (some drugs might have a small global effect size, but large local)
eff.data.heatmap <- eff.data.heatmap %>%
  mutate(mean_eff_zscore_comb = 2*pnorm(q = abs(mean_eff_zscore_comb), lower.tail = F)) %>%
  mutate(mean_eff_zscore_comb = p.adjust(mean_eff_zscore_comb, method = 'fdr')) %>%
  mutate(min_score = ave(mean_eff_zscore_comb, drug, conc_char, FUN = min)) %>%
  filter(min_score < 0.001) %>%
  distinct(drug, target, conc_char, freqCut_dif, IPR)

#eff.data.heatmap2$freqCut_norm_mean[eff.data.heatmap2$mean_eff_zscore_comb > 0.05] <- NA
eff.data.heatmap2 <- eff.data.heatmap %>%
  spread(IPR, freqCut_dif)

eff.data.heatmap2 <- eff.data.heatmap2[,c("drug", "target", "conc_char", iprs)]
eff.data.heatmap2$drug <- gsub(" [(].*[)]", "", eff.data.heatmap2$drug) #shorten name of drug


breaks <- quantile(unlist(eff.data.heatmap2 %>% dplyr::select(contains('IPR'))), c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 100

myBreaks <- c(seq(breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
              seq(breaks[2]/palette_length, breaks[2], length.out=floor(palette_length/2)))

ordered_marks2 <- eff.data.heatmap2 %>% 
  dplyr::select(target) %>%
  distinct() %>%
  arrange(target)

ordered_marks2 <- ordered_marks2$target

annotation_colors_scr$target <- annotation_colors_scr$target[ordered_marks2]

# Only include lowest effective concentrations
# eff.data.heatmap2 <- eff.data.heatmap2 %>%
#   mutate(compound_n = ave(drug, drug, FUN = length))
# all_conc <- eff.data.heatmap2[eff.data.heatmap2$compound_n == "3" & eff.data.heatmap2$conc_char %in% c("1 µM", "10 µM"),]
# eff.data.heatmap2 <- anti_join(eff.data.heatmap2, all_conc)
# 
# eff.data.heatmap2$conc <- 1
# eff.data.heatmap2$conc[eff.data.heatmap2$conc_char == "1 µM"] <- 2
# eff.data.heatmap2$conc[eff.data.heatmap2$conc_char == "10 µM"] <- 3
# 
# for (i in unique(eff.data.heatmap2$drug)) {
#   if (eff.data.heatmap2$compound_n[eff.data.heatmap2$drug == i] == "2") {
#     max_conc <- max(eff.data.heatmap2$conc[eff.data.heatmap2$compound_n == "2" & eff.data.heatmap2$drug == i])
#     remove <- eff.data.heatmap2[eff.data.heatmap2$compound_n == "2" & eff.data.heatmap2$drug == i & eff.data.heatmap2$conc == max_conc,]
#     eff.data.heatmap2 <-anti_join(eff.data.heatmap2, remove)
#   }
# }
# 
# eff.data.heatmap2 <- eff.data.heatmap2 %>%
#   dplyr::select(-compound_n, -conc)

# Add cluster annotation
clusters_drugs <- hclust(dist(eff.data.heatmap2), method = "complete")
clusters_drugs <- cutree(tree = clusters_drugs, k = 6)
clusters_drugs <- data.frame(clusters_drugs)

eff.data.heatmap2.clusters <- cbind(eff.data.heatmap2, clusters_drugs) %>% 
  as_tibble() %>%
  mutate(drug_conc = paste(drug, conc_char, sep = " "))

clusters_drugs <- eff.data.heatmap2.clusters %>%
  distinct(drug_conc, clusters_drugs)

target <- target_set %>%
  rownames_to_column("drug_conc") %>%
  left_join(clusters_drugs) %>%
  na.omit()
rownames(target) <- target$drug_conc
target <- target %>%
  dplyr::select(target, clusters_drugs, conc_char, viab_rep1, viab_rep2)

pdf(paste0(figure_out, "rs20220623_indelfrequency_heatmap_allconc.pdf"), width=20, height=8)
pheatmap(eff.data.heatmap2.clusters %>%
                       arrange(target) %>%
                       column_to_rownames('drug_conc') %>%
                       dplyr::select(contains('IPR')) %>%
                       t(),
           border_color = F,
           cellwidth = 8,
           cellheight = 8,
           breaks = myBreaks,
           cluster_rows = F,
           cluster_cols = T,
           angle_col = 90,
           na_col = "white",
           main = 'efficiency change, all concentrations',
           annotation_col = target,
           annotation_row = chromatin,
           color = colorRampPalette(c("#e07a5f", "#e5e5e5", "#3d405b"))(palette_length),
           annotation_colors = annotation_colors_scr, 
           annotation_legend = T,
           cutree_cols = 4)
dev.off()

pdf(paste0(figure_out, "rs20220623_indelfrequency_heatmap_DMSO.pdf"), width=3, height=8)

breaks <- seq(40,90,0.6)
pheatmap(dmso_freq %>%
             column_to_rownames(var = "IPR"),
         cluster_rows = F,
         cluster_cols = F,
         cellwidth = 8,
         cellheight = 8,
         border_color = NA,
         breaks = breaks,
         color = colorRampPalette(c("white", "black"))(palette_length))
dev.off()

cluster1 = target %>% filter(target == "HDAC") %>% rownames()
cluster1 = cluster1[!grepl("Nullscript", cluster1)]


eff.data.heatmap.cluster1 = eff.data.heatmap2 %>% 
  mutate(drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  filter(drug_conc %in% cluster1)

breaks <- quantile(unlist(eff.data.heatmap.cluster1 %>% 
                            dplyr::select(contains('IPR'))), c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 100

myBreaks <- c(seq(breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
              seq(breaks[2]/palette_length, breaks[2], length.out=floor(palette_length/2)))

# Add cluster annotation
clusters_drugs <- hclust(dist(eff.data.heatmap.cluster1), method = "complete")
clusters_drugs <- cutree(tree = clusters_drugs, k = 4)
clusters_drugs <- data.frame(clusters_drugs)

eff.data.heatmap.cluster1 <- cbind(eff.data.heatmap.cluster1, clusters_drugs) %>% 
  as_tibble()

clusters_drugs <- eff.data.heatmap.cluster1 %>%
  distinct(drug_conc, clusters_drugs)
target.cluster1 <- target_set %>%
  rownames_to_column("drug_conc") %>%
  left_join(clusters_drugs) %>%
  na.omit()
rownames(target.cluster1) <- target.cluster1$drug_conc
target.cluster1 <- target.cluster1 %>%
  dplyr::select(target, clusters_drugs, conc_char)


pdf(paste0(figure_out, "rs20220623_indelfrequency_heatmap_HDAC.pdf"), width=16, height=8)
pheatmap(eff.data.heatmap.cluster1 %>%
                       arrange(target.cluster1) %>%
                       column_to_rownames('drug_conc') %>%
                       dplyr::select(contains('IPR')) %>%
                       t(),
           border_color = F,
           cellwidth = 8,
           cellheight = 8,
           breaks = myBreaks,
           cluster_rows = F,
           cluster_cols = T,
           angle_col = 90,
           na_col = "white",
           main = 'efficiency change, all concentrations',
           annotation_col = target.cluster1,
           annotation_row = chromatin,
           color = colorRampPalette(c("#e07a5f", "#e5e5e5", "#3d405b"))(palette_length),
           annotation_colors = annotation_colors_scr, 
           annotation_legend = T)
dev.off()
```

## Do the different clusters have different effects on different chromatin types?
```{r}
chromatin_ipr <- chromatin %>%
  rownames_to_column("IPR")

eff.data.heatmap3 <- eff.data.heatmap2 %>%
    pivot_longer(cols = contains("IPR"), names_to = "IPR", values_to = "freqCut_dif")

chromatin_effect <- left_join(eff.data.heatmap3, chromatin_ipr) %>%
  #filter(target == "HDAC") %>%
  mutate(freqCut_dif = ave(freqCut_dif, drug, chromatin, conc_char, FUN = mean),
         drug_conc = paste(drug, conc_char, sep = " ")) %>%
  distinct(freqCut_dif, drug, chromatin, conc_char, target, drug_conc)

clusters_drugs = target %>% rownames_to_column(var = "drug_conc")

chromatin_effect <- left_join(chromatin_effect, clusters_drugs)

sigFunc = function(x){
  if(x < 0.001){"***"} 
  else if(x < 0.01){"**"}
  else if(x < 0.05){"*"}
  else{NA}}

pdf(paste0(figure_out, "rs20220623_indelfrequency_3clusters_chromatin.pdf"), width=8, height=4)
ggplot(chromatin_effect,
              aes(x = chromatin, y = freqCut_dif, fill = chromatin)) +
  geom_boxplot(outlier.shape = NA) +
  geom_quasirandom() +
  geom_signif(comparisons = combn(sort(unique(chromatin_effect$chromatin)), 2, simplify = F), 
              map_signif_level=sigFunc, 
              step_increase = 0.1) +
  theme_classic_lines_45() +
  scale_fill_manual(values = chrom_colors) +
  theme(legend.position = "none") +
  facet_wrap(~clusters_drugs)
dev.off()
```
## Plot HDAC inhibitor effects on different chromatin types
```{r}
chromatin_ipr <- chromatin %>%
  rownames_to_column("IPR")

eff.data.heatmap3 <- eff.data.heatmap2.clusters %>%
  pivot_longer(contains("IPR"), names_to = "IPR", values_to = "freqCut_dif")

chromatin_effect <- merge(eff.data.heatmap3, chromatin_ipr) %>%
  #filter(target == "HDAC") %>%
  mutate(freqCut_dif = ave(freqCut_dif, drug, chromatin, conc_char, FUN = mean)) %>%
  distinct(freqCut_dif, drug, chromatin, conc_char, target, clusters_drugs)

sigFunc = function(x){
  if(x < 0.001){"***"} 
  else if(x < 0.01){"**"}
  else if(x < 0.05){"*"}
  else{NA}}

ggplot(chromatin_effect %>%
         filter(target == "HDAC") %>%
         filter(clusters_drugs %in% c(3, 4)),
              aes(x = chromatin, y = freqCut_dif, fill = chromatin)) +
  geom_boxplot(outlier.shape = NA) +
  geom_quasirandom() +
  geom_signif(comparisons = combn(sort(unique(chromatin_effect$chromatin)), 2, simplify = F), 
              map_signif_level=sigFunc, 
              step_increase = 0.1) +
  theme_classic_lines_45() +
  scale_fill_manual(values = chrom_colors) +
  theme(legend.position = "none") +
  facet_wrap(~clusters_drugs)
```



## Aurora A vs. Aurora B
```{r}
aurora_df <- eff.data.heatmap2 %>%
  filter(target == "Aurora Kinase") %>%
  pivot_longer(contains("IPR"), names_to = "IPR", values_to = "freqCut_dif") %>%
  mutate(freqCut_dif = ave(freqCut_dif, drug, FUN = mean)) %>%
  dplyr::select(-IPR) %>%
  distinct()

aurora_targets <- data.frame("drug" = c("Alisertib","Aurora A Inhibitor I", "Barasertib","CCT129202",
                                        "ENMD-2076","Hesperadin","JNJ-7706621","MK-5108","MLN8054","PHA-680632","SNS-314 Mesylate","ZM 447439"),
                             "A_B" = c("A", "A", "B", "A+B+C", "A", "B", "A+B", "A", "A", "A", "A+B+C", "A+B"),
                             "B" = c("No", "No", "Yes", "Yes", "No", "Yes", "Yes", "No", "No", "No", "Yes", "Yes"))
aurora_df <- merge(aurora_df, aurora_targets)

ggplot(aurora_df, 
       aes(x = A_B, y = freqCut_dif)) + 
  geom_boxplot() +
  geom_quasirandom() +
  geom_signif(comparisons = combn(sort(unique(aurora_df$A_B)), 2, simplify = F), 
              map_signif_level=sigFunc, 
              step_increase = 0.1) 

ggplot(aurora_df, 
       aes(x = B, y = freqCut_dif)) + 
  geom_boxplot() +
  geom_quasirandom() +
  geom_signif(comparisons = combn(sort(unique(aurora_df$B)), 2, simplify = F), 
              map_signif_level=sigFunc, 
              step_increase = 0.1) 
```



## Are the changes in indel rate correlated with the starting indel rates?

```{r warning = FALSE}
## Add correlation with DMSO values to heatmap
dmso_freq <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(IPR, freqCut) %>%
  mutate(freqCut = ave(freqCut, IPR, FUN = mean)) %>%
  dplyr::select('dmso_eff' = freqCut, IPR) %>%
  distinct()

corr_dmso <- merge(eff.data.heatmap, dmso_freq) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
    drug_conc = paste(drug, conc_char, sep = " ")) %>%
  mutate(mean_effect = ave(freqCut_dif, drug, conc_char, FUN = mean)) %>%
  mutate(sign = sign(mean_effect))

## Correlation with H3K27me3 (not included in the heatmaps atm, can be added)
h3k27me3 <- indel.data %>%
  distinct(IPR, EZH2.zscore, late_replicating.zscore)

corr_dmso <- merge(corr_dmso, h3k27me3) 

corr_heatmap <- corr_dmso

for (i in unique(corr_heatmap$drug_conc)) {
  
  corr_heatmap$cor_dmso[corr_heatmap$drug_conc == i] <- 
    cor(corr_heatmap$freqCut_dif[corr_heatmap$drug_conc == i],
        corr_heatmap$dmso_eff[corr_heatmap$drug_conc == i])
  
  corr_heatmap$cor_h3k27me3[corr_heatmap$drug_conc == i] <- 
    cor(corr_heatmap$freqCut_dif[corr_heatmap$drug_conc == i],
        corr_heatmap$EZH2.zscore[corr_heatmap$drug_conc == i])
  
  corr_heatmap$cor_repl[corr_heatmap$drug_conc == i] <- 
    cor(corr_heatmap$freqCut_dif[corr_heatmap$drug_conc == i],
        corr_heatmap$late_replicating.zscore[corr_heatmap$drug_conc == i])
}

pdf(paste0(figure_out, "rs20220214_3cluster_corr_DMSO.pdf"), width=5, height=5)
corr_heatmap %>% distinct(drug_conc, cor_dmso) %>% left_join(rownames_to_column(target, var = "drug_conc")) %>%
  ggplot(., aes(as.character(clusters_drugs), cor_dmso)) +
  geom_boxplot(outlier.shape = NA) +
  geom_quasirandom(aes(color = target, size = conc_char)) +
  scale_color_manual(values = annotation_colors_scr$target) + 
  theme_classic_lines_45() +
  theme(legend.position = "none")
dev.off()

## Only include compounds at lowest concentration
corr_dmso <- corr_dmso %>%
  mutate(compound_n = ave(drug, drug, IPR, FUN = length))
all_conc <- corr_dmso[corr_dmso$compound_n == "3" & corr_dmso$conc_char %in% c("1 µM", "10 µM"),]
corr_dmso <- anti_join(corr_dmso, all_conc)

corr_dmso$conc <- 1
corr_dmso$conc[corr_dmso$conc_char == "1 µM"] <- 2
corr_dmso$conc[corr_dmso$conc_char == "10 µM"] <- 3

for (i in unique(corr_dmso$drug)) {
  if (corr_dmso$compound_n[corr_dmso$drug == i] == "2") {
    max_conc <- max(corr_dmso$conc[corr_dmso$compound_n == "2" & corr_dmso$drug == i])
    remove <- corr_dmso[corr_dmso$compound_n == "2" & corr_dmso$drug == i & corr_dmso$conc == max_conc,]
    corr_dmso <-anti_join(corr_dmso, remove)
  }
}

corr_dmso <- corr_dmso %>%
  dplyr::select(-compound_n, -conc)




## Calculate correlations by linear model
corr_dmso <- corr_dmso %>%
  mutate(slope_dmso = NA, p.value_dmso = NA,
          slope_h3k27me3 = NA, p.value_h3k27me3 = NA) 

for (i in unique(corr_dmso$drug_conc)) {
    model.correlation <- lm(formula = freqCut_dif ~ dmso_eff, 
                               data = corr_dmso[corr_dmso$drug_conc == i,]) %>% 
      tidy()
    
    corr_dmso$slope_dmso[corr_dmso$drug_conc == i] <- 
      model.correlation$estimate[2]
    
    corr_dmso$p.value_dmso[corr_dmso$drug_conc == i] <- 
      model.correlation$p.value[2]
}

for (i in unique(corr_dmso$drug_conc)) {
    model.correlation <- lm(formula = freqCut_dif ~ EZH2.zscore, 
                               data = corr_dmso[corr_dmso$drug_conc == i,]) %>% 
      tidy()
    
    corr_dmso$slope_h3k27me3[corr_dmso$drug_conc == i] <- 
      model.correlation$estimate[2]
    
    corr_dmso$p.value_h3k27me3[corr_dmso$drug_conc == i] <- 
      model.correlation$p.value[2]
}



corr_dmso <- corr_dmso %>%
  mutate(p.value_dmso = p.adjust(p.value_dmso, method = "fdr")) #%>%
  # mutate(p.value_h3k27me3 = p.adjust(p.value_h3k27me3, method = "fdr"))

ggplot(corr_dmso %>%
         filter(abs(mean_effect) > 0.025) %>%
         filter(sign == 1) %>%
         distinct(p.value_dmso, drug_conc, sign, target, conc_char),
       aes(y = -log10(p.value_dmso), x = target, 
           #color = as.character(sign), 
           shape = ifelse(p.value_dmso < 0.01, "1", "2"))) +
  geom_quasirandom() +
  #scale_color_manual(values = c("#e07a5f", "#3d405b")) +
  theme_classic_lines_90() +
  geom_hline(yintercept = -log10(0.01), lty = "dashed") #+
  #facet_wrap(~sign, nrow = 2)
  
corr_hits <- corr_dmso %>% filter(drug_conc %in% c("PCI-24781 (Abexinostat) 100 nM", "Fluoro-SAHA 1 µM")) 

corr_h3k27me3_hits <- corr_dmso %>% filter(drug %in% c("EPZ-6438"))
corr_hits <- merge(corr_hits, chromatin)

ggplot(corr_hits %>%
         mutate(freqCut = dmso_eff + freqCut_dif)) +
  geom_abline(lty = "twodash", slope = 1) +
  geom_hline(yintercept = max(dmso_freq$dmso_eff)*100, lty = "dashed") + 
  geom_point(aes(y = freqCut*100, x = dmso_eff*100, fill = chromatin), color = ifelse(corr_hits$freqCut_dif > 0.02, "red", "black"), shape = 21, size = 3) +
  #scale_color_manual(values = c("#e07a5f", "#3d405b")) +
  #stat_cor(method = "pearson") +
  #annotate("rect",xmin=(dmso_freq$dmso_eff + 0.02),xmax=unique(indel.data.freq$max_dmso_freq),ymin=-log10(0.05),ymax=Inf, alpha=0.1, fill="#e07a5f") +
  facet_wrap(~drug_conc) +
  scale_fill_manual(values = chrom_colors) 


ggplot(corr_h3k27me3_hits,
       aes(y = freqCut_dif, x = EZH2.zscore, color = ifelse(sign == 1, "red", "black"))) +
  geom_point() +
  geom_smooth(method = "lm", lty = "dashed") +
  scale_color_manual(values = c("#e07a5f", "#3d405b")) +
  stat_cor(method = "pearson") +
  facet_wrap(~conc_char, scales = "free_y")


# Drug-induced changes per chromatin type
chrom_corr <- indel.data %>%
  distinct(drug, conc_char, freqCut_norm_mean, IPR, target) %>%
  mutate(drug_conc = paste(drug, conc_char, sep = "_"))

drugs_sel <- corr_dmso %>%
  filter(abs(mean_effect) > 0.025) %>%
  #filter(sign == 1) %>%
  distinct(drug_conc, p.value_dmso, p.value_h3k27me3, slope_h3k27me3, sign)

chrom_corr <- merge(chrom_corr, drugs_sel)

chrom_corr$global <- "no"
chrom_corr$global[abs(chrom_corr$p.value_dmso) < 0.01] <- "yes"

ggplot(chrom_corr %>%
         filter(sign == 1) %>%
         filter(target == "HDAC") %>%
         distinct(drug_conc, p.value_dmso, p.value_h3k27me3, target, global, sign),
       aes(x = global, y = -log10(p.value_h3k27me3), shape = ifelse(p.value_h3k27me3 < 0.05, "1", "2"))) +
  #geom_boxplot(outlier.shape = NA) +
  geom_hline(yintercept = -log10(0.05), lty = "dashed") +
  geom_quasirandom() #+
  #facet_wrap(~target) 
```

*Figure 2C: Heatmap of cutting efficiency change induced by HDAC inhibitors at 19 genomic locations.*

**Conclusion: Many different HDAC inhibitors can change cutting efficiency locally. As expected, heterochromatin is more prone to an increase in cutting efficiency.**

## Epistasis per cluster
```{r}
drug_clusters <- eff.data.heatmap2.clusters %>%
  mutate(drug_conc = paste(drug, conc_char, sep = "_")) %>%
  distinct(drug_conc, clusters_drugs, target)

epistasis_cluster <- epistasis_tib %>%
  distinct(drug_conc, feature, indelrate_slope) %>%
  mutate(drug_conc = gsub(" [(].*[)]", "", drug_conc)) %>%
  left_join(drug_clusters) %>%
  na.omit() %>%
  mutate(clusters_drugs = as.character(clusters_drugs)) %>%
  filter(target == "Aurora Kinase")# %>%
  #filter(clusters_drugs %in% c('4','2','5','7'))

ggplot(epistasis_cluster,
       aes(x = clusters_drugs, y = indelrate_slope)) +
  geom_quasirandom() +
  facet_wrap(~feature)

 





epistasis_cluster <- epistasis_cluster %>%
  #filter(feature %in% c("HDAC1.zscore", "HDAC2.zscore", "LMNB1.zscore", "EZH2.zscore")) %>%
  spread(feature, indelrate_slope)

rownames(drug_clusters) <- drug_clusters$drug_conc
drug_clusters <- drug_clusters %>%
  mutate(clusters_drugs = as.character(clusters_drugs)) %>%
  dplyr::select(-drug_conc, -target)

pheatmap(epistasis_cluster %>%
           column_to_rownames(var = "drug_conc") %>%
           dplyr::select(contains("zscore")),
         cluster_rows = T,
         cluster_cols = T,
         cellwidth = 8,
         cellheight = 8,
         annotation_row = drug_clusters,
         annotation_colors = annotation_colors_scr,
         border_color = NA,
         color = colorRampPalette(c("#8c510a", "#f5f5f5","#01665e"))(palette_length))

```




## UMAP
```{r}
eff.data.heatmap3 <- eff.data.heatmap2.clusters %>%
  mutate(drug_conc = paste(drug, conc_char, sep = "_"))

umap_df <- indel.data %>% 
  distinct(freqCut, drug, conc_char, IPR, target) %>% 
  mutate(freqCut = ave(freqCut, drug, conc_char, IPR, FUN = mean)) %>% 
  mutate(drug_conc = paste(drug, conc_char, sep = "_")) %>%
  filter(drug_conc %in% eff.data.heatmap3$drug_conc) %>%
  distinct(drug_conc, target, freqCut, IPR) %>%
  spread(IPR, freqCut)

epistasis_umap <- epistasis_tib %>%
  distinct(drug_conc, feature, indelrate_slope) %>%
  spread(feature, indelrate_slope) %>%
  distinct()

targets <- indel.data %>%
  mutate(drug_conc = paste(drug, conc_char, sep = "_")) %>%
  distinct(drug_conc, target)

epistasis_umap <- merge(epistasis_umap, targets) %>%
  filter(target == "HDAC")

umap_df_umap <- umap::umap(epistasis_umap %>% dplyr::select(contains("zscore")))

layout <- umap_df_umap[["layout"]] 
layout <- data.frame(layout) 
final <- cbind(layout, 'drug_conc' = epistasis_umap$drug_conc, 'target' = epistasis_umap$target, 'EZH2.zscore' = epistasis_umap$EZH2.zscore)

ggplot(final,
       aes(x = X1, y = X2, color = EZH2.zscore)) +
  geom_point() +
  scale_color_viridis()

```




## SA: Correlation of cutting efficiency between replicates of significant hits
We know that the cutting efficiency does not correlate well if all data is taken. However, because of the high noise level and the low effect size of the efficiency changes, it makes more sense to only investigate whether significant hits are robust throughout all replicates. 
```{r warning = FALSE}
# SA: Correlation between technical replicates efficiency: only significant hits
# significant_hits_efficiency <- eff.data %>%
#   filter(pval_adj < 0.05) %>%
#   distinct(drug, conc_char, IPR) %>%
#   mutate(drug_conc = paste(drug, conc_char, IPR, paste = "_"))

# techn_repl <- indel.data %>%
#   mutate(drug_conc = paste(drug, conc_char, IPR, paste = "_")) %>%
#   #filter(drug_conc %in% significant_hits_efficiency$drug_conc) %>%
#   dplyr::distinct(freqCut_norm, drug, tech, replicate, conc_char, sample, IPR) %>%
#   mutate(freqCut_norm = log2(ave(freqCut_norm, drug, tech, replicate, conc_char, IPR, FUN = mean))) %>%
#   unique() %>%
#   mutate(rep = paste(replicate, tech, sep = "_")) %>%
#   dplyr::select(-replicate, -tech) %>%
#   mutate(max = ave(freqCut_norm, drug, conc_char, IPR, FUN = max),
#          min = ave(freqCut_norm, drug, conc_char, IPR, FUN = min)) %>%
#   distinct() %>%
#   spread(rep, freqCut_norm) %>%
#   na.omit() 
#   
# techn_repl$large_effect <- "1_no"
# techn_repl$large_effect[techn_repl$max > 0.3 | techn_repl$min < -0.3] <- "0_yes"
# 
# ggpairs(techn_repl %>% 
#        #filter(conc_char == "1 µM") %>% #1 µM
#        # filter(drug %in% c("Vorinostat (SAHA, MK0683)", "Scriptaid",   "Decitabine" , "(+)-JQ1" ,  "PCI-24781 (Abexinostat)")) %>%
#           dplyr::select(contains("E1")),
#        mapping = aes(color = techn_repl$large_effect),
#                upper = list(continuous = corColor),
#                lower = list(continuous = function(data, mapping, ...) {
#                    ggally_points(data = data, mapping = mapping, alpha = 0.5, size = 0.2) +
#                    geom_abline(slope = 1, lty = "dashed", col = "red")}),
#                diag = list(continuous = function(data, mapping, ...) {
#                    ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
#   # xlim(.4, .7) + 
#   xlab("Cutting efficiency") +
#   ylab("Cutting efficiency") +
#   scale_color_manual(values = colors_diverse)
```

*Figure S2A: Correlation between cutting efficiency of significant hits.*

**Conclusion: The correlation between 'strong' changers of efficiency is OK, not very good, but OK.**



# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

