---
title: "DSB epigenetic drug screen - Figure 1"
author: 
  - name: "Max Trauernicht & Ruben Schep"
    email: "m.trauernicht@nki.nl; r.schep@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
  pdf_document:
    toc: yes
    toc_depth: '4'
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


# Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
$(document).ready(function() {
$items = $('div#TOC li');
$items.each(function(idx) {
num_ul = $(this).parentsUntil('#TOC').length;
$(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
});
});
</script>

```{css, echo = FALSE}
div.sourceCode {
overflow-x: hidden;
}
```

## Libraries

```{r load_packages, results = 'hide', warning = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 


library(tidyverse)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(GGally)
library(pheatmap)
library(RColorBrewer)
library(dendextend)
library(ggrastr)
library(cowplot)
library(ggpubr)
library(scales)

## Select outdir
out.dir = paste0("./figures/", Date, "/")
dir.create(out.dir)
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
knitr::opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

## Functions
```{r load_functions, results = 'hide', warning = FALSE}
p.adjust <- function(p, method = p.adjust.methods, n = length(p))
{
  ## Methods 'Hommel', 'BH', 'BY' and speed improvements
  ## contributed by Gordon Smyth
  method <- match.arg(method)
  if(method == "fdr") method <- "BH"	# back compatibility
  nm <- names(p)
  p <- as.numeric(p)
  p0 <- setNames(p, nm)
  if(all(nna <- !is.na(p))) nna <- TRUE
  p <- p[nna]
  lp <- length(p)
  # stopifnot(n >= lp)
  if (n <= 1) return(p0)
  if (n == 2 && method == "hommel") method <- "hochberg"
  
  p0[nna] <-
    switch(method,
           bonferroni = pmin(1, n * p),
           holm = {
             i <- seq_len(lp)
             o <- order(p)
             ro <- order(o)
             pmin(1, cummax( (n - i + 1L) * p[o] ))[ro]
           },
           hommel = { ## needs n-1 >= 2 in for() below
             if(n > lp) p <- c(p, rep.int(1, n-lp))
             i <- seq_len(n)
             o <- order(p)
             p <- p[o]
             ro <- order(o)
             q <- pa <- rep.int( min(n*p/i), n)
             for (j in (n-1):2) {
               ij <- seq_len(n-j+1)
               i2 <- (n-j+2):n
               q1 <- min(j*p[i2]/(2:j))
               q[ij] <- pmin(j*p[ij], q1)
               q[i2] <- q[n-j+1]
               pa <- pmax(pa,q)
             }
             pmax(pa,p)[if(lp < n) ro[1:lp] else ro]
           },
           hochberg = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             pmin(1, cummin( (n - i + 1L) * p[o] ))[ro]
           },
           BH = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             pmin(1, cummin( n / i * p[o] ))[ro]
           },
           BY = {
             i <- lp:1L
             o <- order(p, decreasing = TRUE)
             ro <- order(o)
             q <- sum(1L/(1L:n))
             pmin(1, cummin(q * n / i * p[o]))[ro]
           },
           none = p)
  p0
}

serialNext = function(prefix, extension){
  i=0
  repeat {
    f = paste(prefix, i , extension, sep=".")
    if(!file.exists(f)){return(f)}
    i=i+1
  }
}

SetFileName <- function(filename, initials, extension) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  filename <- paste0(out.dir, "/", initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename <- serialNext(filename, extension)
  filename
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {
  
  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)
  
  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }
  
  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())
  
  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]
  
  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }
  
  p <- p +
    theme(panel.background = element_rect(fill = corCol))
  
  return(p)
}


theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
    theme(panel.grid.major = element_line(colour = "#adb5bd", 
                                          size = 0.25),
          strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}
theme_classic_lines_45 <- function() {
  theme_pubr(border = T, 
             legend = "top", 
             x.text.angle = 45) +
    theme(panel.grid.major = element_line(colour = "#adb5bd", 
                                          size = 0.25),
          strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, 
             legend = "top", 
             x.text.angle = 90) +
    theme(strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}

theme_classic_lines_90_noborder <- function() {
  theme_pubr(border = F, 
             legend = "top", 
             x.text.angle = 90) +
    theme(strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}

theme_classic_lines_90_noborder <- function() {
  theme_pubr(border = F, legend = "top", x.text.angle = 90) +
    theme(strip.background = element_rect(fill = "white", color = "white")
    )
}

theme_classic_nolines_90 <- function() {
  theme_pubr(border = T, 
             legend = "top", 
             x.text.angle = 90,) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank(),
          strip.background = element_rect(fill = "white", 
                                          color = "white")
    )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")
# colors_drugs <- c("DMSO (neg control)" = "#65BFA4", "DNA-PKi (NHEJ control)" = "#8CA0C4", "drugs" = "#808184", "Mirin (MMEJ control)" = "#5D79AC")
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
#colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51") 
## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")
colors_pathways = c("wt" = "#808184", "other" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "SSTR" = "#007A4C")

target_color = c(HDAC="#157834", HAT="#C6D0A8", Sirtuin="#802155",
                     HMT="#8EC9AB", DNMT="#9A9932", `Histone Demethylase`="#DECD77",
                     HIF="#E2DBD4", JAK="#CB546F", PIM="#45AB9A", `Aurora Kinase`="#A74694",
                     PARP="#D69C81", `Epigenetic Reader Domain`= "#8ACDEF", `DNA-PK`="#779ed7")

changeLevels <- function(object, column = "chr_r", levels = c(paste0("chr", 1:22), "chrX")) {
  object[, column] <- as.character(object[, column])
  object[, column] <- factor(object[, column], levels)
  object
}
```


## Data import

```{r data import, results = 'hide', warning = FALSE}
# Import data from preprocessing script
file = list.files(in.dir, pattern = "episcreen_ratios", full.names = T) %>% tail(n = 1)
indel.data <- readRDS(file)

# Import raw ratios file
file = list.files(in.dir, pattern = "episcreen_raw_ratios", full.names = T) %>% tail(n = 1)
TIF_ratios_all_tib <- readRDS(file)

# Import data from preprocessing script
file.muts = list.files(in.dir, pattern = "episcreen_mutations", full.names = T) %>% tail(n = 1)

IPR_set = c("IPR5", "IPR7")

mutation.data <- readRDS(file.muts) %>% filter(IPR %in% IPR_set)

# Complete data without exlcluding wells <0.5 viability (I need this for plotting)
meta.viab.file = list.files(in.dir, full.names = T)  %>% .[grepl("Viab_Meta", .)] %>% tail(n = 1)
indel.statistics.dt = readRDS(meta.viab.file) %>%
  dplyr::select(ID, drug, target, replicate, viability, conc_char, tech)

chromatin <- readRDS("files_scripts/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(indel.data, barcode))

map_clones = read_table("files_scripts/cl20220321_cured_mapping_clones.txt", col_names = c("chr", "start", "barcode", "clone")) %>%
  filter(chr != "chrY" & clone == "RSTP2_5") %>%
  mutate(chr = factor(chr, levels = c("chr1", "chr2", "chr3", "chr4", "chr5", 
                                      "chr6", "chr7", "chr8", "chr9", "chr10", 
                                      "chr11", "chr12", "chr13", "chr14", "chr15", 
                                      "chr16", "chr17", "chr18", "chr19", "chr20", 
                                      "chr21", "chr22", "chrX")),
         start = start / 1e6) %>%
  left_join(chromatin) %>%
  filter(barcode != "ACCCCTAAAGGCGCTG")

chromsizes = read_table("files_scripts/rs20220602_chromsizes_hg38_UCSC.txt", col_names = T) %>% 
  filter(chr != "chrY") %>%
  mutate(chr = factor(chr, levels = c("chr1", "chr2", "chr3", "chr4", "chr5", 
                                      "chr6", "chr7", "chr8", "chr9", "chr10", 
                                      "chr11", "chr12", "chr13", "chr14", "chr15", 
                                      "chr16", "chr17", "chr18", "chr19", "chr20", 
                                      "chr21", "chr22", "chrX")),
         start = start / 1e6)

centromeres = read_table("files_scripts/rs20220714_UCSC_centromeres_24501022.txt") %>%
  group_by(chr) %>%
  filter(chr != "chrY") %>%
  mutate(chr = factor(chr, levels = c("chr1", "chr2", "chr3", "chr4", "chr5", 
                                      "chr6", "chr7", "chr8", "chr9", "chr10", 
                                      "chr11", "chr12", "chr13", "chr14", "chr15", 
                                      "chr16", "chr17", "chr18", "chr19", "chr20", 
                                      "chr21", "chr22", "chrX")),
         start = start / 1e6,
         end = end / 1e6) %>%
  summarise(start = min(start), # earliest start of centromere
            end = max(end)) # furthest end of centromere

sample_order <- c("100 nM", "1 µM", "10 µM")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
```

# Main figures
## 1A: Clone 5 heatmap of chromatin values
What is the chromatin environment of each IPR?
```{r 1A clone 5 heatmap, warning = FALSE, fig.height=6, fig.width=8}
domain_dt <- readRDS("files_scripts/xv20200915_DDR_data_CHIP.rds") %>%
  distinct(barcode, chromatin)

bc_IPR = indel.data %>% 
  distinct(IPR, barcode)

domain_matrix = bc_IPR %>% left_join(domain_dt) %>% dplyr::select(IPR, chromatin) %>% column_to_rownames(., var = "IPR")

clone5.matrix = indel.data %>% 
  dplyr::select(IPR, ends_with(".zscore")) %>% 
  distinct() %>% 
  rename_with(., ~gsub(".zscore", "", .x)) %>%
  column_to_rownames(var="IPR")

clone5_map_ipr = hclust(dist(clone5.matrix), method = "ward.D")

# Feature clustering based on whole TRIP pool
TRIP_chrom_tib <- readRDS('files_scripts/cl20201026_ChIP_zscore_selection.RDS') # %>% 
# Remove HDAC1 and 2 as they do not reach higher than the 80th percentile of the HDAC 
# districtubion in the TRIP pool for the reporters in Clone 5. 
# dplyr::select(-HDAC1, -HDAC2)

TRIP_chrom = TRIP_chrom_tib %>% dplyr::select(-binsize, -pool) %>% column_to_rownames("ID")
# make clustering with some column rotation for better visualisation
TRIP_chrom_clust = hclust(dist(t(TRIP_chrom)), method = "ward.D")

## Chromatin groups
group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 
                                    'H3K9me2', 'H3K9me3',
                                    'H3K27me3', 'EZH2', 
                                    'CTCF', 'SMC3',
                                    'HDAC3', 'HDAC2', 'HDAC1', 
                                    'H3K4me1', 'H3K4me2', 'H3K4me3', 'H3K27ac',
                                    'H4K5acK8ac', 'H2AFZ', 
                                    'DNAse', 'Dam', 'm5C',
                                    'H3K79me2', 
                                    'TTseq', 
                                    'H3K36me3', 'POL2AS2',
                                    'POL2'),
                        group=factor(c(rep('repressive', 6), 
                                       # rep('repressive', 4), 
                                       rep('insulator', 2),
                                       rep('HDAC', 3), 
                                       rep('euchromatin', 6),
                                       rep('accessibility', 2), 'methylation',
                                       rep('transcribing', 5)
                                       # rep('transcribing', 4)
                        )))

order_vec=rownames(clustering)[nrow(clustering):1]

### Colors
quant_breaks <- quantile(clone5.matrix, c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 100
max_quant = max(abs(quant_breaks[1]), abs(quant_breaks[2]))

breaks.chip <- c(seq(-quant_breaks[2], 0, length.out=ceiling(palette_length/2) + 1), 
                 seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))


colors = colorRampPalette(rev(brewer.pal(n = 11, name="RdBu")))(palette_length)

pheatmap(t(clone5.matrix)[TRIP_chrom_clust$labels, ], 
         cluster_rows = TRIP_chrom_clust,
         annotation_row=clustering,
         annotation_colors=list(chromatin=group_colors),
         color = colors, 
         breaks = breaks.chip,
         annotation_col = domain_matrix,
         cluster_cols = clone5_map_ipr)
```

## 1B: Ideogram and example
```{r 1B chromideograms, fig.height=5, fig.width=6}
p1 = ggplot() +
  geom_segment(data = chromsizes,
               aes(x = chr, xend = chr, y = 0, yend = start),
               lineend = "round", color = "lightgrey", size = 7) +
  geom_segment(data = map_clones,
               aes(x = as.integer(chr) - 0.3, xend = as.integer(chr) + 0.3,
                   y = start, yend = start, color = chromatin),
               size = 3) +
  geom_segment(data = centromeres,
               aes(x = as.integer(chr) - 0.2, xend = as.integer(chr) + 0.2,
                   y = start, yend = start),
               size = 2) +
  geom_segment(data = centromeres,
               aes(x = as.integer(chr) - 0.2, xend = as.integer(chr) + 0.2,
                   y = end, yend = end),
               size = 2) +
  theme_classic_lines_90_noborder() +
  scale_color_manual(values = group_colors) + 
  xlab("Chromosome") + 
  ylab("Position [Mb]") +
  coord_flip()

label_ratio <- filter(indel.data, IPR %in% IPR_set , drug == "DMSO") %>% 
  group_by(IPR) %>%
  summarise(MMEJratio_mean = mean(MMEJratio),
            MMEJratio_sd = sd(MMEJratio),
            freqCut_mean = mean(freqCut),
            freqCut_sd = sd(freqCut)) %>%
  mutate(label_efficiency = paste0("Total indel frequency = ", round(freqCut_mean*100, 1),"% ±", round(freqCut_sd*100, 1),"%"), 
         label_ratio = paste0("MMEJ ratio = ", round(MMEJratio_mean, 2)," ±", round(MMEJratio_sd, 2)))

bc_mut_tib_mean <- mutation.data %>% 
  distinct(IPR, drug, freq_mean,freq_sd, color, mutation, type) %>%
  filter(IPR %in% IPR_set , drug == "DMSO", type %in% c("core", "grouped_indel")) %>% 
  left_join(label_ratio, by = "IPR")

p2 = ggplot(bc_mut_tib_mean, 
            aes(mutation, freq_mean)) + 
  geom_bar(stat = "identity", aes(fill = color)) + 
  # geom_quasirandom(data = bc_mut_tib, aes(mutation, freq_split, fill = color),  color = "black", shape = 21, ) + 
  scale_fill_manual("legend", values = colors_pathways) + 
  scale_color_manual("legend", values = colors_pathways) + 
  scale_y_continuous(name="frequency", labels = percent) +
  scale_x_discrete(name="indel size", breaks = c("<-14", -7, 0, 1, ">2")) + 
  facet_wrap( ~ IPR, ncol = 1) +
  geom_text(aes(x = 7, y = .7, label = label_efficiency)) +
  geom_text(aes(x = 7, y = .6, label = label_ratio)) + 
  geom_errorbar(aes(ymin=freq_mean-freq_sd, ymax=freq_mean+freq_sd), width=.2,
                position=position_dodge(.9)) +
  theme_classic_lines()

plot_grid(p1, p2, nrow = 1, rel_widths = c(3, 2))
```



## 1C: Drugs per target group
Which drugs did we use for the screen and which targets do they have?
```{r 1C drug targets, warning = FALSE, fig.height=4, fig.width=4}
# B: Drugs per target group
## Plot from viability data table (because there all drugs are included)
targets <- indel.statistics.dt %>% 
  ungroup() %>%
  filter(!drug %in% c("Mirin", "DNA-PKi", "DMSO", "PAO")) %>%
  dplyr::distinct(drug, target) %>% 
  count(target, sort = TRUE)

ggplot(targets, 
       aes(y = reorder(target, n), x = n, fill = target)) + 
  geom_bar(stat = "identity")  + 
  scale_fill_manual(values = target_color) + 
  ylab("") +
  theme_pubr(border = T) 
```

***Figure 1B:*** *Number of probed compounds per target group. HDAC = Histone Deacetylase, JAK = Janus Kinase, PARP = Poly ADP Ribose Polymerase, HMT = Histone methyltransferase, PMT = Protein Methyltransferase, DNMT = DNA Methyltransferase,  HAT = Histone Acetyltransferase, HIF = Hypoxia Inducible Factor*

**Conclusion: Most drugs are HDACs, a lot of compounds do not really target chromatin directly.**

## 1D: Schematic

![Scheme of the experimental setup](figures/rs20220707_figure1A-01.png)

***Figure 1D:*** *Experimental procedure. Clonal K562 cell line with 18 integrated pathway reporters (IPRs) is transfected with a gRNA targeting the reporter and treated with a 180 compounds at three concentrations (100 nM, 1µM and 10µM) and incubated for 72 hours. Cell viability is then measured per well and the sequence of the reporters are then amplified for high throughput sequencing. The sequencing then reveals the indel frequency and pathway balance at the 18 genomic sites per compound and concentration.*

# Supplemental figures

## S1A: Experimental timeline

## S1B: Correlate viability data from the 2 replicates
How well does the viability data correlate between the two biological replicates? Did the drugs become less potent after 3 years in the freezer?
```{r S1B viability correlation, warning = FALSE, fig.height=4, fig.width=8}
# A: Correlate viability data from the 2 replicates
indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  distinct(viab_mean, sample, replicate, conc_char, drug, viab_reproducibility,
           indel_freq_filter, pathway_bal_filter) %>%
  mutate(filter_viab = case_when(pathway_bal_filter & viab_reproducibility ~ "Balance filter", 
                                 indel_freq_filter & viab_reproducibility ~ "TIF filter" , 
                                 T ~ "Discarded"),
         filter_viab = factor(filter_viab, levels = c("Discarded", "TIF filter" , "Balance filter")),
         conc_char = factor(conc_char, levels = c("100 nM", "1 µM", "10 µM"))) %>%
  pivot_wider(names_from = replicate, values_from = viab_mean) %>%
  ggplot(., aes(x = E177, y = E1504, color = filter_viab,  shape = viab_reproducibility)) +
  geom_point(size = 4) +
  scale_shape_manual(values=c(18, 16))+
  geom_smooth(aes(group = 1), method = "lm", color = colors_diverse[1]) +
  # geom_segment(x = 0.5, xend = 0.5, y = 0.5, yend = 1.2, linetype = "dashed") +
  # geom_segment(x = 0.5, xend = 1.2, y = 0.5, yend = 0.5, linetype = "dashed") +
  theme_classic_lines_90() +
  scale_color_manual(values = hcl.colors(4, palette = "Lajolla", rev = F)[2:4]) +
  facet_grid(. ~ conc_char)

indel.data %>%
  filter(sample == "Epigenetic Compound") %>%
  distinct(viab_mean, replicate, conc_char, drug, viab_reproducibility,
           indel_freq_filter, pathway_bal_filter) %>%
  mutate(filter_viab = case_when(pathway_bal_filter & viab_reproducibility ~ "Balance filter", 
                                 indel_freq_filter & viab_reproducibility ~ "TIF filter" , 
                                 T ~ "Discarded"),
         filter_viab = factor(filter_viab, levels = c("Discarded", "TIF filter" , "Balance filter")),
         conc_char = factor(conc_char, levels = c("100 nM", "1 µM", "10 µM"))) %>%
  pivot_wider(names_from = replicate, values_from = viab_mean) %>% 
  write_excel_csv(., "Viability_Filters_S1B.csv")

indel.data  %>% 
  filter(!indel_freq_filter | !viab_reproducibility) %>% # freqCut is NA if sample didn't pass filter for viability.
  distinct(drug, conc_char) %>% # unique compound and concentration combination
  pull(conc_char) %>% 
  table() %>%
  knitr::kable() # count compounds per concentration that are removed

indel.data  %>% 
  filter(!indel_freq_filter) %>% # freqCut is NA if sample didn't pass filter for viability.
  distinct(drug, conc_char) %>% # unique compound and concentration combination
  pull(conc_char) %>% 
  table() 
```

*Figure S1A: Correlation of the cell viability between the two biological replicates. Marked in yellow are the compounds that were removed from the analysis.*

**Conclusion: Most drugs that were toxic in rep 1 were also toxic in rep 2, so the trends are similar. However, some drugs were especially more toxic in rep 2. These drugs were removed from the analysis.**


## S1C: Correlation between viability and TIF
Is lower viability causing a decrease in TIF? Can we justify our 0.50 viability cutoff based on the correlation between viability and TIF?

```{r S1C TIF correlation with viability, warning=F, fig.height=5, fig.width=5}
# F: Correlation between viability and MMEJ ratio
# load indel data that has no filter on it.

TIF_ratios_all_tib %>% distinct(exp, viab_norm_split, freqCut_global, sample) %>% 
  ggplot(., aes(x = viab_norm_split, y = freqCut_global, color = sample)) +
  geom_point_rast(raster.dpi = 300) +
  geom_vline(xintercept = 0.25, color = "red") +
  annotate("rect",xmin=0,xmax=0.25,ymin=-Inf,ymax=Inf, alpha=0.1, fill="red") +
  xlab("viable cell count relative to control") +
  ylab("Global TIF") + 
  scale_color_manual(values = colors_diverse) +
  ylim(0, 1) +
  xlim(0, 1.2)
```

## S1D: Correlation between viability and MMEJ ratio
Is lower viability causing a decrease in MMEJ? Can we justify our 0.50 viability cutoff based on the correlation between viability and MMEJ ratio?

```{r S1D MMEJ ratio correlation with viability, warning=F, fig.height=5, fig.width=5}
TIF_ratios_all_tib %>% distinct(exp, viab_norm_split, MMEJratio_global, sample) %>% 
  ggplot(., aes(x = viab_norm_split, y = log2(MMEJratio_global), color = sample)) +
  geom_point_rast(raster.dpi = 300) +
  geom_vline(xintercept = 0.40, color = "red") +
  annotate("rect",xmin=0,xmax=0.40,ymin=-Inf,ymax=Inf, alpha=0.1, fill="red") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.25), limits = c(0.1, 1.2)) +
  scale_color_manual(values = colors_diverse) +
  xlab("viable cell count relative to control") +
  ylab("log2(global MMEJ:NHEJ)") + 
  xlim(0, 1.2)
```


## S1E: Comparison between technical replicates of efficiency in DMSO

```{r S1E comparison for mean indel rate across replicates, warning = FALSE, fig.height=5, fig.width=3}
indel.data %>%
  filter(drug == "DMSO", conc_char == "control") %>%
  mutate(drug = case_when(drug == "DMSO" & plate %in% c(1:6) ~ paste(drug, drug_plate, well, "1", sep = " "), 
                          drug == "DMSO" & plate %in% c(7:12) ~ paste(drug, drug_plate, well, "2", sep = " "),
                          drug == "DMSO" & plate %in% c(13:18) ~ paste(drug, drug_plate, well, "3", sep = " ")))  %>%
  dplyr::distinct(freqCut, well, drug, drug_plate, conc_char, tech, replicate) %>%
  mutate(freqCut = ave(freqCut, well, drug, drug_plate, conc_char, tech, replicate, FUN = mean)) %>%
  distinct() %>%
  mutate(rep = paste(replicate, "tech", tech, sep = "_")) %>%
  ggplot(.,
         aes(x = rep, y = freqCut)) +
  geom_quasirandom(aes(color = replicate)) +
  scale_color_manual(values = c(colors_drugs[8], colors_drugs[7])) +
  theme_classic_lines_90() +
  ylim(0,1)
```

## S1F: Comparison between replicates of efficiency in DMSO

```{r S1F control distribution of indel rates across wells and replicates, warning = FALSE, fig.height=3, fig.width=5}
indel.data %>%
  filter(drug == "DMSO", conc_char == "control") %>%
  mutate(IPR = factor(IPR, levels = clone5_map_ipr$labels[clone5_map_ipr$order])) %>%
  dplyr::distinct(freqCut, IPR, exp, replicate) %>%
  ggplot(.,
         aes(x = IPR, y = freqCut)) +
  geom_quasirandom(aes(color = replicate),dodge.width = 0.75) +
  scale_color_manual(values = c(colors_drugs[8], colors_drugs[7])) +
  theme_classic_lines_90() +
  ylim(0,1)
```


## S1G: Comparison between replicates MMEJ ratio in DMSO

```{r S1G control distribution of MMEJ ratio across wells and replicates, warning = FALSE, fig.height=6, fig.width=10}
indel.data %>%
  filter(drug == "DMSO", conc_char == "control") %>%
  dplyr::distinct(MMEJratio, IPR, exp, replicate) %>%
  mutate(IPR = factor(IPR, levels = clone5_map_ipr$labels[clone5_map_ipr$order])) %>%
  ggplot(.,
         aes(x = IPR, y = MMEJratio)) +
  geom_quasirandom(aes(color = replicate), dodge.width = 0.75) +
  scale_color_manual(values = c(colors_drugs[8], colors_drugs[7])) +
  theme_classic_lines_90()
```

## S1H: Combined TIF z-scores per IPR per concentration
```{r S1H TIF zscores per IPR, fig.height=3, fig.width=8}
zscore_dt = indel.data %>%
  filter(sample %in% c("Epigenetic Compound") & indel_freq_filter) %>%
  distinct(freqCut, IPR, drug, conc_char, tech, sample, replicate, 
           barcode, zscore_bal_comb_rep, zscore_TIF_comb_rep) %>%
  mutate(sample_conc = factor(conc_char, levels = sample_order),
         IPR = factor(IPR, levels = clone5_map_ipr$labels[clone5_map_ipr$order]))

ggplot(zscore_dt,
       aes(x = IPR, y = zscore_TIF_comb_rep, color = sample_conc)) +
  geom_quasirandom_rast(aes(color = sample_conc), dodge.width = 0.75, raster.dpi = 300, size = 1) +
  scale_color_manual(values = conc_colors, guide = "none") +
  geom_hline(linetype = "dashed", yintercept = 2.58) +
  geom_hline(linetype = "dashed", yintercept = -2.58) +
  theme_classic_nolines_90() + 
  ylab("TIF z-score")
```

## S1I: Combined balance z-scores per IPR per concentration
```{r S1I bal zscore per IPR, fig.height=3, fig.width=8}
ggplot(zscore_dt,
       aes(x = IPR, y = zscore_bal_comb_rep, color = sample_conc)) +
  geom_quasirandom_rast(aes(color = sample_conc), dodge.width = 0.75, raster.dpi = 300, size = 1) +
  scale_color_manual(values = conc_colors, guide = "none") +
  geom_hline(linetype = "dashed", yintercept = 2.58) +
  geom_hline(linetype = "dashed", yintercept = -2.58) +
  theme_classic_nolines_90() + 
  ylab("MMEJ:NHEJ z-score")
```

## S1J: Correlation between technical replicates efficiency
How well does the cutting efficiency correlate between the six technical replicates
```{r S1J normalized TIF correlation, warning = FALSE, fig.height=5, fig.width=5}
# C: Correlation between technical replicates efficiency
techn_repl <- indel.data %>%
  filter(sample == "Epigenetic Compound" & 
           indel_freq_filter == T & 
           viab_reproducibility == T,
         filter_zscore_TIF) %>%
  distinct(freqCut_diff, drug, conc_char, replicate, tech, IPR) %>%
  mutate(rep = paste(replicate, tech, sep = "_")) %>%
  dplyr::select(-replicate, -tech) %>%
  spread(rep, freqCut_diff) %>%
  na.omit()

boundaries <- c(0.6,0.7,0.8,0.90)

ggpairs(techn_repl %>% dplyr::select(contains("E1")),
        upper = list(continuous = corColor),
        lower = list(continuous = function(data, mapping, ...) {
          ggally_points(data = data, mapping = mapping, alpha = 0.6, size = 0.5) +
            #   scale_y_continuous(breaks = seq(0.8, 1.2, by = 0.2), limits = c(0.7, 1.2)) +
            #   scale_x_continuous(breaks = seq(0.8, 1.2, by = 0.2), limits = c(0.7, 1.2)) +
            geom_abline(slope = 1, lty = "twodash", col = "grey") +
            geom_smooth(method = "lm", color = "red", lty = "dashed", size = 0.5)}),
        diag = list(continuous = function(data, mapping, ...) {
          ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("deltaTIF") +
  ylab("deltaTIF")
```

*Figure S1J: Correlation of the cutting efficiency between all technical replicates.*

**Conclusion: Only for replicate two the cutting efficiency correlates well. We need to further investigate why replicate 1 does not correlate well.**

## S1K: Correlation between technical replicates MMEJ score
How well does the MMEJ score correlate between the six technical replicates?
```{r S1K normalized MMEJ ratio correlation, warning = FALSE, fig.height=5, fig.width=5}
# B: Correlation between technical replicates MMEJ score
techn_repl <- indel.data %>%
  filter(sample == "Epigenetic Compound" & 
           pathway_bal_filter & 
           viab_reproducibility &
           filter_zscore_bal_comb_rep) %>%
  mutate(rep = paste(replicate, tech, sep = "_")) %>%
  dplyr::select(IPR, rep, MMEJ_log2_FC, drug, conc_char) %>%
  spread(rep, MMEJ_log2_FC) %>%
  na.omit()

ggpairs(techn_repl %>% dplyr::select(contains("E1")),
        upper = list(continuous = corColor),
        lower = list(continuous = function(data, mapping, ...) {
          ggally_points(data = data, mapping = mapping, alpha = 0.6, size = 0.5) +
            #   scale_y_continuous(breaks = seq(-2, 1, by = 1), limits = c(-2, 0.5)) +
            #   scale_x_continuous(breaks = seq(-2, 1, by = 1), limits = c(-2, 0.5)) +
            geom_abline(slope = 1, lty = "twodash", col = "grey") +
            geom_smooth(method = "lm", color = "red", lty = "dashed", size = 0.5)}),
        diag = list(continuous = function(data, mapping, ...) {
          ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("MMEJ ratio change [Log2FC]") +
  ylab("MMEJ ratio change [log2FC]")
```

*Figure S1K: Correlation of the MMEJ score between all technical replicates.*

**Conclusion: The MMEJ scores correlate relatively well, especially within biological replicates. Even the correlation between biological replicates is good.*


## S1L-O: Correlation in MMEJ z-score between biological replicates
How well does the MMEJ z-score correlate between the biologcial replicates before combining them?
```{r S1LMNO MMEJ z-score correlation, warning = FALSE, fig.height=5, fig.width=5}
# D: Correlation between biological replicates MMEJ score
biol_repl_TIF <- indel.data %>%
  filter(sample == "Epigenetic Compound" & 
           viab_reproducibility == TRUE &
           filter_zscore_TIF_comb_rep &
           !is.na(zscore_TIF_tech)) %>%
  dplyr::distinct(barcode, zscore_TIF_tech, 
                  filter_zscore_TIF_comb_rep, 
                  drug, replicate, conc_char) %>%
  pivot_wider(names_from = replicate, 
              values_from = zscore_TIF_tech) %>%
  na.omit()

biol_repl_bal <- indel.data %>%
  filter(sample == "Epigenetic Compound" & 
           viab_reproducibility == TRUE & 
           filter_zscore_bal_comb_rep &
           !is.na(zscore_bal_tech)) %>%
  dplyr::distinct(zscore_bal_tech, barcode,
                  filter_zscore_bal_comb_rep, 
                  drug, replicate, conc_char) %>%
  pivot_wider(names_from = replicate, 
              values_from = zscore_bal_tech) %>%
  na.omit()

# D: Correlation between biological replicates MMEJ score
biol_repl_TIF_glob <- indel.data %>%
  filter(sample == "Epigenetic Compound" & 
           viab_reproducibility == TRUE & 
           filter_zscore_TIF_global_comb_rep &
           !is.na(zscore_TIF_global_tech)) %>%
  dplyr::distinct(zscore_TIF_global_tech, 
                  drug, replicate, conc_char) %>%
  pivot_wider(names_from = replicate, 
              values_from = zscore_TIF_global_tech) %>%
  na.omit()

biol_repl_bal_glob <- indel.data %>%
  filter(sample == "Epigenetic Compound" & 
           viab_reproducibility == TRUE & 
           filter_zscore_bal_global_comb_rep &
           !is.na(zscore_bal_global_tech)) %>%
  dplyr::distinct(zscore_bal_global_tech, 
                  drug, replicate, conc_char) %>%
  pivot_wider(names_from = replicate, 
              values_from = zscore_bal_global_tech) %>%
  na.omit()

p1 = ggplot(biol_repl_TIF,
            aes(x = E1504, y = E177)) +
  geom_point() +
  geom_abline(lty = "twodash") +
  geom_smooth(method = "lm", color = "red", lty = "dashed") +
  stat_cor(method = "pearson") +
  xlab("Screen 2 TIF ratio z-score") +
  ylab("Screen 1 TIF ratio z-score")  +
  theme_classic(base_size = 16) + 
  geom_hline(yintercept = 1.96, linetype = "dashed") +
  geom_hline(yintercept = -1.96, linetype = "dashed") +
  geom_vline(xintercept = 1.96, linetype = "dashed") +
  geom_vline(xintercept = -1.96, linetype = "dashed")


p2 = ggplot(biol_repl_bal,
            aes(x = E1504, y = E177)) +
  geom_point() +
  geom_abline(lty = "twodash") +
  geom_smooth(method = "lm", color = "red", lty = "dashed") +
  stat_cor(method = "pearson") +
  xlab("Screen 2 MMEJ ratio z-score") +
  ylab("Screen 1 MMEJ ratio z-score")  +
  theme_classic(base_size = 16) + 
  geom_hline(yintercept = 1.96, linetype = "dashed") +
  geom_hline(yintercept = -1.96, linetype = "dashed") +
  geom_vline(xintercept = 1.96, linetype = "dashed") +
  geom_vline(xintercept = -1.96, linetype = "dashed")


p3 = ggplot(biol_repl_TIF_glob,
            aes(x = E1504, y = E177)) +
  geom_point() +
  geom_abline(lty = "twodash") +
  geom_smooth(method = "lm", color = "red", lty = "dashed") +
  stat_cor(method = "pearson") +
  xlab("Screen Global 2 TIF ratio z-score") +
  ylab("Screen Global 1 TIF ratio z-score")  +
  theme_classic(base_size = 16) + 
  geom_hline(yintercept = 1.96, linetype = "dashed") +
  geom_hline(yintercept = -1.96, linetype = "dashed") +
  geom_vline(xintercept = 1.96, linetype = "dashed") +
  geom_vline(xintercept = -1.96, linetype = "dashed")


p4 = ggplot(biol_repl_bal_glob,
            aes(x = E1504, y = E177)) +
  geom_point() +
  geom_abline(lty = "twodash") +
  geom_smooth(method = "lm", color = "red", lty = "dashed") +
  stat_cor(method = "pearson") +
  xlab("Screen 2 Global MMEJ ratio z-score") +
  ylab("Screen 1 Global MMEJ ratio z-score") +
  theme_classic(base_size = 16) + 
  geom_hline(yintercept = 1.96, linetype = "dashed") +
  geom_hline(yintercept = -1.96, linetype = "dashed") +
  geom_vline(xintercept = 1.96, linetype = "dashed") +
  geom_vline(xintercept = -1.96, linetype = "dashed")

plot_grid(p1, p2, p3, p4, ncol = 2, nrow = 2, labels = c("L", "M", "N", "O"))
```

*Figure S1L: Correlation of the MMEJ z-scores between the two biological replicates.*

Conclusion: MMEJ z-scores correlate well.

*Figure S1M: Correlation of the cutting efficiencies between the two biological replicates.*

**Conclusion: Cutting efficiencies do not correlate well.**

```{r export}
# setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/")
# filename <- SetFileName("_indel.data", "mt")
# save(indel.data, file = filename)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

