---
title: "Validating hits from the epigenetic screening"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code.folding: show
    highlight: monochrome
    theme: journal
    toc: yes
    toc_float: yes
  word_document:
    toc: yes
  pdf_document:
    toc: yes
  editor_options:
    chunk_output_type: inline
---

# knitr document van Steensel lab

# Epigenetic Screening

# Introduction
This script tries to find out which drugs are able to significantly alter indel pattern formation in RSTP2#5 cells in different chromatin contexts. In the end, the aim is to couple biological relevance to these changes.

## Description of Data
How to make a good rendering table: 
```{r table1, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
| column1	|	column2	|	column3	|	
|----|----|----|
|1	|	2	|	3	|	
|a	|	b	|	c	|	
"
cat(tabl) # output the table in a format good for HTML/PDF/docx conversion
```

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, echo = FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(dplyr)
library(outliers)
library(ggplot2)
library(ggbeeswarm)
library(data.table)
library(ComplexHeatmap)
library(circlize)
library(pheatmap)
library(gridGraphics)
library(gridExtra)
library(RColorBrewer)
library(Pigengene)
library(tidyr)
library(VennDiagram)
library(ggforce)
library(tibble)
library(ggpubr)
library(Laurae)
```

### Custom functions
Functions used thoughout this script.
```{r functions, echo = FALSE, warning = FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}
```


## Data import
```{r Data Import, echo = FALSE, warning = FALSE}
# Import data from 5_HitValidation_DrugScreen.Rmd
setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/data/files_scripts/")
mapping_mean <- get(load("mt20190510_mapping_mean"))
indel.data <- get(load("mt20190502_indel.data"))
mapping_mean_long <- get(load("mt20190510_mapping_mean_long"))
chip_dendogram <- get(load("mt20190510_chip_dendogram"))
mapping_mean_all <- get(load("mt20190329_mapping_mean_all"))
barcode.order <- read.table("barcode_order.txt", header = T)

setwd("/DATA/projects/DSBrepair/scratch/")
delay.df <- read.table("cl20190329_minus7_delay.tsv", header = T)
```


## Compute relevant logratios and divide per concentration
```{r input_prep, echo = FALSE, warning = FALSE}
# Compute mean of logratio per barcode and drug
indel.data$mean.logratio.bc.drug <- 
  ave(indel.data$logratio.platenorm, 
      indel.data$barcode, indel.data$conc, 
      indel.data$Drug, FUN = function(x) mean(x))

# Compute mean of logratio per barcode and target
indel.data$mean.logratio.bc.target <- 
  ave(indel.data$logratio.platenorm, 
      indel.data$barcode, indel.data$conc, 
      indel.data$Target, FUN = function(x) mean(x))

# Compute mean of t-statistic per barcode and target
indel.data$mean.statistic.bc.target <- 
  ave(indel.data$statistic, 
      indel.data$barcode, indel.data$conc, 
      indel.data$Target, FUN = function(x) mean(x))

# Compute mean of t-statistic per barcode and drug
indel.data$mean.statistic.bc.drug <- 
  ave(indel.data$statistic, 
      indel.data$barcode, indel.data$conc,
      indel.data$Drug, FUN = function(x) mean(x))

# Compute mean of efficiency per barcode and drug
indel.data$mean.eff.bc.drug <- 
  ave(indel.data$efficiency.platenorm, 
      indel.data$barcode, indel.data$conc,
      indel.data$Drug, FUN = function(x) mean(x))


# Compute mean and sd of barcode counts per barcode and drug
indel.data$mean.count.bc.drug <- 
  ave(indel.data$rel.counts, 
      indel.data$barcode, indel.data$conc,
      indel.data$Drug, FUN = function(x) mean(x))
indel.data$sd.count.bc.drug <- 
  ave(indel.data$rel.counts, 
      indel.data$barcode, indel.data$conc,
      indel.data$Drug, FUN = function(x) sd(x))

# Split up per concentration
indel.data.high <- indel.data[grep("10um", indel.data$condition),]
indel.data.med <- indel.data[grep("1um", indel.data$condition),]
indel.data.low <- indel.data[grep("100nm", indel.data$condition),]
```





## Bring dataframe to heatmap plotting format
```{r transform_1, echo = FALSE, warning = FALSE}
# Select columns and transform to wide format - efficiency heatmap
## High conc
indel.data.high.eff <- indel.data.high %>% select(barcode, Drug, mean.eff.bc.drug) %>% 
  unique() %>% dcast(barcode ~ Drug, value.var = "mean.eff.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

## Med conc
indel.data.med.eff <- indel.data.med %>% select(barcode, Drug, mean.eff.bc.drug) %>% 
  unique() %>% dcast(barcode ~ Drug, value.var = "mean.eff.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

## Low conc
indel.data.low.eff <- indel.data.low %>% select(barcode, Drug, mean.eff.bc.drug) %>% 
  unique() %>% dcast(barcode ~ Drug, value.var = "mean.eff.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")



# Select columns and transform to wide format - barcode counts
## High conc
indel.data.high.count <- indel.data.high[indel.data.high$sd.count.bc.drug < 1,] %>%
  select(barcode, Drug, mean.count.bc.drug) %>% unique() %>% 
  dcast(barcode ~ Drug, value.var = "mean.count.bc.drug") %>% filter_all(any_vars(!is.na(.))) %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

# indel.data.high.count <- indel.data.high.count[,-118]

## Med conc
indel.data.med.count <- indel.data.med[indel.data.med$sd.count.bc.drug < 1,] %>%
  select(barcode, Drug, mean.count.bc.drug) %>% unique() %>% 
  dcast(barcode ~ Drug, value.var = "mean.count.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

# indel.data.med.count <- indel.data.med.count[,-146]

## Low conc
indel.data.low.count <- indel.data.low[indel.data.low$sd.count.bc.drug < 1,] %>%
  select(barcode, Drug, mean.count.bc.drug) %>% unique() %>% 
  dcast(barcode ~ Drug, value.var = "mean.count.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

# indel.data.low.count <- indel.data.low.count[,-158]



# Same as above, but using t-statistic data
indel.data.high.drug.st <- indel.data.high %>% select(barcode, Drug, mean.statistic.bc.drug) %>%
  unique() %>% dcast(barcode ~ Drug, value.var="mean.statistic.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

indel.data.med.drug.st <- indel.data.med %>% select(barcode, Drug, mean.statistic.bc.drug) %>%
  unique() %>% dcast(barcode ~ Drug, value.var="mean.statistic.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

indel.data.low.drug.st <- indel.data.low %>% select(barcode, Drug, mean.statistic.bc.drug) %>%
  unique() %>% dcast(barcode ~ Drug, value.var="mean.statistic.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")



# Same as above, but using logratio data for each barcode vs each drug
indel.data.high.drug <- indel.data.high %>% select(barcode, Drug, mean.logratio.bc.drug) %>%
  unique() %>% dcast(barcode ~ Drug, value.var="mean.logratio.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

indel.data.med.drug <- indel.data.med %>% select(barcode, Drug, mean.logratio.bc.drug) %>%
  unique() %>% dcast(barcode ~ Drug, value.var="mean.logratio.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

indel.data.low.drug <- indel.data.low %>% select(barcode, Drug, mean.logratio.bc.drug) %>%
  unique() %>% dcast(barcode ~ Drug, value.var="mean.logratio.bc.drug") %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

```



## Annotate heatmap columns and rows
```{r}
# Row annotation: annotate drugs with the target group
target <- indel.data[indel.data$Target != c("Negative Control", "MRN", "DNA-PK"),] %>% 
  select(Drug, Target) %>% unique() %>% remove_rownames %>% 
  column_to_rownames(var="Drug")

# Add type of chr. mark (repressive or active)
type <- mapping_mean_long %>% select(variable, type) %>% unique() %>%
  remove_rownames %>% column_to_rownames(var="variable")

# Add annotation of barcodes (based on clustering in 0_Clone5_Annotation.Rmd)
barcode.annotation <- unique(as.data.frame(indel.data.high$barcode))

## Manually added these tags: check/rewrite this
barcode.annotation$cluster[c(18,2)] <- "Active Promoter/Enhancer"
barcode.annotation$cluster[c(12,8)] <- "H3K9me2/3 Domain"
barcode.annotation$cluster[c(3,14,4,17)] <- "Active Gene Body"
barcode.annotation$cluster[c(5,13,6,15)] <- "LAD"
barcode.annotation$cluster[c(7,16)] <- "No Marks Present"
barcode.annotation$cluster[c(1,10,9,11)] <- "H3K27me3 Domain"

barcode.annotation <- barcode.annotation %>% remove_rownames() %>%
  column_to_rownames("indel.data.high$barcode")


# Change colors of annotations
colors <- brewer.pal(length(unique(barcode.annotation$cluster)), "Pastel2")
names(colors) <- unique(barcode.annotation$cluster)

annotation_colors_scr = list(
  cluster = colors,
  type = c(Active="#669966", Repressive="grey90"),
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43"
             ))



# Order the annotation legend
ordered_marks <- c("Active Promoter/Enhancer", "Active Gene Body", "No Marks Present", "H3K27me3 Domain",  "H3K9me2/3 Domain", "LAD")
annotation_colors_scr$cluster <- annotation_colors_scr$cluster[ordered_marks]

ordered_marks2 <- c("DNA-PK","Epigenetic Reader Domain", "HDAC",  "Sirtuin", "HAT", 
                    "Histone Demethylase", "HMT", "DNMT", 
                    "HIF", "JAK", "PIM",
                    "Aurora Kinase", "PARP")
annotation_colors_scr$target <- annotation_colors_scr$target[ordered_marks2]
```



Next, I want to create heatmaps to visualize the effect of each individual drug on each of the 18 barcodes.


\newpage
# Results
## logratio heatmap of each individual drug vs. barcode
```{r visualization each drug, out.width= "100%", fig.align= "center", echo=FALSE}

## Every drug
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(-1.5,1.5,0.03)

# pheatmap function: logratio heatmap of each individual drug vs. barcode - 10uM drugs added
pheatmap(as.matrix(indel.data.high.drug),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, 
         cluster_rows = chip_dendogram$tree_col, 
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"))
```
## This heatmap contains too much information -> We need to decrease the number of drugs
\newpage
## Only drugs that change the logratio
```{r visualization essential drugs, out.width= "100%", fig.align= "center", echo=FALSE}
## Only drugs that change ratio
# pheatmap function: without DNA-PKi: logratio heatmap of eac a certain fold change
indel.data.high.drug.ess <- indel.data.high.drug[]
indel.data.high.drug.ess[is.na(indel.data.high.drug.ess)] <- 0
indel.data.high.drug.ess <- 
  indel.data.high.drug.ess[, 
                          colSums(indel.data.high.drug.ess > 1.5) >= 1 |
                            colSums(indel.data.high.drug.ess < -0.5) >= 1]
# Keeping the scale in the pheatmap function
myBreaks2 <- c(seq(-0.5, 0, 
                   length.out=ceiling(100/2) + 1), 
              seq(9/100, 
                  9/2, length.out=floor(100/2)))


pheatmap(as.matrix(indel.data.high.drug.ess),
         annotation_col = target,
         annotation_row = barcode.annotation, annotation_legend = T,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks2, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, 
         fontsize = 14, cellwidth = 14, cellheight = 14,
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "Log2(+1/-7 ratio) Change - 10 uM of Drugs Added")


# pheatmap function: without DNA-PKi: logratio heatmap of each individual drug vs. barcode - 1uM drugs added- include only drugs that achieve a certain fold change
indel.data.med.drug.ess <- indel.data.med.drug[]
indel.data.med.drug.ess <- 
  indel.data.med.drug.ess[, 
                          colSums(indel.data.med.drug.ess > 1) >= 1 |
                            colSums(indel.data.med.drug.ess < -0.7) >= 1] 

myBreaks3 <- c(seq(min(as.matrix(indel.data.med.drug.ess)), 0, 
                   length.out=ceiling(100/2) + 1), 
              seq(max(as.matrix(indel.data.med.drug.ess))/100, 
                  max(as.matrix(indel.data.med.drug.ess))/2, length.out=floor(100/2)))


pheatmap(as.matrix(indel.data.med.drug.ess), 
         annotation_col = target,
         annotation_row = barcode.annotation,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks3, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, annotation_legend = F,
         fontsize = 14, cellwidth = 14, cellheight = 14,
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "Log2(+1/-7 ratio) change - 1 uM of Drugs Added")


# pheatmap function: without DNA-PKi: logratio heatmap of each individual drug vs. barcode - 100nM drugs added - include only drugs that achieve a certain fold change
indel.data.low.drug.ess <- indel.data.low.drug[]
indel.data.low.drug.ess <- 
  indel.data.low.drug.ess[, 
                          colSums(indel.data.low.drug.ess > 1) >= 1 |
                            colSums(indel.data.low.drug.ess < -0.5) >= 1]

myBreaks4 <- c(seq(min(as.matrix(indel.data.low.drug.ess)), 0, 
                   length.out=ceiling(100/2) + 1), 
              seq(max(as.matrix(indel.data.low.drug.ess))/100, 
                  max(as.matrix(indel.data.low.drug.ess))/2, length.out=floor(100/2)))


pheatmap(as.matrix(indel.data.low.drug.ess), 
         annotation_col = target, 
         annotation_row = barcode.annotation,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks4, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, annotation_legend = F,
         fontsize = 14, cellwidth = 14, cellheight = 14,
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "Log2(+1/-7 ratio) Change - 100 nM of Drugs Added")

```


```{r visualization efficiency, out.width= "100%", fig.align= "center", echo=FALSE}
# Efficiency change heatmap - 10 uM
myBreaks5 <- seq(-0.1, 0.1, 0.002)

pheatmap(as.matrix(indel.data.high.eff[,-36]), fontsize_col = 8, fontsize_row = 8, cellheight = 8,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks5, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, 
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "efficiency change - 10uM of drugs added")


# Efficiency change heatmap - 10 uM - only essentials
indel.data.high.eff.ess <- indel.data.high.eff
indel.data.high.eff.ess <- 
  indel.data.high.eff.ess[, 
                          colSums(indel.data.high.eff.ess > 0.08) >= 1 |
                            colSums(indel.data.high.eff.ess < -0.2) >= 1]

myBreaks6 <- c(seq(min(as.matrix(indel.data.high.eff.ess))/1.5, 0, 
                   length.out=ceiling(100/2) + 1), 
              seq(max(as.matrix(indel.data.high.eff.ess))/100, 
                  max(as.matrix(indel.data.high.eff.ess))/1.5, length.out=floor(100/2)))

pheatmap(as.matrix(indel.data.high.eff.ess),
         annotation_col = target,   
         annotation_row = barcode.annotation,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks5, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, annotation_legend = F,
         cellwidth = 14, cellheight = 14,
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "efficiency change - 10uM of drugs added")


# Efficiency change heatmap - 1 uM
indel.data.med.eff.ess <- indel.data.med.eff
indel.data.med.eff.ess <- 
  indel.data.med.eff.ess[, 
                         colSums(indel.data.med.eff.ess > 0.0666) >= 1 | 
                           colSums(indel.data.med.eff.ess < -0.1333) >= 1]

myBreaks7 <- c(seq(min(as.matrix(indel.data.med.eff.ess))/1.5, 0, 
                   length.out=ceiling(100/2) + 1), 
              seq(max(as.matrix(indel.data.med.eff.ess))/100, 
                  max(as.matrix(indel.data.med.eff.ess))/1.5, length.out=floor(100/2)))

pheatmap(as.matrix(indel.data.med.eff.ess), 
         annotation_col = target, 
         annotation_row = barcode.annotation,annotation_legend = F,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks5, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, 
         cellwidth = 14, cellheight = 14,
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "efficiency change - 1uM of drugs added")


# Efficiency change heatmap - 100 nM
indel.data.low.eff.ess <- indel.data.low.eff
indel.data.low.eff.ess <- 
  indel.data.low.eff.ess[, 
                         colSums(indel.data.low.eff.ess > 0.0666) >= 1 |
                           colSums(indel.data.low.eff.ess < -0.1333) >= 1]

myBreaks8 <- c(seq(min(as.matrix(indel.data.low.eff.ess))/1.5, 0, 
                   length.out=ceiling(100/2) + 1), 
              seq(max(as.matrix(indel.data.low.eff.ess))/100, 
                  max(as.matrix(indel.data.low.eff.ess))/1.5, length.out=floor(100/2)))

pheatmap(as.matrix(indel.data.low.eff.ess), 
         annotation_col = target,
         annotation_row = barcode.annotation, annotation_legend = F,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks5, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, 
         cellwidth = 14, cellheight = 14,
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "efficiency change - 100nM of drugs added")
```
\newpage
## Read count heatmap
```{r out.width= "100%", fig.align= "center", echo=FALSE}
## Only drugs that change ratio
# pheatmap function: count heatmap, selecting only 
# indel.data.high.count <- na.omit(indel.data.high.count)
# indel.data.high.count.ess <- 
#   indel.data.high.count[, 
#                          colSums(indel.data.high.count > 1.666) >= 1 |
#                            colSums(indel.data.high.count < 0.666) >= 1]
# 
# 
# 
# # Keeping the scale in the pheatmap function
# myBreaks22 <- c(seq(min(as.matrix(indel.data.high.count.ess)), 1, 
#                 length.out=ceiling(100/2) + 1), 
#               seq(1.001, max(as.matrix(indel.data.high.count.ess)), 
#                   length.out=floor(100/2)))
# 
# pheatmap(as.matrix(indel.data.high.count.ess),
#          # annotation_col = target, border_color = "grey60",
#          # annotation_row = barcode.annotation, annotation_legend = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "YlOrRd"))(100),
#          breaks = myBreaks22, # cluster_rows = chip_dendogram$tree_col,
#          annotation_colors = annotation_colors_scr, 
#          main = "barcode count relative to DMSO - 10uM of drugs added")
# 
# 
# # pheatmap function: count heatmap, selecting only 
# indel.data.med.count <- na.omit(indel.data.med.count)
# indel.data.med.count.ess <- 
#   indel.data.med.count[, 
#                          colSums(indel.data.med.count > 1.666) >= 1 |
#                            colSums(indel.data.med.count < 0.666) >= 1]
# 
# 
# 
# # Keeping the scale in the pheatmap function
# myBreaks23 <- c(seq(min(as.matrix(indel.data.med.count.ess)), 1, 
#                 length.out=ceiling(100/2) + 1), 
#               seq(1.001, max(as.matrix(indel.data.med.count.ess)), 
#                   length.out=floor(100/2)))
# 
# pheatmap(as.matrix(indel.data.med.count.ess),
#          annotation_col = target, border_color = "grey60",
#          annotation_row = barcode.annotation, annotation_legend = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "YlOrRd"))(100),
#          breaks = myBreaks23, # cluster_rows = chip_dendogram$tree_col,
#          annotation_colors = annotation_colors_scr, 
#          main = "barcode count relative to DMSO - 1uM of drugs added")
# 
# # pheatmap function: count heatmap, selecting only 
# indel.data.low.count <- na.omit(indel.data.low.count)
# indel.data.low.count.ess <- 
#   indel.data.low.count[, 
#                          colSums(indel.data.low.count > 1.666) >= 1 |
#                            colSums(indel.data.low.count < 0.666) >= 1]
# 
# 
# 
# # Keeping the scale in the pheatmap function
# myBreaks24 <- c(seq(min(as.matrix(indel.data.low.count.ess)), 1, 
#                 length.out=ceiling(100/2) + 1), 
#               seq(1.001, max(as.matrix(indel.data.low.count.ess)), 
#                   length.out=floor(100/2)))
# 
# pheatmap(as.matrix(indel.data.low.count.ess),
#          annotation_col = target, border_color = "grey60",
#          annotation_row = barcode.annotation, annotation_legend = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "YlOrRd"))(100),
#          breaks = myBreaks24, # cluster_rows = chip_dendogram$tree_col,
#          annotation_colors = annotation_colors_scr, 
#          main = "barcode count relative to DMSO - 100nM of drugs added")
```
\newpage
## Heatmap only HDACs
```{r}
# Only select HDAC data, 1 uM concentration
# data2 = logratio_high, data3 = logratio_med, data4 = logratio_low, 
# data5 = efficiency_high, data6 = efficiency_med, data7 = efficiency_low
hdac.logratio <- data7[data7$Target=="HDAC",]
hdac.logratio <- hdac.logratio[,-19]
hdac.logratio <- t(hdac.logratio)

# Generate pheatmap
# Use myBreaks5 for efficiency heatmap, myBreaks1 for logratio heatmap
pheatmap(as.matrix(hdac.logratio), 
         annotation_col = target, 
         annotation_row = barcode.annotation,annotation_legend = F,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks5, cluster_rows = chip_dendogram$tree_col,
         annotation_colors = annotation_colors_scr, 
         cellwidth = 14, cellheight = 14,
         labels_row = c("Integration 15","Integration 1","Integration 5","Integration 6",
                        "Integration 10","Integration 9","Integration 13","Integration 3",
                        "Integration 17","Integration 16","Integration 18", "Integration 4",
                        "Integration 11","Integration 7","Integration 12", "Integration 14",
                        "Integration 8","Integration 2"),
         main = "logratio change - 1uM of drugs added - HDACs only")
```


\newpage
## Efficiency/logratio overlay
```{r visualization overlay efficiency logratio, out.width= "100%", fig.align= "center", echo=FALSE}
# Draw venn diagram to see overlaps between efficiency and logratio outliers
## Calculate the intersections
colnames.logr <- as.data.frame(colnames((indel.data.med.drug.ess)))
setnames(colnames.logr, "colnames((indel.data.med.drug.ess))", "drug")

colnames.eff <- as.data.frame(colnames((indel.data.med.eff.ess)))
setnames(colnames.eff, "colnames((indel.data.med.eff.ess))", "drug")

targets <- target.low[target.low$Target == "HDAC",]
targets <- targets[,-2]

colnames.logr <- unique(colnames.logr[colnames.logr$drug %in% targets,])
colnames.eff <- unique(colnames.eff[colnames.eff$drug %in% targets,])

z <- intersect(colnames.logr, colnames.eff)


# Draw the diagrams
## HDACs overlap
draw.pairwise.venn(length(colnames.logr), 
                   length(colnames.eff), 
                   length(z),
                   category = c("Drugs with altered logratio", "Drugs with altered efficiency"), 
                   fill = c("blue", "red"), cat.pos = c(2,2))




```



Next step is to import the chromatin data to generate a correlation matrix between the logratio change of each drug and the chromatin data. A correlation is calculated by comparing logratio changes of drug A for each barcode with ChIP data of ChIP dataset A for each barcode etc.

## Correlation between ChIP data and logratio changes
```{r correlation chip logratio, out.width= "100%", fig.align= "center", echo=FALSE}
# Import chromatin data
mapping_mean_chip <- mapping_mean
mapping_mean_chip <- tibble::rownames_to_column(as.data.frame(mapping_mean_chip), "ChIP")
chip_data <- melt(mapping_mean_chip, variable.name="barcode", value.name="mean")
chip_data <- dcast(chip_data, barcode ~ ChIP, value.var = "mean")
chip_data <- chip_data[chip_data$barcode != 
                         "ACCCCTAAAGGCGCTG" & chip_data$barcode != "CTCTTAATCGCTGCC", ]
chip_data2 <- chip_data

## In case I want to include the mean.ratio
mean.ratio <- indel.data[indel.data$Drug == "DMSO",]
mean.ratio <- mean.ratio[,c(1,15)]
mean.ratio$ratio <- ave(mean.ratio$ratio, mean.ratio$barcode, FUN = function(x) mean(x))
mean.ratio <- unique(mean.ratio)
chip_data <- merge(chip_data, mean.ratio)


## In case I want to include MMEJ delay 
delay.data <- delay.df[delay.df$clone %in% "5",]
delay.data$fold_dif.mean <- ave(delay.data$fold_difference, delay.data$barcode, FUN= function(x) mean(x))
delay.data <- delay.data[,c(1,7)]
delay.data <- delay.data[!duplicated(delay.data),]
delay.data <- delay.data[c(-1,-11),]
chip_data <- merge(chip_data, delay.data)

chip_data <- chip_data %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

## Ratio is influenced by which ChIP datasets?
mean.ratio2 <- mean.ratio
mean.ratio2 <- mean.ratio2 %>% 
  remove_rownames %>% column_to_rownames(var="barcode")
chip_data2 <- chip_data2 %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

cor_ratio <- cor(mean.ratio2, chip_data2)
chip_data3 <- chip_data2[,14,drop = F]
chip_data3 <- cbind(chip_data3, mean.ratio2)
cor_ratio_test <- cor.test(chip_data3$LMNB1, chip_data3$ratio)
cor_ratio <- as.data.frame(cor_ratio)
cor_ratio <- t(cor_ratio)
cor_ratio <- tibble::rownames_to_column(as.data.frame(cor_ratio), "ChIP")

ggplot(cor_ratio, aes(x = reorder(ChIP,ratio, function(x) -x), y = ratio, fill=ratio)) +
  geom_bar(stat="identity") +
  xlab("ChIP Dataset") + ylab("Pearson Correlation") +
  scale_fill_gradient2(low = "#D67143", high = "#8EBF7A", 
name = "Pearson 
Correlation")+
   theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 12))+
  labs(title = "Correlation Mean Ratio per Barcode vs. Each ChIP Dataset")


## Scatterplot
# DNA methylation
chip_data3 <- tibble::rownames_to_column(as.data.frame(chip_data2))
dnameth.data <- chip_data3[,c(1,18)]
dnameth.data <- cbind(dnameth.data, mean.ratio2)

ggplot(dnameth.data, aes(x=WGBS, y=ratio))+
  geom_point()+
  geom_smooth(method = lm) 

# LMNB1
chip_data3 <- tibble::rownames_to_column(as.data.frame(chip_data2))
lmnb1.data <- chip_data3[,c(1,15)]
lmnb1.data <- cbind(lmnb1.data, mean.ratio2)

ggplot(lmnb1.data, aes(x=LMNB1, y=ratio))+
  geom_point()+
  geom_smooth(method = lm) 

# Calculate correlations
cor_chip_indel.high <- cor(chip_data,indel.data.high.drug)
cor_chip_indel.med <- cor(chip_data,indel.data.med.drug)
cor_chip_indel.low <- cor(chip_data,indel.data.low.drug)

# Calculate correlations for efficiency data
cor_chip_indel.high.eff <- cor(chip_data,indel.data.high.eff)
cor_chip_indel.med.eff <- cor(chip_data,indel.data.med.eff)
cor_chip_indel.low.eff <- cor(chip_data,indel.data.low.eff)

# Take only drugs that have correlation of absolute > 0.5 with at least one ChIP dataset
cor_chip_indel.high <-
  cor_chip_indel.high[,
                      colSums(cor_chip_indel.high > 0.7) >= 1 |
                        colSums(cor_chip_indel.high < -0.7) >= 1]

cor_chip_indel.med <-
  cor_chip_indel.med[,
                      colSums(cor_chip_indel.med > 0.7) >= 1 |
                        colSums(cor_chip_indel.med < -0.7) >= 1]

cor_chip_indel.low <-
  cor_chip_indel.low[,
                      colSums(cor_chip_indel.low > 0.7) >= 1 |
                        colSums(cor_chip_indel.low < -0.7) >= 1]

cor_chip_indel.high.eff <-
  cor_chip_indel.high.eff[,
                      colSums(cor_chip_indel.high.eff > 0.7) >= 1 |
                        colSums(cor_chip_indel.high.eff < -0.7) >= 1]

cor_chip_indel.med.eff <-
  cor_chip_indel.med.eff[,
                      colSums(cor_chip_indel.med.eff > 0.7) >= 1 |
                        colSums(cor_chip_indel.med.eff < -0.7) >= 1]

cor_chip_indel.low.eff <-
  cor_chip_indel.low.eff[,
                      colSums(cor_chip_indel.low.eff > 0.7) >= 1 |
                        colSums(cor_chip_indel.low.eff < -0.7) >= 1]

# plotting the correlations
cor.pheatmap <- 
  list(cor_chip_indel.high, cor_chip_indel.med, cor_chip_indel.low)
cor.pheatmap.eff <- 
  list(cor_chip_indel.high.eff, cor_chip_indel.med.eff, cor_chip_indel.low.eff)



for(i in 1:(length(cor.pheatmap))) {
  data <- cor.pheatmap[[i]]
p <- pheatmap(data, 
         annotation_col = target,
         annotation_row = type,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         annotation_colors = annotation_colors_scr, annotation_legend = F,
         main = paste(i,"correlation heatmap: ChiP data vs. logratio change per drug"))
print(p)
}

for(i in 1:(length(cor.pheatmap.eff))) {
  data <- cor.pheatmap.eff[[i]]
p <- pheatmap(data, 
         annotation_col = target,
         annotation_row = type,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         annotation_colors = annotation_colors_scr, annotation_legend = F,
         main = paste(i,"correlation heatmap: ChiP data vs. efficiency change per drug"))
print(p)
}

```
# There are some interesting drug-induced changes that can be explained by certain chromatin marks - this can be looked at in more detail


## Cluster-specific effects: now, instead of correlating each chromatin mark against drug-induced change, I try to correlate the drug-induced change against the chromatin clusters
```{r}
clusters <- as.data.frame(unique(indel.data$barcode))
clusters$cluster <- "LAD"
clusters$cluster[clusters$`unique(indel.data$barcode)` == "AGAAAATAATATGACG" |
                   clusters$`unique(indel.data$barcode)` == "TTGAACGCGGGCTCGG"] <- "Active_P.E."
clusters$cluster[clusters$`unique(indel.data$barcode)` == "CCGGGGACGTATGCAC" |
                   clusters$`unique(indel.data$barcode)` == "GCTAACATCACGAATC"] <- "H3K9"
clusters$cluster[clusters$`unique(indel.data$barcode)` == "AGGGCGTAAAATATTT" |
                   clusters$`unique(indel.data$barcode)` == "ATACTATATTTAACGG"|
                   clusters$`unique(indel.data$barcode)` == "TATGGCTGTCGGGTAG"|
                   clusters$`unique(indel.data$barcode)` == "TGTCCCTTAGTACTTT"] <- "Active_G.B."
clusters$cluster[clusters$`unique(indel.data$barcode)` == "CATTTCTGATCAATAA" |
                   clusters$`unique(indel.data$barcode)` == "TGGCCAATATTTGTCT"] <- "No"
clusters$cluster[clusters$`unique(indel.data$barcode)` == "ACTGTCGAGTTGTCCG" |
                   clusters$`unique(indel.data$barcode)` == "GAGCGCGTCACCGGGT"|
                   clusters$`unique(indel.data$barcode)` == "CGGCCTGAAGGTCAGG"|
                   clusters$`unique(indel.data$barcode)` == "GCGCACCCTTTAATTG"] <- "H3K27"


clusters2 <- dcast(clusters, `unique(indel.data$barcode)`~cluster)

clusters2 <- clusters2 %>% 
  remove_rownames %>% column_to_rownames(var="unique(indel.data$barcode)")

clusters2[!is.na(clusters2)] <- 1
clusters2[is.na(clusters2)] <- 0

clusters2$Active_G.B. <- as.numeric(clusters2$Active_G.B.)
clusters2$Active_P.E. <- as.numeric(clusters2$Active_P.E.)
clusters2$H3K27 <- as.numeric(clusters2$H3K27)
clusters2$H3K9 <- as.numeric(clusters2$H3K9)
clusters2$LAD <- as.numeric(clusters2$LAD)
clusters2$No <- as.numeric(clusters2$No)


cor.clusters.high <- cor(clusters2, indel.data.high.drug)
cor.clusters.med <- cor(clusters2, indel.data.med.drug)
cor.clusters.low <- cor(clusters2, indel.data.low.drug)

cor.clusters.high.eff <- cor(clusters2, indel.data.high.eff)
cor.clusters.med.eff <- cor(clusters2, indel.data.med.eff)
cor.clusters.low.eff <- cor(clusters2, indel.data.low.eff)

# Take only drugs that have correlation of absolute > 0.5 with at least one ChIP dataset
cor.clusters.high <-
  cor.clusters.high[,
                      colSums(cor.clusters.high > 0.6) >= 1 |
                        colSums(cor.clusters.high < -0.6) >= 1]

cor.clusters.med <-
  cor.clusters.med[,
                      colSums(cor.clusters.med > 0.6) >= 1 |
                        colSums(cor.clusters.med < -0.6) >= 1]

cor.clusters.low <-
  cor.clusters.low[,
                      colSums(cor.clusters.low > 0.6) >= 1 |
                        colSums(cor.clusters.low < -0.6) >= 1]

cor.clusters.high.eff <-
  cor.clusters.high.eff[,
                      colSums(cor.clusters.high.eff > 0.6) >= 1 |
                        colSums(cor.clusters.high.eff < -0.6) >= 1]

cor.clusters.med.eff <-
  cor.clusters.med.eff[,
                      colSums(cor.clusters.med.eff > 0.6) >= 1 |
                        colSums(cor.clusters.med.eff < -0.6) >= 1]

cor.clusters.low.eff <-
  cor.clusters.low.eff[,
                      colSums(cor.clusters.low.eff > 0.6) >= 1 |
                        colSums(cor.clusters.low.eff < -0.6) >= 1]

# plotting the correlations
cor.pheatmap <- 
  list(cor.clusters.high, cor.clusters.med, cor.clusters.low)
cor.pheatmap.eff <- 
  list(cor.clusters.high.eff, cor.clusters.med.eff, cor.clusters.low.eff)



for(i in 1:(length(cor.pheatmap))) {
  data <- cor.pheatmap[[i]]
p <- pheatmap(data, 
         annotation_col = target,
         annotation_row = type,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         annotation_colors = annotation_colors_scr, annotation_legend = F,
         main = paste(i,"correlation heatmap: ChiP data vs. logratio change per drug"))
print(p)
}

for(i in 1:(length(cor.pheatmap.eff))) {
  data <- cor.pheatmap.eff[[i]]
p <- pheatmap(data, 
         annotation_col = target,
         annotation_row = type,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         annotation_colors = annotation_colors_scr, annotation_legend = F,
         main = paste(i,"correlation heatmap: ChiP data vs. efficiency change per drug"))
print(p)
}

# Boxplot
mean.ratio <- indel.data[indel.data$Drug == "CTPB" & indel.data$conc == "10um",]
mean.ratio <- mean.ratio[,c(1,25)]
mean.ratio$logratio.platenorm <- ave(mean.ratio$logratio.platenorm, mean.ratio$barcode, FUN = function(x) mean(x))
mean.ratio <- unique(mean.ratio)
mean.ratio <- mean.ratio %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

# Combine dfs
clusters <- clusters %>% 
  remove_rownames %>% column_to_rownames(var="unique(indel.data$barcode)")
mean.ratio2 <- cbind(mean.ratio, clusters)


# Boxplot
ggplot(mean.ratio2, aes(x=cluster, y=logratio.platenorm, fill = cluster))+
  geom_boxplot()+
  scale_fill_brewer(palette = "Pastel2")+
  scale_y_continuous(breaks=c(-0.5,0,0.5))+ theme_classic2()
  
## Statistical analysis: are the groups significantly different?
mean.ratio <- indel.data[indel.data$Drug == "AT9283" & indel.data$conc == "1um",]
mean.ratio <- mean.ratio[,c(1,12,25)]
mean.ratio$logratio.platenorm <- ave(mean.ratio$logratio.platenorm, mean.ratio$rep, mean.ratio$barcode, FUN = function(x) mean(x))
mean.ratio <- unique(mean.ratio)
clusters2 <- tibble::rownames_to_column(as.data.frame(clusters), "barcode")
mean.ratio <- merge(mean.ratio, clusters2)
t.test(mean.ratio$logratio.platenorm[mean.ratio$cluster == "Active_G.B."],
                                     mean.ratio$logratio.platenorm[mean.ratio$cluster == "Active_P.E."])$p.value


anova <- aov(logratio.platenorm ~ cluster, data = mean.ratio)
summary.aov(anova)
tukey <- TukeyHSD(anova)
```
# Again, some interesting stuff


## HDAC vs Integration 17
```{r HDAC vs Integration 17, out.width= "100%", fig.align= "center", echo=FALSE}
# select HDAC data
hdac <- data3[data3$Target == "HDAC",]
hdac <- hdac[,-19]
hdac <- hdac[rowSums(hdac > 1) >= 1 | rowSums(hdac < -0.7) >= 1,] 
hdac <- t(hdac)

# calculate correlations
cor_hdac <- cor(hdac, chip_data)

pheatmap(cor_hdac,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100))
```



## Correlation plots individual drugs
```{r}
## In case I want to include the mean.ratio
mean.ratio <- indel.data[indel.data$Drug == "Decitabine",]
mean.ratio <- mean.ratio[,c(1,25)]
mean.ratio$logratio.platenorm <- ave(mean.ratio$logratio.platenorm, mean.ratio$barcode, FUN = function(x) mean(x))
mean.ratio <- unique(mean.ratio)


## Ratio is influenced by which ChIP datasets?
mean.ratio2 <- mean.ratio
mean.ratio2 <- mean.ratio2 %>% 
  remove_rownames %>% column_to_rownames(var="barcode")

 # chip_data <- chip_data %>%
 # remove_rownames %>% column_to_rownames(var="rowname")

cor_ratio <- cor(mean.ratio2, chip_data)
cor_ratio <- as.data.frame(cor_ratio)
cor_ratio <- t(cor_ratio)
cor_ratio <- tibble::rownames_to_column(as.data.frame(cor_ratio), "ChIP")

ggplot(cor_ratio, aes(x = reorder(ChIP,logratio.platenorm, function(x) -x), y = logratio.platenorm, fill=logratio.platenorm)) +
  geom_bar(stat="identity") +
  xlab("ChIP Dataset") + ylab("Pearson Correlation") +
  scale_fill_gradient2(low = "#D67143", high = "#8EBF7A", 
name = "Pearson 
Correlation")+
   theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 12))+
  labs(title = "Correlation MC-1293 (1 uM) Logratio Change per Barcode vs. Each ChIP Dataset")

# Scatterplot
mean.ratio <- tibble::rownames_to_column(as.data.frame(mean.ratio2))
chip_data <- tibble::rownames_to_column(as.data.frame(chip_data))
hdac1.data <- chip_data[,c(1,18)]
hdac.1.data <- cbind(hdac1.data, mean.ratio2)

ggplot(hdac.1.data, aes(x=WGBS, y=logratio.platenorm))+
  geom_point()+
  geom_smooth(method = lm) +
  theme_classic2()

# Statistical significance?
x <- cor.test(hdac.1.data$WGBS, hdac.1.data$logratio.platenorm)

```



## HDAC correlation
```{r}
## In case I want to include the mean.ratio
mean.ratio <- indel.data[indel.data$Drug == "AR-42" | indel.data$Drug == "Vorinostat (SAHA, MK0683)" |
                           indel.data$Drug == "PCI-24781 (Abexinostat)" |
                           indel.data$Drug == "Givinostat (ITF2357)" |
                           indel.data$Drug == "Scriptaid" |
                           indel.data$Drug == "DMSO" & indel.data$conc == "1um",]
mean.ratio <- mean.ratio[,c(1,6,25)]
mean.ratio$logratio.platenorm <- ave(mean.ratio$logratio.platenorm, mean.ratio$Drug, mean.ratio$barcode, FUN = function(x) mean(x))
mean.ratio <- unique(mean.ratio)
setnames(mean.ratio, "barcode", "Var2")

## Ratio is influenced by which ChIP datasets?
mean.ratio2 <- mean.ratio
mean.ratio2 <- dcast(mean.ratio2, Var2 ~ Drug)
mean.ratio2 <- mean.ratio2 %>% 
  remove_rownames %>% column_to_rownames(var="Var2")
mean.ratio2 <- hdac[,colSums(hdac > 1) >= 1 | colSums(hdac < -0.7) >= 1] 

chip_data <- chip_data %>%
  remove_rownames %>% column_to_rownames(var="rowname")

cor_ratio <- cor(mean.ratio2, chip_data)
cor_ratio <- as.data.frame(cor_ratio)
cor_ratio <- t(cor_ratio)
cor_ratio <- tibble::rownames_to_column(as.data.frame(cor_ratio), "ChIP")
cor_ratio <- melt(cor_ratio)


ggplot(cor_ratio, aes(x = reorder(ChIP, value, function(x) -mean(x)), y = value, fill = value)) +
  geom_boxplot() +
  geom_jitter(aes(color=variable))+
  xlab("ChIP Dataset") + ylab("Pearson Correlation") +
   theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 12))+
  labs(title = "Correlation HDACs Logratio Change per Barcode vs. Each ChIP Dataset")


# Scatterplot
chip_data <- tibble::rownames_to_column(as.data.frame(chip_data))
h3k27.data <- chip_data[,c(1,4)]
h3k27.data <- cbind(h3k27.data, mean.ratio2)
h3k27.data <- melt(h3k27.data, id.vars = c("rowname","H3K27me3"))



ggplot(h3k27.data, aes(x=H3K27me3, y=value, color = variable))+
  geom_point()+
  geom_smooth(method = lm) +
  scale_color_brewer(palette = "Set2")+
  theme_classic2()

```

\newpage
## MC-1293 HDAC1 correlation 
```{r}
# mc1293.indel.data <- indel.data.med.drug[,73, drop = F]
# mc1293.indel.data <- tibble::rownames_to_column(as.data.frame(mc1293.indel.data), "barcode")
# barcode.order2 <- barcode.order[,-2]
# mc1293.indel.data <- merge(mc1293.indel.data, barcode.order2)
# mc1293.indel.data <- mc1293.indel.data[,-1]
# 
# chip_data2 <- chip_data[,10, drop = F]
# chip_data2 <- tibble::rownames_to_column(as.data.frame(chip_data2), "barcode")
# chip_data2 <- merge(chip_data2, barcode.order2)
# chip_data2 <- chip_data2[,-1]
# 
# 
# 
# mc1293.hdac1 <- merge(mc1293.indel.data, chip_data2)
# mc1293.hdac1 <- melt(mc1293.hdac1)
# 
# ggplot(mc1293.hdac1, aes(x = reorder(integration, value), y = value, fill = variable, group = variable))+
#   geom_bar(stat="identity", position = "dodge")+ 
#   theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 12))
```


\newpage
## Drug dose response
```{r drug dose responses, out.width= "100%", fig.align= "center", echo=FALSE}
# Select all drugs that are changed (in either high, med, or low concentration)
# names.high <- colnames(indel.data.high.drug.ess)
# names.med <- colnames(indel.data.med.drug.ess)
# names.low <- colnames(indel.data.low.drug.ess)
# indel.data.dose <- indel.data[indel.data$Drug %in% names.high |
#                                 indel.data$Drug %in% names.med |
#                                 indel.data$Drug %in% names.low, ]
# 
# # Plot drug dose response per drug for all barcodes
# # indel.data.dose$conc <- factor(indel.data.dose$conc, levels=c("100nm", "1um", "10um"))
# # for (i in c(1:6)){
# # p<- ggplot(data = indel.data.dose, aes(x = conc, y = logratio.mean, color = barcode)) +
# #   geom_point(alpha = 0.5) + 
# #   labs(title = "dose response curve of effective drugs", x = "concentration", y = "log2(+1/-7 ratio) change") +
# #   stat_summary(fun.y=mean, aes(group=1), geom="line", colour="#993300", alpha = 0.5) +
# #   stat_summary(fun.y=mean, aes(group=1), geom="point", colour="#993300", size=3, shape=4, alpha = 0.5)+
# #   facet_wrap_paginate(~Drug, nrow = 4, ncol = 4, page = i)
# # print(p)
# # }
# 
# # Or directly in one plot
# ggplot(data = indel.data.dose, aes(x = conc, y = logratio.mean, color = barcode)) +
#   geom_point(alpha = 0.5) + 
#   labs(title = "dose response curve of effective drugs", x = "concentration", y = "log2(+1/-7 ratio) change") +
#   stat_summary(fun.y=mean, aes(group=1), geom="line", colour="#993300", alpha = 0.5) +
#   stat_summary(fun.y=mean, aes(group=1), geom="point", colour="#993300", size=3, shape=4, alpha = 0.5)+
#   facet_wrap(~Drug, nrow = 8, ncol = 8)
# 
# 
# 
# # Look at individual drugs more carefully
# ## Select drug
# Drug <- "Resveratrol"
# for (i in Drug) {
# p <- ggplot(data = indel.data.dose[indel.data.dose$Drug == Drug,], aes(x = conc, y = logratio.mean, group = barcode, color = barcode)) +
#   geom_point(alpha = 0.5) + geom_line(alpha = 0.5, linetype = "dashed") +
#   labs(title = paste("dose response curve of", i), x = "concentration", y = "log2(+1/-7 ratio) change")
# print(p)
# }
# 
# 
# # Dose curves for ChIP correlations
# ## Compute all corralations
# indel.data.cor <- indel.data[indel.data$Drug %in% names.high |
#                                 indel.data$Drug %in% names.med |
#                                 indel.data$Drug %in% names.low, ]
# indel.data.cor$mean.logratio.bc.drug <- ave(indel.data.cor$logratio.platenorm, 
#                                             indel.data.cor$barcode, 
#                                             indel.data.cor$Drug, 
#                                             FUN = function(x) mean(x))
# indel.data.cor.high <- indel.data.cor[indel.data.cor$conc %in% "10um",]
# indel.data.cor.med <- indel.data.cor[indel.data.cor$conc %in% "1um",]
# indel.data.cor.low <- indel.data.cor[indel.data.cor$conc %in% "100nm",]
# indel.data.cor.high <- indel.data.cor.high[,c(1,6,41)]
# indel.data.cor.med <- indel.data.cor.med[,c(1,6,41)]
# indel.data.cor.low <- indel.data.cor.low[,c(1,6,41)]
# indel.data.cor.high <- indel.data.cor.high[!duplicated(indel.data.cor.high), ]
# indel.data.cor.med <- indel.data.cor.med[!duplicated(indel.data.cor.med), ]
# indel.data.cor.low <- indel.data.cor.low[!duplicated(indel.data.cor.low), ]
# indel.data.cor.high <- dcast(indel.data.cor.high, barcode ~ Drug, value.var="mean.logratio.bc.drug")
# indel.data.cor.med <- dcast(indel.data.cor.med, barcode ~ Drug, value.var="mean.logratio.bc.drug")
# indel.data.cor.low <- dcast(indel.data.cor.low, barcode ~ Drug, value.var="mean.logratio.bc.drug")
# row.names(indel.data.cor.high) <- indel.data.cor.high$barcode
# indel.data.cor.high <- indel.data.cor.high[,-1]
# row.names(indel.data.cor.med) <- indel.data.cor.med$barcode
# indel.data.cor.med <- indel.data.cor.med[,-1]
# row.names(indel.data.cor.low) <- indel.data.cor.low$barcode
# indel.data.cor.low <- indel.data.cor.low[,-1]
# 
# # Calculate correlations
# cor_chip_indel.high <- cor(chip_data,indel.data.cor.high)
# cor_chip_indel.med <- cor(chip_data,indel.data.cor.med)
# cor_chip_indel.low <- cor(chip_data,indel.data.cor.low)
# 
# ## Convert dfs to long format
# dose_chip_high <- tibble::rownames_to_column(as.data.frame(cor_chip_indel.high), "ChIP")
# dose_chip_high <- melt(dose_chip_high, variable.name = "Drug", value.name = "cor")
# dose_chip_high$conc <- "10um"
# 
# dose_chip_med <- tibble::rownames_to_column(as.data.frame(cor_chip_indel.med), "ChIP")
# dose_chip_med <- melt(dose_chip_med, variable.name = "Drug", value.name = "cor")
# dose_chip_med$conc <- "1um"
# 
# dose_chip_low <- tibble::rownames_to_column(as.data.frame(cor_chip_indel.low), "ChIP")
# dose_chip_low <- melt(dose_chip_low, variable.name = "Drug", value.name = "cor")
# dose_chip_low$conc <- "100nm"
# 
# ## Combine all dfs to one
# dose_chip <- rbind(dose_chip_high, dose_chip_med, dose_chip_low)
# 
# ## Visualize the dose response of the correlations
# dose_chip$conc <- factor(dose_chip$conc, levels=c("100nm", "1um", "10um"))
# ggplot(data = dose_chip, aes(x = conc, y = cor, group = ChIP, color = ChIP)) +
#   geom_point(alpha = 0.5) + geom_line(alpha = 0.5, linetype = "dashed")+
#   labs(title = "correlation behavior per concentration", x = "concentration", y = "correlation ChIP vs logratio change") +
#   facet_wrap(~Drug, nrow = 7, ncol = 7)
# 
# # Look at individual drugs more carefully
# ## Select drug
# Drug <- "Resveratrol"
# for (i in Drug) {
# p <- ggplot(data = dose_chip[dose_chip$Drug == Drug,], aes(x = conc, y = logratio.mean, group = barcode, color = barcode)) +
#   geom_point(alpha = 0.5) + geom_line(alpha = 0.5, linetype = "dashed") +
#   labs(title = paste("dose response curve of", i), x = "concentration", y = "log2(+1/-7 ratio) change")
# print(p)
# }


```




## Exporting potential data. 
```{r}
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

