---
title: "DSB epigenetic drug screen - Figure 4 Epistasis"
author: 
  - name: "Max Trauernicht & Ruben Schep"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
  pdf_document:
    toc: yes
    toc_depth: '4'
---

*knitr document van Steensel lab*


# Introduction
Clone 5 with ~18 sgRNA target site IPRs was cut and treated with 160 epigenetic drugs. The repair outcomes of that experiment will be visualized here. I previously processed the raw sequencing data, and calculated the repair outcomes. In this script, the main results plots will be generated.


## Setup {.tabset}z

<!-- little HTML script to do indentation of the table of contents -->
<script>
$(document).ready(function() {
$items = $('div#TOC li');
$items.each(function(idx) {
num_ul = $(this).parentsUntil('#TOC').length;
$(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
});

});
</script>

```{css, echo = FALSE}
div.sourceCode {
overflow-x: hidden;
}
```

# Libraries
```{r libraries}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

# libraries:
library(knitr)
library(tidyverse)
library(data.table)
# library(parallel)
# library(gtools)
library(pheatmap)
library(ggbeeswarm)
# library(reshape2)
# library(gridExtra)
library(broom)
# library(ggrepel)
# library(dendextend)
# library(Hmisc)
library(GGally)
library(RColorBrewer)
library(colorspace)
library(magrittr)
library(ggpubr)
library(pls)
# library(rstatix)

## Select outdir
out.dir = paste0("./figures/", Date, "/")
dir.create(out.dir)
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)

```

# Load libraries (Just need a data table with all the data and )
```{r}
#Chromatin data (it needs a barcode column)
# Import data from preprocessing script
in.dir = list.dirs(path = "/DATA/projects/DSBrepair/data/R") %>% .[grepl("episcreen", .)] %>% tail(n = 1)
file = list.files(in.dir, pattern = "ratios", full.names = T) %>% tail(n = 1)
indel.data <- readRDS(file) %>% filter(pathway_bal_filter == TRUE & viab_reproducibility == TRUE) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug)) %>%
  # We remove the HDAC1 and HDAC2 chromatin features as we do not have a sufficient dynamic range in
  # our clone to make strong conclusions. The max z-score we get are < 80% CI of the z-score in the
  # whole pool.
  select(-HDAC1, -HDAC2)


TRIP_chrom_tib <- readRDS('/DATA/projects/DSBrepair/data/R/cl20201026_ChIP_zscore_selection.RDS') %>% 
  # Same for the TRIP chromatin features
  select(-HDAC1, -HDAC2)

TRIP_chrom = TRIP_chrom_tib %>% select(-binsize, -pool) %>% column_to_rownames("ID")
# make clustering with some column rotation for better visualisation
TRIP_chrom_clust = dendextend::rotate(hclust(dist(t(TRIP_chrom))), c(3:1, 6:4, 9:10, 14, 16, 15, 13:11, 7:8, 17:18, 20:23, 19))



figure_out = "/DATA/projects/DSBrepair/scratch/episcreen/"
chromatin.features <- colnames(indel.data)[grepl("[.]zscore", colnames(indel.data))]
chrom_data = indel.data %>% dplyr::select(barcode, all_of(chromatin.features)) %>% distinct()  %>%
  rename_with(., ~gsub(".zscore", "", .x))

chromatin.features.short = gsub(".zscore", "", chromatin.features)
```


```{r correlation between features TRIP wide, echo = FALSE, eval = FALSE}
max_chrom_tib_C5 = chrom_data %>% 
  as.tibble() %>% 
  select(-barcode) %>%
  rename_with(., ~gsub(".zscore", "", .x)) %>% 
  summarise_each(., funs(max(., na.rm=TRUE))) %>% 
  pivot_longer(cols = everything(), names_to = "name", values_to = "max")

maxmin_chrom_tib_C5 = chrom_data %>% 
  as.tibble() %>% 
  select(-barcode) %>%
  rename_with(., ~gsub(".zscore", "", .x)) %>% 
  summarise_each(., funs(min(., na.rm=TRUE))) %>% 
  pivot_longer(cols = everything(), names_to = "name", values_to = "min") %>%
  left_join(max_chrom_tib_C5)

chrom_tib = TRIP_chrom_tib %>% 
  select(-binsize, -pool, -ID) %>% 
  distinct() %>% 
  pivot_longer(cols = everything()) %>% 
  left_join(maxmin_chrom_tib_C5) %>%
    group_by(name) %>%
    mutate(max_CI = ecdf(value)(max),
           min_CI = ecdf(value)(min)) %>%
  distinct(name, max_CI, min_CI)

chrom_tib %>% filter(max_CI < 0.8)

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {
  
  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)
  
  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }
  
  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())
  
  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]
  
  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }
  
  p <- p +
    theme(panel.background = element_rect(fill = corCol))
  
  return(p)
}
boundaries <- c(-0.7, -.5,0.5,0.7)
pdf("/DATA/projects/DSBrepair/git/EpiScreen/rs20220722_chrom_cor_clones.pdf", width=20, height=20)
ggpairs(chrom_tib[, labels(TRIP_chrom_clust) %>% gsub(".zscore", "", .)],
        upper = list(continuous = corColor),
        # lower = list(continuous = "density"),
        lower = list(continuous = function(data, mapping, ...) {
          ggally_density(data = data, mapping = mapping) +
            geom_abline(slope = 1, lty = "twodash", col = "grey")
          # ggally_smooth(data = data, mapping = mapping, method = "lm", color = "red", lty = "dashed", size = 0.5)
        }),
        diag = list(continuous = function(data, mapping, ...) {
          ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("Chromatin") +
  ylab("Chromatin")
dev.off()
```


```{r}
colors_drugs <- c("#FA8D62", "#65BFA4", "#8CA0C4", "#808184", "#F6D289", "#E5E5E5", "#ACACAC", "#737373")
conc_control_colors = c("control" = colors_drugs[2], "100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")
conc_colors = c("100 nM" = colors_drugs[7], "1 µM" = colors_drugs[4], "10 µM" = "#000000")

group_colors = c("Euchromatin" = "#f7941d",
                 "other-heterochromatin" = "#838687",
                 "Triple Heterochromatin" = "#662d91",
                 "H3K27me3" = "#d21f8a")

viab_colors = sequential_hcl(11, "Peach", rev = TRUE)
names(viab_colors) = as.numeric(seq(0, 1, by = 0.1))

annotation_colors_scr = list(
  chromatin = group_colors,
  target = c(HDAC="#60988D", HAT="#C6D0A8", Sirtuin="#8FCAAC",
             HMT="#FDCDAC", DNMT="#D69C81", `Histone Demethylase`="#FFF2AE",
             HIF="#E3DCD5", JAK="#C8D7E2", PIM="#9AB8C4", `Aurora Kinase`="#F4CAE4",
             PARP="#CB546F", `Epigenetic Reader Domain`= "#476D61", `DNA-PK`="#E07A43", 
             `Negative Control` = "#65BFA4", `MRN` = "#5D79AC"),
  conc_char = conc_colors,
  conc_char_ctrl = conc_control_colors,
  clusters_drugs = sequential_hcl(7, "Plasma"),
  # viab_rep1 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  # viab_rep2 = c('#ffeee9', '#ffdcd4', '#ffcbbe', '#ffb8a8', '#fda693', '#f9947e', '#f4826a', '#ef6f56', '#e95b43', '#e24530'),
  viab = viab_colors,
  TIF = colorRampPalette(c("white", "black"))(20),
  MMEJratio = c('#e1251b', '#d32c2d', '#c5313c', '#b6354a', '#a73857', '#963a65', '#843d72', '#6e3e7f', '#53408d', '#26419a'),
  group = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
            transcribing="#FBB040", HDAC="#aac9e0", accessibility='#cebc85',
            methylation='#7dc98f'))


ordered_marks2 <- c("Negative Control", "MRN", "DNA-PK","Epigenetic Reader Domain", "HDAC",  "Sirtuin", "HAT", 
                    "Histone Demethylase", "HMT", "DNMT", 
                    "HIF", "JAK", "PIM",
                    "Aurora Kinase", "PARP")

#Epistatic interaction colors
slope.colors <- tibble(color = c("#8c510a", "#f5f5f5","#01665e"),label = c("negative","none","positive"), feature = "epistasis")

annotation_colors_scr$target <- annotation_colors_scr$target[ordered_marks2]

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2', 'H3K9me3', 'H3K27me3', 'EZH2', 
                                    'CTCF', 'SMC3',
                                    'HDAC3', 
                                    'H3K4me1', 'H3K4me2', 'H3K4me3', 'H3K27ac', 'H4K5acK8ac', 'H2AFZ', 
                                    'DNAse', 'Dam', 
                                    'm5C',
                                    'H3K79me2', 'TTseq',  'H3K36me3', 'POL2AS2','POL2'),
                        group=factor(c(rep('repressive', 6), 
                                       rep('insulator', 2),
                                       rep('HDAC', 1), 
                                       rep('euchromatin', 6),
                                       rep('accessibility', 2), 
                                       'methylation',
                                       rep('transcribing', 5)),
                        levels=c('transcribing', 'accessibility',
                                 'methylation', 'euchromatin',
                                 'HDAC', 'insulator', 'repressive')))

```

## Prepare data

```{r}
# Data table and compute log2_ratio (This line needs to be adjusted)
filtered_drug_conc = indel.data %>%
  filter(abs(mean_eff_zscore_comb) > 1.96 | abs(MMEJ_zscore_comb) > 1.96 | drug == "DMSO") %>%
  mutate(drug_conc = paste(drug, conc_char,sep = " "),
         effect_on_indelrate = ifelse(abs(mean_eff_zscore_comb) > 1.96, TRUE, FALSE),
         effect_on_ratio = ifelse(abs(MMEJ_zscore_comb) > 1.96, TRUE, FALSE)) %>%
  distinct(drug, conc_char, drug_conc, effect_on_indelrate, effect_on_ratio) %>%
  mutate(counts = ave(drug_conc, drug, conc_char, FUN = function(x) length(x)),
         keep = ifelse(counts > 1 & !effect_on_indelrate | counts > 1 & !effect_on_ratio, FALSE, TRUE)) %>%
  filter(keep) %>% 
  distinct(drug, conc_char, drug_conc, effect_on_indelrate, effect_on_ratio)


ratios_tib <- indel.data %>%
  mutate(drug_conc = paste(drug, conc_char, sep = " ")) %>%
  semi_join(filtered_drug_conc) %>%
  distinct(MMEJratio, freqCut, drug, conc_char, drug_conc, target, barcode) %>%
  mutate(MMEJratio = ave(MMEJratio, drug, conc_char, barcode, FUN = mean)) %>%
  mutate(MMEJratio.log2 = log2(MMEJratio)) %>%
  mutate(freqCut = (ave(freqCut, drug, conc_char, barcode, FUN = mean))) %>%
  mutate(freqCut.log2 = log2(freqCut)) %>%
  distinct()

dmso_ratios_tib <- indel.data %>%
  filter(drug == "DMSO") %>%
  distinct(MMEJratio, freqCut, barcode) %>%
  mutate(DMSO.ratio = ave(MMEJratio, barcode, FUN = mean)) %>%
  mutate(DMSO.ratio.log2 = log2(DMSO.ratio)) %>%
  mutate(DMSO.freqCut = ave(freqCut, barcode, FUN = mean)) %>%
  mutate(DMSO.freqCut.log2 = log2(DMSO.freqCut)) %>%
  dplyr::select(-MMEJratio, -freqCut) %>%
  distinct()

ratios_tib <- merge(ratios_tib, dmso_ratios_tib)
```

The log2 distance ratio in our case is positive if it promotes NHEJ and negative if it promotes NHEJ


# 2nd step check compute distances and plot them
```{r}
# Each replicate has a different value (compute differences by replicate)
ratios_tib_dist <- ratios_tib %>% 
  dplyr::group_by(barcode, drug, conc_char) %>% 
  mutate(log2.dist.ratio = MMEJratio.log2 - DMSO.ratio.log2, 
         log2.dist.freqCut = freqCut.log2 - DMSO.freqCut.log2) %>%
  left_join(chrom_data)
```


# Extract slopes for all genes
```{r}
# Perform analysis across all features for three test genes
ratios_tib_dist %<>% mutate(drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  filter(!is.na(MMEJratio))

slope.protein.features <- tibble(drug_conc = NA, feature = NA,  term = NA, 
                                 indelrate = NA, p.value.indelrate = NA, 
                                 ratio = NA, p.value.ratio = NA)

for (i in unique(ratios_tib_dist$drug_conc)) {
  for (j in chromatin.features.short) {
    model_tib <- ratios_tib_dist %>% filter(drug_conc == i) 
    
    # Indel rates
    model.indelrate.log2 <- lm(formula = log2.dist.freqCut ~ unlist(model_tib[j]), 
                               data = model_tib) %>% 
      tidy()
    
    # Pathway balance
    model.ratio.log2 <- lm(formula = log2.dist.ratio ~ unlist(model_tib[j]), 
                           data = model_tib) %>% 
      tidy()
    
    slope.protein.features <- slope.protein.features %>% 
      add_row(drug_conc = i, 
              feature = j, 
              term = model.indelrate.log2 %>%
                pull(term),
              indelrate = model.indelrate.log2 %>% 
                pull(estimate), 
              p.value.indelrate = model.indelrate.log2 %>% 
                pull(p.value),
              ratio = model.ratio.log2 %>% 
                pull(estimate), 
              p.value.ratio = model.ratio.log2 %>% 
                pull(p.value))
  }
}

# Remove the 1st NA column
slope.protein.features %<>% 
  filter(!is.na(feature)) %>%
  mutate(term = ifelse(term == "(Intercept)", "intercept", "slope"))
```

```{r}
CCD_tib = ratios_tib_dist %>% select(barcode, drug_conc, log2.dist.freqCut, log2.dist.ratio, all_of(chromatin.features.short)) %>% ungroup()

#Create an empty dt with CCDs of DDR proteins
drug_conc_CCDs_dt <- tibble(var = NA, drug_conc = NA, num_comp = NA, r.squared = NA, adj.r.squared = NA,p.value = NA)
# chrom_formula = as.formula(paste("y ~ ", paste(chromatin.features, collapse= "+")))

for (i in unique(CCD_tib$drug_conc)){
  for (j in c("log2.dist.freqCut", "log2.dist.ratio")) {
    drug_conc_dt <- filter(CCD_tib, drug_conc == i)
    # Run a model per drug_conc
    set.seed(1)
    chrom_formula = reformulate(chromatin.features.short, response = j)
    PCR_model_DDR_test <- pcr(chrom_formula,
                              data=drug_conc_dt , 
                              validation="CV")
    
    pcr_pred <- predict(PCR_model_DDR_test, 
                        drug_conc_dt, ncomp = 3)
    
    combined.dt <- tibble(measured = drug_conc_dt %>% pull(j), 
                          predicted = as.numeric(pcr_pred))
    
    pred_vs_estim <- lm(formula = measured ~ predicted, 
                        data = combined.dt) %>% 
      glance()
    
    drug_conc_CCDs_dt <- drug_conc_CCDs_dt %>% 
      add_row(var = j,
              drug_conc = i, 
              r.squared = pred_vs_estim %>% 
                pull(r.squared), 
              adj.r.squared = pred_vs_estim %>% 
                pull(adj.r.squared), 
              p.value = pred_vs_estim %>% 
                pull(p.value))
  }
}

#Correct model to adjust for multiple testing correction
adj_p.value_KO_model <- drug_conc_CCDs_dt %>% 
  dplyr::select(var, num_comp, p.value, drug_conc) %>% 
  group_by(var) %>%
  mutate(p.adj = p.adjust(p.value, method = "BH")) %>% 
  dplyr::select(var, drug_conc,p.value,p.adj)

signif_hits = adj_p.value_KO_model %>% filter(p.adj < 0.05) %>% 
  distinct(drug_conc, var) %>% 
  ungroup() %>% 
  mutate(val = TRUE,
         var = gsub("log2.dist", "signif", var)) %>%
  pivot_wider(names_from = var, 
              values_from = val, 
              values_fill = FALSE) 
```


```{r epistasis filtering}
# Calculate postition where the linear model crosses y = 0, we do this by -intercept / slope
epistasis_tib = slope.protein.features %>% 
  distinct(drug_conc, term, feature, indelrate, ratio) %>% 
  pivot_wider(., names_from = term, values_from = c(indelrate, ratio)) %>%
  mutate(feature = gsub(".zscore", "", feature),
         x_intercept_indelrate = -indelrate_intercept / indelrate_slope,
         x_intercept_ratio = -ratio_intercept / ratio_slope)

effect_compounds = ratios_tib_dist %>%
  group_by(drug_conc)  %>% 
  mutate(indelrate_pos = sum(log2.dist.freqCut > 0),
         indelrate_effect = ifelse(indelrate_pos >= 17, "indel_rate_increase", 
                                   ifelse(indelrate_pos <= 1, "indel_rate_decrease", "mixed")),
         ratio_pos = sum(log2.dist.ratio > 0),
         ratio_effect = ifelse(ratio_pos >= 17, "NHEJ_increase", 
                               ifelse(ratio_pos <= 1, "MMEJ_increase", "mixed"))) %>%
  distinct(drug_conc, indelrate_effect, ratio_effect) %>% 
  left_join(distinct(filtered_drug_conc, drug_conc, effect_on_indelrate, effect_on_ratio))


epistasis_tib %<>% left_join(effect_compounds)  %>% 
  left_join(signif_hits) %>%
  mutate(
    # x_intercept_indelrate = ifelse(p.value.indelrate < 0.05,
    #                                x_intercept_indelrate, NA),
    # x_intercept_ratio = ifelse(p.value.ratio < 0.05,
    #                            x_intercept_ratio, NA),
    indelrate_slope_plot = ifelse(x_intercept_indelrate < 1 & effect_on_indelrate & signif.freqCut, indelrate_slope, NA),
    ratio_slope_plot = ifelse(x_intercept_ratio < 1 & effect_on_ratio & signif.ratio, ratio_slope, NA)) %>%
  distinct(drug_conc, feature, indelrate_slope_plot, ratio_slope_plot)
```


```{r prepare heatmaps, fig.width= 10, fig.height= 20}
clone5.matrix = indel.data %>% 
  dplyr::select(IPR, ends_with(".zscore")) %>% 
  distinct() %>% 
  rename_with(., ~gsub(".zscore", "", .x)) %>%
  column_to_rownames(var="IPR")

clone5_map = hclust(dist(t(clone5.matrix)))
clone5_map_ipr = dendextend::rotate(hclust(dist(clone5.matrix)), c(1:2,5:8,3:4, 15:14, 17, 16, 18 ,13,10:12,9))

chromatin <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(distinct(indel.data, barcode, IPR)) 

chromatin <- chromatin %>%
  filter(IPR != "<NA>") %>% 
  dplyr::select(-barcode) %>%
  column_to_rownames(var="IPR")

# Set annotation for heatmap
viab_drug = indel.data %>% filter(drug != "PAO") %>% 
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>% 
  distinct(viab_mean, drug_conc, replicate) %>% group_by(drug_conc) %>%
  summarise(viab = mean(viab_mean))

# Row annotation: annotate drugs with the target group
target <- indel.data %>% 
  # filter(!target %in% c("Negative Control", "MRN", "DNA-PK")) %>%
  mutate(drug = gsub(" [(].*[)]", "", drug),
         drug_conc = paste(drug, conc_char, sep = " ")) %>%
  left_join(viab_drug) %>%
  distinct(drug_conc, conc_char, target, viab) %>% 
  column_to_rownames(var="drug_conc")

TIF_IPR = indel.data %>% filter(drug == "DMSO") %>%
  distinct(replicate, tech, freqCut,MMEJratio, IPR, well, plate) %>%
  group_by(IPR) %>%
  summarise(TIF = mean(freqCut),
            MMEJratio = mean(MMEJratio))

chromatin_TIF = chromatin %>% rownames_to_column(var = "IPR") %>%
  left_join(TIF_IPR) %>%
  column_to_rownames("IPR")
```

```{r Synergies chromatin - Balance all, fig.width= 10, fig.height= 20}
all_compounds_epistasis = epistasis_tib %>%
  dplyr::select(drug_conc, ratio_slope_plot, feature) %>%
  group_by(drug_conc) %>%
  mutate(onlyNA = sum(is.na(ratio_slope_plot))) %>% 
  filter(onlyNA != 25) %>%
  reshape2::dcast(drug_conc ~ feature, value.var = "ratio_slope_plot", fill = 0) %>% 
  column_to_rownames(var = "drug_conc")

# change these numbers according to where you want to set the cutoff
quant_breaks <- quantile(all_compounds_epistasis, c(.01,.99), na.rm = T)
palette_length <- 100
breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
            seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
colors[which(quant_breaks == 0)-1] = "gray90"

target_ctrl = rename(target, conc_char_ctrl = conc_char)

pheatmap(all_compounds_epistasis[, TRIP_chrom_clust$labels],
         border_color = F,
         cellwidth = 8,
         cellheight = 8,
         breaks = breaks,
         clustering_method = "ward.D",
         cluster_cols = TRIP_chrom_clust,
         annotation_row = target_ctrl,
         annotation_col = clustering,
         main = "Chromatin synergies - MMEJ ratio all",
         color = colors,
         annotation_colors = annotation_colors_scr, 
         annotation_legend = T)
```

```{r Synergies chromatin - Balance per target, fig.width= 10, fig.height= 10}
targets = target %>% 
  pull(target) %>% 
  unique()

for (i in targets) {
  target_drugs = rownames(filter(target, target %in% i))
  
  # Make matrix with drugs from selected target
  mmej_synergy_matrix = all_compounds_epistasis %>% 
    rownames_to_column(var = "drug_conc") %>%
    filter(drug_conc %in% target_drugs) %>%
    column_to_rownames(var = "drug_conc")
  
  cluster_rows = ifelse(nrow(mmej_synergy_matrix) > 1, TRUE, FALSE)
  
  if(nrow(mmej_synergy_matrix) > 0) {
    # change these numbers according to where you want to set the cutoff
    quant_breaks <- quantile(mmej_synergy_matrix, c(.01,.99), na.rm = T) 
    palette_length <- 100
    breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
                seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
    slope.colors.temp = slope.colors
    if(all(mmej_synergy_matrix <= 0)) {
      breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length)))
      slope.colors.temp = filter(slope.colors, label != "positive")
    }
    if(all(mmej_synergy_matrix >= 0)) {
      breaks <- c(seq(0, quant_breaks[2], length.out=ceiling(palette_length)))
      slope.colors.temp = filter(slope.colors, label != "negative")
    }
    colors = colorRampPalette(slope.colors.temp %>% pull(color))(length(breaks))
    colors[which(quant_breaks == 0)-1] = "gray90"
    
    # heatmap
    p = pheatmap(mmej_synergy_matrix,
                 border_color = F,
                 cellwidth = 8,
                 cellheight = 8,
                 cluster_rows = cluster_rows,
                 breaks = breaks,
                 # cluster_cols = TRIP_chrom_clust,
                 annotation_row = target_ctrl,
                 
                 clustering_method = "ward.D",
                 main = paste("Chromatin synergies - MMEJ ratio per target :" , i, sep = " "),
                 color = colors,
                 annotation_colors = annotation_colors_scr, 
                 annotation_legend = T)
    print(p)
  }
}
```

```{r Z-score - Balance all, fig.width= 10, fig.height= 20}
ritios_tib_epi_effects = indel.data %>%
  mutate(drug_conc_ctrl = paste(drug, conc_char, sep = " ")) %>%
  filter(drug_conc_ctrl %in% rownames(all_compounds_epistasis)) %>%
  distinct(IPR, MMEJ_zscore_comb, drug_conc_ctrl, drug, conc_char) 

mmejzscore_matrix = ritios_tib_epi_effects %>%
  distinct(drug_conc_ctrl, IPR, MMEJ_zscore_comb) %>% 
  pivot_wider(names_from = IPR, values_from = MMEJ_zscore_comb) %>%
  column_to_rownames(var = "drug_conc_ctrl")

# clone5_map_ipr_TIF_Bal = hclust(dist(column_to_rownames(TIF_IPR, var = "IPR"))) %>% rotate(., c(1, 3, 2))

quant_breaks <- quantile(mmejzscore_matrix, c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 1000
breaks <- breaks <- c(seq(quant_breaks[1], -1, 
                          length.out=ceiling(palette_length/2) + 1), 
                      seq(-quant_breaks[1]/palette_length, 
                          -quant_breaks[1], 
                          length.out=floor(palette_length/2)))
colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
colors[-which(breaks < -1.96 | breaks > 1.96 )] = "gray90"

pheatmap(mmejzscore_matrix[, clone5_map_ipr$labels],
         border_color = F,
         cellwidth = 8,
         cellheight = 8,
         breaks = breaks,
         cluster_cols = clone5_map_ipr,
         annotation_row = target_ctrl,
         
         clustering_method = "ward.D",
         main = "Z-score - MMEJ ratio all",
         color = colors,
         annotation_colors = annotation_colors_scr, 
         annotation_legend = T,
         annotation_col = chromatin_TIF)
```

```{r Z-score - Balance per target, fig.width= 10, fig.height= 10}
targets = target %>% 
  pull(target) %>% 
  unique()

for (i in targets) {
  target_drugs = rownames(filter(target, target %in% i))
  
  # Make matrix with drugs from selected target
  mmejzscore_matrix = ritios_tib_epi_effects %>%
    distinct(drug_conc_ctrl, IPR, MMEJ_zscore_comb) %>% 
    filter(drug_conc_ctrl %in% target_drugs) %>%
    pivot_wider(names_from = IPR, values_from = MMEJ_zscore_comb) %>%
    column_to_rownames(var = "drug_conc_ctrl")
  
  cluster_rows = ifelse(nrow(mmejzscore_matrix) > 1, TRUE, FALSE)
  if(nrow(mmejzscore_matrix) > 0) {
    # change these numbers according to where you want to set the cutoff
    quant_breaks <- quantile(mmejzscore_matrix, c(.01,.99), na.rm = T) 
    palette_length <- 100
    breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
                seq(-quant_breaks[1]/palette_length, -quant_breaks[1], length.out=floor(palette_length/2)))
    colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
    colors[which(quant_breaks == 0)-1] = "gray90"
    
    # heatmap
    p = pheatmap(mmejzscore_matrix[, clone5_map_ipr$labels],
                 border_color = F,
                 cellwidth = 8,
                 cellheight = 8,
                 cluster_cols = clone5_map_ipr,
                 cluster_rows = cluster_rows,
                 
                 clustering_method = "ward.D",
                 breaks = breaks,
                 annotation_row = target_ctrl,
                 main = paste("Z-score - MMEJ ratio per target :" , i, sep = " "),
                 color = colors,
                 annotation_colors = annotation_colors_scr, 
                 annotation_legend = T,
                 annotation_col = chromatin_TIF)
    print(p)
  }
}
```

```{r Synergies chromatin - TIF all, fig.width= 10, fig.height= 20}
all_compounds_epistasis_TIF = epistasis_tib %>%
  dplyr::select(drug_conc, indelrate_slope_plot, feature) %>%
  group_by(drug_conc) %>%
  mutate(onlyNA = sum(is.na(indelrate_slope_plot))) %>% 
  filter(onlyNA != 25) %>%
  reshape2::dcast(drug_conc ~ feature, value.var = "indelrate_slope_plot", fill = 0) %>% 
  column_to_rownames(var = "drug_conc")



# change these numbers according to where you want to set the cutoff
quant_breaks <- quantile(all_compounds_epistasis_TIF, c(.01,.99), na.rm = T)
palette_length <- 100
breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
            seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
colors[which(quant_breaks == 0)-1] = "gray90"

pheatmap(all_compounds_epistasis_TIF[, TRIP_chrom_clust$labels],
         border_color = F,
         cellwidth = 8,
         cellheight = 8,
         breaks = breaks,
         cluster_cols = TRIP_chrom_clust,
         clustering_method = "ward.D",
         annotation_row = target_ctrl,
         main = "Chromatin synergies - TIF all",
         color = colors,
         annotation_colors = annotation_colors_scr, 
         annotation_legend = T)
```

```{r Synergies chromatin - TIF per target, fig.width= 10, fig.height= 10}
targets = target %>% 
  pull(target) %>% 
  unique()

for (i in targets) {
  target_drugs = rownames(filter(target, target %in% i))
  
  # Make matrix with drugs from selected target
  TIF_synergy_matrix = all_compounds_epistasis_TIF %>% 
    rownames_to_column(var = "drug_conc") %>%
    filter(drug_conc %in% target_drugs) %>%
    column_to_rownames(var = "drug_conc")
  
  cluster_rows = ifelse(nrow(TIF_synergy_matrix) > 1, TRUE, FALSE)
  
  if(nrow(TIF_synergy_matrix) > 0) {
    # change these numbers according to where you want to set the cutoff
    quant_breaks <- quantile(TIF_synergy_matrix, c(.01,.99), na.rm = T) 
    palette_length <- 100
    breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
                seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
    slope.colors.temp = slope.colors
    if(all(TIF_synergy_matrix <= 0)) {
      breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length)))
      slope.colors.temp = filter(slope.colors, label != "positive")
    }
    if(all(TIF_synergy_matrix >= 0)) {
      breaks <- c(seq(0, quant_breaks[2], length.out=ceiling(palette_length)))
      slope.colors.temp = filter(slope.colors, label != "negative")
    }
    colors = colorRampPalette(slope.colors.temp %>% pull(color))(length(breaks))
    colors[which(quant_breaks == 0)-1] = "gray90"
    
    # heatmap
    p = pheatmap(TIF_synergy_matrix[, TRIP_chrom_clust$labels],
                 border_color = F,
                 cellwidth = 8,
                 cellheight = 8,
                 cluster_rows = cluster_rows,
                 breaks = breaks,
                 cluster_cols = TRIP_chrom_clust,
                 annotation_row = target_ctrl,
                 clustering_method = "ward.D",
                 main = paste("Chromatin synergies - TIF per target :" , i, sep = " "),
                 color = colors,
                 annotation_colors = annotation_colors_scr, 
                 annotation_legend = T)
    print(p)
  }
}
```

```{r Z-score - TIF all, fig.width= 10, fig.height= 20}
ritios_tib_epi_effects = indel.data %>%
  mutate(drug_conc_ctrl = paste(drug, conc_char, sep = " ")) %>%
  filter(drug_conc_ctrl %in% rownames(all_compounds_epistasis)) %>%
  distinct(IPR, mean_eff_zscore_comb, drug_conc_ctrl, drug, conc_char) 

mmejzscore_matrix = ritios_tib_epi_effects %>%
  distinct(drug_conc_ctrl, IPR, mean_eff_zscore_comb) %>% 
  pivot_wider(names_from = IPR, values_from = mean_eff_zscore_comb) %>%
  column_to_rownames(var = "drug_conc_ctrl")


quant_breaks <- quantile(mmejzscore_matrix, c(.01,.99), na.rm = T) # change these numbers according to where you want to set the cutoff
palette_length <- 1000
breaks <- breaks <- c(seq(quant_breaks[1], -1, 
                          length.out=ceiling(palette_length/2) + 1), 
                      seq(1, 
                          quant_breaks[2], 
                          length.out=floor(palette_length/2)))
colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
colors[-which(breaks < -1.96 | breaks > 1.96 )] = "gray90"

pheatmap(mmejzscore_matrix[, clone5_map_ipr$labels],
         border_color = F,
         cellwidth = 8,
         cellheight = 8,
         breaks = breaks,
         cluster_cols = clone5_map_ipr,
         annotation_row = target_ctrl,
         main = "Z-score - TIF all",
         color = colors,
         annotation_colors = annotation_colors_scr, 
         annotation_legend = T,
         annotation_col = chromatin_TIF)
```

```{r Z-score - TIF per target, fig.width= 10, fig.height= 10}
targets = target %>% 
  pull(target) %>% 
  unique()

for (i in targets) {
  target_drugs = rownames(filter(target, target %in% i))
  
  # Make matrix with drugs from selected target
  mmejzscore_matrix = ritios_tib_epi_effects %>%
    distinct(drug_conc_ctrl, IPR, mean_eff_zscore_comb) %>% 
    filter(drug_conc_ctrl %in% target_drugs) %>%
    pivot_wider(names_from = IPR, values_from = mean_eff_zscore_comb) %>%
    column_to_rownames(var = "drug_conc_ctrl")
  
  cluster_rows = ifelse(nrow(mmejzscore_matrix) > 1, TRUE, FALSE)
  if(nrow(mmejzscore_matrix) > 0) {
    # change these numbers according to where you want to set the cutoff
    quant_breaks <- quantile(mmejzscore_matrix, c(.01,.99), na.rm = T) 
    palette_length <- 100
    breaks <- c(seq(quant_breaks[1], 0, length.out=ceiling(palette_length/2) + 1), 
                seq(quant_breaks[2]/palette_length, quant_breaks[2], length.out=floor(palette_length/2)))
    colors = colorRampPalette(slope.colors %>% pull(color))(length(breaks))
    colors[which(quant_breaks == 0)-1] = "gray90"
    
    # heatmap
    p = pheatmap(mmejzscore_matrix[, clone5_map_ipr$labels],
                 border_color = F,
                 cellwidth = 8,
                 cellheight = 8,
                 cluster_cols = clone5_map_ipr,
                 cluster_rows = cluster_rows,
                 breaks = breaks,
                 annotation_row = target_ctrl,
                 main = paste("Z-score - TIF per target :" , i, sep = " "),
                 color = colors,
                 annotation_colors = annotation_colors_scr, annotation_legend = T,
                 annotation_col = chromatin_TIF)
    print(p)
  }
}
```


```{r, fig.width= 10, fig.height= 8}
# ratio_targets = slope.features.tib %>% left_join(rownames_to_column(target, var = "drug")) %>% 
#   distinct(drug, conc, ratio_slope_plot, target, feature, conc) %>% 
#   filter(ratio_slope_plot != 0) %>% 
#   group_by(target, conc) %>% 
#   mutate(sign = ifelse(ratio_slope_plot < 0, "increase NHEJ", "increase MMEJ"),
#         total = length(target)) %>%
#   group_by(conc, sign, target) %>%
#   mutate(n = length(sign)) %>%
#   ungroup() %>%
#   distinct(target, total, sign, conc, n)
# 
# ggplot(ratio_targets,
#        aes(y = reorder(target, -total), x = n, fill = sign)) +
#   geom_bar(stat = "identity") +
#   facet_wrap(~conc)+
#   scale_fill_manual(values = c("#e07a5f", "#3d405b", "#C2C2C2"))


ratio_features = epistasis_tib %>% left_join(rownames_to_column(target, var = "drug_conc")) %>% 
  distinct(conc_char, ratio_slope_plot, target, feature, drug_conc) %>% 
  filter(!is.na(ratio_slope_plot)) %>% 
  mutate(sign = ifelse(ratio_slope_plot < 0, "increase NHEJ", "increase MMEJ")) %>%
  group_by(conc_char, feature, sign) %>% 
  mutate(total = length(target)) %>%
  group_by(conc_char, sign, feature, target) %>%
  mutate(n = length(target),
         fraction = n/total) %>%
  distinct(conc_char, feature, total, sign, n, target, fraction)

ggplot(ratio_features,
       aes(y = reorder(feature, -total), x = n, fill = target)) +
  geom_bar(stat = "identity") +
  facet_grid(conc_char ~ sign)+
  scale_fill_manual(values = annotation_colors_scr$target)
```

```{r, fig.width= 10, fig.height= 8}
ratio_features = epistasis_tib %>% left_join(rownames_to_column(target, var = "drug_conc")) %>% 
  distinct(conc_char, indelrate_slope_plot, target, feature, drug_conc) %>% 
  filter(!is.na(indelrate_slope_plot)) %>% 
  mutate(sign = ifelse(indelrate_slope_plot < 0, "decrease TIF", "increase TIF")) %>%
  group_by(conc_char, feature, sign) %>% 
  mutate(total = length(target)) %>%
  group_by(conc_char, sign, feature, target) %>%
  mutate(n = length(target),
         fraction = n/total) %>%
  distinct(conc_char, feature, total, sign, n, target, fraction)

ggplot(ratio_features,
       aes(y = reorder(feature, -total), x = n, fill = target)) +
  geom_bar(stat = "identity") +
  facet_grid(conc_char ~ sign)+
  scale_fill_manual(values = annotation_colors_scr$target)
```
