---
title: "KDM6-KD_indel_analysis"
author: "Max Trauernicht & Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  pdf_document:
    toc: yes
  editor_options:
    chunk_output_type: console
  html_document:
    theme: journal
    highlight: monochrome
    toc: yes
    toc_float: yes
    code_folding: show
---

knitr document van Steensel lab

# Preprocessing of the indel data
# Introduction

In this script I want to prepocess the raw data from the KDM6-KD experiment into one dataframe containing all information: +1/-7 ratio, efficiency, etc.


## Description of Data

For this analysis we need the mapping and the indel data of the TRIP integrations. These 
files are obtained with the crispr_trip.snake script that C. Leemans edited. This data 
contains the genomic locations of the TRIP integrations (hg38) and the indel frequencies 
at each integration.

The mutations were called by counting the distance between two constant regions. These
were separated by barcode. The barcodes were also filtered on the starcode, to pick out
the most abundant, and considered real, ones.

Mutations files : *genuine_mapped.table

| barcode  | type | score | 
| ------- | --------- | ----- | 
| TTCTATTCGCACACAA | ins | 1 |
| TTTCCCACATCAGGAG | wt | 0 |
| CCATAGTAGTGATTAC | del | -4 |

# Data importing and processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, message=FALSE, warnings=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

# libraries:
# library(devtools)
# install_github("lawremi/rtracklayer")
library(tidyr)
library(plyr)
library(dplyr)
library(data.table)
library(ggplot2)
library(car)
library(parallel)
library(gtools)
library(rtracklayer)
library(platetools)
library(pheatmap)
library(tibble)
library(ggbeeswarm)
library(RColorBrewer)
library(ggpubr)
```
## Custom functions
Functions used include all functions that have been used previously by Ruben (some of them might not be needed anymore, but are kept anyway). Functions 'CallTrueBarcodes' & 'RemoveLowReads' have been added by Max.
```{r functions, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), filename)
  filename
}

CallTrueBarcodes <- function(df) {
  df <- df[df$barcode %in% barcodes.list,]
  df
}

RemoveLowReads <- function(reads) {
  if (sum(reads$count) < 2000) {
    reads <- NULL
  } else {
    reads 
  } 
}

corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}

Read_Counts <- function(x) {
  #This function sums all the reads per barcode
  tmp <- x %>% group_by(barcode) %>% summarise(sum = sum(count))
  tmp$V2 <- as.character(tmp$barcode)
  colnames(tmp) <- c("barcode", "sum")
  tmp
}


MutationsToDataframe <- function(mutations) {
  # This function does
  tmp <- dcast(mutations, formula = barcode ~ indel, value.var = "count", fun = sum)
  tmp
}

RownamesList <- function(list) {
  rownames(list) <- list$"V2"
  list
}
  # n <- colnames(tmp)
  # names(tmp) <- n

IndelsOnly <- function(x) {
  #This function removes the "not_clear and "point_mut" from the indels.
  tmp <- x[!x$indel %in% c("not_clear", "wt_point_mut"), ]
  tmp
}

Not_Clear_Counts <- function(x){
  tmp <- x[x$V3 == "not_clear", ]
  tmp <- tmp[, c("V2", "V1")]
  colnames(tmp) <- c("barcode", "not_clear")
  tmp
}

Point_Mut_Counts <- function(x){
  tmp <- x[x$V3 == "wt_point_mut", ]
  tmp <- tmp[, c("V2", "V1")]
  colnames(tmp) <- c("barcode", "point_mut")
  tmp
}

RenameBarcodes <- function(list, pool) {
# This function renames the barcodes in the samples to barcode.pool to remove ambuguity between the cell pools
    for (i in pool) {
    list[[i]]$V2 <- paste(list[[i]]$V2, deparse(substitute(pool)), sep = ".")
  }
  list
}

NAToZero <- function(DT) {
  # This function changes NA to 0 in a data.table by name :
  for (j in names(DT))
    set(DT,which(is.na(DT[[j]])),j,0)
  return(DT)
  }

RemoveBarcodeFromList_RN <- function(x, barcode) {
  tmp <- x[!rownames(x)==barcode, ]
  tmp
}

RemoveBarcodeFromList_barcode <- function(x, barcode) {
  tmp <- x[!x$barcode==barcode, ]
  tmp
}

CalculateRatio <- function(mutations, indel) {
  # This function calculates the proportion of any indel over all the indels using the 
  # dataframe created with MutationsToDataframe.
  if (as.character(indel) %in% names(mutations)) {
    indel_reads <- mutations[,as.character(indel)]
    indel_reads <- as.matrix(indel_reads)
    rownames(indel_reads) <- rownames(mutations)
    
    ratio=ifelse(((indel_reads==0) | mutations[, "0"]==0), NA, 
                 as.numeric(mutations[,as.character(indel)]/(rowSums(mutations)-mutations[, "0"])))
                 
    ratio=round(ratio,3)
    
    df <- data.frame(barcode = as.character(row.names(mutations)),
             ratio = ratio,
             stringsAsFactors = FALSE)
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel)))
    df
  } else {
    df <- data.frame(barcode = as.character(row.names(mutations)),
             ratio = NA, 
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel)))
    df
  }
}

CalculatePercentage <- function(mutations, indel) {
  # This function calculates the proportion of any indel over all the indels using the 
  # dataframe created with MutationsToDataframe.
  if (as.character(indel) %in% names(mutations)) {
    indel_reads <- mutations[,as.character(indel)]
    indel_reads <- as.matrix(indel_reads)
    rownames(indel_reads) <- rownames(mutations)
    
    percentage=ifelse((indel_reads==0), 0, 
                 as.numeric(mutations[,as.character(indel)]/(rowSums(mutations))*100))
                 
    percentage=round(percentage,3)
    
    df <- data.frame(barcode = as.character(row.names(mutations)),
             percentage = percentage,
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("percentage.", as.character(indel)))
    df
  } else {
    df <- data.frame(barcode = as.character(row.names(mutations)),
             ratio = NA,
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("percentage.", as.character(indel)))
    df
  }
}


CalculateRatio_internal <- function(mutations, indel1, indel2){
   # This function calculates the proportion of any indel over all the indels using the 
  # dataframe created with MutationsToDataframe.
  if (as.character(indel1) %in% names(mutations) & as.character(indel2) %in% names(mutations)) {
    indel1_reads <- mutations[,as.character(indel1)]
    indel1_reads <- as.matrix(indel1_reads)
    rownames(indel1_reads) <- rownames(mutations)
    
    indel2_reads <- mutations[,as.character(indel2)]
    indel2_reads <- as.matrix(indel2_reads)
    rownames(indel2_reads)=rownames(mutations)
    
    ratio=ifelse(((indel1_reads==0 & indel2_reads==0) | indel2_reads==0), NA, 
                 as.numeric(mutations[,as.character(indel1)]/mutations[,as.character(indel2)]))
                 
    ratio=round(ratio,3)
    
    df <- data.frame(barcode = row.names(mutations),
             ratio = ratio,
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel1), "_", as.character(indel2)))
    df
  } else {
    df <- data.frame(barcode = row.names(mutations),
             ratio = NA, 
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel1), "_", as.character(indel2)))
    df
  }
}

CalculateEfficiencies <- function(efficiencies) {
  # This function calculates the efficiency of the cut and repair at each barcode
  # using the dataframe created with MutationsToDataframe.
  data.frame(barcode = row.names(efficiencies),
             efficiency = (1 - efficiencies[, "0"] / rowSums(efficiencies)) * 100)
}

CalculateInsertions <- function(insertions) {
  # This function calculates the efficiency of the cut and repair at each barcode
  # using the dataframe created with MutationsToDataframe.
  data.frame(barcode = row.names(insertions),
             insertions = (rowSums(insertions[, -grep('^[-0]',colnames(insertions))]) / rowSums(insertions)) * 100)
}

CalculateDeletions <- function(deletions) {
  # This function calculates the efficiency of the cut and repair at each barcode
  # using the dataframe created with MutationsToDataframe.
  data.frame(barcode = row.names(deletions),
             deletions = (rowSums(deletions[ ,grep('-',colnames(deletions))]) / rowSums(deletions)) * 100)
}
```

## Data import
Data import from mapping, files were generated on 21.01.2019.
```{r import, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Set directory to the mutation output folder of the CRISPR-TRIP snakemake script
setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/data/KDM6_KD/processed/indelPCR_counts/")

# Import files in list and make individual tables
file.list <- list.files(pattern='*[.]count.table')
df.list <- lapply(file.list, fread)

# Import RSTP2 clone 5 barcode list (in total 20 barcodes)
barcodes.list <- c("AGGGCGTAAAATATTT", "TATGGCTGTCGGGTAG", "TGTCCCTTAGTACTTT", "AGAAAATAATATGACG","CGGCCTGAAGGTCAGG","TTGAACGCGGGCTCGG","GCTAACATCACGAATC","GCGCACCCTTTAATTG","ACTGTCGAGTTGTCCG","CCGGGGACGTATGCAC","TCTTTTGAGGAGCTGA","ATATCGTTGCTGGAGA","CATCCACCACACTTCA","ACCCCTAAAGGCGCTG","ATACTATATTTAACGG","CATTTCTGATCAATAA","CTCTTAATCGCTGCC","GAGCGCGTCACCGGGT","GTACCTCTCGATAGTG","TGGCCAATATTTGTCT")

  
# rename the lists
names(df.list) <- gsub('(*?)[.]ra.*', '\\1', file.list)

# these are the samples
head(names(df.list))

# count the sample number
n.samples <- length(df.list)
```


### Select true barcodes and exclude low read counts
```{r preprocessing, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Call true barcodes
df.list <- mclapply(df.list, CallTrueBarcodes, mc.cores = 10)

# Remove samples with low read counts
df.list <- mclapply(df.list, RemoveLowReads, mc.cores = 10)

# Remove empty dfs
df.list <- Filter(Negate(function(x) is.null(unlist(x))), df.list)
```



### Some data pre-processing
Set everything in a dataframe that contains the barcodes, indel ratios, and efficiencies.
```{r indeldataframe, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Total reads per barcode
total.reads.list <- mclapply(df.list, Read_Counts, mc.cores = 10)

# Now remove not clear samples from all the mutations lists before we make the indel dataframes
indel.list <- mclapply(df.list, IndelsOnly, mc.cores = 10)

# Transform to long format
mutations.list <- mclapply(indel.list, MutationsToDataframe, mc.cores = 10)
```

# Analysis
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Put the lists into one df
mutations.dt.list <- copy(mutations.list)
mutations.dt.list <- lapply(mutations.dt.list, setDT, keep.rownames = TRUE)
indels.df <- ldply(mutations.dt.list, data.frame)
```

# Add identifiers and compute indel ratios
```{r calculate_ratios, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Assigning identifiers to df
indels.df <- indels.df %>% 
  mutate(.id = gsub(".count.table", "\\1", .id),
         rep = gsub(".*rep([1-3])", "\\1", .id),
         siRNA = gsub("(.*?)_rep[1-3]","\\1", .id))

# Let's plot the distribution of the amount of reads per drug
indels.df <- mutate_all(indels.df, funs(replace(., is.na(.), 0)))
indels.df$read.count <- rowSums(indels.df[,c(4:97)])
indels.df$all.indels <- rowSums(indels.df[,c(4:60,62:97)])
ggplot(indels.df, aes(read.count)) + geom_density(fill = "#3D9F83") + theme_bw() + 
  labs(title = "read count distribution of all conditions")
ggplot(indels.df, aes(read.count)) + geom_density(fill = "#3D9F83") + theme_bw() + 
  labs(title = "read distribution per siRNA") + 
  facet_wrap(~siRNA)
ggplot(indels.df, aes(read.count)) + geom_density(fill = "#3D9F83") + theme_bw() + 
  labs(title = "read distribution per barcode") + 
  facet_wrap(~barcode)

# Filter out noisy data
indels.df$read.count1.7 <- rowSums(indels.df[,c(54,62)])
indels.df <- indels.df[indels.df$read.count1.7 > 30,] %>% 
  select(-read.count1.7)


# Calculate NHEJ/MMEJ ratios and efficiency
indels.df <- indels.df %>%
  mutate(MMEJratio = X.7 / all.indels,
         freqCut = 1 - (X0 / (X0 + all.indels)))
indels_df <- indels.df
indels.df <- indels.df %>%
  select(.id, barcode, siRNA, rep, MMEJratio, freqCut) %>%
  setnames(".id", "id")


# Remove barcodes with abnormal counts from analysis
indels.df <- indels.df[indels.df$barcode != "ACCCCTAAAGGCGCTG",]

# Exclude NA rows
indels.df <- indels.df[rowSums(is.na(indels.df)) != ncol(indels.df), ]

# Compute mean per condition and normalize against scrambled control
scr_ctrl <- indels.df %>%
  filter(siRNA == "Scr") %>%
  mutate(MMEJratio_mean_Scr = ave(MMEJratio, barcode, FUN = mean)) %>%
  mutate(freqCut_mean_Scr = ave(freqCut, barcode, FUN = mean)) %>%
  mutate(MMEJratio_Scr = ave(MMEJratio, barcode, rep, FUN = mean)) %>%
  mutate(freqCut_Scr = ave(freqCut, barcode, rep, FUN = mean)) %>%
  distinct(MMEJratio_Scr, freqCut_Scr, MMEJratio_mean_Scr, freqCut_mean_Scr, barcode, rep)

indels.df <- merge(indels.df, scr_ctrl) %>%
  mutate(freqCut_mean = ave(freqCut, siRNA, barcode, FUN = mean),
         MMEJratio_mean = ave(MMEJratio, siRNA, barcode, FUN = mean)) %>%
  mutate(freqCut_norm = log2(freqCut / freqCut_Scr),
         MMEJratio_norm = log2(MMEJratio / MMEJratio_Scr)) %>%
  mutate(freqCut_mean_norm = log2(freqCut_mean / freqCut_mean_Scr),
         MMEJratio_mean_norm = log2(MMEJratio_mean / MMEJratio_mean_Scr)) %>%
  dplyr::select(-freqCut_Scr, -MMEJratio_Scr, -freqCut_mean_Scr, -MMEJratio_mean_Scr)
```




# Make first plots
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Import GSK J4 data from drug screen
gsk.data <- readRDS("/DATA/projects/DSBrepair/data/R/rs20220201_episcreen/rs20220201_episcreen_ratios.RDS") %>%
  filter(drug == "GSK J4 HCl") %>%
  distinct(MMEJ_ratio_norm, IPR) %>%
  mutate(MMEJ_ratio_norm = log2(ave(MMEJ_ratio_norm, IPR, FUN = mean))) %>%
  distinct() %>%
  setnames("MMEJ_ratio_norm", "gsk_mmej_ratio_norm")

# Add IPR
indels.df.heatmap <- readRDS("/DATA/projects/DSBrepair/data/R/rs20200519_clone5_newdoms_chromatin.RDS") %>%
  filter(IPR != "IPR1") %>%
  distinct(barcode, IPR) %>%
  left_join(indels.df)

compare_hdm <- indels.df.heatmap %>%
  filter(str_detect(siRNA, "KDM")) %>%
  distinct(MMEJratio_norm, id, IPR)

compare_hdm <- merge(compare_hdm, gsk.data)

ggplot(compare_hdm, 
       aes(x = gsk_mmej_ratio_norm, y = MMEJratio_norm)) +
  geom_quasirandom() +
  facet_wrap(~id) +
  theme_classic_lines_90()

# Replicate correlation
compare_hdm <- compare_hdm %>%
  spread(id, MMEJratio_norm) %>%
  dplyr::select(-gsk_mmej_ratio_norm)

boundaries <- c(0.7,0.75,0.8,0.85)
ggpairs(compare_hdm %>% dplyr::select(contains("KDM6B")),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data, mapping = mapping, alpha = 0.6, size = 0.5) +
                   scale_y_continuous(breaks = c(-2, -1, 0), limits = c(-1.9, 0.5)) +
                   scale_x_continuous(breaks = c(-2, -1, 0), limits = c(-1.9, 0.5)) +
                   geom_abline(slope = 1, lty = "twodash", col = "grey") +
                   geom_smooth(method = "lm", color = "red", lty = "dashed", size = 0.5)}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red")})) +
  xlab("MMEJ ratio change (log2)") +
  ylab("MMEJ ratio change (log2)")



# Add chromatin type
indels.df.heatmap <- readRDS(
  "/DATA/usr/x.vergara/XV_ChIPIndels/XV20200902_DDR_RS_revision/XV20200902_DDR_RS_revision/data/xv20200915_DDR_data_CHIP.rds") %>% 
  distinct(barcode, chromatin) %>%
  left_join(indels.df.heatmap) %>%
  na.omit() %>% 
  select(-barcode)

# Add barcode annotation
barcode_ann <- indels.df.heatmap %>%
  distinct(IPR, chromatin) %>%
  column_to_rownames("IPR")

# Add chromatin colors
# Change colors of annotations
chrom_colors = c("Euchromatin" = "#EAEEE8",
                  "other-heterochromatin" = "#a5a58d",  
                     "H3K27me3" = "#CAD6C2", 
                  "Triple Heterochromatin" = "#6b705c")

# colors <- brewer.pal(length(unique(barcode.annotation$cluster)), "Pastel2")
# names(colors) <- unique(barcode.annotation$cluster)

annotation_colors_scr = list(
  chromatin = chrom_colors)


## Generate MMEJscore heatmaps of id vs. barcode
indels.df.heatmap <- indels.df.heatmap %>%
  distinct(siRNA, IPR, MMEJratio_norm) %>%
  spread(siRNA, MMEJratio_norm) %>%
  column_to_rownames("IPR")

myBreaks1 <- seq(-3,3,0.06)

pheatmap(indels.df.heatmap,
         color = colorRampPalette(c("red", "white", "blue"))(100),
         breaks = myBreaks1, 
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8, 
         annotation_row = barcode_ann,
         annotation_colors = annotation_colors_scr, 
         main = "MMEJscore of all samples per barcode")


## Generate freqCut heatmaps of id vs. barcode
indels.df.heatmap <- indels.df %>%
  dcast(id ~ barcode, value.var = "freqCut") %>%
  column_to_rownames("id")

myBreaks1 <- seq(0,01,0.01)

pheatmap(as.matrix(indels.df.heatmap),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         breaks = myBreaks1, border_color = "black", 
         cellwidth = 12, cellheight = 10,annotation_col = barcode_ann,
         angle_col = 90, main = "freqCut of all samples per barcode")

## Plotting MMEJscore distribution
ggplot(indels.df, aes(x = id, y = MMEJscore)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "MMEJscore distribution per sample") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))

## Plotting freqCut distribution
ggplot(indels.df, aes(x = id, y = freqCut)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "freqCut distribution per sample") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))


## Plot separately for each rep
indels.df.rep1 <- indels.df[indels.df$rep == 1,]
indels.df.rep2 <- indels.df[indels.df$rep == 2,]
indels.df.rep3 <- indels.df[indels.df$rep == 3,]
ggplot(indels.df.rep1, aes(x = id, y = MMEJscore)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "MMEJscore distribution per sample - only rep 1") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))

ggplot(indels.df.rep2, aes(x = id, y = MMEJscore)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "MMEJscore distribution per sample - only rep 2") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))

ggplot(indels.df.rep3, aes(x = id, y = MMEJscore)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "MMEJscore distribution per sample - only rep 3") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))

ggplot(indels.df.rep1, aes(x = id, y = freqCut)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "freqCut distribution per sample - only rep 1") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))

ggplot(indels.df.rep2, aes(x = id, y = freqCut)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "freqCut distribution per sample - only rep 2") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))

ggplot(indels.df.rep3, aes(x = id, y = freqCut)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "freqCut distribution per sample - only rep 3") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))

## Generate MMEJscore heatmaps of id vs. barcode - only rep1
indels.df.heatmap <- indels.df.rep1 %>%
  dcast(id ~ barcode, value.var = "MMEJscore") %>%
  column_to_rownames("id")

myBreaks1 <- seq(0,0.3,0.003)

pheatmap(as.matrix(indels.df.heatmap),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         breaks = myBreaks1, border_color = "black", 
         cellwidth = 12, cellheight = 10,annotation_col = barcode_ann,
         angle_col = 90, main = "MMEJscore of rep1 samples per barcode")

## Do the same for rep2

indels.df.heatmap <- indels.df.rep2 %>%
  dcast(id ~ barcode, value.var = "MMEJscore") %>%
  column_to_rownames("id")

myBreaks1 <- seq(0,0.3,0.003)

pheatmap(as.matrix(indels.df.heatmap),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         breaks = myBreaks1, border_color = "black", 
         cellwidth = 12, cellheight = 10,annotation_col = barcode_ann,
         angle_col = 90, main = "MMEJscore of rep2 samples per barcode")

## Do the same for rep2

indels.df.heatmap <- indels.df.rep3 %>%
  dcast(id ~ barcode, value.var = "MMEJscore") %>%
  column_to_rownames("id")

myBreaks1 <- seq(0,0.3,0.003)

pheatmap(as.matrix(indels.df.heatmap),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         breaks = myBreaks1, border_color = "black", 
         cellwidth = 12, cellheight = 10,annotation_col = barcode_ann,
         angle_col = 90, main = "MMEJscore of rep3 samples per barcode")

indels.df.heatmap <- indels.df %>%
  select(siRNA, barcode, MMEJscore) %>%
  mutate(MMEJscore = ave(MMEJscore, siRNA, barcode, FUN = function(x) mean(x))) %>%
  unique() %>%
  dcast(siRNA ~ barcode, value.var = "MMEJscore") %>%
  column_to_rownames("siRNA")

myBreaks1 <- seq(0,0.4,0.004)

pheatmap(as.matrix(indels.df.heatmap),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "black", 
         cellwidth = 12, cellheight = 10,annotation_col = barcode_ann,
         angle_col = 90, main = "MMEJscore of rep3 samples per barcode")
```


# Generate final overview
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
my_comparisons <- list(c("KDM6A", "Scr"), c("KDM6B", "Scr"), c("PolQ", "Scr"))

# Boxplot comparing the 4 conditions
ggplot(indels.df, aes(x = siRNA, y = MMEJscore)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "MMEJscore distribution per condition") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))+
  stat_compare_means(comparisons = my_comparisons)+ 
  stat_compare_means(label.y = 0.5)



ggplot(indels.df, aes(x = siRNA, y = freqCut)) +
  geom_point() + geom_boxplot() + theme_bw() +
  labs(title = "freqCut distribution per condition") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10))+
  stat_compare_means(comparisons = my_comparisons)+ 
  stat_compare_means(label.y = 1)
```

# Make beeswarm indel plots per siRNA condition
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
indels_df <- indels_df[,c(1:96)]
indels_df_long <- melt(indels_df, id.vars = c("barcode", ".id"))
indels_df_long$variable <- gsub("X", "", indels_df_long$variable)
indels_df_long$variable <- gsub("[.]", "-", indels_df_long$variable)
indels_df_long <- indels_df_long[indels_df_long$variable != 'Inf.',]
indels_df_long$variable <- as.numeric(indels_df_long$variable)
indels_df_long <- indels_df_long[indels_df_long$variable < 5 & indels_df_long$variable > -10,]
indels_df_long <- indels_df_long[!is.na(indels_df_long$variable),]
indels_df_long$value <- ave(indels_df_long$value, indels_df_long$barcode, indels_df_long$.id, 
                               FUN = function(x) x/sum(x))

# Plot it in a beeswarm plot
ggplot(indels_df_long, aes(variable, value, group = variable)) + geom_point() + geom_boxplot() + theme_bw() +
  facet_wrap(~`.id`) + ylab("indel percentage per barcode") + xlab("sample")
```










## Processed data export
```{r export}

```

### Conclusions 
Looks all fine! Now, we can generate indel plots and look in detail at the ratios.

### Session Info
```{r session_info}
sessionInfo()
getwd()
date()
paste("Run time: ",format(Sys.time()-StartTime))

```

