---
title: "Indels Preprocessing"
author: "Max Trauernicht & Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

knitr document van Steensel lab

# Preprocessing of the indel data
# Introduction

In this script I want to prepocess the raw data from the epigenetic screening into one dataframe containing all information: +1/-7 ratio, efficiency, etc. This should all be integrated with the drug and the drug target group corresponding to the well. 
This will be used for plotting indel patterns, calculating ratios later on and such.

* Efficiency (All mutations / Total or (Total - WT sequences) / Total)  
* +1 / -7  


## Description of Data

For this analysis we need the mapping and the indel data of the TRIP integrations. These 
files are obtained with the crispr_trip.snake script that C. Leemans edited. This data 
contains the genomic locations of the TRIP integrations (hg38) and the indel frequencies 
at each integration.

The mutations were called by counting the distance between two constant regions. These
were separated by barcode. The barcodes were also filtered on the starcode, to pick out
the most abundant, and considered real, ones.

Mutations files : *genuine_mapped.table

| barcode  | type | score | 
| ------- | --------- | ----- | 
| TTCTATTCGCACACAA | ins | 1 |
| TTTCCCACATCAGGAG | wt | 0 |
| CCATAGTAGTGATTAC | del | -4 |

# Data importing and processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, message=FALSE, warnings=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

# libraries:
# library(devtools)
# install_github("lawremi/rtracklayer")
library(tidyr)
library(plyr)
library(dplyr)
library(data.table)
library(ggplot2)
library(car)
library(parallel)
library(gtools)
library(rtracklayer)
library(platetools)
```
## Custom functions
Functions used include all functions that have been used previously by Ruben (some of them might not be needed anymore, but are kept anyway). Functions 'CallTrueBarcodes' & 'RemoveLowReads' have been added by Max.
```{r functions}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), filename)
  filename
}

CallTrueBarcodes <- function(df) {
  df <- df[df$V2 %in% barcodes.list,]
  df
}

RemoveLowReads <- function(reads) {
  if (sum(reads$V1) < 2000) {
    reads <- NULL
  } else {
    reads 
  } 
}

Read_Counts <- function(x) {
  #This function sums all the reads per barcode
  tmp <- x %>% group_by(V2) %>% summarise(sum = sum(V1))
  tmp$V2 <- as.character(tmp$V2)
  colnames(tmp) <- c("barcode", "sum")
  tmp
}


MutationsToDataframe <- function(mutations) {
  # This function does
  tmp <- dcast(mutations, formula = V2 ~ V4, value.var = "V1", fun = sum)
  tmp
}

RownamesList <- function(list) {
  rownames(list) <- list$"V2"
  list
}
  # n <- colnames(tmp)
  # names(tmp) <- n

IndelsOnly <- function(x) {
  #This function removes the "not_clear and "point_mut" from the indels.
  tmp <- x[!x$V3 %in% c("not_clear", "wt_point_mut"), ]
  tmp
}

Not_Clear_Counts <- function(x){
  tmp <- x[x$V3 == "not_clear", ]
  tmp <- tmp[, c("V2", "V1")]
  colnames(tmp) <- c("barcode", "not_clear")
  tmp
}

Point_Mut_Counts <- function(x){
  tmp <- x[x$V3 == "wt_point_mut", ]
  tmp <- tmp[, c("V2", "V1")]
  colnames(tmp) <- c("barcode", "point_mut")
  tmp
}

RenameBarcodes <- function(list, pool) {
# This function renames the barcodes in the samples to barcode.pool to remove ambuguity between the cell pools
    for (i in pool) {
    list[[i]]$V2 <- paste(list[[i]]$V2, deparse(substitute(pool)), sep = ".")
  }
  list
}

NAToZero <- function(DT) {
  # This function changes NA to 0 in a data.table by name :
  for (j in names(DT))
    set(DT,which(is.na(DT[[j]])),j,0)
  return(DT)
  }

RemoveBarcodeFromList_RN <- function(x, barcode) {
  tmp <- x[!rownames(x)==barcode, ]
  tmp
}

RemoveBarcodeFromList_barcode <- function(x, barcode) {
  tmp <- x[!x$barcode==barcode, ]
  tmp
}

CalculateRatio <- function(mutations, indel) {
  # This function calculates the proportion of any indel over all the indels using the 
  # dataframe created with MutationsToDataframe.
  if (as.character(indel) %in% names(mutations)) {
    indel_reads <- mutations[,as.character(indel)]
    indel_reads <- as.matrix(indel_reads)
    rownames(indel_reads) <- rownames(mutations)
    
    ratio=ifelse(((indel_reads==0) | mutations[, "0"]==0), NA, 
                 as.numeric(mutations[,as.character(indel)]/(rowSums(mutations)-mutations[, "0"])))
                 
    ratio=round(ratio,3)
    
    df <- data.frame(barcode = as.character(row.names(mutations)),
             ratio = ratio,
             stringsAsFactors = FALSE)
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel)))
    df
  } else {
    df <- data.frame(barcode = as.character(row.names(mutations)),
             ratio = NA, 
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel)))
    df
  }
}

CalculatePercentage <- function(mutations, indel) {
  # This function calculates the proportion of any indel over all the indels using the 
  # dataframe created with MutationsToDataframe.
  if (as.character(indel) %in% names(mutations)) {
    indel_reads <- mutations[,as.character(indel)]
    indel_reads <- as.matrix(indel_reads)
    rownames(indel_reads) <- rownames(mutations)
    
    percentage=ifelse((indel_reads==0), 0, 
                 as.numeric(mutations[,as.character(indel)]/(rowSums(mutations))*100))
                 
    percentage=round(percentage,3)
    
    df <- data.frame(barcode = as.character(row.names(mutations)),
             percentage = percentage,
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("percentage.", as.character(indel)))
    df
  } else {
    df <- data.frame(barcode = as.character(row.names(mutations)),
             ratio = NA,
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("percentage.", as.character(indel)))
    df
  }
}


CalculateRatio_internal <- function(mutations, indel1, indel2){
   # This function calculates the proportion of any indel over all the indels using the 
  # dataframe created with MutationsToDataframe.
  if (as.character(indel1) %in% names(mutations) & as.character(indel2) %in% names(mutations)) {
    indel1_reads <- mutations[,as.character(indel1)]
    indel1_reads <- as.matrix(indel1_reads)
    rownames(indel1_reads) <- rownames(mutations)
    
    indel2_reads <- mutations[,as.character(indel2)]
    indel2_reads <- as.matrix(indel2_reads)
    rownames(indel2_reads)=rownames(mutations)
    
    ratio=ifelse(((indel1_reads==0 & indel2_reads==0) | indel2_reads==0), NA, 
                 as.numeric(mutations[,as.character(indel1)]/mutations[,as.character(indel2)]))
                 
    ratio=round(ratio,3)
    
    df <- data.frame(barcode = row.names(mutations),
             ratio = ratio,
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel1), "_", as.character(indel2)))
    df
  } else {
    df <- data.frame(barcode = row.names(mutations),
             ratio = NA, 
             stringsAsFactors = FALSE) 
    colnames(df) <- c("barcode", paste0("ratio.", as.character(indel1), "_", as.character(indel2)))
    df
  }
}

CalculateEfficiencies <- function(efficiencies) {
  # This function calculates the efficiency of the cut and repair at each barcode
  # using the dataframe created with MutationsToDataframe.
  data.frame(barcode = row.names(efficiencies),
             efficiency = (1 - efficiencies[, "0"] / rowSums(efficiencies)) * 100)
}

CalculateInsertions <- function(insertions) {
  # This function calculates the efficiency of the cut and repair at each barcode
  # using the dataframe created with MutationsToDataframe.
  data.frame(barcode = row.names(insertions),
             insertions = (rowSums(insertions[, -grep('^[-0]',colnames(insertions))]) / rowSums(insertions)) * 100)
}

CalculateDeletions <- function(deletions) {
  # This function calculates the efficiency of the cut and repair at each barcode
  # using the dataframe created with MutationsToDataframe.
  data.frame(barcode = row.names(deletions),
             deletions = (rowSums(deletions[ ,grep('-',colnames(deletions))]) / rowSums(deletions)) * 100)
}
```

## Data import
Data import from mapping, files were generated on 21.01.2019.
```{r import}
# Set directory to the mutation output folder of the CRISPR-TRIP snakemake script
setwd("/DATA/projects/DSBrepair/data/5280_DrugScreen/processed_190305/indelPCR_counts/")

# Import files in list and make individual tables
file.list <- list.files(pattern='*[.]count.table')
df.list <- lapply(file.list, fread)

# Import RSTP2 clone 5 barcode list (in total 20 barcodes)
barcodes.list <- c("AGGGCGTAAAATATTT", "TATGGCTGTCGGGTAG", "TGTCCCTTAGTACTTT", "AGAAAATAATATGACG","CGGCCTGAAGGTCAGG","TTGAACGCGGGCTCGG","GCTAACATCACGAATC","GCGCACCCTTTAATTG","ACTGTCGAGTTGTCCG","CCGGGGACGTATGCAC","TCTTTTGAGGAGCTGA","ATATCGTTGCTGGAGA","CATCCACCACACTTCA","ACCCCTAAAGGCGCTG","ATACTATATTTAACGG","CATTTCTGATCAATAA","CTCTTAATCGCTGCC","GAGCGCGTCACCGGGT","GTACCTCTCGATAGTG","TGGCCAATATTTGTCT")

# Import drug list 
setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/")
drugs.list <- read.delim("mt20190123_EpigeneticDrugList.txt")
drugs.list <- drugs.list %>%
  dplyr::select(-IC50..ÂµM.) %>%
  setnames("Number", "well")
  
  
# rename the lists
names(df.list) <- gsub('(*?)[.]ra.*', '\\1', file.list)

# these are the samples
head(names(df.list))

# count the sample number
n.samples <- length(df.list)

# Import viability data and the mean +1/-7 ratio from previous experiments
viability.data <- read.delim("mt20190124_viabilityData") %>% setnames(".id", "id")
viability <- read.csv2("EpiscreenAllPlates.csv")
```


## Visualize viability data
```{r viability_overview, echo = FALSE, warning = FALSE}
# Compute relative viability per plate
viability.df <- melt(viability) %>% 
  mutate(value = ave(value, variable, FUN = function(x) x/max(x)))

for (i in unique(viability.df$variable)) {
p <- raw_grid(data = viability.df$value[viability.df$variable == i],
       well = viability.df$X[viability.df$variable == i],
       plate_id = viability.df$variable[viability.df$variable == i])
print(p)
}
```



### dplyr::select true barcodes and exclude low read counts
```{r preprocessing}
# Call true barcodes
df.list <- mclapply(df.list, CallTrueBarcodes, mc.cores = 10)

# Remove samples with low read counts
df.list <- mclapply(df.list, RemoveLowReads, mc.cores = 10)

# Remove empty dfs
df.list <- Filter(Negate(function(x) is.null(unlist(x))), df.list)
```



### Some data pre-processing
Set everything in a dataframe that contains the barcodes, indel ratios, and efficiencies.
```{r indeldataframe}
# Total reads per barcode
total.reads.list <- mclapply(df.list, Read_Counts, mc.cores = 10)

# Now remove not clear samples from all the mutations lists before we make the indel dataframes
indel.list <- mclapply(df.list, IndelsOnly, mc.cores = 10)

# Transform to long format
mutations.list <- mclapply(indel.list, MutationsToDataframe, mc.cores = 10)
```

# Analysis
```{r}
# Put the lists into one df
mutations.dt.list <- copy(mutations.list)
mutations.dt.list <- lapply(mutations.dt.list, setDT, keep.rownames = TRUE)
indels.df <- ldply(mutations.dt.list, data.frame)
```

# Add identifiers and compute indel ratios
```{r calculate_ratios}
RSTP2_2000_indels.df <- indels.df
# Assigning identifiers to df
RSTP2_2000_indels.df <- RSTP2_2000_indels.df %>% 
  mutate(.id = gsub(".count.table", "\\1", .id),
         rep = gsub(".*m_rep(.*?)_.*", "R\\1", .id),
         conc = gsub("_rep.*", "\\1", .id),
         well = gsub(".*plate.*_(.*?)","\\1", .id))



# Import the drugs and match with the df
RSTP2_2000_indels.df <- merge(drugs.list, RSTP2_2000_indels.df) %>%
  mutate(well = gsub("_[0-9]{3}", "", well))

# Let's plot the distribution of the amount of reads per drug
RSTP2_2000_indels.df <- mutate_all(RSTP2_2000_indels.df, funs(replace(., is.na(.), 0)))
RSTP2_2000_indels.df$read.count <- rowSums(RSTP2_2000_indels.df[,c(6:118)])
RSTP2_2000_indels.df$all.indels <- rowSums(RSTP2_2000_indels.df[,c(6:53,55:118)])
ggplot(RSTP2_2000_indels.df, aes(read.count)) + geom_density(fill = "#3D9F83") + theme_bw() + 
  labs(title = "read distribution per barcode") + 
  facet_wrap(~V2)

# Filter out noisy data
RSTP2_2000_indels.df$read.count1.7 <- rowSums(RSTP2_2000_indels.df[,c(18,25)])


# Calculate NHEJ/MMEJ ratios and efficiency
RSTP2_2000_indels.df <- RSTP2_2000_indels.df %>%
  mutate(freqMMEJ = X.7 / all.indels,
         freqNHEJ = X1 / all.indels,
         NHEJMMEJratio = freqMMEJ / freqNHEJ,
         MMEJscore = freqMMEJ / (freqMMEJ + freqNHEJ),
         NHEJscore = freqNHEJ / (freqNHEJ + freqMMEJ),
         freqCut = 1 - (X0 / (X0 + all.indels)))
RSTP2_2000_indels.df <- RSTP2_2000_indels.df %>%
  dplyr::select(Drug, Target, well, V2, rep, conc, freqMMEJ, freqNHEJ,
         NHEJMMEJratio, MMEJscore, NHEJscore, freqCut, .id, read.count1.7)
setnames(RSTP2_2000_indels.df, c("Drug", "Target", "V2", ".id"), c("drug", "target", "barcode", "id"))

# Exclude NA rows
RSTP2_2000_indels.df <- RSTP2_2000_indels.df[rowSums(is.na(RSTP2_2000_indels.df)) != ncol(RSTP2_2000_indels.df), ]

# Add sample identifier
RSTP2_2000_indels.df$sample <- "drug"
RSTP2_2000_indels.df$sample[RSTP2_2000_indels.df$drug == "DMSO"] <- "DMSO"
RSTP2_2000_indels.df$sample[RSTP2_2000_indels.df$drug == "DNA-PKi"] <- "DNA-PKi"
RSTP2_2000_indels.df$sample[RSTP2_2000_indels.df$drug == "Mirin"] <- "Mirin"
```

## Processed data export
The files will be saved in the processed data folder.
```{r export}
setwd("/DATA/usr/m.trauernicht/projects/EpiScreen/files_scripts/")
# The mutations list that can be loaded for the indel spectra plots.
filename <- SetFileName("_drug_screen_mutations_rep1", "mt")
save(RSTP2_2000_indels.df, file = filename)
```

### Conclusions 
Looks all fine! Now, we can generate indel plots and look in detail at the ratios.

### Session Info
```{r session_info}
sessionInfo()
getwd()
date()
paste("Run time: ",format(Sys.time()-StartTime))

```

